{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MHC class II genotype vs CH risk"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- note: this is the script to reproduce the figures from the report\n",
    "- more complete analysis is provided in additional scripts "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n",
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n"
     ]
    }
   ],
   "source": [
    "# imported packages\n",
    "import random\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import matplotlib.ticker as plticker\n",
    "from matplotlib.ticker import LinearLocator, FormatStrFormatter\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "import scipy.special\n",
    "import scipy.integrate as it\n",
    "from scipy import integrate\n",
    "from scipy.interpolate import interp1d\n",
    "from scipy.stats import kde\n",
    "import copy\n",
    "import glob, os\n",
    "import re\n",
    "# from sklearn import datasets, linear_model\n",
    "import pandas as pd\n",
    "from decimal import *\n",
    "from operator import itemgetter    \n",
    "from collections import OrderedDict\n",
    "import timeit\n",
    "import time \n",
    "import csv\n",
    "import seaborn as sns \n",
    "import scipy as sp\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from matplotlib.ticker import LogLocator\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy.stats import kruskal\n",
    "from myfunctions import transform_format\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "timestr = time.strftime(\"%Y%m%d\") # get current date "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify font for plotting \n",
    "plt.rcParams.update({'font.sans-serif':'Verdana'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import batch data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Examined file: batch_21_annotated_data\n",
      "Examined file: batch_18_annotated_data\n",
      "Examined file: batch_15_annotated_data\n",
      "Examined file: batch_11_annotated_data\n",
      "Examined file: batch_16_annotated_data\n",
      "Examined file: batch_12_annotated_data\n",
      "Examined file: batch_13_annotated_data\n",
      "Examined file: batch_17_annotated_data\n",
      "Examined file: batch_14_annotated_data\n",
      "Examined file: batch_19_annotated_data\n",
      "Examined file: batch_20_annotated_data\n",
      "Number of samples with variants examined: 76718\n"
     ]
    }
   ],
   "source": [
    "\n",
    "folder_path = '/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/ukb_hotspot_calls_annotated'  # Path to ukb files from Hamish \n",
    "\n",
    "# Get all files in the folder \n",
    "files = glob.glob(os.path.join(folder_path, '*.txt'))  \n",
    "# print('Files examined are:', files)\n",
    "\n",
    "# Initialize an empty dictionary \n",
    "dataframes = {}\n",
    "\n",
    "# Iterate through the CSV files and read each one with pandas\n",
    "for csv_file in files:\n",
    "    \n",
    "    df = pd.read_csv(csv_file, sep = '\\t')\n",
    "    dataframes[csv_file] = df\n",
    "\n",
    "# Extract dataframes \n",
    "for file_name, df in dataframes.items():\n",
    "    \n",
    "    variable_name = file_name.split('/')[8].split('.')[0] + '_data'  # Removing the file extension\n",
    "    print('Examined file:', variable_name)\n",
    "    \n",
    "    df['batch'] = variable_name # add column to indicate source \n",
    "    globals()[variable_name] = df  # assign dataframe to the variable \n",
    "\n",
    "\n",
    "# Concat all into one df\n",
    "dfs_to_concat = [v for k, v in globals().items() if k.endswith('_data') and isinstance(v, pd.DataFrame)]\n",
    "\n",
    "# Row bind all batch dataframes into a single one \n",
    "batch_all = pd.concat(dfs_to_concat, ignore_index=True)\n",
    "batch_all = batch_all.dropna(subset=['batch']) # remove rows which are read incorrectly (w/o batch number)\n",
    "\n",
    "print('Number of samples with variants examined:', batch_all.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples carrying more than a single read with the variant sequence: 2823\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>end_position</th>\n",
       "      <th>VAF</th>\n",
       "      <th>var_depth</th>\n",
       "      <th>depth</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Codons</th>\n",
       "      <th>batch</th>\n",
       "      <th>alt_variant</th>\n",
       "      <th>ref_variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>2100538</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7674894</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>2</td>\n",
       "      <td>55</td>\n",
       "      <td>R/*</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>*</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>2100679</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25234373</td>\n",
       "      <td>0.048780</td>\n",
       "      <td>4</td>\n",
       "      <td>82</td>\n",
       "      <td>R/H</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>cGc/cAc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>H</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2100706</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25234307</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>2</td>\n",
       "      <td>103</td>\n",
       "      <td>P/L</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>cCg/cTg</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>L</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>2100777</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240313</td>\n",
       "      <td>0.020833</td>\n",
       "      <td>2</td>\n",
       "      <td>96</td>\n",
       "      <td>R/*</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>*</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>2100788</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240420</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>8</td>\n",
       "      <td>61</td>\n",
       "      <td>Y/C</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>tAc/tGc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>C</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>2101006</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240420</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>2</td>\n",
       "      <td>72</td>\n",
       "      <td>Y/S</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>tAc/tCc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>2101361</td>\n",
       "      <td>chr17</td>\n",
       "      <td>76736877</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>2</td>\n",
       "      <td>78</td>\n",
       "      <td>P/H</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>cCc/cAc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>H</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>2101637</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240313</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>2</td>\n",
       "      <td>92</td>\n",
       "      <td>R/*</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>*</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>2101708</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7675088</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>2</td>\n",
       "      <td>59</td>\n",
       "      <td>R/H</td>\n",
       "      <td>TP53</td>\n",
       "      <td>cGc/cAc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>H</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>2102011</td>\n",
       "      <td>chr15</td>\n",
       "      <td>90088702</td>\n",
       "      <td>0.097561</td>\n",
       "      <td>4</td>\n",
       "      <td>41</td>\n",
       "      <td>R/Q</td>\n",
       "      <td>IDH2</td>\n",
       "      <td>cGg/cAg</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_ID chromosome  end_position       VAF  var_depth  depth  \\\n",
       "41     2100538      chr17       7674894  0.036364          2     55   \n",
       "56     2100679       chr2      25234373  0.048780          4     82   \n",
       "59     2100706       chr2      25234307  0.019417          2    103   \n",
       "64     2100777       chr2      25240313  0.020833          2     96   \n",
       "65     2100788       chr2      25240420  0.131148          8     61   \n",
       "83     2101006       chr2      25240420  0.027778          2     72   \n",
       "109    2101361      chr17      76736877  0.025641          2     78   \n",
       "126    2101637       chr2      25240313  0.021739          2     92   \n",
       "130    2101708      chr17       7675088  0.033898          2     59   \n",
       "154    2102011      chr15      90088702  0.097561          4     41   \n",
       "\n",
       "    Amino_acids  SYMBOL   Codons                    batch alt_variant  \\\n",
       "41          R/*    TP53  Cga/Tga  batch_21_annotated_data           *   \n",
       "56          R/H  DNMT3A  cGc/cAc  batch_21_annotated_data           H   \n",
       "59          P/L  DNMT3A  cCg/cTg  batch_21_annotated_data           L   \n",
       "64          R/*  DNMT3A  Cga/Tga  batch_21_annotated_data           *   \n",
       "65          Y/C  DNMT3A  tAc/tGc  batch_21_annotated_data           C   \n",
       "83          Y/S  DNMT3A  tAc/tCc  batch_21_annotated_data           S   \n",
       "109         P/H   SRSF2  cCc/cAc  batch_21_annotated_data           H   \n",
       "126         R/*  DNMT3A  Cga/Tga  batch_21_annotated_data           *   \n",
       "130         R/H    TP53  cGc/cAc  batch_21_annotated_data           H   \n",
       "154         R/Q    IDH2  cGg/cAg  batch_21_annotated_data           Q   \n",
       "\n",
       "    ref_variant  \n",
       "41            R  \n",
       "56            R  \n",
       "59            P  \n",
       "64            R  \n",
       "65            Y  \n",
       "83            Y  \n",
       "109           P  \n",
       "126           R  \n",
       "130           R  \n",
       "154           R  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Change numerical variables to integers\n",
    "\n",
    "batch_all['end_position'] = batch_all['end_position'].astype(int)\n",
    "batch_all['position'] = batch_all['position'].astype(int)\n",
    "batch_all['sample_ID'] = batch_all['sample_ID'].astype(int)\n",
    "\n",
    "# Subset and create new useful columns\n",
    "\n",
    "batch_all = batch_all[['sample_ID', 'chromosome', 'end_position', 'VAF', 'var_depth', 'depth', 'Amino_acids', 'SYMBOL', 'Codons', 'batch']]\n",
    "batch_all['alt_variant'] = batch_all['Amino_acids'].str.split('/', expand = True)[1] # alternative (CH) variant\n",
    "batch_all['ref_variant'] = batch_all['Amino_acids'].str.split('/', expand = True)[0] # reference variant \n",
    "\n",
    "# There are some cases where there is no change in amino acids, for now save as NaN \n",
    "\n",
    "batch_all['alt_variant'].fillna(batch_all['Amino_acids'], inplace=True)\n",
    "batch_all['ref_variant'].fillna(batch_all['Amino_acids'], inplace=True)\n",
    "\n",
    "# Exclude data with singletons (likely errors)\n",
    "\n",
    "batch_all_ns = batch_all[batch_all['var_depth'] >= 2]\n",
    "\n",
    "# Number of samples with two variant reads or more \n",
    "print('Number of samples carrying more than a single read with the variant sequence:', batch_all_ns.shape[0])\n",
    "# okay so we have 2823 of these but then note that only 2249 are with CH variants \n",
    "# so like 20% of >= 2 reads are not v likely to be real? should conslut with Jamie \n",
    "\n",
    "batch_all_ns.head(n = 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Import indexes tested in each batch \n",
    "folder_path = '/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/ukb_hotspot_calls/batch_ids'  # Path to ukb files from Hamish \n",
    "\n",
    "# Get all files in the folder \n",
    "files_ids = glob.glob(os.path.join(folder_path, '*.tsv'))  \n",
    "\n",
    "# Read each file one by one \n",
    "\n",
    "indexes = {}\n",
    "\n",
    "# Iterate through the CSV files and read each one with pandas\n",
    "for file in files_ids:\n",
    "    \n",
    "    id = pd.read_csv(file, sep = '\\t')\n",
    "    id = id.rename(columns={'batch ID': 'sample_ID'})\n",
    "    id['sample_ID'] = id['sample_ID'].str.split('_', n = 1).str[0]\n",
    "    indexes[file] = id\n",
    "\n",
    "for file_name, df in indexes.items():\n",
    "    \n",
    "    variable_name = file_name.split('/')[9].split('.')[0]    # Remove file extension\n",
    "    globals()[variable_name] = df  # Assign the DataFrame to a variable with the file name\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate samples with genetic variants carried"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples which have been correctly annotated: 2818\n",
      "Number of variants identified in batches analysed: 89\n",
      "Number of sites identified in batches analysed: 41\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Find out how many gene_variants were called:\n",
    "\n",
    "# don't show warnings \n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# identify variants called \n",
    "batch_all_ns['variant_coord'] = batch_all_ns['chromosome'].astype(str) + \"_\" + batch_all_ns['end_position'].astype(str) # specific position in the genome \n",
    "batch_all_ns['variant_coord'] = batch_all_ns['variant_coord'].astype('category')\n",
    "batch_all_ns['variant_coord_pos'] = batch_all_ns['variant_coord'].astype(str) +  \"_\" + batch_all_ns['SYMBOL'].astype(str) + \"_\" + batch_all_ns['ref_variant'].astype(str) + \"_\" + batch_all_ns['alt_variant'].astype(str) # change to a specific aa\n",
    "batch_all_ns['variant_coord_pos'] = batch_all_ns['variant_coord_pos'].astype('category')\n",
    "\n",
    "# remove samples that have not been annotated (you can tell from the coordinate what is likely but these could be different mutations)\n",
    "batch_all_ns = batch_all_ns.dropna(subset=['SYMBOL']) # remove column where gene is not known \n",
    "print('Number of samples which have been correctly annotated:', batch_all_ns.shape[0]) # but at each site, you are getting reads modified to sth else \n",
    "# ok so sth worked wrong with annotation in only 5 cases > that looks good \n",
    "\n",
    "# identify the number of variants in a specific position\n",
    "num_variants = pd.DataFrame(batch_all_ns['variant_coord_pos'].value_counts())\n",
    "num_variants = num_variants[num_variants['count']!=0]\n",
    "num_variants = num_variants.sort_values(by = 'count')\n",
    "num_variants['variant_coord_pos'] = num_variants.index\n",
    "num_variants = num_variants.reset_index(drop=True)\n",
    "num_variants = num_variants.sort_values(by = 'variant_coord_pos')\n",
    "print('Number of variants identified in batches analysed:', num_variants.shape[0]) # but at each site, you are getting reads modiifed to sth else \n",
    "\n",
    "# identify the number of positions we looked at\n",
    "num_sites = pd.DataFrame(batch_all_ns['variant_coord'].value_counts())\n",
    "num_sites = num_sites[num_sites['count']!=0]\n",
    "num_sites = num_sites.sort_values(by = 'count')\n",
    "num_sites['variant_coord'] = num_sites.index\n",
    "num_sites = num_sites.reset_index(drop=True)\n",
    "num_sites = num_sites.sort_values(by = 'variant_coord')\n",
    "print('Number of sites identified in batches analysed:', num_sites.shape[0]) # okay see so you only found 37 sites \n",
    "\n",
    "# we have more variants than sites because for each site, we look at any change from the reference sequence (any possible variant)\n",
    "# at the same time, because we have this data, we can check if we have more variants that \"random\" mutations\n",
    "\n",
    "# save coordinates to a df\n",
    "pd.DataFrame(num_variants.variant_coord_pos.unique().tolist()).to_csv('/Users/barbarawalkowiak/Desktop/variants_annotated.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants which have been investigated: 42\n",
      "Number of sites which have been investigated: 36\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Annotation (using df with variant names and genomic coordinates)\n",
    "\n",
    "# read in the df with coordinates\n",
    "coord_gene_var = pd.read_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/ch_variants_coordinates_tp53_added_nts.csv')\n",
    "\n",
    "# all coordinates identified in the batches  \n",
    "coord_out = num_sites['variant_coord'].tolist()\n",
    "\n",
    "# intersection (annotate)\n",
    "coord_gene_var['variant_coord'] = coord_gene_var['chromosome'] + \"_\" + coord_gene_var['end_position'].astype(str) # find cariant coordinates\n",
    "coord_gene_var['SYMBOL'] = coord_gene_var['gene_var'].str.split('_').str[0] # find gene analysed\n",
    "coord_gene_var['ref_variant'] = coord_gene_var['gene_var'].str.split('_').str[1].str[0] # reference sequence variant \n",
    "coord_gene_var['alt_variant'] = coord_gene_var['gene_var'].str.split('_').str[1].str[-1] # alternative sequence (CH / mutation) variant\n",
    "coord_gene_var['variant_coord_pos'] = coord_gene_var['variant_coord'].astype(str) + \"_\" + coord_gene_var['SYMBOL'].astype(str) + \"_\" + coord_gene_var['ref_variant'].astype(str) + \"_\" + coord_gene_var['alt_variant'].astype(str) # specific mutation \n",
    "\n",
    "print('Number of variants which have been investigated:', len(coord_gene_var.variant_coord_pos.unique()))\n",
    "print('Number of sites which have been investigated:',len(coord_gene_var.variant_coord.unique()))\n",
    "\n",
    "# NB we removed one variant bc the coordinates were incorrect\n",
    "\n",
    "# NOTE: \n",
    "# TP53_R249S was searched for but not found in any of the batches so far \n",
    "# there have been 5 samples identified (in DNMT3A, at 2 different sites) with variants that I did not originally search for \n",
    "# in addition, some of the variants were offset by one base (could have been a deletion / insertion)\n",
    "# this is why we have differences in the number of sites in the two datasets \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants identified with annotations: 32\n",
      "Number of samples with a mutation in a CH-relevant position (with >2 reads): 2818\n",
      "Number of patients with a mutation in a CH-relevant position (with >2 reads): 2761\n",
      "Number of samples with a CH hotspot mutation (with > 2 reads): 2249\n",
      "Number of patients with a CH hotspot mutation (with > 2 reads): 2214\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# identify CH variants typed in the dataset \n",
    "gene_vars = coord_gene_var[['variant_coord_pos', 'gene_var']]\n",
    "\n",
    "# subset batch df to only include variants which were successfully identified \n",
    "batch_gene_vars = pd.merge(batch_all_ns, gene_vars, on = 'variant_coord_pos', how = 'inner')\n",
    "batch_gene_vars['gene_var'] = batch_gene_vars['gene_var'].astype('category')\n",
    "batch_gene_vars['gene_var'].value_counts()\n",
    "batch_gene_vars.rename(columns = {'sample_ID' : 'Person_ID'}, inplace = True)\n",
    "\n",
    "gene_vars_count = pd.DataFrame(batch_gene_vars['gene_var'].value_counts())\n",
    "gene_vars_count['gene_var'] = gene_vars_count.index\n",
    "gene_vars_count = gene_vars_count.reset_index(drop=True)\n",
    "gene_vars_sorted = gene_vars_count.sort_values(by = 'count', ascending=False)\n",
    "\n",
    "print('Number of variants identified with annotations:', gene_vars_sorted.shape[0])\n",
    "# so we will have 30 variants to look at bc the rest was not mapped correctly \n",
    "\n",
    "print('Number of samples with a mutation in a CH-relevant position (with >2 reads):', batch_all_ns.shape[0])\n",
    "print('Number of patients with a mutation in a CH-relevant position (with >2 reads):', len(batch_all_ns.sample_ID.unique()))\n",
    "print('Number of samples with a CH hotspot mutation (with > 2 reads):', batch_gene_vars.shape[0])\n",
    "print('Number of patients with a CH hotspot mutation (with > 2 reads):', len(batch_gene_vars.Person_ID.unique()))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add age data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals for whom age data is available: 459385\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# load the age dataset \n",
    "age_data = pd.read_csv('~/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/2022-05-20_healthy_pheno.tsv', sep = '\\t')\n",
    "\n",
    "age_df = age_data[['ID_v0', 'Age.when.attended.assessment.centre_v0']]\n",
    "age_df.columns.values[0] = 'Person_ID'\n",
    "age_df.columns.values[1] = 'age'\n",
    "print('Number of individuals for whom age data is available:', age_df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sample with available CH variant: 2249\n",
      "Number of sample with available CH variant and age: 1986\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# add age data to the rest of the data \n",
    "batch_gene_vars.gene_var = batch_gene_vars.gene_var.astype(str)\n",
    "batch_gene_vars.gene_var = batch_gene_vars.gene_var.astype('category') # remove previous categories we will not be looking at\n",
    "batch_gene_age = pd.merge(batch_gene_vars, age_df, on = 'Person_ID') # add age to everyone who has a variant\n",
    "print(\"Number of sample with available CH variant:\", batch_gene_vars.shape[0]) # note we are using the filtered dataframe here \n",
    "print(\"Number of sample with available CH variant and age:\", batch_gene_age.shape[0]) # these are healthy / non-cancer cases (can think if we want to filter out the cancer ones, but tbf probably yes)\n",
    "# note: this is throwing away patients who have been diagnosed with cancer (this dataset is for healthy individuals only)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load MHC class II genotype data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with MHC II genotype imputation data 488377\n",
      "Number of samples with MHC II genotype confidently imputed 378317\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# add MHC II genotype data to CH cases\n",
    "\n",
    "# path to file\n",
    "file_hla = \"/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/UKBB/ukb_typed_hla_with_ids_07_threshold_processed.txt\"\n",
    "\n",
    "# get the header \n",
    "header = pd.read_csv(file_hla, sep='\\t', nrows=1, header=None).values.tolist()\n",
    "head = [item for sublist in header for item in sublist]\n",
    "\n",
    "# get the actual dataframe \n",
    "df = pd.read_csv(file_hla, skiprows = 1, sep = ' ', header = None)\n",
    "\n",
    "# add columns \n",
    "df.columns = head \n",
    "\n",
    "# subset data for HLA-I and HLA-II class alleles \n",
    "# we are only interested in HLA-I for the momnet \n",
    "df_hla1 = df.filter(regex='^(Person_|A_|B_|C_)') # 488377 cases \n",
    "df_hla2 = df.filter(regex='^(Person_|D)') # 488377 cases \n",
    "\n",
    "df_clean_hla2 = df_hla2[~df_hla2.isin([0.5]).any(axis=1)] # exclude people for whom we lack full genotype data\n",
    "df_clean_hla2 = df_clean_hla2[~df_clean_hla2.isin([1.5]).any(axis=1)] # exclude people for whom we lack full genotype data\n",
    "\n",
    "# we will analyse DR and DP / DQ alleles separately because the \"rules\" are different (ie alpha / beta chain combinations etc.)\n",
    "# NOTE: I am removing samples for which we don't have correct allele annotation\n",
    "df_hla2_dp = df_clean_hla2.filter(regex='^(Person_|DP)')\n",
    "df_hla2_dq = df_clean_hla2.filter(regex='^(Person_|DQ)')\n",
    "df_hla2_dr = df_clean_hla2.filter(regex='^(Person_|DR)')\n",
    "\n",
    "print(\"Number of samples with MHC II genotype imputation data\", df_hla2.shape[0])\n",
    "print(\"Number of samples with MHC II genotype confidently imputed\", df_clean_hla2.shape[0]) # not ideal as this really throws out a lot of the data we have actually "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# okay, a problem with DP / DQ is that we have predictions for COMBINATIONS\n",
    "# therefore, we need to create columns with combinations and use them for when we make the prediction\n",
    "\n",
    "# what we can do is create additional columns and set the value to 1 if someone has both alleles (who knows if they will have the combination\n",
    "# but if predictions are available for it in NetMHCII we can assume that maybe it's legit)\n",
    "\n",
    "from itertools import combinations\n",
    "\n",
    "# combinations for DQ\n",
    "cc = list(combinations(df_hla2_dq.columns[1:],2)) # possible combinations of these columns (first is Person_ID)\n",
    "df_hla2_dq_comb = pd.concat([df_hla2_dq[c[0]].multiply(df_hla2_dq[c[1]]) for c in cc], axis=1, keys=cc) \n",
    "# we are multiplying\n",
    "# in this way, if someone has >= 1 in both alleles they will get a positive number\n",
    "# even if they have 2 alleles in one allele but 0 in the other, they will get a zero\n",
    "\n",
    "df_hla2_dq_comb.columns = df_hla2_dq_comb.columns.map('-'.join) # join with a '-'\n",
    "\n",
    "# repeat the same for DP \n",
    "cc = list(combinations(df_hla2_dp.columns[1:],2)) # possible combinations of these columns (first is Person_ID)\n",
    "df_hla2_dp_comb = pd.concat([df_hla2_dp[c[0]].multiply(df_hla2_dp[c[1]]) for c in cc], axis=1, keys=cc) \n",
    "# we are multiplying\n",
    "# in this way, if someone has >= 1 in both alleles they will get a positive number\n",
    "# even if they have 2 alleles in one allele but 0 in the other, they will get a zero\n",
    "\n",
    "df_hla2_dp_comb.columns = df_hla2_dp_comb.columns.map('-'.join) # join with a '-'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# okay, now: we do not want to predict stuff for 2 alpha or 2 beta chains because these are not created so we can remove these columns\n",
    "# we also want to rename columns from XXB-XXA to XXA-XXB\n",
    "def is_valid_column_name(col):\n",
    "    parts = col.split('-') # two part of the name of the column\n",
    "    return 'A' in parts[0] and 'B' in parts[1] or 'B' in parts[0] and 'A' in parts[1]\n",
    "\n",
    "def switch_parts(col):\n",
    "    parts = col.split('-') \n",
    "    new_name = parts[1] + '-' + parts[0]\n",
    "    return new_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# remove columns with BB or AA combinations \n",
    "filtered_columns_dp = [col for col in df_hla2_dp_comb.columns if is_valid_column_name(col)]\n",
    "filtered_columns_dq = [col for col in df_hla2_dq_comb.columns if is_valid_column_name(col)]\n",
    "\n",
    "# Create a new DataFrame with the filtered and reordered columns\n",
    "new_df_dp = df_hla2_dp_comb[filtered_columns_dp]\n",
    "new_df_dq = df_hla2_dq_comb[filtered_columns_dq]\n",
    "\n",
    "renamed_columns_dp = [switch_parts(col) for col in new_df_dp.columns]\n",
    "renamed_columns_dq = [switch_parts(col) for col in new_df_dq.columns]\n",
    "\n",
    "new_df_dp.columns = renamed_columns_dp\n",
    "new_df_dq.columns = renamed_columns_dq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>DQB1_201</th>\n",
       "      <th>DQB1_202</th>\n",
       "      <th>DQB1_301</th>\n",
       "      <th>DQB1_302</th>\n",
       "      <th>DQB1_303</th>\n",
       "      <th>DQB1_304</th>\n",
       "      <th>DQB1_401</th>\n",
       "      <th>DQB1_402</th>\n",
       "      <th>DQB1_501</th>\n",
       "      <th>...</th>\n",
       "      <th>DQA1_102-DQB1_609</th>\n",
       "      <th>DQA1_103-DQB1_609</th>\n",
       "      <th>DQA1_201-DQB1_609</th>\n",
       "      <th>DQA1_301-DQB1_609</th>\n",
       "      <th>DQA1_302-DQB1_609</th>\n",
       "      <th>DQA1_303-DQB1_609</th>\n",
       "      <th>DQA1_401-DQB1_609</th>\n",
       "      <th>DQA1_501-DQB1_609</th>\n",
       "      <th>DQA1_505-DQB1_609</th>\n",
       "      <th>DQA1_601-DQB1_609</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4860169</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3381323</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2805252</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3318036</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4120291</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488370</th>\n",
       "      <td>4478244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488373</th>\n",
       "      <td>3573995</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488374</th>\n",
       "      <td>1206874</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488375</th>\n",
       "      <td>3025735</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488376</th>\n",
       "      <td>1381182</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>378317 rows × 236 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Person_ID  DQB1_201  DQB1_202  DQB1_301  DQB1_302  DQB1_303  DQB1_304  \\\n",
       "1         4860169       1.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2         3381323       0.0       0.0       0.0       1.0       0.0       0.0   \n",
       "3         2805252       0.0       0.0       1.0       0.0       0.0       0.0   \n",
       "5         3318036       0.0       1.0       0.0       0.0       0.0       0.0   \n",
       "6         4120291       0.0       1.0       0.0       0.0       0.0       0.0   \n",
       "...           ...       ...       ...       ...       ...       ...       ...   \n",
       "488370    4478244       0.0       0.0       1.0       0.0       0.0       0.0   \n",
       "488373    3573995       1.0       0.0       0.0       0.0       0.0       0.0   \n",
       "488374    1206874       0.0       0.0       1.0       0.0       0.0       0.0   \n",
       "488375    3025735       0.0       0.0       1.0       0.0       0.0       0.0   \n",
       "488376    1381182       1.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "        DQB1_401  DQB1_402  DQB1_501  ...  DQA1_102-DQB1_609  \\\n",
       "1            0.0       0.0       0.0  ...                0.0   \n",
       "2            0.0       0.0       1.0  ...                0.0   \n",
       "3            0.0       0.0       0.0  ...                0.0   \n",
       "5            0.0       0.0       0.0  ...                0.0   \n",
       "6            0.0       0.0       1.0  ...                0.0   \n",
       "...          ...       ...       ...  ...                ...   \n",
       "488370       0.0       0.0       0.0  ...                0.0   \n",
       "488373       0.0       0.0       0.0  ...                0.0   \n",
       "488374       0.0       0.0       0.0  ...                0.0   \n",
       "488375       0.0       1.0       0.0  ...                0.0   \n",
       "488376       0.0       0.0       0.0  ...                0.0   \n",
       "\n",
       "        DQA1_103-DQB1_609  DQA1_201-DQB1_609  DQA1_301-DQB1_609  \\\n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "5                     0.0                0.0                0.0   \n",
       "6                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "488370                0.0                0.0                0.0   \n",
       "488373                0.0                0.0                0.0   \n",
       "488374                0.0                0.0                0.0   \n",
       "488375                0.0                0.0                0.0   \n",
       "488376                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_302-DQB1_609  DQA1_303-DQB1_609  DQA1_401-DQB1_609  \\\n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "5                     0.0                0.0                0.0   \n",
       "6                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "488370                0.0                0.0                0.0   \n",
       "488373                0.0                0.0                0.0   \n",
       "488374                0.0                0.0                0.0   \n",
       "488375                0.0                0.0                0.0   \n",
       "488376                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_501-DQB1_609  DQA1_505-DQB1_609  DQA1_601-DQB1_609  \n",
       "1                     0.0                0.0                0.0  \n",
       "2                     0.0                0.0                0.0  \n",
       "3                     0.0                0.0                0.0  \n",
       "5                     0.0                0.0                0.0  \n",
       "6                     0.0                0.0                0.0  \n",
       "...                   ...                ...                ...  \n",
       "488370                0.0                0.0                0.0  \n",
       "488373                0.0                0.0                0.0  \n",
       "488374                0.0                0.0                0.0  \n",
       "488375                0.0                0.0                0.0  \n",
       "488376                0.0                0.0                0.0  \n",
       "\n",
       "[378317 rows x 236 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# once we have selected the correct columns, combine with other data\n",
    "\n",
    "# DP\n",
    "df_hla2_dp_all = pd.concat([df_hla2_dp, new_df_dp], axis = 1)\n",
    "df_hla2_dp_all.replace(2, 1, inplace=True) # if someone got a 2 (2 * 1), replace to 1\n",
    "df_hla2_dp_all.replace(4, 1, inplace=True) # if someone got a 2 (2 * 2), replace to 1\n",
    "\n",
    "# let's remove columns where noone has the combination (it's possible that some combinations are never seen in our dataset)\n",
    "colsums = pd.DataFrame(df_hla2_dp_all.sum()).reset_index()\n",
    "colsums.rename(columns={'index': 'col_name', 0:'col_sum'}, inplace=True)\n",
    "col_to_retain = colsums[colsums['col_sum']>=1]['col_name'] # these are the columns to retain \n",
    "df_hla2_dp_all = df_hla2_dp_all[col_to_retain]\n",
    "df_hla2_dp_all \n",
    "\n",
    "# DQ\n",
    "df_hla2_dq_all = pd.concat([df_hla2_dq, new_df_dq], axis = 1)\n",
    "df_hla2_dq_all.replace(2, 1, inplace=True) # if someone got a 2 (2 * 1), replace to 1\n",
    "df_hla2_dq_all.replace(4, 1, inplace=True) # if someone got a 2 (2 * 1), replace to 1\n",
    "\n",
    "# let's remove columns where noone has the combination (it's possible that some combinations are never seen in our dataset)\n",
    "colsums = pd.DataFrame(df_hla2_dq_all.sum()).reset_index()\n",
    "colsums.rename(columns={'index': 'col_name', 0:'col_sum'}, inplace=True)\n",
    "col_to_retain = colsums[colsums['col_sum']>=1]['col_name'] # these are the columns to retain \n",
    "df_hla2_dq_all = df_hla2_dq_all[col_to_retain]\n",
    "df_hla2_dq_all "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load MHC binding affinity prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Pos</th>\n",
       "      <th>HLA</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>Of</th>\n",
       "      <th>Core</th>\n",
       "      <th>Core_Rel</th>\n",
       "      <th>Inverted</th>\n",
       "      <th>Identity</th>\n",
       "      <th>Score_EL</th>\n",
       "      <th>%Rank_EL</th>\n",
       "      <th>Exp_Bind</th>\n",
       "      <th>BindLevel</th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>genotype</th>\n",
       "      <th>gene_var</th>\n",
       "      <th>Score_BA</th>\n",
       "      <th>%Rank_BA</th>\n",
       "      <th>gene_var_gt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>81173</td>\n",
       "      <td>1</td>\n",
       "      <td>DRB1_0101</td>\n",
       "      <td>IRHLFALLKEYFACV</td>\n",
       "      <td>4</td>\n",
       "      <td>FALLKEYFA</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_9</td>\n",
       "      <td>0.035335</td>\n",
       "      <td>14.93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>81178</td>\n",
       "      <td>1</td>\n",
       "      <td>DRB1_0102</td>\n",
       "      <td>SVPVIRHLFALLKEY</td>\n",
       "      <td>4</td>\n",
       "      <td>IRHLFALLK</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_5</td>\n",
       "      <td>0.075223</td>\n",
       "      <td>15.46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>81191</td>\n",
       "      <td>1</td>\n",
       "      <td>DRB1_0103</td>\n",
       "      <td>IRHLFALLKEYFACV</td>\n",
       "      <td>4</td>\n",
       "      <td>FALLKEYFA</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_9</td>\n",
       "      <td>0.015829</td>\n",
       "      <td>21.55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>81196</td>\n",
       "      <td>1</td>\n",
       "      <td>DRB1_0301</td>\n",
       "      <td>SVPVIRHLFALLKEY</td>\n",
       "      <td>4</td>\n",
       "      <td>IRHLFALLK</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_5</td>\n",
       "      <td>0.003383</td>\n",
       "      <td>38.95</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>81206</td>\n",
       "      <td>1</td>\n",
       "      <td>DRB1_0302</td>\n",
       "      <td>VPVIRHLFALLKEYF</td>\n",
       "      <td>3</td>\n",
       "      <td>IRHLFALLK</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_6</td>\n",
       "      <td>0.003467</td>\n",
       "      <td>51.72</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  Pos        HLA          Peptide  Of       Core  Core_Rel  \\\n",
       "0       81173    1  DRB1_0101  IRHLFALLKEYFACV   4  FALLKEYFA      0.99   \n",
       "1       81178    1  DRB1_0102  SVPVIRHLFALLKEY   4  IRHLFALLK      0.96   \n",
       "2       81191    1  DRB1_0103  IRHLFALLKEYFACV   4  FALLKEYFA      0.99   \n",
       "3       81196    1  DRB1_0301  SVPVIRHLFALLKEY   4  IRHLFALLK      0.92   \n",
       "4       81206    1  DRB1_0302  VPVIRHLFALLKEYF   3  IRHLFALLK      0.89   \n",
       "\n",
       "   Inverted Identity  Score_EL  %Rank_EL  Exp_Bind BindLevel    gene variant  \\\n",
       "0         0    seq_9  0.035335     14.93       NaN       NaN  DNMT3A   P904L   \n",
       "1         0    seq_5  0.075223     15.46       NaN       NaN  DNMT3A   P904L   \n",
       "2         0    seq_9  0.015829     21.55       NaN       NaN  DNMT3A   P904L   \n",
       "3         0    seq_5  0.003383     38.95       NaN       NaN  DNMT3A   P904L   \n",
       "4         0    seq_6  0.003467     51.72       NaN       NaN  DNMT3A   P904L   \n",
       "\n",
       "  genotype      gene_var  Score_BA  %Rank_BA      gene_var_gt  \n",
       "0       ch  DNMT3A_P904L       NaN       NaN  DNMT3A_P904L_ch  \n",
       "1       ch  DNMT3A_P904L       NaN       NaN  DNMT3A_P904L_ch  \n",
       "2       ch  DNMT3A_P904L       NaN       NaN  DNMT3A_P904L_ch  \n",
       "3       ch  DNMT3A_P904L       NaN       NaN  DNMT3A_P904L_ch  \n",
       "4       ch  DNMT3A_P904L       NaN       NaN  DNMT3A_P904L_ch  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# parameters\n",
    "param = '%Rank_EL' # let's do it on EL ranks first \n",
    "\n",
    "# NOTE: In the original file (sub), I was missing a score for DNMT3A_R736C_ch so I generated it separately with the same code and will concatenate the files here \n",
    "# load the file with binding predictions across variants\n",
    "pred_file_dpq_sub = \"/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/NetMHCII_out/scores/20240211_NetMHC_HLA_UKBB_with_affinities_DP_DQ_bestscores.csv\"\n",
    "pred_file_dpq_r736c = \"/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/NetMHCII_out/scores/20240215_NetMHC_HLA_UKBB_with_affinities_DP_DQ_R736C_bestscores.csv\"\n",
    "pred_file_dpq_stop = \"/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/NetMHCII_out/scores/20240214_NetMHC_HLA_UKBB_with_affinities_DP_DQ_stopcodons_bestscores.csv\"\n",
    "pred_method = pred_file_dpq_sub.split('_out')[0] # all are with the same method so does not matter which file you end up using \n",
    " \n",
    "# organize file with prediction scores for DP and DQ alleles \n",
    "pred_filename_dpq_sub = pred_file_dpq_sub.split('/')[2].split('.')[0]\n",
    "pred_filename_dpq_stop = pred_file_dpq_stop.split('/')[2].split('.')[0]\n",
    "pred_df_dpq_sub = pd.read_csv(pred_file_dpq_sub)\n",
    "pred_df_dpq_r736c = pd.read_csv(pred_file_dpq_r736c)\n",
    "pred_df_dpq_stop = pd.read_csv(pred_file_dpq_stop)\n",
    "\n",
    "# In the sub file, I had some predictions for STOP-codon-containing variants, I don't want these so remove them\n",
    "pred_df_dpq_sub = pred_df_dpq_sub[~pred_df_dpq_sub['gene_var'].str.contains('\\*')]\n",
    "\n",
    "# concat the files \n",
    "pred_df_dpq_stop.drop(columns=['Affinity (nM)'], inplace=True)\n",
    "pred_df_dpq_r736c.drop(columns=['Affinity (nM)'], inplace=True)\n",
    "pred_df_dpq = pd.concat([pred_df_dpq_sub, pred_df_dpq_stop, pred_df_dpq_r736c], ignore_index = True)\n",
    "pred_df_dpq['gene_var_gt'] = pred_df_dpq['gene'] + '_' + pred_df_dpq['variant'] + '_' + pred_df_dpq['genotype'] # create a column that includes all genotype data\n",
    "\n",
    "# organize file with prediction scores for DR alleles \n",
    "pred_file_dr_sub = \"/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/NetMHCII_out/scores/20240214_NetMHC_HLA_UKBB_with_affinities_DR_bestscores.csv\"\n",
    "pred_file_dr_r736c = \"/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/NetMHCII_out/scores/20240215_NetMHC_HLA_UKBB_with_affinities_DR_R736C_bestscores.csv\"\n",
    "pred_file_dr_stop = \"/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/NetMHCII_out/scores/20240214_NetMHC_HLA_UKBB_with_affinities_DR_stopcodons_bestscores.csv\"\n",
    "\n",
    "# organize file with prediction scores for DR alleles \n",
    "pred_filename_dr_sub = pred_file_dr_sub.split('/')[2].split('.')[0]\n",
    "pred_filename_dr_stop = pred_file_dr_stop.split('/')[2].split('.')[0]\n",
    "pred_df_dr_sub = pd.read_csv(pred_file_dr_sub)\n",
    "pred_df_dr_r736c = pd.read_csv(pred_file_dr_r736c)\n",
    "pred_df_dr_stop = pd.read_csv(pred_file_dr_stop)\n",
    "\n",
    "# In the sub file, I had some predictions for STOP-codon-containing variants, I don't want these so remove them\n",
    "pred_df_dr_sub = pred_df_dr_sub[~pred_df_dr_sub['gene_var'].str.contains('\\*')]\n",
    "\n",
    "# remove affinity columns\n",
    "pred_df_dr_stop.drop(columns=['Affinity (nM)'], inplace=True)\n",
    "pred_df_dr_r736c.drop(columns=['Affinity(nM)'], inplace=True)\n",
    "\n",
    "# concat the files \n",
    "pred_df_dr = pd.concat([pred_df_dr_sub, pred_df_dr_stop, pred_df_dr_r736c], ignore_index = True)\n",
    "pred_df_dr['gene_var_gt'] = pred_df_dr['gene'] + '_' + pred_df_dr['variant'] + '_' + pred_df_dr['genotype'] # create a column that includes all genotype data\n",
    "pred_df_dr.head(n = 5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# we need to move the format of the DP/DQ predictions to something that matches our UKBB genotyping and everything else \n",
    "def transform_format_DPQ(input_string):\n",
    "    # Define a regular expression pattern to match the input format\n",
    "    pattern = re.compile(r'HLA-(\\w{3})(\\d{1})(\\d{4})-(\\w{3})(\\d{1})(\\d{4})') # okay so this is the pattern we are trying to match\n",
    "\n",
    "    # check if there is a match\n",
    "    match = pattern.match(input_string)\n",
    "\n",
    "    # if match, apply transformation\n",
    "    if match:\n",
    "        group1 = match.group(1) # we are not including HLA annotations, this is the name of the first allele   \n",
    "        group2 = int(match.group(2)) # there will be no zeroes, leave as it is \n",
    "        group3 = int(match.group(3)) # remove zeros at the start \n",
    "        group4 = match.group(4) # name of the second allele in the combination\n",
    "        group5 = int(match.group(5)) # there will be no zeroes, leave as it is \n",
    "        group6 = int(match.group(6)) # remove zeros at the start \n",
    "\n",
    "        # Format the output string\n",
    "        output_string = f'{group1}{group2}_{group3}-{group4}{group5}_{group6}' # stitch back \n",
    "\n",
    "        return output_string # return transformed string \n",
    "\n",
    "    # if no much, return original string \n",
    "    return 0\n",
    "\n",
    "def transform_format_DR(input_string):\n",
    "    # Define a regular expression pattern to match the input format\n",
    "    pattern = re.compile(r'(\\w{3})(\\d{1})_(\\d+)') # okay so this is the pattern we are trying to match\n",
    "\n",
    "    # check if there is a match\n",
    "    match = pattern.match(input_string)\n",
    "\n",
    "    # if match, apply transformation\n",
    "    if match:\n",
    "        group1 = match.group(1) # we are not including HLA annotations, this is the name of the first allele   \n",
    "        group2 = int(match.group(2)) # there will be no zeroes, leave as it is \n",
    "        group3 = int(match.group(3)) # remove zeros at the start \n",
    "\n",
    "        # Format the output string\n",
    "        output_string = f'{group1}{group2}_{group3}' # stitch back \n",
    "\n",
    "        return output_string # return transformed string \n",
    "\n",
    "    # if no much, return original string \n",
    "    return 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Pos</th>\n",
       "      <th>HLA</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>Of</th>\n",
       "      <th>Core</th>\n",
       "      <th>Core_Rel</th>\n",
       "      <th>Inverted</th>\n",
       "      <th>Identity</th>\n",
       "      <th>Score_EL</th>\n",
       "      <th>...</th>\n",
       "      <th>Score_BA</th>\n",
       "      <th>%Rank_BA</th>\n",
       "      <th>Affinity (nM)</th>\n",
       "      <th>BindLevel</th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>genotype</th>\n",
       "      <th>gene_var</th>\n",
       "      <th>gene_var_gt</th>\n",
       "      <th>HLA_formatted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>181708</td>\n",
       "      <td>1</td>\n",
       "      <td>HLA-DPA10103-DPB10101</td>\n",
       "      <td>PVIRHLFALLKEYFA</td>\n",
       "      <td>3</td>\n",
       "      <td>RHLFALLKE</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_7</td>\n",
       "      <td>0.112253</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "      <td>DPA1_103-DPB1_101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>727345</td>\n",
       "      <td>1</td>\n",
       "      <td>HLA-DPA10103-DPB10201</td>\n",
       "      <td>SVPVIRHLFALLKEY</td>\n",
       "      <td>3</td>\n",
       "      <td>VIRHLFALL</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_5</td>\n",
       "      <td>0.219174</td>\n",
       "      <td>...</td>\n",
       "      <td>&lt;=</td>\n",
       "      <td>WB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "      <td>DPA1_103-DPB1_201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>249261</td>\n",
       "      <td>1</td>\n",
       "      <td>HLA-DPA10103-DPB10202</td>\n",
       "      <td>WSVPVIRHLFALLKE</td>\n",
       "      <td>4</td>\n",
       "      <td>VIRHLFALL</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_4</td>\n",
       "      <td>0.248802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.565826</td>\n",
       "      <td>4.66</td>\n",
       "      <td>109.69</td>\n",
       "      <td>&lt;=</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "      <td>DPA1_103-DPB1_202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>655999</td>\n",
       "      <td>1</td>\n",
       "      <td>HLA-DPA10103-DPB10301</td>\n",
       "      <td>GRSWSVPVIRHLFAL</td>\n",
       "      <td>1</td>\n",
       "      <td>RSWSVPVIR</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_1</td>\n",
       "      <td>0.012210</td>\n",
       "      <td>...</td>\n",
       "      <td>0.558070</td>\n",
       "      <td>2.21</td>\n",
       "      <td>119.29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "      <td>DPA1_103-DPB1_301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>248910</td>\n",
       "      <td>1</td>\n",
       "      <td>HLA-DPA10103-DPB10401</td>\n",
       "      <td>WSVPVIRHLFALLKE</td>\n",
       "      <td>4</td>\n",
       "      <td>VIRHLFALL</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0</td>\n",
       "      <td>seq_4</td>\n",
       "      <td>0.053899</td>\n",
       "      <td>...</td>\n",
       "      <td>0.490329</td>\n",
       "      <td>6.20</td>\n",
       "      <td>248.27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P904L</td>\n",
       "      <td>ch</td>\n",
       "      <td>DNMT3A_P904L</td>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "      <td>DPA1_103-DPB1_401</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  Pos                    HLA          Peptide  Of       Core  \\\n",
       "0      181708    1  HLA-DPA10103-DPB10101  PVIRHLFALLKEYFA   3  RHLFALLKE   \n",
       "1      727345    1  HLA-DPA10103-DPB10201  SVPVIRHLFALLKEY   3  VIRHLFALL   \n",
       "2      249261    1  HLA-DPA10103-DPB10202  WSVPVIRHLFALLKE   4  VIRHLFALL   \n",
       "3      655999    1  HLA-DPA10103-DPB10301  GRSWSVPVIRHLFAL   1  RSWSVPVIR   \n",
       "4      248910    1  HLA-DPA10103-DPB10401  WSVPVIRHLFALLKE   4  VIRHLFALL   \n",
       "\n",
       "   Core_Rel  Inverted Identity  Score_EL  ...  Score_BA  %Rank_BA  \\\n",
       "0      0.87         0    seq_7  0.112253  ...       NaN       NaN   \n",
       "1      1.00         0    seq_5  0.219174  ...        <=        WB   \n",
       "2      0.98         0    seq_4  0.248802  ...  0.565826      4.66   \n",
       "3      0.64         0    seq_1  0.012210  ...  0.558070      2.21   \n",
       "4      0.98         0    seq_4  0.053899  ...  0.490329      6.20   \n",
       "\n",
       "  Affinity (nM) BindLevel    gene variant genotype      gene_var  \\\n",
       "0           NaN       NaN  DNMT3A   P904L       ch  DNMT3A_P904L   \n",
       "1           NaN       NaN  DNMT3A   P904L       ch  DNMT3A_P904L   \n",
       "2        109.69        <=  DNMT3A   P904L       ch  DNMT3A_P904L   \n",
       "3        119.29       NaN  DNMT3A   P904L       ch  DNMT3A_P904L   \n",
       "4        248.27       NaN  DNMT3A   P904L       ch  DNMT3A_P904L   \n",
       "\n",
       "       gene_var_gt      HLA_formatted  \n",
       "0  DNMT3A_P904L_ch  DPA1_103-DPB1_101  \n",
       "1  DNMT3A_P904L_ch  DPA1_103-DPB1_201  \n",
       "2  DNMT3A_P904L_ch  DPA1_103-DPB1_202  \n",
       "3  DNMT3A_P904L_ch  DPA1_103-DPB1_301  \n",
       "4  DNMT3A_P904L_ch  DPA1_103-DPB1_401  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change format of HLA allele naming \n",
    "pred_df_dpq['HLA_formatted'] = pred_df_dpq['HLA'].map(transform_format_DPQ)\n",
    "pred_df_dr['HLA_formatted'] = pred_df_dr['HLA'].map(transform_format_DR)\n",
    "\n",
    "# combine all predictions\n",
    "pred_df_all = pd.concat([pred_df_dpq, pred_df_dr], axis = 0)\n",
    "pred_df_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with available CH variant: 2249\n",
      "Number of samples with annotated variants CH variant and HLA genotype (DR): 1746\n",
      "Number of samples with annotated variants CH variant and HLA genotype (DP): 1746\n",
      "Number of samples with annotated variants CH variant and HLA genotype (DQ): 1746\n",
      "Number of samples with available CH variant and age: 1986\n",
      "Number of samples with annotated variants CH variant and age and HLA genotype (DR): 1539\n",
      "Number of samples with annotated variants CH variant and age and HLA genotype (DP): 1539\n",
      "Number of samples with annotated variants CH variant and age and HLA genotype (DQ): 1539\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# now you need to merge the data that you want to use with the allele data to get predictions\n",
    "# most comprehensive dataset we have is batch_gene_age (includes genetic variant someone carries and age)\n",
    "# Find IDs of CH-affected persons and match to their HLA genotype \n",
    "\n",
    "batch_gene_vars.rename(columns={'sample_ID': 'Person_ID'}, inplace=True)\n",
    "batch_gene_age.rename(columns={'sample_ID': 'Person_ID'}, inplace=True)\n",
    "ids_batch = batch_gene_age['Person_ID']\n",
    "hla_batch_ids_dr = df_hla2_dr[df_hla2_dr['Person_ID'].isin(ids_batch)] \n",
    "\n",
    "# add HLA genotype data to samples with annotated CH variant and age\n",
    "batch_gene_hla_dr = pd.merge(batch_gene_vars, df_hla2_dr, on='Person_ID')\n",
    "batch_gene_hla_dp = pd.merge(batch_gene_vars, df_hla2_dp_all, on='Person_ID')\n",
    "batch_gene_hla_dq = pd.merge(batch_gene_vars, df_hla2_dq_all, on='Person_ID')\n",
    "\n",
    "batch_gene_age_hla_dr = pd.merge(batch_gene_age, df_hla2_dr, on='Person_ID')\n",
    "batch_gene_age_hla_dp = pd.merge(batch_gene_age, df_hla2_dp_all, on='Person_ID')\n",
    "batch_gene_age_hla_dq = pd.merge(batch_gene_age, df_hla2_dq_all, on='Person_ID')\n",
    "\n",
    "print(\"Number of samples with available CH variant:\", batch_gene_vars.shape[0])\n",
    "print('Number of samples with annotated variants CH variant and HLA genotype (DR):', batch_gene_hla_dr.shape[0]) # 1 HLA genotype missing \n",
    "print('Number of samples with annotated variants CH variant and HLA genotype (DP):', batch_gene_hla_dp.shape[0]) # 1 HLA genotype missing \n",
    "print('Number of samples with annotated variants CH variant and HLA genotype (DQ):', batch_gene_hla_dq.shape[0]) # 1 HLA genotype missing \n",
    "print(\"Number of samples with available CH variant and age:\", batch_gene_age.shape[0])\n",
    "print('Number of samples with annotated variants CH variant and age and HLA genotype (DR):', batch_gene_age_hla_dr.shape[0]) # 1 HLA genotype missing \n",
    "print('Number of samples with annotated variants CH variant and age and HLA genotype (DP):', batch_gene_age_hla_dp.shape[0]) # 1 HLA genotype missing \n",
    "print('Number of samples with annotated variants CH variant and age and HLA genotype (DQ):', batch_gene_age_hla_dq.shape[0]) # 1 HLA genotype missing \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dataframe with all relevant genotype (DRB1 + DQ combinations + DP combinations)\n",
    "batch_gene_hla_dp_comb = pd.concat([batch_gene_hla_dp.iloc[:, 0], batch_gene_hla_dp.iloc[:, 14], batch_gene_hla_dp.iloc[:, 56:]], axis = 1)\n",
    "batch_gene_hla_dq_comb = pd.concat([batch_gene_hla_dq.iloc[:, 0], batch_gene_hla_dq.iloc[:, 14], batch_gene_hla_dq.iloc[:, 47:]], axis = 1)\n",
    "batch_gene_hla_all = pd.merge(batch_gene_hla_dr, batch_gene_hla_dp_comb, on = ['Person_ID', 'gene_var'])                             \n",
    "batch_gene_hla_all = pd.merge(batch_gene_hla_all, batch_gene_hla_dq_comb, on = ['Person_ID', 'gene_var']) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>end_position</th>\n",
       "      <th>VAF</th>\n",
       "      <th>var_depth</th>\n",
       "      <th>depth</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Codons</th>\n",
       "      <th>batch</th>\n",
       "      <th>...</th>\n",
       "      <th>DQA1_103-DQB1_609</th>\n",
       "      <th>DQA1_201-DQB1_609</th>\n",
       "      <th>DQA1_301-DQB1_609</th>\n",
       "      <th>DQA1_302-DQB1_609</th>\n",
       "      <th>DQA1_303-DQB1_609</th>\n",
       "      <th>DQA1_401-DQB1_609</th>\n",
       "      <th>DQA1_501-DQB1_609</th>\n",
       "      <th>DQA1_505-DQB1_609</th>\n",
       "      <th>DQA1_601-DQB1_609</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2100538</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7674894</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>2</td>\n",
       "      <td>55</td>\n",
       "      <td>R/*</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1641209</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7674894</td>\n",
       "      <td>0.260000</td>\n",
       "      <td>13</td>\n",
       "      <td>50</td>\n",
       "      <td>R/*</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_16_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1471938</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7674894</td>\n",
       "      <td>0.032787</td>\n",
       "      <td>2</td>\n",
       "      <td>61</td>\n",
       "      <td>R/*</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_14_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1482881</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7674894</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>2</td>\n",
       "      <td>70</td>\n",
       "      <td>R/*</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_14_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1946555</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7674894</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>2</td>\n",
       "      <td>39</td>\n",
       "      <td>R/*</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_19_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1534</th>\n",
       "      <td>1619263</td>\n",
       "      <td>chr15</td>\n",
       "      <td>90088606</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "      <td>R/K</td>\n",
       "      <td>IDH2</td>\n",
       "      <td>aGg/aAg</td>\n",
       "      <td>batch_16_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>56.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1535</th>\n",
       "      <td>1456018</td>\n",
       "      <td>chr15</td>\n",
       "      <td>90088606</td>\n",
       "      <td>0.032258</td>\n",
       "      <td>2</td>\n",
       "      <td>62</td>\n",
       "      <td>R/K</td>\n",
       "      <td>IDH2</td>\n",
       "      <td>aGg/aAg</td>\n",
       "      <td>batch_14_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>64.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>2045629</td>\n",
       "      <td>chr15</td>\n",
       "      <td>90088606</td>\n",
       "      <td>0.031746</td>\n",
       "      <td>2</td>\n",
       "      <td>63</td>\n",
       "      <td>R/K</td>\n",
       "      <td>IDH2</td>\n",
       "      <td>aGg/aAg</td>\n",
       "      <td>batch_20_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>1740102</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114716126</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>2</td>\n",
       "      <td>33</td>\n",
       "      <td>G/D</td>\n",
       "      <td>NRAS</td>\n",
       "      <td>gGt/gAt</td>\n",
       "      <td>batch_17_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1538</th>\n",
       "      <td>1929818</td>\n",
       "      <td>chr4</td>\n",
       "      <td>54733155</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>D/V</td>\n",
       "      <td>KIT</td>\n",
       "      <td>gAc/gTc</td>\n",
       "      <td>batch_19_annotated_data</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1539 rows × 464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Person_ID chromosome  end_position       VAF  var_depth  depth  \\\n",
       "0       2100538      chr17       7674894  0.036364          2     55   \n",
       "1       1641209      chr17       7674894  0.260000         13     50   \n",
       "2       1471938      chr17       7674894  0.032787          2     61   \n",
       "3       1482881      chr17       7674894  0.028571          2     70   \n",
       "4       1946555      chr17       7674894  0.051282          2     39   \n",
       "...         ...        ...           ...       ...        ...    ...   \n",
       "1534    1619263      chr15      90088606  0.031250          2     64   \n",
       "1535    1456018      chr15      90088606  0.032258          2     62   \n",
       "1536    2045629      chr15      90088606  0.031746          2     63   \n",
       "1537    1740102       chr1     114716126  0.060606          2     33   \n",
       "1538    1929818       chr4      54733155  0.066667          2     30   \n",
       "\n",
       "     Amino_acids SYMBOL   Codons                    batch  ...  \\\n",
       "0            R/*   TP53  Cga/Tga  batch_21_annotated_data  ...   \n",
       "1            R/*   TP53  Cga/Tga  batch_16_annotated_data  ...   \n",
       "2            R/*   TP53  Cga/Tga  batch_14_annotated_data  ...   \n",
       "3            R/*   TP53  Cga/Tga  batch_14_annotated_data  ...   \n",
       "4            R/*   TP53  Cga/Tga  batch_19_annotated_data  ...   \n",
       "...          ...    ...      ...                      ...  ...   \n",
       "1534         R/K   IDH2  aGg/aAg  batch_16_annotated_data  ...   \n",
       "1535         R/K   IDH2  aGg/aAg  batch_14_annotated_data  ...   \n",
       "1536         R/K   IDH2  aGg/aAg  batch_20_annotated_data  ...   \n",
       "1537         G/D   NRAS  gGt/gAt  batch_17_annotated_data  ...   \n",
       "1538         D/V    KIT  gAc/gTc  batch_19_annotated_data  ...   \n",
       "\n",
       "     DQA1_103-DQB1_609 DQA1_201-DQB1_609 DQA1_301-DQB1_609 DQA1_302-DQB1_609  \\\n",
       "0                  0.0               0.0               0.0               0.0   \n",
       "1                  0.0               0.0               0.0               0.0   \n",
       "2                  0.0               0.0               0.0               0.0   \n",
       "3                  0.0               0.0               0.0               0.0   \n",
       "4                  0.0               0.0               0.0               0.0   \n",
       "...                ...               ...               ...               ...   \n",
       "1534               0.0               0.0               0.0               0.0   \n",
       "1535               0.0               0.0               0.0               0.0   \n",
       "1536               0.0               0.0               0.0               0.0   \n",
       "1537               0.0               0.0               0.0               0.0   \n",
       "1538               0.0               0.0               0.0               0.0   \n",
       "\n",
       "     DQA1_303-DQB1_609  DQA1_401-DQB1_609  DQA1_501-DQB1_609  \\\n",
       "0                  0.0                0.0                0.0   \n",
       "1                  0.0                0.0                0.0   \n",
       "2                  0.0                0.0                0.0   \n",
       "3                  0.0                0.0                0.0   \n",
       "4                  0.0                0.0                0.0   \n",
       "...                ...                ...                ...   \n",
       "1534               0.0                0.0                0.0   \n",
       "1535               0.0                0.0                0.0   \n",
       "1536               0.0                0.0                0.0   \n",
       "1537               0.0                0.0                1.0   \n",
       "1538               0.0                0.0                0.0   \n",
       "\n",
       "      DQA1_505-DQB1_609  DQA1_601-DQB1_609   age  \n",
       "0                   0.0                0.0  51.0  \n",
       "1                   0.0                0.0  60.0  \n",
       "2                   0.0                0.0  48.0  \n",
       "3                   0.0                0.0  69.0  \n",
       "4                   0.0                0.0  66.0  \n",
       "...                 ...                ...   ...  \n",
       "1534                0.0                0.0  56.0  \n",
       "1535                0.0                0.0  64.0  \n",
       "1536                0.0                0.0  61.0  \n",
       "1537                0.0                0.0  67.0  \n",
       "1538                0.0                0.0  66.0  \n",
       "\n",
       "[1539 rows x 464 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add age data\n",
    "batch_gene_age_hla_all = pd.merge(batch_gene_hla_all, age_df, on = 'Person_ID')\n",
    "batch_gene_age_hla_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants to examine (common variants): 20\n",
      "Number of samples with a CH hotspot mutation (with > 2 reads): 2249\n",
      "Number of patients with a CH hotspot mutation (with > 2 reads): 2214\n",
      "Number of samples with a CH hotspot mutation (with > 2 reads) (common variants): 1478\n",
      "Number of patients with a CH hotspot mutation (with > 2 reads) (common variants): 1455\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Filtering 2: we will only be looking at variants for which we have more than 10 samples \n",
    "gene_vars_count = pd.DataFrame(batch_gene_vars['gene_var'].value_counts())\n",
    "gene_vars_count['gene_var'] = gene_vars_count.index\n",
    "gene_vars_count = gene_vars_count.reset_index(drop=True)\n",
    "gene_vars_sorted = gene_vars_count.sort_values(by = 'count', ascending=False)\n",
    "\n",
    "variants_to_examine = gene_vars_sorted[gene_vars_sorted['count'] >= 15].gene_var.tolist()\n",
    "print('Number of variants to examine (common variants):', len(variants_to_examine))\n",
    "\n",
    "# filter the dataframe to only include CH variants with minimum 10 samples \n",
    "# we will be using this dataset going forward \n",
    "batch_gene_age_hla_dr_10 = batch_gene_age_hla_dr[batch_gene_age_hla_dr['gene_var'].isin(variants_to_examine)]\n",
    "batch_gene_age_hla_dp_10 = batch_gene_age_hla_dp[batch_gene_age_hla_dp['gene_var'].isin(variants_to_examine)]\n",
    "batch_gene_age_hla_dq_10 = batch_gene_age_hla_dq[batch_gene_age_hla_dq['gene_var'].isin(variants_to_examine)]\n",
    "batch_gene_age_hla_all_10 = batch_gene_age_hla_all[batch_gene_age_hla_all['gene_var'].isin(variants_to_examine)]\n",
    "print('Number of samples with a CH hotspot mutation (with > 2 reads):', batch_gene_vars.shape[0])\n",
    "print('Number of patients with a CH hotspot mutation (with > 2 reads):', len(batch_gene_vars.Person_ID.unique()))\n",
    "print('Number of samples with a CH hotspot mutation (with > 2 reads) (common variants):', batch_gene_age_hla_all_10.shape[0])\n",
    "print('Number of patients with a CH hotspot mutation (with > 2 reads) (common variants):', len(batch_gene_age_hla_all_10.Person_ID.unique()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reset number of categories of variants\n",
    "batch_gene_age_hla_dr_10['gene_var'] = batch_gene_age_hla_dr_10['gene_var'].astype(str)\n",
    "batch_gene_age_hla_dr_10['gene_var'] = batch_gene_age_hla_dr_10['gene_var'].astype('category')\n",
    "batch_gene_age_hla_dp_10['gene_var'] = batch_gene_age_hla_dp_10['gene_var'].astype(str)\n",
    "batch_gene_age_hla_dp_10['gene_var'] = batch_gene_age_hla_dp_10['gene_var'].astype('category')\n",
    "batch_gene_age_hla_dq_10['gene_var'] = batch_gene_age_hla_dq_10['gene_var'].astype(str)\n",
    "batch_gene_age_hla_dq_10['gene_var'] = batch_gene_age_hla_dq_10['gene_var'].astype('category')\n",
    "\n",
    "batch_gene_age_hla_all_10['gene_var'] = batch_gene_age_hla_all_10['gene_var'].astype(str)\n",
    "batch_gene_age_hla_all_10['gene_var'] = batch_gene_age_hla_all_10['gene_var'].astype('category')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add scores based on MHC class II genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# oki basically we need to figure out what is wrong with this function\n",
    "\n",
    "# define function to find the best score for the variant that is carried (ie present in the person)\n",
    "def find_best_score_for_variant_carried(row, df, param):\n",
    "\n",
    "    '''\n",
    "    This functiion is applied to the CH dataset\n",
    "    The df is the dataset with predictions for given allele and genetic variant \n",
    "    parameter is what to base this prediction on (here will only be using %Rank_EL)\n",
    "    '''\n",
    "    \n",
    "    # these are the values of rows to seach for \n",
    "    row_values = pd.to_numeric(row[1:-1], errors='coerce')\n",
    "    \n",
    "    # find alleles which are present \n",
    "    hla = row.index[1:-1][row_values >= 1]  \n",
    "    \n",
    "    # find values corresponding to these alleles \n",
    "    vals = df.loc[df['gene_var'] == row['gene_var'], hla].values.flatten() \n",
    "    \n",
    "    # it can be that nothing was found e.g., bc there were no predictions made for this variant\n",
    "    if vals.size == 0: \n",
    "        value = None\n",
    "    \n",
    "    else: \n",
    "    \n",
    "        # note here we will only be using '%Rank_EL'\n",
    "        if param == \"Aff_nM\":\n",
    "            value = min(vals) # highest affinity corresponds to lowest value (see plots above)\n",
    "        elif param == \"%Rank_EL\":\n",
    "            value = min(vals) # the best rank is the lowest one (indicates peptide in top x % of binders)\n",
    "        elif param == \"Score_EL\":\n",
    "            value = max(vals) # the best score is the highest one \n",
    "        elif param == \"%Rank_BA\":\n",
    "            value = min(vals) # the best rank is the lowest one (indicates peptide in top x % of binders)\n",
    "        elif param == \"Score_BA\":\n",
    "            value = max(vals) # the best score is the highest one\n",
    "        else:\n",
    "            print('Incorrect parameter provided') \n",
    "    \n",
    "    return value  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# do this for DR alleles \n",
    "param = '%Rank_EL'\n",
    "\n",
    "# DR\n",
    "pred_sub_dr = pred_df_dr[['HLA_formatted', 'gene_var_gt', param]]\n",
    "pred_sub_dr_wide = pd.pivot(pred_sub_dr, index='gene_var_gt', columns='HLA_formatted', values=param)\n",
    "pred_sub_dr_wide = pred_sub_dr_wide.reset_index() # this is to make sure that you have the gene_var column in there too\n",
    "\n",
    "hla_ukbb_dr = batch_gene_age_hla_dr_10.filter(regex='\\d').columns.tolist() # relevant HLAs\n",
    "hla_intersect_dr = pred_sub_dr_wide.columns[pred_sub_dr_wide.columns.isin(hla_ukbb_dr)] # HLA in the UKBB which I have predictions for \n",
    "hla_intersect_dr_list = hla_intersect_dr.tolist() \n",
    "\n",
    "# prepare gene variants names to match names in the Patient file \n",
    "pred_sub_dr = pred_sub_dr_wide[hla_intersect_dr_list + pred_sub_dr_wide.columns[pred_sub_dr_wide.columns.str.contains('gene_var')].tolist()]\n",
    "pred_sub_dr = pred_sub_dr[pred_sub_dr['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "pred_sub_dr['gene_var'] = pred_sub_dr['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "pred_sub_dr['gene_var'] = pred_sub_dr['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "pred_sub_dr['gene_var'] = pred_sub_dr['gene_var'].astype(str)\n",
    "\n",
    "# subset batch_gene_age_hla file \n",
    "ch_hla_sub_dr = batch_gene_age_hla_dr_10[hla_intersect_dr_list + batch_gene_age_hla_dr_10.columns[batch_gene_age_hla_dr_10.columns.str.contains('gene_var')].tolist()]\n",
    "ch_hla_sub_dr = pd.concat([batch_gene_age_hla_dr_10[\"Person_ID\"], ch_hla_sub_dr], axis=1) # add CH cases \n",
    "ch_hla_sub_dr['score'] = ch_hla_sub_dr.apply(find_best_score_for_variant_carried, df=pred_sub_dr, param=param, axis=1) # add score for the parameter\n",
    "ch_hla_scores_dr = ch_hla_sub_dr.dropna() # remove NA (incorrectly annotated cases)\n",
    "\n",
    "# merge scores with VAF and age\n",
    "age_vaf_dr = batch_gene_age_hla_dr_10[['Person_ID', 'VAF', 'var_depth', 'age', 'gene_var']]\n",
    "ch_hla_merge_dr = pd.merge(ch_hla_scores_dr, age_vaf_dr, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "# now add the columns with VAF and age \n",
    "col_to_select = ['Person_ID', 'gene_var', 'score', 'age', 'var_depth', 'VAF'] # subset the data \n",
    "ch_hla_merge_sub_dr = ch_hla_merge_dr[col_to_select]\n",
    "ch_hla_merge_sub_dr['log_score'] = -1*np.log10(ch_hla_merge_sub_dr['score']) # convert score to -log10(score)\n",
    "ch_hla_merge_sub_dr['allele_type'] = 'DR'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# DP\n",
    "pred_sub_dpq = pred_df_dpq[['HLA_formatted', 'gene_var_gt', param]]\n",
    "pred_sub_dpq = pred_sub_dpq[~pred_sub_dpq.duplicated()] # remove duplicate columns \n",
    "pred_sub_dpq_reset = pred_sub_dpq.reset_index()\n",
    "pred_sub_dpq_wide = pd.pivot(pred_sub_dpq_reset, index='gene_var_gt', columns='HLA_formatted', values=param)\n",
    "pred_sub_dpq_wide = pred_sub_dpq_wide.reset_index() # this is to make sure that you have the gene_var column in there too\n",
    "\n",
    "hla_ukbb_dp = batch_gene_age_hla_dp_10.filter(regex='\\d').columns.tolist() # relevant HLAs\n",
    "hla_intersect_dp = pred_sub_dpq_wide.columns[pred_sub_dpq_wide.columns.isin(hla_ukbb_dp)] # HLA in the UKBB which I have predictions for \n",
    "hla_intersect_dp_list = hla_intersect_dp.tolist() \n",
    "\n",
    "# prepare gene variants names to match names in the Patient file \n",
    "pred_sub_dp = pred_sub_dpq_wide[hla_intersect_dp_list + pred_sub_dpq_wide.columns[pred_sub_dpq_wide.columns.str.contains('gene_var')].tolist()]\n",
    "pred_sub_dp = pred_sub_dp[pred_sub_dp['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "pred_sub_dp['gene_var'] = pred_sub_dp['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "pred_sub_dp['gene_var'] = pred_sub_dp['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "pred_sub_dp['gene_var'] = pred_sub_dp['gene_var'].astype(str)\n",
    "\n",
    "# subset batch_gene_age_hla file \n",
    "ch_hla_sub_dp = batch_gene_age_hla_dp_10[hla_intersect_dp_list + batch_gene_age_hla_dp_10.columns[batch_gene_age_hla_dp_10.columns.str.contains('gene_var')].tolist()]\n",
    "ch_hla_sub_dp = pd.concat([batch_gene_age_hla_dp_10[\"Person_ID\"], ch_hla_sub_dp], axis=1) # add CH cases \n",
    "ch_hla_sub_dp['score'] = ch_hla_sub_dp.apply(find_best_score_for_variant_carried, df=pred_sub_dp, param=param, axis=1) # add score for the parameter\n",
    "ch_hla_scores_dp = ch_hla_sub_dp.dropna() # remove NA (incorrectly annotated cases)\n",
    "\n",
    "# merge scores with VAF and age\n",
    "age_vaf_dp = batch_gene_age_hla_dp[['Person_ID', 'VAF', 'var_depth', 'age', 'gene_var']]\n",
    "ch_hla_merge_dp = pd.merge(ch_hla_scores_dp, age_vaf_dp, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "# now add the columns with VAF and age \n",
    "col_to_select = ['Person_ID', 'gene_var', 'score', 'age', 'var_depth', 'VAF'] # subset the data \n",
    "ch_hla_merge_sub_dp = ch_hla_merge_dp[col_to_select]\n",
    "ch_hla_merge_sub_dp['log_score'] = -1*np.log10(ch_hla_merge_sub_dp['score']) # convert score to -log10(score)\n",
    "ch_hla_merge_sub_dp['allele_type'] = 'DP'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# DQ\n",
    "hla_ukbb_dq = batch_gene_age_hla_dq_10.filter(regex='\\d').columns.tolist() # relevant HLAs\n",
    "hla_intersect_dq = pred_sub_dpq_wide.columns[pred_sub_dpq_wide.columns.isin(hla_ukbb_dq)] # HLA in the UKBB which I have predictions for \n",
    "hla_intersect_dq_list = hla_intersect_dq.tolist() \n",
    "\n",
    "# prepare gene variants names to match names in the Patient file \n",
    "pred_sub_dq = pred_sub_dpq_wide[hla_intersect_dq_list + pred_sub_dpq_wide.columns[pred_sub_dpq_wide.columns.str.contains('gene_var')].tolist()]\n",
    "pred_sub_dq = pred_sub_dq[pred_sub_dq['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "pred_sub_dq['gene_var'] = pred_sub_dq['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "pred_sub_dq['gene_var'] = pred_sub_dq['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "pred_sub_dq['gene_var'] = pred_sub_dq['gene_var'].astype(str)\n",
    "\n",
    "# subset batch_gene_age_hla file \n",
    "ch_hla_sub_dq = batch_gene_age_hla_dq_10[hla_intersect_dq_list + batch_gene_age_hla_dq_10.columns[batch_gene_age_hla_dq_10.columns.str.contains('gene_var')].tolist()]\n",
    "ch_hla_sub_dq = pd.concat([batch_gene_age_hla_dq_10[\"Person_ID\"], ch_hla_sub_dq], axis=1) # add CH cases \n",
    "ch_hla_sub_dq['score'] = ch_hla_sub_dq.apply(find_best_score_for_variant_carried, df=pred_sub_dq, param=param, axis=1) # add score for the parameter\n",
    "ch_hla_scores_dq = ch_hla_sub_dq.dropna() # remove NA (incorrectly annotated cases)\n",
    "\n",
    "# merge scores with VAF and age\n",
    "age_vaf_dq = batch_gene_age_hla_dq[['Person_ID', 'VAF', 'var_depth', 'age', 'gene_var']]\n",
    "ch_hla_merge_dq = pd.merge(ch_hla_scores_dq, age_vaf_dq, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "# now add the columns with VAF and age \n",
    "col_to_select = ['Person_ID', 'gene_var', 'score', 'age', 'var_depth', 'VAF'] # subset the data \n",
    "ch_hla_merge_sub_dq = ch_hla_merge_dq[col_to_select]\n",
    "ch_hla_merge_sub_dq['log_score'] = -1*np.log10(ch_hla_merge_sub_dq['score']) # convert score to -log10(score)\n",
    "ch_hla_merge_sub_dq['allele_type'] = 'DQ'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>gene_var</th>\n",
       "      <th>score</th>\n",
       "      <th>age</th>\n",
       "      <th>var_depth</th>\n",
       "      <th>VAF</th>\n",
       "      <th>log_score</th>\n",
       "      <th>allele_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2100679</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>7.50</td>\n",
       "      <td>69.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.048780</td>\n",
       "      <td>-0.875061</td>\n",
       "      <td>DR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2110983</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>3.01</td>\n",
       "      <td>63.0</td>\n",
       "      <td>5</td>\n",
       "      <td>0.089286</td>\n",
       "      <td>-0.478566</td>\n",
       "      <td>DR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2116112</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>11.18</td>\n",
       "      <td>60.0</td>\n",
       "      <td>5</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>-1.048442</td>\n",
       "      <td>DR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2118726</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>3.01</td>\n",
       "      <td>57.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025316</td>\n",
       "      <td>-0.478566</td>\n",
       "      <td>DR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2120797</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>11.18</td>\n",
       "      <td>45.0</td>\n",
       "      <td>6</td>\n",
       "      <td>0.127660</td>\n",
       "      <td>-1.048442</td>\n",
       "      <td>DR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1473</th>\n",
       "      <td>1415885</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>3.39</td>\n",
       "      <td>63.0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.422222</td>\n",
       "      <td>-0.530200</td>\n",
       "      <td>DQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1474</th>\n",
       "      <td>1435102</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>1.43</td>\n",
       "      <td>69.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.048780</td>\n",
       "      <td>-0.155336</td>\n",
       "      <td>DQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1475</th>\n",
       "      <td>1444475</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>26.02</td>\n",
       "      <td>70.0</td>\n",
       "      <td>5</td>\n",
       "      <td>0.294118</td>\n",
       "      <td>-1.415307</td>\n",
       "      <td>DQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>1445954</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>1.43</td>\n",
       "      <td>62.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>-0.155336</td>\n",
       "      <td>DQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1477</th>\n",
       "      <td>1921630</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>5.18</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>-0.714330</td>\n",
       "      <td>DQ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4434 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Person_ID      gene_var  score   age  var_depth       VAF  log_score  \\\n",
       "0       2100679  DNMT3A_R882H   7.50  69.0          4  0.048780  -0.875061   \n",
       "1       2110983  DNMT3A_R882H   3.01  63.0          5  0.089286  -0.478566   \n",
       "2       2116112  DNMT3A_R882H  11.18  60.0          5  0.066667  -1.048442   \n",
       "3       2118726  DNMT3A_R882H   3.01  57.0          2  0.025316  -0.478566   \n",
       "4       2120797  DNMT3A_R882H  11.18  45.0          6  0.127660  -1.048442   \n",
       "...         ...           ...    ...   ...        ...       ...        ...   \n",
       "1473    1415885   SF3B1_K666N   3.39  63.0         19  0.422222  -0.530200   \n",
       "1474    1435102   SF3B1_K666N   1.43  69.0          2  0.048780  -0.155336   \n",
       "1475    1444475   SF3B1_K666N  26.02  70.0          5  0.294118  -1.415307   \n",
       "1476    1445954   SF3B1_K666N   1.43  62.0          2  0.055556  -0.155336   \n",
       "1477    1921630   SF3B1_K666N   5.18  56.0          2  0.074074  -0.714330   \n",
       "\n",
       "     allele_type  \n",
       "0             DR  \n",
       "1             DR  \n",
       "2             DR  \n",
       "3             DR  \n",
       "4             DR  \n",
       "...          ...  \n",
       "1473          DQ  \n",
       "1474          DQ  \n",
       "1475          DQ  \n",
       "1476          DQ  \n",
       "1477          DQ  \n",
       "\n",
       "[4434 rows x 8 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we can actually combine all these scores now \n",
    "ch_hla_merge_sub_all = pd.concat([ch_hla_merge_sub_dr, ch_hla_merge_sub_dp, ch_hla_merge_sub_dq], axis = 0)\n",
    "ch_hla_merge_sub_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define function to find the best score for the variant that is carried (ie present in the person)\n",
    "def find_best_score_for_variant_carried_all_hla(row, df, param):\n",
    "\n",
    "    '''\n",
    "    This functiion is applied to the CH dataset\n",
    "    The df is the dataset with predictions for given allele and genetic variant \n",
    "    parameter is what to base this prediction on (here will only be using %Rank_EL)\n",
    "    '''\n",
    "    \n",
    "    # these are the values of rows to seach for \n",
    "    row_values = pd.to_numeric(row[1:-1], errors='coerce')\n",
    "    \n",
    "    # find alleles which are present (NOTE this is changed from the previous function bc formatting is different)\n",
    "    hla = row.index[1:-1][row_values >= 1]  \n",
    "    \n",
    "    # find values corresponding to these alleles \n",
    "    vals = df.loc[df['gene_var'] == row['gene_var'], hla].values.flatten() \n",
    "    \n",
    "    # it can be that nothing was found e.g., bc there were no predictions made for this variant\n",
    "    if vals.size == 0: \n",
    "        value = None\n",
    "    \n",
    "    else: \n",
    "    \n",
    "        # note here we will only be using '%Rank_EL'\n",
    "        if param == \"Aff_nM\":\n",
    "            value = min(vals) # highest affinity corresponds to lowest value (see plots above)\n",
    "        elif param == \"%Rank_EL\":\n",
    "            value = min(vals) # the best rank is the lowest one (indicates peptide in top x % of binders)\n",
    "        elif param == \"Score_EL\":\n",
    "            value = max(vals) # the best score is the highest one \n",
    "        elif param == \"%Rank_BA\":\n",
    "            value = min(vals) # the best rank is the lowest one (indicates peptide in top x % of binders)\n",
    "        elif param == \"Score_BA\":\n",
    "            value = max(vals) # the best score is the highest one\n",
    "        else:\n",
    "            print('Incorrect parameter provided') \n",
    "    \n",
    "    return value  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of alleles and allele combinations for which we have predictions: 441\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# add a single score based on all of the MHC class II alleles someone has (not breaking down into DR, DP, DQ)\n",
    "\n",
    "pred_sub_all = pred_df_all[['HLA_formatted', 'gene_var_gt', param]]\n",
    "pred_sub_all = pred_sub_all[~pred_sub_all.duplicated()] # remove duplicate columns \n",
    "pred_sub_all_reset = pred_sub_all.reset_index()\n",
    "pred_sub_all_wide = pd.pivot(pred_sub_all_reset, index='gene_var_gt', columns='HLA_formatted', values=param).reset_index()\n",
    "\n",
    "hla_ukbb_all = batch_gene_age_hla_all_10.columns[15:].tolist() # this includes DP / DQ and in combinations only\n",
    "hla_intersect_all = pred_sub_all_wide.columns[pred_sub_all_wide.columns.isin(hla_ukbb_all)] # HLA in the UKBB which I have predictions for \n",
    "hla_intersect_all_list = hla_intersect_all.tolist() \n",
    "\n",
    "# prepare gene variants names to match names in the Patient file \n",
    "pred_sub_all = pred_sub_all_wide[hla_intersect_all_list + pred_sub_all_wide.columns[pred_sub_all_wide.columns.str.contains('gene_var')].tolist()]\n",
    "pred_sub_all = pred_sub_all[pred_sub_all['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "pred_sub_all['gene_var'] = pred_sub_all['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "pred_sub_all['gene_var'] = pred_sub_all['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "pred_sub_all['gene_var'] = pred_sub_all['gene_var'].astype(str)\n",
    "\n",
    "# subset batch_gene_age_hla file \n",
    "ch_hla_sub_all = batch_gene_age_hla_all_10[hla_intersect_all_list + batch_gene_age_hla_all_10.columns[batch_gene_age_hla_all.columns.str.contains('gene_var')].tolist()]\n",
    "ch_hla_sub_all = pd.concat([batch_gene_age_hla_all_10[\"Person_ID\"], ch_hla_sub_all], axis=1) # add CH cases \n",
    "ch_hla_sub_all['score'] = ch_hla_sub_all.apply(find_best_score_for_variant_carried_all_hla, df=pred_sub_all, param=param, axis=1) # add score for the parameter\n",
    "ch_hla_scores_all = ch_hla_sub_all.dropna() # remove NA (incorrectly annotated cases)\n",
    "\n",
    "# merge scores with VAF and age\n",
    "age_vaf_all = batch_gene_age_hla_all[['Person_ID', 'VAF', 'var_depth', 'age', 'gene_var']]\n",
    "ch_hla_merge_all = pd.merge(ch_hla_scores_all, age_vaf_all, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "# now add the columns with VAF and age \n",
    "col_to_select = ['Person_ID', 'gene_var', 'score', 'age', 'var_depth', 'VAF'] # subset the data \n",
    "ch_hla_merge_sub_all = ch_hla_merge_all[col_to_select]\n",
    "ch_hla_merge_sub_all['log_score'] = -1*np.log10(ch_hla_merge_sub_all['score']) # convert score to -log10(score)\n",
    "\n",
    "# filter out rare variants\n",
    "ch_hla_merge_sub_all = ch_hla_merge_sub_all[ch_hla_merge_sub_all['gene_var'].isin(variants_to_examine)]\n",
    "print('Number of alleles and allele combinations for which we have predictions:', len(hla_intersect_all_list))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# SANITY CHECK: \n",
    "ch_hla_merge_sub_by_allele = pd.concat([ch_hla_merge_sub_dq, ch_hla_merge_sub_dp, ch_hla_merge_sub_dr], axis = 0)\n",
    "ch_by_allele = pd.DataFrame(ch_hla_merge_sub_by_allele.groupby(['Person_ID', 'gene_var']).log_score.max()) # find best score across alleles\n",
    "ch_sub = ch_hla_merge_sub_all[['Person_ID', 'gene_var', 'log_score']].sort_values(by = 'Person_ID') # compare to score added \n",
    "ch_diff = pd.merge(ch_by_allele, ch_sub, on = ['Person_ID', 'gene_var'])\n",
    "ch_diff['diff'] = ch_diff['log_score_x'] - ch_diff['log_score_y']\n",
    "ch_diff['diff'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add columns to change gene var formatting and add number of individuals observed for each variant\n",
    "ch_hla_merge_sub_all['gene_var2'] = ch_hla_merge_sub_all['gene_var'].str.replace('_', '\\n')\n",
    "\n",
    "# add counts\n",
    "df_carriers_counts = ch_hla_merge_sub_all['gene_var'].value_counts().reset_index()\n",
    "df_carriers_counts.columns.values[1] = 'nr_carriers'\n",
    "df_carriers_counts.nr_carriers = df_carriers_counts.nr_carriers.astype(str)\n",
    "\n",
    "ch_hla_merge_sub_all = pd.merge(ch_hla_merge_sub_all, df_carriers_counts, on = 'gene_var')\n",
    "ch_hla_merge_sub_all['gene_var3'] = ch_hla_merge_sub_all['gene_var2'] + '\\n(n=' + ch_hla_merge_sub_all['nr_carriers'] + ')'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Okay, first, we need to get a score column so that will be a log score which matches the gene var someone has \n",
    "def get_score(row):\n",
    "    \n",
    "    category = row['gene_var']\n",
    "    return row[f'score_{category}']\n",
    "\n",
    "# split into a desired number of groups\n",
    "def split_into_equal_groups(df, num_groups):\n",
    "\n",
    "    # first, assign a rank to everyone based on score\n",
    "    # the highest score = the lowest rank (ie highest score means you get rank 1)\n",
    "    # if two people have the same score, assign consecutive ranks \n",
    "    df['rank'] = df['log_score'].rank(method='first', ascending=False)\n",
    "    \n",
    "    # assign groups based on rank \n",
    "\n",
    "    # first, determine the number of samples\n",
    "    total_samples = len(df)\n",
    "\n",
    "    # now, determine the number of samples in each group\n",
    "    samples_per_group = total_samples // num_groups\n",
    "    remainder = total_samples % num_groups\n",
    "    group_sizes = [samples_per_group + 1 if i < remainder else samples_per_group for i in range(num_groups)] # add one person if there is a reminder \n",
    "    \n",
    "    # now, assign the group based on the rank \n",
    "    df_sort = df.sort_values(by = 'rank')\n",
    "    group_assignments = []\n",
    "    group_number = 1\n",
    "    start = 0\n",
    "    for size in group_sizes:\n",
    "        end = start + size\n",
    "        group_assignments.extend([group_number] * (end - start))\n",
    "        start = end\n",
    "        group_number += 1\n",
    "\n",
    "    # Add a new column 'group' to the DataFrame indicating the group assignment for each row\n",
    "    df_sort['group'] = group_assignments\n",
    "    \n",
    "    return df_sort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define function to plot\n",
    "def plot_ranks(data, x, y, hue, colors, order, plt_type, nr, hue_order = None, name = None):\n",
    "\n",
    "    score = '-log10(%Rank_EL)'\n",
    "\n",
    "    plt.figure(figsize = (16, 4))\n",
    "\n",
    "    if hue_order is None:\n",
    "        if plt_type == 'box':\n",
    "            p = sns.boxplot(x=x, y=y, hue=hue, data=data, showfliers=False, palette=colors, order=order)\n",
    "            plt.xticks(rotation=90, fontsize = 12)\n",
    "            plt.yticks(fontsize = 13)\n",
    "            plt.xlabel('CH hotspot variant carried', fontsize = 14)\n",
    "            plt.ylabel(r'$\\log_{10}$(min %EL rank)', fontsize = 14)\n",
    "            plt.title('MHC II', fontsize = 16)\n",
    "            plt.ylim(-2.5, 2.5)\n",
    "            plt.legend(title = 'Variant presentation', loc = 'upper right', fontsize = 13, title_fontsize = 14)\n",
    "            plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_dist_{hue}_{y}_{plt_type}_{nr}_{name}.pdf', bbox_inches='tight')\n",
    "\n",
    "        if plt_type == 'jitter':\n",
    "            p = sns.stripplot(x=x, y=y, hue=hue, data=data, palette=colors, order=order, size = 5, alpha = 0.4)\n",
    "            plt.xticks(rotation=90, fontsize = 12)\n",
    "            plt.yticks(fontsize = 13)\n",
    "            plt.xlabel('CH hotspot variant carried', fontsize = 14)\n",
    "            plt.ylabel(r'$\\log_{10}$(min %EL rank)', fontsize = 14)\n",
    "            plt.title('MHC II', fontsize = 16)\n",
    "            plt.legend(title = 'Variant presentation', markerscale = 2, loc = 'upper right', fontsize = 13, title_fontsize = 14)\n",
    "            plt.ylim(-2.5, 2.5)\n",
    "            legend = plt.gca().get_legend()\n",
    "            for handle, color in zip(legend.legendHandles, colors):\n",
    "                handle.set_color(color)\n",
    "            for lh in legend.legendHandles:\n",
    "                lh.set_alpha(1)\n",
    "            plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_dist_{hue}_{y}_{plt_type}_{nr}_{name}.pdf', bbox_inches='tight')\n",
    "\n",
    "    else:\n",
    "        if plt_type == 'box':\n",
    "            p = sns.boxplot(x=x, y=y, hue=hue, data=data, showfliers=False, palette=colors, order=order, hue_order=hue_order)\n",
    "            plt.xticks(rotation=90, fontsize = 12)\n",
    "            plt.yticks(fontsize = 13)\n",
    "            plt.xlabel('CH hotspot variant', fontsize = 14)\n",
    "            plt.ylabel(r'$\\log_{10}$(min %EL rank)', fontsize = 14)\n",
    "            plt.title('MHC II', fontsize = 16)\n",
    "            plt.ylim(-2.5, 2.5)\n",
    "            plt.legend(title = 'Variant presentation', loc = 'upper right', fontsize = 13, title_fontsize = 14)\n",
    "            plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_dist_{hue}_{y}_{plt_type}_{nr}_{name}.pdf', bbox_inches='tight')\n",
    "\n",
    "        if plt_type == 'jitter':\n",
    "            p = sns.stripplot(x=x, y=y, hue=hue, data=data, palette=colors, order=order, hue_order=hue_order, size = 5, alpha = 0.4)\n",
    "            plt.xticks(rotation=90, fontsize = 12)\n",
    "            plt.yticks(fontsize = 13)\n",
    "            plt.xlabel('CH hotspot variant', fontsize = 14)\n",
    "            plt.ylabel(r'$\\log_{10}$(min %EL rank)', fontsize = 14)\n",
    "            plt.title('MHC II', fontsize = 16)\n",
    "            plt.ylim(-2.5, 2.5)\n",
    "            plt.legend(title = 'Variant presentation', loc = 'upper right', markerscale = 2, fontsize = 13, title_fontsize = 14)\n",
    "            legend = plt.gca().get_legend()\n",
    "            for handle, color in zip(legend.legendHandles, colors):\n",
    "                handle.set_color(color)\n",
    "            for lh in legend.legendHandles:\n",
    "                lh.set_alpha(1)\n",
    "            plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_dist_{hue}_{y}_{plt_type}_{nr}_{name}.pdf', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# specify colors to use for plotting\n",
    "col0 = '#910e0e'\n",
    "col1 = '#ed5d5d'\n",
    "col1b = '#fab9b2'\n",
    "col2 = '#65e6f5'\n",
    "col3 = '#1091af'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# specify order\n",
    "ch_hla_merge_sub_all['median_score'] = ch_hla_merge_sub_all.groupby('gene_var')['log_score'].transform('median')\n",
    "ch_hla_merge_sub_all = ch_hla_merge_sub_all.sort_values(by='median_score', ascending = False)\n",
    "order = ch_hla_merge_sub_all.gene_var2.unique()\n",
    "\n",
    "# split df into equal groups\n",
    "\n",
    "# apply function separately to each category \n",
    "df_groups2_eqsize = ch_hla_merge_sub_all.groupby('gene_var2').apply(lambda x: split_into_equal_groups(x, num_groups=2))\n",
    "df_groups3_eqsize = ch_hla_merge_sub_all.groupby('gene_var2').apply(lambda x: split_into_equal_groups(x, num_groups=3))\n",
    "df_groups4_eqsize = ch_hla_merge_sub_all.groupby('gene_var2').apply(lambda x: split_into_equal_groups(x, num_groups=4))\n",
    "df_groups5_eqsize = ch_hla_merge_sub_all.groupby('gene_var2').apply(lambda x: split_into_equal_groups(x, num_groups=5))\n",
    "\n",
    "# add a more descriptive name for the group\n",
    "df_groups2_eqsize['group_name'] =  df_groups2_eqsize['group'].map({1: 'top', 2: 'bottom'})\n",
    "df_groups3_eqsize['group_name'] =  df_groups3_eqsize['group'].map({1: 'top', 2: 'mid', 3: 'bottom'})\n",
    "df_groups4_eqsize['group_name'] =  df_groups4_eqsize['group'].map({1: 'top', 2: 'mid-high', 3: 'mid-low', 4: 'bottom'})\n",
    "df_groups5_eqsize['group_name'] =  df_groups5_eqsize['group'].map({1: 'top', 2: 'mid-high', 3: 'mid', 4: 'mid-low', 5: 'bottom'})\n",
    "\n",
    "# specify order of levels in group descriptors\n",
    "cat_type2 = pd.CategoricalDtype(categories=['top', 'bottom'], ordered=True)\n",
    "df_groups2_eqsize['group_name'] = df_groups2_eqsize['group_name'].astype(cat_type2)\n",
    "df_groups2_eqsize['group_name'] = df_groups2_eqsize['group_name'].astype('category')\n",
    "\n",
    "cat_type3 = pd.CategoricalDtype(categories=['top', 'mid', 'bottom'], ordered=True)\n",
    "df_groups3_eqsize['group_name'] = df_groups3_eqsize['group_name'].astype(cat_type3)\n",
    "df_groups3_eqsize['group_name'] = df_groups3_eqsize['group_name'].astype('category')\n",
    "\n",
    "cat_type4 = pd.CategoricalDtype(categories=['top', 'mid-high', 'mid-low', 'bottom'], ordered=True)\n",
    "df_groups4_eqsize['group_name'] = df_groups4_eqsize['group_name'].astype(cat_type4)\n",
    "df_groups4_eqsize['group_name'] = df_groups4_eqsize['group_name'].astype('category')\n",
    "\n",
    "cat_type5 = pd.CategoricalDtype(categories=['top', 'mid-high', 'mid', 'mid-low', 'bottom'], ordered=True)\n",
    "df_groups5_eqsize['group_name'] = df_groups5_eqsize['group_name'].astype(cat_type5)\n",
    "df_groups5_eqsize['group_name'] = df_groups5_eqsize['group_name'].astype('category')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# create plots\n",
    "colors2 = [col0, col3]\n",
    "hue_order2 = ['top', 'bottom']\n",
    "scores_groups2_eqsize = df_groups2_eqsize.sort_values(by = 'log_score', ascending = False)\n",
    "\n",
    "colors3 = [col0, col1b, col3]\n",
    "hue_order3 = ['top','mid', 'bottom']\n",
    "scores_groups3_eqsize = df_groups3_eqsize.sort_values(by = 'log_score', ascending = False)\n",
    "\n",
    "colors4 = [col0, col1, col2, col3]\n",
    "hue_order4 = ['top','mid-high', 'mid-low', 'bottom']\n",
    "scores_groups4_eqsize = df_groups4_eqsize.sort_values(by = 'log_score', ascending = False)\n",
    "\n",
    "colors5 = [col0, col1, col1b, col2, col3]\n",
    "hue_order5 = ['top', 'mid-high', 'mid', 'mid-low', 'bottom']\n",
    "scores_groups5_eqsize = df_groups5_eqsize.sort_values(by = 'log_score', ascending = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify order for plotting \n",
    "ch_hla_merge_sub_all['median_score'] = ch_hla_merge_sub_all.groupby('gene_var')['log_score'].transform('median')\n",
    "ch_hla_merge_sub_all = ch_hla_merge_sub_all.sort_values(by='median_score', ascending = False)\n",
    "order_3 = ch_hla_merge_sub_all.gene_var3.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# add column with VAF data \n",
    "df_groups2_eqsize['VAF_percent'] = df_groups2_eqsize['VAF'] * 100\n",
    "df_groups3_eqsize['VAF_percent'] = df_groups3_eqsize['VAF'] * 100\n",
    "df_groups4_eqsize['VAF_percent'] = df_groups4_eqsize['VAF'] * 100\n",
    "df_groups5_eqsize['VAF_percent'] = df_groups5_eqsize['VAF'] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define function to plot cumulative probability\n",
    "def plot_cumulative(data, x, y, hue, colors, xlabel, ylabel, var, score, score_title, name):\n",
    "\n",
    "    var_name = var.split('_')[0:2]\n",
    "    var_name = ' '.join(var_name)\n",
    "    nr_cases = data.shape[0]\n",
    "\n",
    "    plt.figure(figsize=(4,4)) \n",
    "    sns.lineplot(x=x, y=y, data=data, hue=hue, palette=colors, legend = True)\n",
    "    sns.scatterplot(x=x, y=y, data=data, hue=hue, palette=colors, legend = False, alpha = 0.5, linewidth=0)\n",
    "\n",
    "    # specify title and axes labels \n",
    "    plt.title(f'{var_name} (n={nr_cases})', fontsize = 12)\n",
    "    plt.xlabel(xlabel, fontsize = 10)\n",
    "    plt.ylabel(ylabel, fontsize = 10)\n",
    "\n",
    "    # convert both axes to log scale \n",
    "    plt.gca().set_xscale('log')\n",
    "    plt.gca().set_yscale('log')\n",
    "\n",
    "    # adjust legend \n",
    "    if var == 'DNMT3A_R326C':\n",
    "        legend = plt.legend()\n",
    "        legend.prop.set_size(10) \n",
    "        plt.legend(title = 'Variant presentation', loc = 'lower left', fontsize = 10, title_fontsize = 11)\n",
    "    else:\n",
    "        plt.legend().remove()\n",
    "\n",
    "    # specify ticks on the x axis\n",
    "    x_ticks = [1, 3, 10, 30]\n",
    "    plt.xticks(x_ticks, x_ticks, fontsize = 9)\n",
    "\n",
    "    y_ticks = [1, 0.3, 0.1, 0.03]\n",
    "    plt.yticks(y_ticks, y_ticks, fontsize = 9)\n",
    "    \n",
    "    plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_VAF_{hue}_{name}_{score}_{var}.pdf', bbox_inches='tight')\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# define a new function just to do it for one variant \n",
    "def find_best_score_one_variant_log_ref_ch(row, df, param, variant):\n",
    "    \n",
    "    '''\n",
    "    The same function but log scores \n",
    "    the only allowed parameters are %Rank_EL and %Rank_BA\n",
    "    we want -1 * log(score) so the highest score if the most \"immunogenic\" (best binding) one\n",
    "    '''\n",
    "    \n",
    "    hlas = row.index[1:-1][row[1:-1] >= 1] # select alleles which each Person (row) carrie\n",
    "    \n",
    "    scores = {} # initialise empty dictionaries\n",
    "\n",
    "    if param == \"%Rank_EL\":\n",
    "\n",
    "        # take maximum of the negative log score \n",
    "        best_value = max(-1 * np.log10(df.loc[df['gene_var_gt'] == variant, hlas].values[0]))\n",
    "        scores[f'score_{variant}'] = best_value\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"%Rank_BA\":\n",
    "        \n",
    "        # take maxium of the negative log score \n",
    "        best_value = max(-1 * np.log10(df.loc[df['gene_var_gt'] == variant, hlas].values[0]))\n",
    "        scores[f'score_{variant}'] = best_value\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Aff_nM\":\n",
    "        \n",
    "        # take maxium of the negative log score (low aff = more immunogenic)\n",
    "        best_value = max(-1 * np.log10(df.loc[df['gene_var_gt'] == variant, hlas].values[0]))\n",
    "        scores[f'score_{variant}'] = best_value\n",
    "        return pd.Series(scores)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define a function to split people into groups based on MHC presentation score \n",
    "\n",
    "def split_equal_groups_by_varscore(df, num_groups, var):\n",
    "\n",
    "    '''\n",
    "    first, assign a rank to everyone based on score\n",
    "    the highest score = the lowest rank (ie highest score means you get rank 1)\n",
    "    if two people have the same score, assign consecutive ranks \n",
    "    '''\n",
    "    \n",
    "    # first, determine the number of samples\n",
    "    total_samples = len(df)\n",
    "\n",
    "    # now, determine the number of samples in each group\n",
    "    samples_per_group = total_samples // num_groups\n",
    "    remainder = total_samples % num_groups\n",
    "    group_sizes = [samples_per_group + 1 if i < remainder else samples_per_group for i in range(num_groups)] # add one person if there is a reminder \n",
    "    \n",
    "    # now, assign the group based on the rank \n",
    "    group_assignments = []\n",
    "    group_number = 1\n",
    "    start = 0\n",
    "    for size in group_sizes:\n",
    "        end = start + size\n",
    "        group_assignments.extend([group_number] * (end - start))\n",
    "        start = end\n",
    "        group_number += 1\n",
    "\n",
    "    # Add a new column 'group' to the DataFrame indicating the group assignment for each row\n",
    "    return group_assignments\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define functions to compare groups\n",
    "\n",
    "def percentage_difference(group, nr_groups):\n",
    "\n",
    "    x_a = group.loc[group[f'groups_{nr_groups}'] == 1, 'CH_variant_positive'].iloc[0]  # Value of column X where Y is 'a'\n",
    "    x_b = group.loc[group[f'groups_{nr_groups}'] == nr_groups, 'CH_variant_positive'].iloc[0]  # Value of column X where Y is 'b'\n",
    "    \n",
    "    result = (x_b - x_a) / x_a * 100\n",
    "    return result\n",
    "\n",
    "def normalized_difference(group, nr_groups):\n",
    "\n",
    "    x_a = group.loc[group[f'groups_{nr_groups}'] == 1, 'CH_variant_positive'].iloc[0]  # Value of column X where Y is 'a'\n",
    "    x_b = group.loc[group[f'groups_{nr_groups}'] == nr_groups, 'CH_variant_positive'].iloc[0]  # Value of column X where Y is 'b'\n",
    "    num_observations = group['count']   # Number of observations in this category\n",
    "    \n",
    "    result = (x_b - x_a) / num_observations * 100\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batches that were examined for variants: {11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21}\n",
      "Number of samples examined for variants: 99597\n"
     ]
    }
   ],
   "source": [
    "# NOTE: we can only do this for people we actually screened for CH\n",
    "# find IDs of these individuals who were screened \n",
    "batch_all['batch_number'] = batch_all['batch'].apply(lambda x: [int(num) for num in re.findall(r'\\d+', x)])\n",
    "batch_numbers_examined = set(batch_all['batch_number'].sum()) \n",
    "print('Batches that were examined for variants:', batch_numbers_examined)\n",
    "\n",
    "all_dataframes = [f'batch_{i}_ids' for i in range(11, 61)]\n",
    "selected_ids = []\n",
    "\n",
    "# Iterate through all df name \n",
    "for df_name in all_dataframes:\n",
    "    number = int(re.search(r'\\d+', df_name).group())\n",
    "    # check if number matches numbers of batches examined \n",
    "    if number in batch_numbers_examined:        \n",
    "        selected_ids.append(globals()[df_name])\n",
    "\n",
    "ids_examined = pd.concat(selected_ids, ignore_index=True)\n",
    "print('Number of samples examined for variants:', ids_examined.shape[0])\n",
    "\n",
    "ids_examined_list = ids_examined['sample_ID'].tolist()\n",
    "ids_examined_list = [int(id) for id in ids_examined_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# create df with combinations\n",
    "df_hla2_dp_all_combinations = pd.concat([df_hla2_dp_all.iloc[:,0], df_hla2_dp_all.iloc[:,42:]], axis = 1)\n",
    "df_hla2_dq_all_combinations = pd.concat([df_hla2_dq_all.iloc[:,0], df_hla2_dq_all.iloc[:,33:]], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now, get a dataframe for everyone who was genotyped and their full genotype\n",
    "df_hla2_drp = pd.merge(df_hla2_dr, df_hla2_dp_all_combinations, on = 'Person_ID')\n",
    "df_hla2_drpq = pd.merge(df_hla2_drp, df_hla2_dq_all_combinations, on = 'Person_ID')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset for individuals who were screened\n",
    "df_hla2_drpq_screened = df_hla2_drpq[df_hla2_drpq['Person_ID'].isin(ids_examined_list)]\n",
    "df_hla2_drpq_screened = df_hla2_drpq_screened.drop(columns=['DRB5_9901', 'DRB4_9901', 'DRB3_9901']) # we don't consider these alleles here "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# cases who have CH with a genotyped variant (from one of the ones we were looking at) are in batch_gene_vars \n",
    "# add a single score based on all of the MHC class II alleles someone has (not breaking down into DR, DP, DQ)\n",
    "\n",
    "pred_sub_all = pred_df_all[['HLA_formatted', 'gene_var_gt', param]]\n",
    "pred_sub_all = pred_sub_all[~pred_sub_all.duplicated()] # remove duplicate columns \n",
    "pred_sub_all_reset = pred_sub_all.reset_index()\n",
    "pred_sub_all_wide = pd.pivot(pred_sub_all_reset, index='gene_var_gt', columns='HLA_formatted', values=param).reset_index()\n",
    "\n",
    "hla_ukbb_all = batch_gene_age_hla_all_10.columns[15:].tolist() # this includes DP / DQ and in combinations only\n",
    "hla_intersect_all = pred_sub_all_wide.columns[pred_sub_all_wide.columns.isin(hla_ukbb_all)] # HLA in the UKBB which I have predictions for \n",
    "hla_intersect_all_list = hla_intersect_all.tolist() \n",
    "\n",
    "# prepare gene variants names to match names in the Patient file \n",
    "pred_sub_all = pred_sub_all_wide[hla_intersect_all_list + pred_sub_all_wide.columns[pred_sub_all_wide.columns.str.contains('gene_var')].tolist()]\n",
    "pred_sub_all = pred_sub_all[pred_sub_all['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "pred_sub_all['gene_var'] = pred_sub_all['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "pred_sub_all['gene_var'] = pred_sub_all['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "pred_sub_all['gene_var'] = pred_sub_all['gene_var'].astype(str)\n",
    "\n",
    "# note: for most people, we DONT KNOW if they have CH or not so you have to only consider those you actually screened\n",
    "ch_sampleids = batch_gene_age_hla_all_10.Person_ID.tolist()\n",
    "df_clean_hla2_screened = df_clean_hla2[df_clean_hla2['Person_ID'].isin(ids_examined_list)]\n",
    "ukbb_no_ch = df_clean_hla2_screened[~df_clean_hla2_screened['Person_ID'].isin(ch_sampleids)]\n",
    "hla_ukbb = ukbb_no_ch.filter(regex='\\d').columns # identify HLA alleles carried by non-CH-individuals in the UKBB data\n",
    "\n",
    "# prepare gene variants names to match names in the Patient file \n",
    "pred_sub_all = pred_sub_all_wide[hla_intersect_all_list + pred_sub_all_wide.columns[pred_sub_all_wide.columns.str.contains('gene_var')].tolist()]\n",
    "pred_sub_all = pred_sub_all[pred_sub_all['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "pred_sub_all['gene_var'] = pred_sub_all['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "pred_sub_all['gene_var'] = pred_sub_all['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "pred_sub_all['gene_var'] = pred_sub_all['gene_var'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loop for all variants\n",
    "\n",
    "scores_2groups = {}\n",
    "scores_for_each_var_ukbb = {}\n",
    "\n",
    "for var in variants_to_examine:\n",
    "    \n",
    "    # for everyone in the UKBB, add binding score for each variant based on their MHC genotype\n",
    "    df_all_ukbb_scores = pd.concat([df_hla2_drpq_screened, df_hla2_drpq_screened.apply(find_best_score_one_variant_log_ref_ch, df=pred_sub_all, param=param, variant = f'{var}_ch', axis=1)], axis=1)\n",
    "    scores_for_each_var_ukbb[var] = df_all_ukbb_scores[f'score_{var}_ch'] # add this so we can later correlate with immunogenicity\n",
    "\n",
    "    # order the df by rank\n",
    "    df_all_ukbb_scores['rank'] = df_all_ukbb_scores[f'score_{var}_ch'].rank(method='first', ascending=False)\n",
    "    df_all_ukbb_scores_sort = df_all_ukbb_scores.sort_values(by = 'rank')\n",
    "\n",
    "    # split people into 2 groups for each variant (how strongly they bind)\n",
    "    df_all_ukbb_scores_sort['groups_2'] = split_equal_groups_by_varscore(df_all_ukbb_scores_sort, num_groups = 2, var = var)\n",
    "\n",
    "    # okay now we need to add a column to indicate if that person has CH driven by this variant\n",
    "    # identify IDs of people with the variant\n",
    "    ids_var = batch_gene_vars[batch_gene_vars['gene_var']==var].Person_ID.tolist()\n",
    "    df_all_ukbb_scores_sort[f'{var}_positive'] = df_all_ukbb_scores_sort.Person_ID.isin(ids_var) # added True if someone has the variant, False otherwise \n",
    "\n",
    "    # calculate % of people in each of the 5 or 10 groups (as % of total CH positive cases)\n",
    "    counts2 = df_all_ukbb_scores_sort.groupby('groups_2')[f'{var}_positive'].sum().reset_index()\n",
    "    counts2['percentage_all_cases'] = counts2[f'{var}_positive'] / counts2[f'{var}_positive'].sum() * 100\n",
    "    \n",
    "    # add data to the dictionary \n",
    "    scores_2groups[var] = counts2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# get dfs\n",
    "scores_2groups_df = pd.DataFrame()\n",
    "\n",
    "for key, value in scores_2groups.items():\n",
    "    temp_df = scores_2groups[key]\n",
    "    temp_df['gene_var'] = key\n",
    "    temp_df.rename(columns = {f'{key}_positive':'CH_variant_positive'}, inplace = True)\n",
    "    scores_2groups_df = pd.concat([scores_2groups_df, temp_df], axis=0)\n",
    "\n",
    "genevar_counts = pd.DataFrame(batch_gene_vars.gene_var.value_counts().reset_index())\n",
    "\n",
    "# I think one way we can analyse this is for each variant, plot the difference between top and bottom group, normalized by the total nr of people\n",
    "scores_2groups_counts = pd.merge(scores_2groups_df, genevar_counts, on = 'gene_var')\n",
    "\n",
    "# Group by Category and apply the custom calculation function\n",
    "result_2groups1 = pd.DataFrame(scores_2groups_counts.groupby('gene_var').apply(percentage_difference, nr_groups = 2))\n",
    "result_2groups2 = pd.DataFrame(scores_2groups_counts.groupby('gene_var').apply(normalized_difference, nr_groups = 2))\n",
    "result_2groups = pd.merge(result_2groups1, result_2groups2, on = 'gene_var')\n",
    "result_2groups = result_2groups.groupby('gene_var').first().reset_index()\n",
    "result_2groups.columns.values[1] = 'percentage_diff'\n",
    "result_2groups.columns.values[2] = 'norm_diff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "order_percent2 = result_2groups.sort_values(by = 'percentage_diff').gene_var.tolist()\n",
    "order_norm2 = result_2groups.sort_values(by = 'norm_diff').gene_var.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# each group had 45237 observations (we have 90474 people in total and we split them into 2)\n",
    "# note group 1 has 38599 ppl and group 2 has 38598 ppl but this is negligible difference here \n",
    "scores_2groups_df['CH_variant_negative'] = (int(df_hla2_drpq_screened.shape[0]/2) - scores_2groups_df['CH_variant_positive']).astype(int)\n",
    "scores_2groups_df['total'] = int(df_hla2_drpq_screened.shape[0]/2)\n",
    "scores_2groups_df['Group'] = scores_2groups_df['groups_2'].map({1: 'better presenters', 2: 'worse presenters'})\n",
    "scores_2groups_df['gene_var2'] = scores_2groups_df['gene_var'].str.replace('_', '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simpler plot \n",
    "order_percent2 = result_2groups.sort_values(by = 'percentage_diff').gene_var.tolist()\n",
    "order_norm2 = result_2groups.sort_values(by = 'norm_diff').gene_var.tolist()\n",
    "order_percent2 = [x.replace('_', '\\n') for x in order_percent2]\n",
    "order_norm2 = [x.replace('_', '\\n') for x in order_norm2]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "order_by_total = scores_2groups_df.groupby('gene_var2')['CH_variant_positive'].sum().reset_index()\n",
    "order_by_total = order_by_total.sort_values(by = 'CH_variant_positive', ascending = False).gene_var2.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test 4: split by absolute threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>score</th>\n",
       "      <th>gene_var</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "      <td>-0.607455</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>-0.607455</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>-0.397940</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>-0.301030</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>-0.652246</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77192</th>\n",
       "      <td>378301</td>\n",
       "      <td>-0.155336</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77193</th>\n",
       "      <td>378303</td>\n",
       "      <td>-1.097951</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77194</th>\n",
       "      <td>378305</td>\n",
       "      <td>-0.705008</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77195</th>\n",
       "      <td>378311</td>\n",
       "      <td>-0.155336</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77196</th>\n",
       "      <td>378316</td>\n",
       "      <td>-0.584331</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1543940 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index     score      gene_var\n",
       "0           6 -0.607455  DNMT3A_R882H\n",
       "1           7 -0.607455  DNMT3A_R882H\n",
       "2           8 -0.397940  DNMT3A_R882H\n",
       "3          12 -0.301030  DNMT3A_R882H\n",
       "4          13 -0.652246  DNMT3A_R882H\n",
       "...       ...       ...           ...\n",
       "77192  378301 -0.155336   SF3B1_K666N\n",
       "77193  378303 -1.097951   SF3B1_K666N\n",
       "77194  378305 -0.705008   SF3B1_K666N\n",
       "77195  378311 -0.155336   SF3B1_K666N\n",
       "77196  378316 -0.584331   SF3B1_K666N\n",
       "\n",
       "[1543940 rows x 3 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a dataframe with scores for everyone in the UKBB \n",
    "scores_each_var_df = pd.DataFrame()\n",
    "\n",
    "for key, value in scores_for_each_var_ukbb.items():\n",
    "    temp_df = scores_for_each_var_ukbb[key].reset_index()\n",
    "    temp_df['gene_var'] = key\n",
    "    temp_df.rename(columns = {f'score_{key}_ch':'score'}, inplace = True)\n",
    "    scores_each_var_df = pd.concat([scores_each_var_df, temp_df], axis=0)\n",
    "\n",
    "scores_each_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SF3B1\\nK700E',\n",
       " 'DNMT3A\\nR729W',\n",
       " 'SRSF2\\nP95R',\n",
       " 'SRSF2\\nP95H',\n",
       " 'TP53\\nR248Q',\n",
       " 'DNMT3A\\nR736H',\n",
       " 'JAK2\\nV617F',\n",
       " 'DNMT3A\\nR882H',\n",
       " 'IDH2\\nR140Q',\n",
       " 'DNMT3A\\nY735C',\n",
       " 'SF3B1\\nK666N',\n",
       " 'GNB1\\nK57E',\n",
       " 'DNMT3A\\nP904L',\n",
       " 'DNMT3A\\nR736C',\n",
       " 'DNMT3A\\nR771*',\n",
       " 'DNMT3A\\nR326C',\n",
       " 'DNMT3A\\nR882C',\n",
       " 'TP53\\nR175H',\n",
       " 'DNMT3A\\nR320*',\n",
       " 'DNMT3A\\nR598*']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop rows with non-numeric score columns\n",
    "scores_each_var_df['score'] = pd.to_numeric(scores_each_var_df['score'], errors='coerce')\n",
    "scores_each_var_df = scores_each_var_df.dropna(subset=['score'])\n",
    "scores_each_var_df['gene_var2'] = scores_each_var_df['gene_var'].str.replace('_', '\\n')\n",
    "\n",
    "# specify order (for all)\n",
    "scores_each_var_df['median_score'] = scores_each_var_df.groupby('gene_var2').score.transform('median')\n",
    "order_all = scores_each_var_df.sort_values(by = 'median_score', ascending = False).gene_var2.unique().tolist()\n",
    "order_all\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_var2</th>\n",
       "      <th>no binding</th>\n",
       "      <th>weak binding</th>\n",
       "      <th>strong binding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNMT3A\\nP904L</td>\n",
       "      <td>40488</td>\n",
       "      <td>36709</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNMT3A\\nR320*</td>\n",
       "      <td>73302</td>\n",
       "      <td>3895</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNMT3A\\nR326C</td>\n",
       "      <td>39445</td>\n",
       "      <td>37752</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNMT3A\\nR598*</td>\n",
       "      <td>77197</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNMT3A\\nR729W</td>\n",
       "      <td>7685</td>\n",
       "      <td>15749</td>\n",
       "      <td>53763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DNMT3A\\nR736C</td>\n",
       "      <td>53292</td>\n",
       "      <td>22558</td>\n",
       "      <td>1347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DNMT3A\\nR736H</td>\n",
       "      <td>1628</td>\n",
       "      <td>51598</td>\n",
       "      <td>23971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DNMT3A\\nR771*</td>\n",
       "      <td>44615</td>\n",
       "      <td>32582</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DNMT3A\\nR882C</td>\n",
       "      <td>75653</td>\n",
       "      <td>1544</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>409</td>\n",
       "      <td>76463</td>\n",
       "      <td>325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>DNMT3A\\nY735C</td>\n",
       "      <td>10864</td>\n",
       "      <td>42432</td>\n",
       "      <td>23901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GNB1\\nK57E</td>\n",
       "      <td>16924</td>\n",
       "      <td>59203</td>\n",
       "      <td>1070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>IDH2\\nR140Q</td>\n",
       "      <td>16000</td>\n",
       "      <td>48301</td>\n",
       "      <td>12896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>JAK2\\nV617F</td>\n",
       "      <td>12862</td>\n",
       "      <td>64335</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SF3B1\\nK666N</td>\n",
       "      <td>33250</td>\n",
       "      <td>39226</td>\n",
       "      <td>4721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SF3B1\\nK700E</td>\n",
       "      <td>9</td>\n",
       "      <td>7831</td>\n",
       "      <td>69357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SRSF2\\nP95H</td>\n",
       "      <td>4131</td>\n",
       "      <td>27722</td>\n",
       "      <td>45344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRSF2\\nP95R</td>\n",
       "      <td>4032</td>\n",
       "      <td>27821</td>\n",
       "      <td>45344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TP53\\nR175H</td>\n",
       "      <td>60255</td>\n",
       "      <td>14672</td>\n",
       "      <td>2270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TP53\\nR248Q</td>\n",
       "      <td>1259</td>\n",
       "      <td>31846</td>\n",
       "      <td>44092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_var2  no binding  weak binding  strong binding\n",
       "0   DNMT3A\\nP904L       40488         36709               0\n",
       "1   DNMT3A\\nR320*       73302          3895               0\n",
       "2   DNMT3A\\nR326C       39445         37752               0\n",
       "3   DNMT3A\\nR598*       77197             0               0\n",
       "4   DNMT3A\\nR729W        7685         15749           53763\n",
       "5   DNMT3A\\nR736C       53292         22558            1347\n",
       "6   DNMT3A\\nR736H        1628         51598           23971\n",
       "7   DNMT3A\\nR771*       44615         32582               0\n",
       "8   DNMT3A\\nR882C       75653          1544               0\n",
       "9   DNMT3A\\nR882H         409         76463             325\n",
       "10  DNMT3A\\nY735C       10864         42432           23901\n",
       "11     GNB1\\nK57E       16924         59203            1070\n",
       "12    IDH2\\nR140Q       16000         48301           12896\n",
       "13    JAK2\\nV617F       12862         64335               0\n",
       "14   SF3B1\\nK666N       33250         39226            4721\n",
       "15   SF3B1\\nK700E           9          7831           69357\n",
       "16    SRSF2\\nP95H        4131         27722           45344\n",
       "17    SRSF2\\nP95R        4032         27821           45344\n",
       "18    TP53\\nR175H       60255         14672            2270\n",
       "19    TP53\\nR248Q        1259         31846           44092"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# okay now for each variant, let's first calculate people at each category \n",
    "bins = [-float('inf'), -1*np.log10(5), -1*np.log10(1), float('inf')]\n",
    "labels = ['no binding', 'weak binding', 'strong binding']\n",
    "\n",
    "# Categorize scores into bins\n",
    "scores_each_var_df['score_bin'] = pd.cut(scores_each_var_df['score'], bins=bins, labels=labels)\n",
    "\n",
    "# Group by category and score bins, then count observations\n",
    "grouped_df = scores_each_var_df.groupby(['gene_var2', 'score_bin']).size().unstack(fill_value=0)\n",
    "\n",
    "# Create a new DataFrame with desired columns\n",
    "scores_each_var_df_bins = grouped_df.reset_index()\n",
    "scores_each_var_df_bins.columns.name = None  # Remove the name of columns index\n",
    "scores_each_var_df_bins.columns = ['gene_var2', 'no binding', 'weak binding', 'strong binding']\n",
    "\n",
    "scores_each_var_df_bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_var2</th>\n",
       "      <th>no binding_population</th>\n",
       "      <th>weak binding_population</th>\n",
       "      <th>strong binding_population</th>\n",
       "      <th>no binding_carriers</th>\n",
       "      <th>weak binding_carriers</th>\n",
       "      <th>strong binding_carriers</th>\n",
       "      <th>no binding_noncarriers</th>\n",
       "      <th>weak binding_noncarriers</th>\n",
       "      <th>strong binding_noncarriers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNMT3A\\nP904L</td>\n",
       "      <td>40488</td>\n",
       "      <td>36709</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>40440</td>\n",
       "      <td>36669</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNMT3A\\nR320*</td>\n",
       "      <td>73302</td>\n",
       "      <td>3895</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>73244</td>\n",
       "      <td>3894</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNMT3A\\nR326C</td>\n",
       "      <td>39445</td>\n",
       "      <td>37752</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>39408</td>\n",
       "      <td>37714</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNMT3A\\nR598*</td>\n",
       "      <td>77197</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>77133</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNMT3A\\nR729W</td>\n",
       "      <td>7685</td>\n",
       "      <td>15749</td>\n",
       "      <td>53763</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>36</td>\n",
       "      <td>7684</td>\n",
       "      <td>15737</td>\n",
       "      <td>53727</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_var2  no binding_population  weak binding_population  \\\n",
       "0  DNMT3A\\nP904L                  40488                    36709   \n",
       "1  DNMT3A\\nR320*                  73302                     3895   \n",
       "2  DNMT3A\\nR326C                  39445                    37752   \n",
       "3  DNMT3A\\nR598*                  77197                        0   \n",
       "4  DNMT3A\\nR729W                   7685                    15749   \n",
       "\n",
       "   strong binding_population  no binding_carriers  weak binding_carriers  \\\n",
       "0                          0                   48                     40   \n",
       "1                          0                   58                      1   \n",
       "2                          0                   37                     38   \n",
       "3                          0                   64                      0   \n",
       "4                      53763                    1                     12   \n",
       "\n",
       "   strong binding_carriers  no binding_noncarriers  weak binding_noncarriers  \\\n",
       "0                        0                   40440                     36669   \n",
       "1                        0                   73244                      3894   \n",
       "2                        0                   39408                     37714   \n",
       "3                        0                   77133                         0   \n",
       "4                       36                    7684                     15737   \n",
       "\n",
       "   strong binding_noncarriers  \n",
       "0                           0  \n",
       "1                           0  \n",
       "2                           0  \n",
       "3                           0  \n",
       "4                       53727  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# make sure you are using the same bins and labels\n",
    "bins = [-float('inf'), -1*np.log10(5), -1*np.log10(1), float('inf')]\n",
    "labels = ['no binding', 'weak binding', 'strong binding']\n",
    "\n",
    "# okay, now we are doing the same thing for carriers \n",
    "ch_hla_merge_sub_all['gene_var2'] = ch_hla_merge_sub_all['gene_var'].str.replace('_', '\\n')\n",
    "scores_each_var_df_carriers = ch_hla_merge_sub_all[['Person_ID', 'gene_var2', 'log_score']]\n",
    "scores_each_var_df_carriers.rename(columns = {'log_score':'score'}, inplace = True)\n",
    "\n",
    "# add score bins\n",
    "scores_each_var_df_carriers['score_bin'] = pd.cut(scores_each_var_df_carriers['score'], bins=bins, labels=labels)\n",
    "\n",
    "grouped_df_carriers = scores_each_var_df_carriers.groupby(['gene_var2', 'score_bin']).size().unstack(fill_value=0)\n",
    "scores_each_var_df_bins_carriers = grouped_df_carriers.reset_index()\n",
    "scores_each_var_df_bins_carriers.columns.name = None  # Remove the name of columns index\n",
    "scores_each_var_df_bins_carriers.columns = ['gene_var2', 'no binding', 'weak binding', 'strong binding']\n",
    "\n",
    "# merge all together and calculate values for non-carriers\n",
    "merged_df = scores_each_var_df_bins.merge(scores_each_var_df_bins_carriers, on='gene_var2', suffixes=('_population', '_carriers'))\n",
    "merged_df['no binding_noncarriers'] = merged_df['no binding_population'] - merged_df['no binding_carriers']\n",
    "merged_df['weak binding_noncarriers'] = merged_df['weak binding_population'] - merged_df['weak binding_carriers']\n",
    "merged_df['strong binding_noncarriers'] = merged_df['strong binding_population'] - merged_df['strong binding_carriers']\n",
    "\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# do mann-whitney test to compare population (non-carriers) and carriers scores \n",
    "pvalues_mwu = {}\n",
    "cat_series = pd.Series(['Carrier', 'Non-carrier'])\n",
    "df_scores_all_plot = pd.DataFrame()\n",
    "\n",
    "for var in order_all:\n",
    "    carrier_scores = scores_each_var_df_carriers[scores_each_var_df_carriers['gene_var2']==var].score.tolist()\n",
    "    all_scores = scores_each_var_df[scores_each_var_df['gene_var2']==var].score.tolist()\n",
    "\n",
    "    # Remove elements from x only once if they appear once in y\n",
    "    from collections import Counter \n",
    "    counts = Counter(carrier_scores)\n",
    "\n",
    "    noncarrier_scores = []\n",
    "    for x in all_scores:\n",
    "        if counts[x]:\n",
    "            counts[x] -= 1\n",
    "        else:\n",
    "            noncarrier_scores.append(x)\n",
    "\n",
    "    scores_series = pd.Series(carrier_scores + noncarrier_scores).reset_index(drop=True)\n",
    "    repetitions = [len(carrier_scores), len(noncarrier_scores)]\n",
    "    cat_series_rep = pd.Series(cat_series.repeat(repetitions)).reset_index(drop=True)\n",
    "    df_x = pd.DataFrame({'score': scores_series, 'CH status': cat_series_rep})\n",
    "    df_x['gene_var2'] = var\n",
    "\n",
    "    df_scores_all_plot = pd.concat([df_scores_all_plot, df_x], axis = 0)\n",
    "\n",
    "    statistic, p_value = mannwhitneyu(carrier_scores, noncarrier_scores)\n",
    "    pvalues_mwu[var] = p_value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do a nice visualisation to compare these \n",
    "# the easiest I can think about is doing % of people in each group\n",
    "# and maybe having carriers / non-carriers side by side or sth\n",
    "merged_df['no binding_carriers_percent'] = merged_df['no binding_carriers'] / (merged_df['no binding_carriers'] + merged_df['weak binding_carriers'] + merged_df['strong binding_carriers']) * 100\n",
    "merged_df['weak binding_carriers_percent'] = merged_df['weak binding_carriers'] / (merged_df['no binding_carriers'] + merged_df['weak binding_carriers'] + merged_df['strong binding_carriers']) * 100\n",
    "merged_df['strong binding_carriers_percent'] = merged_df['strong binding_carriers'] / (merged_df['no binding_carriers'] + merged_df['weak binding_carriers'] + merged_df['strong binding_carriers']) * 100\n",
    "\n",
    "merged_df['no binding_noncarriers_percent'] = merged_df['no binding_noncarriers'] / (merged_df['no binding_noncarriers'] + merged_df['weak binding_noncarriers'] + merged_df['strong binding_noncarriers']) * 100\n",
    "merged_df['weak binding_noncarriers_percent'] = merged_df['weak binding_noncarriers'] / (merged_df['no binding_noncarriers'] + merged_df['weak binding_noncarriers'] + merged_df['strong binding_noncarriers']) * 100\n",
    "merged_df['strong binding_noncarriers_percent'] = merged_df['strong binding_noncarriers'] / (merged_df['no binding_noncarriers'] + merged_df['weak binding_noncarriers'] + merged_df['strong binding_noncarriers']) * 100\n",
    "\n",
    "merged_df_percents = merged_df[['gene_var2', 'no binding_carriers_percent', 'weak binding_carriers_percent', 'strong binding_carriers_percent',\n",
    "                                'no binding_noncarriers_percent', 'weak binding_noncarriers_percent', 'strong binding_noncarriers_percent']]\n",
    "\n",
    "merged_df_percents_melt = pd.melt(merged_df_percents, id_vars = 'gene_var2')\n",
    "merged_df_percents_melt[['binding level', 'group', 'drop']] = merged_df_percents_melt['variable'].str.split('_', expand = True)\n",
    "merged_df_percents_melt['CH status'] = merged_df_percents_melt['group'].map({'carriers' : 'Carrier', 'noncarriers': 'Non-carrier'})\n",
    "\n",
    "# Prepare df\n",
    "stacked_df_carriers = merged_df_percents_melt[merged_df_percents_melt['group'] == 'carriers'].pivot_table(\n",
    "    index='gene_var2', columns='binding level', values='value', aggfunc='sum').reset_index()\n",
    "stacked_df_noncarriers = merged_df_percents_melt[merged_df_percents_melt['group'] == 'noncarriers'].pivot_table(\n",
    "    index='gene_var2', columns='binding level', values='value', aggfunc='sum').reset_index()\n",
    "bar_width = 0.4\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABVsAAAH5CAYAAACfwu6KAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/OQEPoAAAACXBIWXMAAA9hAAAPYQGoP6dpAADbQ0lEQVR4nOzdd3hT5f//8VfSQjer7CF7ypC9ZCMIMmXvPRwfpuw9lI2AoqAoQ1REkCVDUEBEQRkyBUFRluw9C23v3x/+yNfSUtL2pE2b5+O6cl3knDv3eb+5c5L0nTv3sRljjAAAAAAAAAAAcWJP6AAAAAAAAAAAICmg2AoAAAAAAAAAFqDYCgAAAAAAAAAWoNgKAAAAAAAAABag2AoAAAAAAAAAFqDYCgAAAAAAAAAWoNgKAAAAAAAAABag2AoAAAAAAAAAFqDYCgAAAAAAAAAWoNgKAAAAS+TIkUM2m002m03r1q2Ltm3Tpk0dbUePHh1h39atW2Wz2ZQnT56nHvNRH2fOnHlim+PHj2vAgAEqVaqUgoODlSxZMqVOnVqlS5dWr1699N1338kY41SOUsQ8/3v7bx6jR4+Osk2OHDmcPg4AAAASH++EDgAAAABJz5tvvqm6detGue/333/XihUrXB5DWFiYBg8erLffflthYWFKmTKlihUrpoCAAF2+fFkHDhzQ7t279c4776ho0aLav3+/U/3WqlVLFy9e1E8//aRLly6pePHieuaZZ1SgQAFHmwIFCqhhw4Y6c+aM9uzZo+DgYD3//PNKnz69q9IFAACAG6DYCgAAAEvlzZtXP/30kzZv3qzq1atH2j958mSFh4crb968On78uMvi6NChgz799FOlSJFCM2fOVLt27eTl5eXYf+/ePX311VcaP368Tp486XS/H3zwgSSpZs2a+u6779SrVy917NgxQpuWLVuqZcuWWrx4sdq1a6fChQtr5cqVVqQFAAAAN8YyAgAAALDUwIEDJUnjxo2LtO/s2bNavHixypcvr8qVK7sshoULF+rTTz+Vj4+PNm/erI4dO0YotEqSn5+f2rRpo3379ql///4uiwUAAACeg2IrAAAALFWzZk2VKVNGW7du1Y8//hhh37Rp0/TgwQMNGzbMpTGMHz9ekvTGG2+oZMmS0bb18fHRiBEjXBoPAAAAPAPFVgAAAFhuyJAhkiLObr169ao+/PBDPffcc3rppZdcduzDhw/rjz/+kCT16NHDZccBAAAAHsearQAAALBcw4YN9eyzz+qbb77R7t27VapUKb3zzju6ffu2hg4d6lQf58+fV6NGjWJ87F27dkmSnnnmGWXLli3GjwcAAABii2IrAAAALGez2TR48GC1a9dO48eP12effaZ33nlH+fPnV5MmTZzq486dO1q1alWMj33x4kVJUsaMGWP8WAAAACAuWEYAAAAALtGqVSvlzJlTq1evVq9evXTlyhUNGTJEdrtzH0Fz584tY0y0t+jYbDYr0gAAAACcRrEVAAAALuHl5aWBAwfKGKOPPvpIOXLkUJs2bVx+3LRp00r6dxkCAAAAID5RbAUAAIDLdOrUSZkyZZIkDRw4UN7erl/FqlixYpKkkydP6syZMy4/HgAAAPAIxVYAAAC4jI+Pj9555x299tpr6ty5c7wcs3jx4sqQIYMk6b333ouXYwIAAAASxVYAAAC4WJMmTfTuu+/Kx8cnXo5nt9s1YMAASdL06dP1yy+/RNs+NDRUn3zySXyEBgAAgCSOYisAAACSnN69e6t69eoKCQlRjRo19P777yskJCRCm5CQEC1fvlylSpXStGnTEihSAAAAJCUUWwEAAJDkeHt7a9WqVWrRooVu376tV199VcHBwapUqZJeeukllS1bVqlTp1bTpk21f/9+5cyZ0+m+u3fvrkaNGunAgQOSpFmzZqlRo0ZasmSJo82SJUvUqFEjzZgxQ5J06NAhNWrUSN27d7c0TwAAALgX11+hAAAAAEgAgYGBWrJkiXr37q358+frxx9/1IEDB3T37l2lTp1axYsXV6VKlfTyyy+rTJkyTve7ceNGnTx50nH/119/1a+//qrnnnvOse3o0aNatWqV4/6VK1e0atUqZc+e3ZLcAAAA4J5sxhiT0EEAAAAAAAAAQGLHMgIAAAAAAAAAYAG3L7YOGTJEgYGBT9x/7NgxNW3aVGnTppWvr6+effZZzZw5U+Hh4ZHahoSEaPTo0cqTJ498fHyUMWNGderUSefOnXNlCgAAAAAAAAA8gFsuI3D9+nWdOnVKmzdv1uDBg+Xt7a3bt29Hanfy5EmVLFlSktSjRw+lS5dOGzdu1Pr16zVw4EBNmjQpQvtmzZpp2bJlatWqlcqWLauTJ09q7ty5ypw5s/bu3augoKB4yQ8AAAAAAABA0uOWxdbRo0drzJgxjvsBAQFRFlu7du2q+fPna+/evSpWrJhj+8svv6w1a9boxIkTypYtmyRp+/btqlSpkvr166dp06Y52q5evVoNGzbUW2+9pSFDhrgwKwAAAAAAAABJmVsuI9C6dWutWbNGa9asUeHChaNsY4zRypUrVa5cuQiFVknq2bOnQkNDI1wBdsWKFY59/9WgQQNlzpxZy5cvtzgLAAAAAAAAAJ7ELYut+fLlU7169VSvXj0FBwdH2ebs2bO6cuWKChUqFGlfkSJFJEn79+93bNu/f7+SJUumPHnyRGpfuHBhHTp0KMp1XgEAAAAAAADAGd4JHUBsPbqoVcqUKSPte1Sg/e+Fr86dO6cUKVLIZrNF2T4kJERXr15V2rRpozxeSEiIQkJCHPfDw8N19epVBQcHR9knAAAAAAAAgCczxujWrVvKnDmz7Ha3nBMaY4m22Hr//n1Jkr+/f6R9yZIlkyTdu3cvQvuo2j6p/eMmTJgQYR1ZAAAAAAAAAHF3+vRpZc2aNaHDsESiLbb6+vpKksLCwiLte/jwoSTJz88vQvtHBVpn2j9uyJAh6tevn+P+jRs39Mwzz+j06dNKkSJFzBNwwqK8MXuStT9+xvHv/e++HaPHFnu9b4zau4Kn5St5Xs6elq/keTl7Wr6SdP+5HDFq77vvb8e/Q+fMiNFjvXv2iVF7V4jLGCfGfCXPe157Wr6S5+XsaflKvFY/DWM8I0aPdYcx9rR8Jc/LmfP46RL7GBccsy9G7Y+Mes7x75FrTsfosWPrZ4tR+/h08+ZNZcuWTUFBQQkdimUSbbE1Q4YMkqRr165F2nf58uUIbR79+++//5YxJtLP/i9fvqzkyZMrderUTzyej4+PfHx8Im1PkSKFy4qtfvaYLU/w3zgC/38xOjaPTSielq/keTl7Wr6S5+XsaflKUvIY/tTF9z9xh8YwZ283yDkuY5wY85U873ntaflKnpezp+Ur8Vr9NIxx4htjT8tX8rycOY+fLrGPsd0nMEbt/ztOPv4xK0y6yxhHJykt0Zloi63ZsmVT6tSptXPnzkj7Hm0rWrSoY1vRokW1ZcsWHTx4MML28PBw7dq1S4UKFZKXl5frAwcAIBFrE5AsoUMAAAAAALeVaFeetdlsatCggfbt26eDBw9G2Pfpp5/Ky8tLDRo0cGxr1KiRJGnRokUR2q5du1bXr19X48aNXR4zAAAAkJgV8bbF6AYAAOBp3HJm62+//abffvtNknTp0iWFhYVp2bJlkqQcOXKoVKlSkqSRI0dq1apVeuGFF9StWzelTZtWmzZt0tq1a9WnTx/lzJnT0WfVqlXVoEEDTZ8+XZcuXVLJkiV16tQpzZkzRzly5FCvXr3iP1EAAAAAsFhcfoVAkRzuKm/zjTFqH7MVLd2PJ/6ayNPGGEmXWxZbly5dqjFjxkTY1qxZM0lShw4dtGDBAklSrly59OOPP2rYsGGaPXu27ty5o9y5c2vatGnq06dPpH6XLFmiN998U5999pk+//xzpU6dWk2aNNFbb72lVKlSuTgrAAAAAAAQ34Zlaxmj9pNcFAcAz+CWxdbRo0dr9OjRTrUtVKiQVqxY4VRbPz8/jR8/XuPHj49DdAAAAAAAAAAQWaJdsxUAAAAAAAAA3IlbzmwFAAAAAABAzLH2MpCwmNkKAAAAAAAAABag2AoAAAAAAAAAFqDYCgAAAAAAAAAWYM1WAAAAwIO0CUiW0CEAAAAkWcxsBQAAAAAAAAALMLMVAAAAAAAAidawbC1j1H6Si+IAJGa2AgAAAAAAAIAlKLYCAAAAAAAAgAVYRgAAAACIpSLetoQOAQAAAG6Ema0AAAAAAAAAYAGKrQAAAAAAAABgAYqtAAAAAAAAAGABiq0AAAAAAAAAYAEukAUAAACP1iYgWUKHAAAAgCSCma0AAAAAAAAAYAGKrQAAAAAAAABgAYqtAAAAAAAAAGABiq0AAAAAAAAAYAGKrQAAAAAAAABgAYqtAAAAAAAAAGABiq0AAAAAAAAAYAGKrQAAAAAAAABgAYqtAAAAAAAAAGABiq0AAAAAAAAAYAGKrQAAAAAAAABgAe+EDgAAAAAAAAD/J2/zjTFqf9pFcQCIOWa2AgAAAAAAAIAFKLYCAAAAAAAAgAUotgIAAAAAAACABSi2AgAAAAAAAIAFuEAWAABANNoEJEvoEAAAAAAkEsxsBQAAAAAAAAALUGwFAAAAAAAAAAtQbAUAAAAAAAAAC1BsBQAAAAAAAAALUGwFAAAAAAAAAAtQbAUAAAAAAAAAC1BsBQAAAAAAAAALUGwFAAAAAAAAAAtQbAUAAAAAAAAAC1BsBQAAAAAAAAALeCd0AAAAAElVEW9bQocAAAAAIB4xsxUAAAAAAAAALECxFQAAAAAAAAAsQLEVAAAAAAAAACxAsRUAAAAAAAAALJDoi613797VpEmTVLRoUQUGBipjxoyqUaOG1q5dG6ntsWPH1LRpU6VNm1a+vr569tlnNXPmTIWHhydA5AAAAAAAAACSEu+EDiAuQkJCVLlyZe3bt08tWrRQx44ddePGDX3++eeqV6+eJk6cqEGDBkmSTp48qQoVKkiSevTooXTp0mnjxo3q06eP/vnnH02aNCkhUwEAAAAAAACQyCXqYuvSpUu1Z88eTZs2Tf369XNsHzhwoIoWLarx48frjTfekJeXl8aNG6dr165p7969KlasmCSpT58+evnllzV9+nS9/vrrypYtW0KlAgAAAAAAACCRS9TLCJw7d06SVLhw4QjbAwICVLJkSd29e1ehoaEyxmjlypUqV66co9D6SM+ePRUaGqpVq1bFW9wAAAAAAAAAkp5EXWytUaOGbDabBg8erD///NOx/c6dO/rxxx9Vv359+fj46OzZs7py5YoKFSoUqY8iRYpIkvbv3x/tsUJCQnTz5s0INwAAAAAAAAB4JFEXW0uWLKlPPvlEJ06cUP78+dWoUSPNnz9fdevWVZ48efTRRx9J+r8ZsClTpozUR3BwcIQ2TzJhwgSlTJnScWPJAQAAAAAAAAD/laiLrZKUNm1apU2bVu3bt9fVq1fVtWtX/fDDDypXrpz8/f0lSffv35ckx/3/SpYsmSTp3r170R5nyJAhunHjhuN2+vRpizMBAAAAAAAAkJgl6gtkbdmyRfXq1dPs2bPVvXt3SdLZs2c1ZcoUTZkyRb/88os2b94sX19fSVJYWFikPh4+fChJ8vPzi/ZYPj4+8vHxsTgDAAAAAAAAAElFoi62TpgwQalSpVK3bt0c27JkyaIZM2bIGKNZs2bphx9+UI4cOSRJ165di9TH5cuXJUkZMmSIl5gBAAAAAAAAJE2JehmB06dPy263y2azRdqXO3duSdL58+eVLVs2pU6dWjt37ozU7tG2okWLujZYAAAAAAAAAElaoi62lihRQhcvXtQnn3wSYXtoaKiWLl0qLy8vlS5dWjabTQ0aNNC+fft08ODBCG0//fRTeXl5qUGDBvEZOgAAAAAAAIAkJlEvIzB27Fht2rRJHTp00OrVq1WuXDndunVLX375pX777Te9+eabypkzpyRp5MiRWrVqlV544QV169ZNadOm1aZNm7R27Vr16dPH0Q4AAAAAACCxGpatZYzaT3JRHICnStTF1ty5c2vv3r0aN26cNmzYoJUrV8rf31+lS5fW5MmT9dJLLzna5sqVSz/++KOGDRum2bNn686dO8qdO7emTZumPn36JFwSAAAAAAAAAJKERF1slaSsWbNq7ty5TrUtVKiQVqxY4eKIAAAAAAAAAHiiRL1mKwAAAAAAAAC4C4qtAAAAAAAAAGABiq0AAAAAAAAAYAGKrQAAAAAAAABgAYqtAAAAAAAAAGABiq0AAAAAAAAAYAGKrQAAAAAAAABgAYqtAAAAAAAAAGABiq0AAAAAAAAAYAHvuHZw5coVXbp0SdevX1eqVKmUPn16pUmTxorYAAAAAAAAACDRiHGxNSQkREuWLNG6dev07bff6vr165HapEqVSi+88ILq1q2rFi1ayMfHx4pYAQAAAAAAAMBtOV1sDQsL07vvvqvJkyfr3Llz8vX1VaVKlVS8eHFlyZJFAQEBun37ts6cOaNff/1VK1eu1JdffqkhQ4Zo0KBBeu211+Tl5eXKXAAAAAAAAAAgwThVbP3jjz/UunVr7d69W8WKFdOkSZPUpEkT+fn5PfExt2/f1uLFizV16lT17dtXn376qT777DPlzp3bsuABAAAAAAAAwF04dYGs4sWL69ixY/rwww+1d+9etW3bNtpCqyQFBgaqZ8+eOnz4sIYPH64DBw6oRIkSlgQNAAAAAAAAAO7GqWJrcHCwdu7cqS5dushms8XoAD4+PhozZoy2b9+uwMDAWAUJAAAAAAAAAO7OqWUEdu7cqYwZM8bpQCVLltQPP/wQpz4AAAAAAAAAwF05NbM1roXWR3LlymVJPwAAAAAAAADgbpwqtgIAAAAAAAAAoufUMgLRCQ0N1cKFC7V161ZdvHhRfn5+Kly4sDp27Kg8efJYESMAAAAAAAAAuL04FVtDQ0NVpUoV7dy5U8YYx/bVq1dr2rRpWr9+vapWrRrXGAEAAAAAAADA7cVpGYEvv/xSO3bsULdu3fTnn3/q/v37unXrlnbt2qVixYpp4MCBVsUJAAAAAAAAAG4tTsXWw4cPy2azacqUKcqZM6eSJ0+ugIAAlSxZUv369dPBgwetihMAAAAAAAAA3JrTxdZ79+5F2pYyZUpJ0rFjxyLtO378uFKkSBGH0AAAAAAAAAAg8XB6zdbcuXNr1KhR6tq1q7y8vCRJDRo00LBhw1SrVi3Vr19fWbJkUWhoqI4ePaoNGzaoS5cuLgscAAAAAAAAANyJ08XWWrVq6bXXXtO0adM0fvx4NW/eXPnz59eSJUvUq1cvLVq06P869fZWy5YtNW3aNJcEDQAAAAAAAADuxuli64IFC/TGG29o6NChatmypSZPnqwJEybo5ZdfVuPGjfXHH3/o4sWL8vX1Vb58+RQUFOTKuAEASUwRb1tChwAgjjiPgcRvWLaWMWo/yUVxAACQWMXoAlmFCxfW6tWrtX37dvn7++vFF19UjRo1tHv3buXNm1cVK1ZUyZIlKbQCAAAAAAAA8DgxKrY+UqFCBW3btk2rV6/W5cuXVa5cOTVr1izKC2UBAAAAAAAAgCeIVbH1kZdeekn79u3TwoULtXfvXhUuXFjdu3fX2bNnrYoPAAAAAAAAABKFGBVbb9y4offff1/9+/fXq6++qpkzZ+rChQtq27atjh49qqlTp2r16tXKly+fBg0apGvXrrkqbgAAAAAAAABwK05fIOvAgQOqVq2arl+/LmOMY/vo0aP1ww8/qHDhwurVq5e6dOmiqVOnavr06frwww919epVlwQOAAAAAAAAAO7E6Zmt/fr1U0hIiD766CMdPHhQBw8e1Mcff6zQ0FD169fP0S4gIECjRo3SH3/8oQ4dOrgkaAAAAAAAAABwN07PbN27d69atWqljh07OrY9++yz2r59u1asWBGpfbp06fT2229bEiQAAAAAAAAAuDuni60BAQFRLglw/fp1+fn5WRoUACBxahOQLKFDAAAAAAAgwThdbK1Zs6YWLVqkrl27qmTJkpKkPXv2aOXKlWrVqpXLAgQAAAAAAACAxMDpYuv06dN16NAhffzxx/r4448d24sWLarp06e7JDgAAAAAAAAASCycLramTp1au3bt0q+//qpjx45JkvLkyeOY5QoAAAAASHh5m2+MUfvTLooDAABP5HSx9ZHixYurePHirogFAAAAAAAAABIte0IHAAAAAAAAAABJgVMzW2fNmqVevXrF+WDLly9XkyZN4twPAAAAXKNNQLKEDgEAAABItJya2dqnTx8NGjRI4eHhsT7QlClT1LJly1g/HgAAAAAAAADcmVPF1po1a2rKlCkqV66cfvzxxxgdYM+ePSpTpowGDRqk6tWrxypIAAAAAAAAAHB3ThVbN27cqClTpujQoUOqXLmyKlasqLlz5+rkyZNRtr906ZLmzJmj559/XmXKlNGhQ4c0depUffPNN5YGDwAAAAAAAADuwqk1WyWpf//+at26tSZOnKh58+Zpx44dstlsSp06tbJnz67AwEDdvXtX586d07lz52SMkZ+fn/73v/9p0KBBypQpkyvzAAAAAAAAAIAE5XSxVZIyZcqkmTNnasKECdq8ebM2bdqkP/74Q5cuXdK5c+eUKlUqPffcc2ratKleeOEFVatWTf7+/q6KHQAAAAAAAADcRoyKrY/4+/urXr16qlevntXxAB6FKz4DAAAAcZe3+cYYtT/tojgAAHBqzVYAAAAAAAAAQPQotgIAAAAAAACABZJMsXXDhg1q2rSpsmTJouTJkytVqlTatWtXhDbHjh1T06ZNlTZtWvn6+urZZ5/VzJkzFR4enkBRAwAAAAAAAEgqYrVmqzsJDQ1Vly5dtGjRIpUoUULdunVT6tSpdeLECSVL9n/rYZ48eVIVKlSQJPXo0UPp0qXTxo0b1adPH/3zzz+aNGlSQqUAAAAAAAAAIAlI9MXWMWPGaNGiRRo9erRGjRr1xHbjxo3TtWvXtHfvXhUrVkyS1KdPH7388suaPn26Xn/9dWXLli2+wgYAAAAAAACQxCTqZQQuXbqkyZMnq3Llyho5cuQT2xljtHLlSpUrV85RaH2kZ8+eCg0N1apVq1wdLgAAAAAAAIAkLFHPbF2+fLkePHig//3vf7LZbHrw4IHu3Lmj1KlTR2h39uxZXblyRYUKFYrUR5EiRSRJ+/fvj/ZYISEhCgkJcdy/efOmBRkAAAAAAAAASCoS9czW3bt3S5J+/fVXFSpUSH5+fkqTJo1SpEihLl266Nq1a5Kkc+fOSZJSpkwZqY/g4OAIbZ5kwoQJSpkypePGkgMAAAAAAAAA/itOM1tPnjyp7777TufPn1eZMmVUs2ZNSf/OEr1x44bKli0rHx8fSwKNyqMC6bfffqsOHTooV65cunfvnr7++mt9/PHH2rdvn3bu3Kn79+9Lkvz9/SP18egiWvfu3Yv2WEOGDFG/fv0c92/evEnBFQAAAAAAAIBDrIutgwYN0vTp0xUWFiabzab+/fs7iq2//vqrunTpooULF6pt27aWBfu4Bw8eSJJ27twpm83m2N6+fXu9+uqrev/997Vu3TplzpxZkhQWFhapj4cPH0qS/Pz8oj2Wj4+PSwvHAAAAAAAAABK3WC0jMG/ePE2ZMkWVK1fWF198IWNMhP1t27ZVcHCwli5dakmQT5ImTRpJ0p07dyLtq1+/viTpyJEjypAhgyQ5lhX4r8uXL0uSow0AAAAAAAAAxEasiq3vv/++cubMqQ0bNqhZs2aR9nt7e6tatWr65Zdf4hxgdJ577jlJ0oEDByLtu3r1qiQpKChI2bJlU+rUqbVz585I7R5tK1q0qOsCBQAAAAAAAJDkxarYevToUVWtWtWx3mlU0qdP7yh4ukqLFi3k5eWl0aNHKzQ01LE9PDxc8+fPl7e3t+rUqSObzaYGDRpo3759OnjwYIQ+Pv30U3l5ealBgwYujRUAAAAAAABA0harNVu9vLzk5eUVbZu///5bKVKkiFVQzsqVK5fGjh2rYcOGqUSJEmrRooUkaeXKldq9e7cmT56sXLlySZJGjhypVatW6YUXXlC3bt2UNm1abdq0SWvXrlWfPn2UM2dOl8YKAAAAAAAAIGmLVbG1cOHCUf4k/5EzZ85oy5YtKl++fKwDc9bQoUOVK1cuTZ8+XW+++absdrtKliypr7/+Wi+99JKjXa5cufTjjz9q2LBhmj17tu7cuaPcuXNr2rRp6tOnj8vjBAAAAAC4n2HZWsao/SQXxQEASBpiVWzt1q2bunTporfeektDhw6NsO+vv/5S69atde/ePXXs2NGKGJ+qZcuWatny6W+QhQoV0ooVK+IhIgAAAAAAAACeJlZrtnbq1Elt27bV8OHDlSdPHtlsNn355ZcqXry48uXLp59//lmNGzdWmzZtrI4XAAAAAAAAANxSrIqtkrRo0SJ9+OGHCgoKkiSdPHlS+/fvV86cOTV16lQtXbrUsiABAAAAAAAAwN3FahmBR7p06aIuXbro3r17unbtmoKCghzFVwAAAAAAAADwJHEqtj7i5+cnPz8/K7oCAAAAAAAAgEQpVsXWU6dOPbWN3W5XYGCgUqVKFZtDAAAAAAAAAECiEqtia44cOWSz2ZxqGxwcrAYNGmjs2LHKnDlzbA4HAAAAAAAAAG4vVsXWggUL6siRI6pSpYp8fX0j7T906JCuXLmiypUr659//tHHH3+sDRs2aNeuXcqUKVOcgwYAAAAAAAAAdxOrYmvu3Ll19+5dbdmyJcr927ZtU9WqVdWrVy/VrVtX69evV4MGDfTWW2/pnXfeiVPAAAAAAAAAAOCOYlVs3bFjh1566aUn7q9cubIqVaqkKVOmqG7duqpTp45efPFFrV+/PtaBAgAAAEBs5G2+MUbtT7soDgAAkPTZY/OgGzduKGXKlNG2KVq0qPbs2eO4nzt3bp09ezY2hwMAAAAAAAAAtxerma3Zs2fXDz/8EG2bkydPKjw83HH/2rVr8vLyis3hAAAAAJdh1iMAAACsEquZrW3bttW+ffvUp08f3b59O9L+FStWaMOGDSpdurRj24kTJ5QuXbrYRwoAAAAAAAAAbixWM1sHDRqkb775RrNmzdInn3yi0qVLK0uWLLp586YOHz6s33//XT4+Ppo4caIk6c6dO9q9e7dq165tafAAAAAAAAAA4C5iVWz19fXVli1bNG7cOL3//vvauPH/fnpls9lUrVo1TZo0SSVLlpQkJU+eXIcOHVKqVKksCRoAAAAAAAAA3E2siq2S5OPjo/Hjx2vcuHE6duyYrly5Il9fX+XNm1dBQUER2iZLlky5c+eOc7AAAAAAEJ+GZWsZo/aTXBQHAABIHGJdbH3EZrMpf/78VsQCAAAAAAAAAIlWnIut169f14MHD564P3369HE9BAAAAAAAAAC4vVgXWydNmqSpU6fq6tWrT2xjs9kUGhoa20MAAAAAAAAAQKIRq2LrggULNGTIEKVOnVqZM2fW2bNnVaVKFcf+vXv3yhjjuEAWAAAAAABIeKxDDACuZY/Ng+bNm6csWbLojz/+UOfOnWWz2bRp0yZt2bJFW7Zs0dtvv6379+9r5syZVscLAAAAAAAAAG4pVsXWw4cPq3r16kqdOrVSpkwpSbpy5Ypjf+fOnVWwYEFNmTLFmigBAAAAAAAAwM3Fqth6584dpUmTRpIcxdZTp05FaFOxYkVt3bo1btEBAAAAAAAAQCIRq2Jr6tSpde3aNUlS/vz5ZYzRjz/+GKFNSEhItBfPAgAAAAAAAICkJFbF1kyZMun48eOSpLJlyyp16tSaNm2aY3br33//rVWrVilPnjzWRQoAAAAAAAAAbsw7Ng8qXLiwvvzyS92/f1++vr4aMGCAhg4dqnz58ilTpkz6559/9PDhQ82ePdvqeAEAAAAA8Gh5m2+MUfvTLooDABBZrGa2fvLJJ7p79658fX0lSYMHD9aECROUOXNmXbx4UYUKFdKSJUvUokULS4MFAAAAAAAAAHcVq5mtNptNXl5eEbYNGjRIgwYNsiQoAAAAAAAAAEhsYjWztXPnzvryyy+jbTNp0iS9++67sQoKAAAAAAAAABKbWBVbFyxYoF27dkXbZv/+/ZoyZUqsggIAAAAAAACAxCZWxVZnBAQE6OLFi67qHgAAAAAAAADcikuKrUePHtXatWuVJUsWV3QPAAAAAAAAAG7H6Qtk5cqVK8L9Dz/8UMuWLYvU7sGDBzp37pwkaeTIkXEMDwAAAAAAAAASB6eLrf/8849sNpskyWaz6fbt27p//37kDr29lTdvXrVo0ULDhw+3LlIAAAAAAAAAcGNOF1v/W1i12+3q27evJk+e7JKgAAAAAAAAACCxcbrY+l9//fWXUqZMaXUsAAAAAAAAAJBoxarYmj17dqvjAAAAAAAAAIBELVbFVkm6c+eOPvvsMx07dkxXr16VMSZSG5vNpo8++ihOAQIAAAAAAABAYhCrYuvevXtVp04dXb58Ocoi6yMUWwEAAAAAAAB4ilgVW/v166dLly6pe/fuqlOnjoKDg5UsWTKrYwMAAAAAAACARCNWxdZffvlFdevW1Zw5c6yOBwAAAAAAAAASJXtsHuTn56fcuXNbHQsAAAAAAAAAJFqxKrZWrlxZe/bssToWAAAAAAAAAEi0YlVs7du3r3bs2KENGzZYHQ8AAAAAAAAAJEqxWrN1yZIlSpUqlbp376569eo9sZ3NZtPs2bNjHRwAAADgzoZlaxmj9pNcFAcAAADcQ6yKrY8ujHXt2rVoL5JFsRUAYq+Ity2hQwAAAAAAADEQq2Lrli1brI4DAAAAAAAAABK1WBVbq1SpYnUcAAAAAAAAAJCoxeoCWe7uzTfflM1m0+uvvx5p3y+//KIXX3xRqVKlkr+/v0qVKqXFixcnQJQAAAAAAAAAkpJYF1tPnz6trl276plnnlHy5Mk1cOBAx77p06erevXqOnTokCVBxsRnn32mESNGRLlv9+7dqlKlio4cOaJ+/fpp3LhxSp48udq1a6f3338/niMFAAAAAAAAkJTEahmBv//+W2XLltXly5dVoEABnTlzJsL++vXra8CAAVq8eLEmTpxoSaDO+P7779WpUyd17NhR8+fPj7R/8ODBSpYsmX766SdlyZJFktSnTx+VL19ew4YNU4cOHeTv7x9v8QJx4WkXT/K0fAEAAAAAQOITq5mtI0eO1JUrV7RhwwYdPnw40v68efOqTJkyWrduXZwDdNbRo0fVuHFjtW7dWiNHjoy0/9q1a9q6davq16/vKLRKkpeXl7p27apr165p8+bN8RYvAAAAAAAAgKQlVsXWb7/9VnXq1NELL7zwxDbPPfec/vzzz1gHFhMXL15U3bp1VaJECX3wwQdRtjl8+LDCwsJUqFChSPuKFCkiSdq/f/8TjxESEqKbN29GuAEAAAAAAADAI7FaRuDKlSvKmTNntG3u378vY0ysgoqJu3fvqn79+goICNDy5cuVLFmyKNudO3dOkpQyZcpI+4KDgyO0icqECRM0ZswYCyIG4EnaBET9mgQAAAAAAJKeWM1szZAhg86ePRttm927dytbtmyxCspZ4eHhatOmjU6fPq21a9dGWUh95P79+5IU5Zqsjwq09+7de+LjhwwZohs3bjhup0+fjmP0AAAAAAAAAJKSWM1srVWrlj7//HOdPXs2wvqnj3z44Yc6fPiwXn/99TgHGJ0vv/xSa9eu1bJly2S32x0X6jp//rwk6c6dOzpz5oxSpEghX19fSVJYWFikfh4+fChJ8vPze+KxfHx85OPjY3UKAAAAAAAAAJKIWM1sHTFihHx8fFS5cmV98sknkqSzZ89q1apV6tSpk1599VWlSJFC/fv3tzTYx927d08PHz5Uw4YNlS1bNsetfPnykqQFCxYoW7Zsmj59ujJkyCDp3wtlPe7y5cuS5GgDAAAAAAAAADEVq5mt2bNn18aNG9WiRQt16NBBkrRkyRItWbJExhhlyJBBX3zxhbJnz25psI+rWbOm1qxZE2n7xYsX1aVLF7300kvq2bOn8uXLp7Rp08put2vnzp2R2j/aVrRoUZfGCwAAAAAAACDpilWxVZJKlSql33//XWvXrtWOHTt09epVBQUFqWTJkmrcuHG0P8m3StasWZU1a9ZI2//++29JUo4cOVSvXj3H9kqVKmnTpk06f/68MmbMKOnfZQWWLFmioKAg1ahRw+UxAwAAAAAAAEiaYl1slSRvb281bNhQDRs2tCoel5o4caKqVKmiihUrqlOnTvL399eKFSu0a9cuzZgxQ4GBgQkdIgAAAAAAAIBEKlZrtn7//ffKlSuXBg4cGOX+yZMnq0SJEtqzZ0+cgrNauXLltGXLFuXKlUuTJ0/WsGHDdOfOHS1YsEC9e/dO6PAAAAAAAAAAJGKxmtn68ccf659//lGfPn2i3N++fXuNHDlSc+fO1QcffBCX+GIlR44cMsZEua9ChQratGlTPEcEAAAAAAAAIKmL1czWzZs3q3r16sqcOXOU+zNmzKgaNWpo27ZtcQoOAAAAAAAAABKLWBVbL168qLx580bbJnfu3Dp16lSsggIAAAAAAACAxCZWxdY0adLo3Llz0ba5cOECF5wCAAAAAAAA4DFiVWytUqWK1q5dq2PHjkW5//jx4/r6669Vrly5OAUHAAAAAAAAAIlFrIqtQ4cOVVhYmMqXL69Jkybp559/1qlTp7Rr1y5NnjxZFSpUUEhIiPr37291vAAAAAAAAADglrxj86CiRYvq888/V/v27TV06NAI+4wxSpYsmWbPnq0qVapYEiQAAAAAAAAAuLtYFVslqXHjxjpx4oQWLFigXbt26fr160qRIoWKFy+uNm3aKEeOHBaGCQAAAAAAAADuLVbF1h07digoKEiFCxfWgAEDrI4JAAAAAAAAABKdWF8g6+2337Y6FgAAAAAAAABItGJVbPXx8VFQUJDVsQAAAAAAAABAohWrZQQqVKigQ4cOWR0LgCSuTUCyhA4BAAAAAADAZWI1s3XgwIHaunWrtm3bZnU8AAAAAAAAAJAoxWpm6549e5QpUyZ16NBBr7zySrRtBw4cGKvAAAAAAAAAACAxiVWxdfDgwVH++3E2m41iKwAAAAAAAACPEKti6/z5862OAwAAAAAAAAAStVgVWzt06GB1HAAAAAAAAACQqMXqAlkAAAAAAAAAgIhiXWy9deuWxo8fr8qVKytfvnyaOHGiY9+HH36ozp0766+//rIkSAAAAAAAAABwd7FaRuDy5cuqWLGijh8/roCAAN29e1dXr1517C9VqpR69OihnDlzasSIEZYFCwAAAAAAAADuKlYzW0eNGqXjx4/rww8/1M2bN2WMibC/ePHiKlq0qFavXm1JkAAAAAAAAADg7mJVbP36669VrVo1denSRTabLco2pUuX1tGjR+MUHAAAAAAAAAAkFrFaRuD8+fOqX79+tG1sNpsePnwYq6CAmCriHXXRHwAAAAAAAIgvsZrZmiZNGl25ciXaNvv371emTJliFRQAAAAAAAAAJDaxKrZWqVJF33zzjW7cuBHl/jVr1mjXrl2qVq1anIIDAAAAAAAAgMQiVsXW4cOH6/79+6pVq5Z++OEHSdLt27e1f/9+jRkzRi1btlTy5MnVv39/S4MFAAAAAAAAAHcVqzVbCxcurC+//FJt27ZV1apVJUlz587V3LlzZYyRr6+vFi5cqGeffdbKWAEAAAAAAADAbcWq2CpJL730kk6cOKFFixZpx44dunr1qoKCglSyZEl17NhRmTNntjJOAAAAAAAAAHBrsSq23rp1SyEhIUqbNq169+6t3r17Wx0XAAAAAAAAACQqMVqzdffu3SpZsqRSpUqlDBkyKFeuXFq8eLGrYgMAAAAAAACARMPpma2nT59WzZo1dfPmTQUEBMjb21t///23OnTooDt37qhHjx6ujBMAAAAAAAAA3JrTM1vnzJmjmzdvavz48bp27ZquXbumHTt2KGPGjBowYICuX7/uwjABAAAAAAAAwL05XWxduXKlChQooKFDh8rb+98JsWXLltX06dN1+/ZtffPNNy4LEgAAAAAAAADcndPF1r///luVK1eOtL1GjRqSpOPHj1sXFQAAAAAAAAAkMk6v2Xrv3j0FBQVF2h4cHCxJun37tnVRwSO1CUiW0CEAAAAAAAAAseZ0sVWSbDbbE7eFh4dbExEAAIAbydt8Y4zan3ZRHAAAAADcX4yKrcuXL9ehQ4ec3mez2bR27drYRwcAAAAAAAAAiUSMiq1//fWX/vrrL6f3RTUTFgAAAAAAAACSIqeLrU8qsgIAAAAAAAAAYlBszZ49uyvjAAAAAAAAAIBEzZ7QAQAAAAAAAABAUkCxFQAAAAAAAAAsQLEVAAAAAAAAACxAsRUAAAAAAAAALECxFQAAAAAAAAAsQLEVAAAAAAAAACxAsRUAAAAAAAAALECxFQAAAAAAAAAs4J3QAQAAAABIHIZlaxmj9pNcFAcAAIC7StQzW69cuaLhw4fr+eefV5YsWeTv76/8+fOrV69eunDhQqT2v/zyi1588UWlSpVK/v7+KlWqlBYvXpwAkQMAAAAAAABIahL1zNaDBw9qwoQJql69unr06KGgoCDt3btX7733nr766ivt27dPadOmlSTt3r1bVapUUfr06dWvXz8FBARo+fLlateunW7duqVXXnklgbMBAAAAAAAAkJgl6mJr7ty5dfToUeXNmzfC9hIlSqhfv36aN2+eBg8eLEkaPHiwkiVLpp9++klZsmSRJPXp00fly5fXsGHD1KFDB/n7+8d7DgAAAAAAAACShkRdbM2WLVuU22vVqiVJ+vvvvyVJ165d09atW9WiRQtHoVWSvLy81LVrV/Xo0UObN29WvXr1XB4zAADwHKxvCQAAAHiWRL1m65OcOnVKkhyF1cOHDyssLEyFChWK1LZIkSKSpP3790fbZ0hIiG7evBnhBgAAAAAAAACPJMli68yZM2Wz2dS0aVNJ0rlz5yRJKVOmjNQ2ODg4QpsnmTBhglKmTOm4PWlWLQAAAAAAAADPlOSKrVOmTNE333yj3r17q2DBgpKk+/fvS1KUa7ImS5ZMknTv3r1o+x0yZIhu3LjhuJ0+fdriyAEAAAAAAAAkZol6zdbHzZ07V4MGDVL9+vU1ZcoUx3ZfX19JUlhYWKTHPHz4UJLk5+cXbd8+Pj7y8fGxMFoAAAAAAAAASUmSmdk6bdo0vfLKK6pfv76+/PJLeXv/Xx05Q4YMkv69UNbjLl++HKENAAAAAAAAAMRGoi+2GmPUv39/vfHGG+rcubO++uqrSDNQCxcuLLvdrp07d0Z6/KNtRYsWjZd4AQAAAAAAACRNibrYevfuXTVp0kRvv/22xowZo3nz5snLyytSuzRp0qhSpUratGmTzp8/79geFhamJUuWKCgoSDVq1IjP0AEAAAAAAAAkMYl6zdYBAwZoxYoVqlWrltKnT685c+ZEatOgQQNlzpxZEydOVJUqVVSxYkV16tRJ/v7+WrFihXbt2qUZM2YoMDAwATIAAAAAAAAAkFQk6mLrnTt3JEkbN27Uxo0bo2xToEABZc6cWeXKldOWLVs0atQoTZ48WQ8fPlTBggW1YMECdejQIT7DBgAAAAAAAJAEJepi64IFC7RgwQKn21eoUEGbNm1yXUAAAACAm8vbPOpJCk9y2kVxAAAAJEWJes1WAAAAAAAAAHAXiXpmKwAAidmwbC1j1H6Si+IAAAAAAFiDma0AAAAAAAAAYAGKrQAAAAAAAABgAYqtAAAAAAAAAGAB1mwFACAOuKo3AAAAAOARZrYCAAAAAAAAgAUotgIAAAAAAACABSi2AgAAAAAAAIAFWLMVAAAAlhiWrWWM2k9yURwAAABAQmFmKwAAAAAAAABYgGIrAAAAAAAAAFiAYisAAAAAAAAAWIBiKwAAAAAAAABYgGIrAAAAAAAAAFiAYisAAAAAAAAAWIBiKwAAAAAAAABYgGIrAAAAAAAAAFiAYisAAAAAAAAAWIBiKwAAAAAAAABYwDuhAwAAAID7yNt8Y4zan3ZRHAAAAEBiRLE1iSribUvoEAAAAAAAAACPQrEVScKwbC1j1H6Si+IAAAAAAACA52LNVgAAAAAAAACwAMVWAAAAAAAAALAAxVYAAAAAAAAAsADFVgAAAAAAAACwAMVWAAAAAAAAALAAxVYAAAAAAAAAsIB3QgeAJ2sTkCyhQwAAAAAAAADgJGa2AgAAAAAAAIAFKLYCAAAAAAAAgAUotgIAAAAAAACABVizFUhAeZtvjFH70y6KAwAAAAAAAHHHzFYAAAAAAAAAsADFVgAAAAAAAACwAMsIAEgUhmVrGaP2k1wUBwAAAAAAwJNQbAUSKYqPAAAAAAAA7oVlBAAAAAAAAADAAhRbAQAAAAAAAMACFFsBAAAAAAAAwAIUWwEAAAAAAADAAhRbAQAAAAAAAMACFFsBAAAAAAAAwAIUWwEAAAAAAADAAhRbAQAAAAAAAMAC3gkdAPBI3uYbY9T+tIvigOswxojOsGwtY9R+koviAAAAAAAgtjyu2BoeHq533nlHH374of744w8FBASoWrVqeuutt5QvX76EDg8AEjUK6gAAAAAAT+ZxxdZBgwZp6tSpqlOnjrp27apLly5p7ty5qlixovbu3ats2bIldIgAIImZngAAAAAAJDYeVWw9efKk3n77bTVp0kTLli1zbG/WrJlKlCih8ePHa+7cuQkYIQAAAAAAAIDEyqMukLV69WqFhYWpR48eEbY/99xzKleunL766qsEigwAAAAAAABAYudRM1v3798vSSpUqFCkfUWKFNGOHTt09uxZZcmSJb5DAwAgyWNpDAAAAABJnUcVW8+dOydJSpkyZaR9wcHBjjZRFVtDQkIUEhLiuH/jxg1J0s2bN10RqiTpfnh4jNo/+E8soffvx+ix3i7Mw1nhIbdj1P6///chd2/F+rEJydNy9rR8pbidx4kxZ08cY0/LuWDDmP0K5Egiz1fyvDH2tHwlz8vZ0/KVPC9nT8tX8rycPS1fyfNy9rR8Jc/L2dPyfZJHsRljEjgS69hMUsrmKWrUqKHNmzcrLCxMdnvEFRRGjRqlsWPHatu2bapUqVKkx44ePVpjxoyJr1ABAAAAAAAAj3D69GllzZo1ocOwhEfNbPX19ZWkKIutDx8+lCT5+flF+dghQ4aoX79+jvvh4eG6evWqgoODZbPZXBQxAAAAAAAAkDQZY3Tr1i1lzpw5oUOxjEcVWzNkyCBJunbtmtKnTx9h3+XLlyO0eZyPj498fHwibEuVKpX1QQIAAAAAAAAeIqrlPhMz+9ObJB1FixaVJO3cuTPSvp07dyp16tRJZsoyAAAAAAAAgPjlUcXWBg0ayG6365NPPomw/cCBAzp06JAaNWrEkgAAAAAAAAAAYsWjLpAlSb1799asWbPUoEEDVa9eXZcuXdIHH3yg0NBQ7dmzRzlz5kzoEAEAAAAAAAAkQh5XbA0LC9OMGTP00Ucf6c8//1RAQICqVq2qN998UwULFkzo8AAAAAAAAAAkUh5XbAUAAAAAAAAAV/CoNVsBAAAAAAAAwFUotgIAAAAAAACABSi2AgAAAAAAAIAFKLYCAAAAAAAAgAUotiLRCAkJSegQAMQR53FE27Zt0/nz5xM6DEsxxkkfY4ykxhOf056Ws6flK3lezp6WryfyxDH2xJyTCoqtcGuhoaH6+uuv1a5dO6VPnz6hw4k358+f17Zt2xI6jHjjafnGxb1793To0CHt3LlTJ06c0IMHD57Ydvbs2Zo7d248Rhc1Tz2PnVGtWjWtXr3acf/evXtatGiR/vnnnwSMKuYY45i7ffu2zp8/r7CwsGjbLV++XF9//XU8RfVkjLFnOHLkiMaOHZvQYcQLT3xOe1rOnpav5Hk5e1q+nsgTx9gTc06SDBKd3bt3m06dOj21Xd++fc2OHTsc92/cuGH69u1rjhw5EqntTz/9ZPr27WtpnLEVGhpq1q9fbzp27GhSp05t7Ha7SZ48ualdu3ZChxZnhw4dMt27dzcvvvii6dy5s1mzZk2U7ebMmWPsdns8R2c9T8vXlTZt2mSqV69ukidPbux2u+Pm4+Njmjdvbn7++edIj+nTp49JnTp1AkSbtM9jK9lsNjN37lzH/TNnzhi73W7Wrl2bgFE5hzGOud9++8107tzZZMmSxXEOe3l5mezZs5uBAweav/76K9JjunfvbtKlSxf/wRrG2BMtW7YsSb8fe+Jz2tNy9rR8jfG8nD0tX0/kiWPsiTkndd4JXexFzP39999auHChPv7442jbzZgxQwUKFFC5cuUkSbdu3dLMmTNVs2ZNFShQIELbAwcOaObMmZo+fbrL4o5OeHi4Nm/erKVLl2rFihW6evWqJOn5559Xly5d1KBBA6VKlSpBYrPKoUOHVK5cOd29e1fp0qXTt99+qwULFqhKlSr69NNPlSlTpoQO0VKelq8rvfnmmxo5cqRSpkypNm3aqGjRogoKCtKtW7d04MABrVq1SitWrNCMGTP06quvOh6XNm1a3blzJ97i9ITzOD4YYxI6hCdijGNv0aJF6tGjh0JDQ1WpUiU1bdrUcR4fPHhQ06dP15w5c7Ro0SI1bNjQ8bhcuXLp+vXr8RYnYxzRnj17NHv27Kd+5sqVK1eM+rXZbPrzzz/jEprTHo2hM27duuV025s3b8rf31/e3t4RtgUGBspuj/jjudDQUN29e1cpUqRwun+reOJz2tNy9rR8Jc/L2dPytZI7vz/9lyeOsSfm7EkotrqJvXv3Ot32xIkTsT6Ou/0R//333+uLL77Q8uXLdfnyZSVPnly1a9dWhQoVNHjwYPXu3Vsvv/xyQodpiWHDhikgIEA///yznn32Wd27d0/z5s3T0KFDVaFCBX3zzTfKly9fQodpGU/L11W2bNmiESNGqFGjRpo/f75SpkwZqc2tW7fUsWNH9e7dW6VLl9Zzzz2nrVu36v3331fhwoVdHqMnnceeijGOm19//VVdunRR6dKltXjx4ij/8Pnrr7/Utm1btWrVSvv371fevHl1/PhxLVy4MMZ/KMUGYxw1Z7/gLlSokGw221P72717ty5cuOBUW6ukTZvWJcdLnTq13n//fXXv3l2SdPbsWT3zzDNas2aN6tatG6HtRx99pFdfffWpy2ZYyROf056Ws6flK3lezp6Wrys8/v50//59fffddypdurRb/ETdE8fYE3P2RBRb3USpUqWc/iBsjInXD+mukjlzZl24cMHx4tKsWTM1aNBAQUFB+vPPPzV48OCEDtFSu3fvVrdu3fTss89Kkvz8/PS///1PNWvW1Isvvqhq1arp+++/V548eRI4Umt4Wr6uMmPGDOXNm1dLly6NMHvov4KCgrRkyRI9++yzqlixouOP2RQpUmjmzJkujc/TzmNPxBjH3ZQpU5QhQwZt3LhRgYGBUbbJmTOnNmzYoIIFC6pixYpKliyZzp8/Ly8vLy1dutSl8XnaGLviC+6nrau7adMmjR8/XhcuXFD27Nk1cOBAp2OIq1WrVul///ufTp8+rZYtW0b7Refhw4e1fPlyp/qN6gt8d/lS39Oe05Ln5exp+Uqel7On5esqj78/nTlzRs8884xGjRoV6Yux+OaJY+yJOXsqiq1uYv/+/WrXrp0OHjyo3r17q1ixYk9su2vXLr3//vvxGJ1rnD9/XunTp9f06dPVrFkzJUuWLKFDcqmbN28qQ4YMkbYXLFhQP/74o2rWrKlq1aq59EJR+/fvj/a5ZSV3yDcp2LFjh3r06PHEQusjyZIlU8uWLfXWW2+pd+/eyps3r1q3bh3lTFgredp57Ap37951/Gzo2rVrkv6drfykn/6mSZMm3mKTGGMrbN26VW3btn1iofWRoKAgtWnTRtOmTVODBg2UN29ede/eXblz53ZpfJ42xvH5BfeaNWv05ptvateuXcqTJ48++ugjtWvX7qmv6VaqX7++JKlRo0Zq06ZNtH9cL1++3OliqzvztOe05Hk5e1q+kufl7Gn5xhd3mrTliWPsiTl7KvvTmyA+FClSRFOnTpUk1axZUx06dHjirVq1agkcrTV69+6tZMmSOa6y17FjR61bt04PHz6MdZ99+/bVypUrdeXKFQsjtUaOHDl0+PDhKPdlzZpVmzdvlp+fn2rXrq2LFy9afvz169ercuXKlvf7JAmdb1LxpKJ1VNKnTy8vLy9Nnz5dr7zyissLrZJrzmN3FxYWpgsXLig8PPyJbbZu3arOnTs71V///v2VLl06pUuXzvFlSOvWrR3bHr/FN08cY6tduXJFzzzzjFNts2XLJrvdrq+++kqTJk1yeaFV8rwx3r9/v4oUKSLp39w//vjjJ97+uw52TCxdulTPPfecGjZsqDt37mjx4sU6cuSIOnXqFK+F1kcKFizo9KxTd5mdGhee9pyWPC9nT8tX8rycPS3fFStWqFWrVho4cKDjb6ODBw+qSpUq8vf3V3BwsFq0aKE//vgjgSO1jqeNseSZOXushLs2Fx538uRJkyNHDrNly5Zo261fv97kyJHjqf3ZbDYzfPhws2fPHrNnzx6zfv16Y7PZzKxZsxzbHt2GDh2aYFef/eGHH8zrr79uMmfObGw2m0mdOrWpX7++sdvt5quvvopRX76+vo4rPBcqVMj07NnTfPbZZ+bMmTMuit55HTp0MGnTpjX37t17YptTp06ZbNmyOa4474zQ0FCzevVqM3nyZDNnzhxz8ODBSG3mzJljvL29Tf78+WMdf0y5Kt+cOXPG6JYrVy6rUnqqr776yrRs2dIMGDDAXLhwwRhjzIEDB0zlypWNn5+fSZMmjWnevLk5fvy4031mz57d9OrVy6m2vXr1cuq1wRWsPI/d2YgRI0xgYKDjCqFNmjQxv/32W6R2c+bMceo5PXr06BjfEoqnjPEjoaGh5vz58yYsLOyJbbZs2WI6der01L4yZ85sBg4c6NRxBw4caLJkyeJ0nFbypDHetGmTsdvtZu3atdG2W7ZsmdPvT2FhYWbhwoWmQIECxmazmZIlSya5/7f/stlsZu7cuY77Z86cMTabLcr/U2dfE63mSc/pRzwtZ0/L1xjPy9kT8l25cqWx2WzG19fX+Pn5mfz585uzZ8+aDBkymJw5c5rWrVubunXrmuTJk5vg4GDzxx9/xPpY0b1WJxRPGOPHeWLOnoZiaxJms9kchcdHt6i2/Xd7QgoPDzdbtmwxr7zyismQIYOx2WwmZcqUpm3btmbFihXRFu0eefDggdmxY4eZPHmyadSokUmXLp0jt5w5c5oOHTqYjz76KEbFLqvs27fPDB482OzduzfadkeOHDEFCxY0OXPmfGqf169fN6VLl3aM4aNcO3bsaB4+fGiMMWbAgAHGZrOZypUrmytXrliSizNcka8xxrz00kumXr16jlvNmjWNzWYzZcqUibD9v7f44KoPSV27djXBwcGO4u2TnD9/3qRJk8Z0797dinRizYrz2F3Nnz/f2Gw2U7ZsWTNw4EDTqlUrExAQYPz9/c3ixYsjtE2owkJ8SMpj/IjVRfVWrVqZrFmzmjt37kTb7tatWyZLliymTZs2sY7dCnEZ47t375quXbuaPn36mPDw8Ce2GzlypOnevbsJCQlxRQpPZfUX3HPmzDE5c+Y0NpvNVKxY0axfv96iSN3Xk4qt69ati9Q2oV8TPeF163GelrOn5WuM5+VsVb5nzpwxJ0+ejLbNqVOn4nWyTpUqVUyhQoXMjRs3zL1798yLL75oypYtaypUqBAhr/3795vAwEDTunXrWB/LHYutj3jac9oYz8zZU9iMSQK/FUKUFi5cGOPHdOjQwQWRxFx4eLi2bt2qJUuWaOXKlbp8+bICAgJ069atGPd17Ngxbd++Xdu3b9dPP/2kY8eOyWazKWPGjKpcubI+//xzF2QQP/r06aN3331Xw4YNU5MmTWS327Vq1Sq99dZb6tmzp06fPq1ly5apXbt2mjdvXpJcE+bRIu9ff/11gi7yXrVqVV26dEk7duxQ8uTJ1bhxY127dk1eXl767rvv5OvrK0k6cOCAKlasqAYNGujTTz99ar9//vmnihUrpuzZs2vevHkqX758pDY7d+5Uly5ddPr0ae3fv185c+a0PL/YiO15HNMrr9tsNv35559xCdUpJUuWVEBAQIR1hs+fP6/27dvru+++0/Tp09W7d29J0ty5c+P9ytsJIbZjfOrUqRgfy9mf4cfVggUL1LlzZ5UpU0ZVqlTR6dOntXr1ahlj9MEHH6hNmzaOts6O86+//qoyZcqoQoUKWrx4sbJlyxapzZkzZ9S2bVvt2LFDu3btUtGiRS3PLTZiOsYffvihevbsqcWLF6tVq1ZP7Hfbtm2qWrWqFixYoPbt27sq/Hhjt9uVKVMm9evXz+nlnkqUKOHiqJxz7do1eXt7KygoKEaPs9vtKlq0qOPcvH//vr799luVKVMm0hWuT506pYMHD7rFa6KVnzETi9jmfP78eVWqVEl58+bV2rVrn7jWY6dOnfTTTz9p+/btCbLczePiMsYhISHy8fGJ9G93xxg7l+++fftUsmRJ1atXT6tWrXpiu169emn27Nk6cOCA42K/rpQ2bVr17dtXw4YNkyRt2bJFNWvW1Ny5c9W1a9cIbf/3v/9p6dKlunDhglN9P35RyIsXL6pu3bqaOXOmKlasGOVj3OH9ifPYM96fkrQELvbCCbdu3TJ79+41Bw4ciHaWSFIVGhpqNmzYYLp06RLnvm7fvm0+//xzU7x48XifzdupUyezdetWS/vMmTOnadmyZaTt77zzjiO/cePGWXpMZ7ki36i4y7ezwcHBZvz48Y77mzdvNna73Xz44YeR2r7++usmffr0Tve9YcMGExwcbOx2u8mdO7dp3Lixad++vWncuLHJkyePsdvtJjg42Hz77beW5OIKMTmP3XX2clBQkJkyZUqk7WFhYaZLly7Gbrebd955xxjj/CyunDlzmty5c5tBgwaZixcvWh5zfIrJGD/pVxbR3eJLiRIlTKVKlSJsO3funHnhhReM3W43M2bMcGyPyWy9+fPnGx8fH5MsWTJTo0YN07dvXzNy5EjTt29fU7NmTZM8eXLj4+NjPvnkE0vzsZIzY1y7dm1TpEgRp/qrVKlSvJ2/rvbfX5c87ZZQvya6ePGiefDggeP+ggULTNasWR1xlS1b1vz8889O95c9e3aTI0eOGN3cjZWfMf/rgw8+MNWqVbO0T6vEJOexY8cau91udu/eHW27v/76y/j4+JjJkydbFaZlYpLvuHHjTNeuXR33u3btmmCfo+OCMX6y119/3QQEBJhz585F2+7mzZsmbdq05o033rAqzGj5+PiYWbNmOe7//fffxmazmTVr1kRqO3PmTOPn5+d034nt165R4TyOXkhIiLl8+bLj/uXLlxPsl0P4P8xsdSMPHjzQ/v37lSlTJmXNmlWSNGbMGE2ePFn379+XJGXMmFGzZ89Wo0aNEjDSxOPcuXP68ccf9eOPP2r79u3av3+/wsLClDt3bj3//POqVKmSOnXqFC+x2O12zZkzR927d5ckXb58Wc2bN9eECRNUtmzZWPXp6+urCRMmqG/fvhG2HzhwQM8995ymTp2qfv36xTn22HBFvlE5e/assmXLluAzW319fTVlyhT973//kySdPHlSOXPm1OrVq1WvXr0IbWfNmqXBgwfr7t27Tvd/+fJlzZkzR+vXr9fRo0d18+ZNBQUFKX/+/HrxxRf16quvusVMA1dwl9nLmTJlUp8+fTRo0KAo9/fu3VvvvvuuPvroI4WEhDg149Fut6tkyZI6fPiw7Ha7unfvrgEDBihTpkyuSMFtTJ06NcLsmevXr+vNN99U586dVbBgwSgf079//3iJLUWKFBo5cqTeeOONCNvDw8PVvXt3zZ8/XzNnztTrr78e4xnMR48e1dSpU7V+/XqdO3fOsT1Dhgx68cUXNXDgwCfmn1ikT59eXbp00YQJE57adtSoUfrggw8i/F8ktNu3b+v48ePy9vZW4cKFnb5qszv/mujcuXOqX7++fv31VwUEBGjKlCnKkiWLGjRooLJly6pq1aq6e/euvvrqK129elU///yzChcuHC+xJVVDhgzR5MmT3WImb1yUK1dOQUFB2rRp01PbNmnSRBcuXND27dvjITLXuHz5sgoXLqxNmzbpwYMHeumll3TkyBGlTp06oUNzGU8b40KFCqlYsWJO/bKxa9eu+uWXX3TgwAGXx5U/f37Vrl1bs2bNkvTvZ46//vpLmTNnlp+fX4S2/fv316pVq5y+UJY7vz+5gieex2PHjtX+/fu1fPlySdLLL7+s4sWLa8SIEQkcmWeL/8uhIkq///67atSooXPnzslms2nQoEHKnz+/xowZo2bNmql69eq6d++eFi1apGbNmmnr1q1PnPb/X+Hh4Vq6dKm+/vprHTp0SBcuXND9+/fl5+endOnSqVChQqpXr55atGiRIFfHPX36tNatW6ds2bI5iih3797VpEmTtGvXLvn5+almzZrq0qWLkidP/tT+Dh8+7Cis/vjjj/rrr7/k7e2t5557TpUqVdKQIUP0/PPPR/p5W0IICQnR1q1bdeXKlVj38eDBAwUEBETaHhwcLEkqUKBArPu2mhX5urPs2bPr+PHjjvvZsmXT8ePHlTlz5khtT548GeX26KRNm1bDhw/X8OHD4xyr1aw+jx/nbLHD1XLnzq1t27Y9sdg6c+ZMPXz4UN27d1edOnWc7rdbt25q0KCBJk6cqDlz5mjOnDnq3LmzBg0aFOXPzROC1WP8eCHz7NmzevPNN/Xyyy8naEFdkgICAqIskNjtds2bN08BAQHq3bu3AgMDY9x3gQIFNG/ePEnSrVu3HF+apEiRIs5xx5VVY3zjxg1lyZLFqWOmS5dOV69etST+mLL6C253/sN0zJgx+vXXX9WxY0fZ7Xb169dPmTJlUvXq1fXtt9862o0cOVLPPfecxo0bpy+++CIBI7aGq9+b3JHVOR85ckQDBgxw6tilS5fWpEmT4hR/TFmdb9q0aTVp0iT17t1bDx8+1NSpU92uQMMYxy3fv/76S507d3bq2IUKFYq318Lnn39eX375paZNm6ZkyZLJbrcrd+7ckdoZY7Rq1aoYfTHrzu9PEuexFTn3799fhQsX1g8//KAHDx5o3759Ti1XBxdL4Jm1+P/atm1r/Pz8zNixY8348eNNmjRpTLp06czLL78cod3du3dNvnz5zEsvvfTUPi9dumRKlChhbDabeeaZZ0zt2rVN69atTYcOHUzr1q1N7dq1Tdq0aY3NZjPFihV76gV4rLZ//36TIkUKx88VGjdubB4+fGhKlSpl7Ha7yZw5swkKCjI2m82ULl3a3L59+6l9Puorf/78ZuzYsea777576gVJ4ktMrtobkz6HDx9u9uzZE+G2fv16Y7PZzKxZsyLtiy+uyDcq7rKMQOfOnU3GjBkj/EQzKuHh4SZ37tyW/XT24cOHCfrzc1ecx49zlzEeNWqU8fLyMkeOHIm2XevWrZ3+Gdbj58m5c+dM7969jZ+fn/Hx8THdunUzf/75Z5xjjwtPGmNjjKlYsaKpW7dutG1eeeUVkyxZMtOgQQO3/LldTFk5xpkyZTIjRoxw6rhDhgwxGTJksCoNpx09etRkyZLF2O124+XlZYYOHWoWLlxobDabad68uZkzZ455++23TfHixY23t7fZvn17nI4XHh5uLly4EKtzwwrZsmUz7du3d9x/7733nrjMTf/+/U3GjBmd6nfMmDFm3Lhx5tChQ5bFapX4eN2KzuDBg+P9tcEVOfv6+jqWx3maWbNmGR8fn7im4TSr861ataqpVq2aqVq1qrHb7cbb29tUq1bNcXMHjHHc8w0MDIywHFB0pk2bZgICAuKahlP27dtn3njjDXPs2LFo233++ecmVapUZv78+fESl6txHlv3/vT111+b4sWLm8KFC5sNGza4OAs4g2Krm8iQIYN55ZVXHPc//fRTY7PZzMKFCyO1HTJkiEmbNu1T++zYsaNJlSrVU9dx/Pbbb02aNGlMx44dYx54HDRv3twEBwebb775xmzdutXky5fPdOrUyWTPnt1x1efw8HDHHwUDBgx4ap+NGzd2XMUvc+bMplmzZmbmzJlm9+7dJjQ01NUpRctVxdbo1oT77/74XoPHVcXWmBSW47PA7MoPSTt37jRt2rQxjRo1MgsWLHBsHzJkiPHx8TF2u93kypUrQa5+7Yrz+HHuUoi7c+eOOXr0qLly5Uq07R4+fGgmTZpkRo8e/dQ+Hz9PHvlv0TVZsmSmQ4cOT31uuYonjbExrimqG/Pv2r6ff/65adOmjSlWrJjJmDGjSZUqlcmUKZMpWrSoadmypVm8eLF5+PChFWnEiJVj3LRpU1OgQIGnvuc+ePDA5MqVy6kvj63mii+4n2T06NEmTZo0jsLu888/H69ffBpjTPLkySMUF44fPx7tWoDJkiVzql+bzWa8vb2Nl5eXadq0qdm3b59lMcdVfLxuRSchiq2uyDlPnjymd+/eTh3/lVdeidd1ea3Od+vWrWbr1q1m8eLFJkWKFCZVqlTm888/d2x3B4xx3PMtXry4ad26tVPHb9q0qSlcuHCccnAH1apVi/L1fseOHaZly5amdOnSpmbNmmbatGnxPkmJ8zjuOefIkcPkzJnT5MiRw1FgfnQ/Z86c8ZESnoBiq5tIlixZhEWx//zzT2Oz2czXX38dqe0777xjvL29n9pnunTpzNChQ506/tChQ026dOmcD9gCmTJlMgMHDnTcX7ZsmbHb7Wbq1KmR2jZr1ixGLxZ//PGHWbhwoenevbspVKiQsdvtJiAgwFSrVs0MHz7crF+/3ly/ft2SPJzliuLjggULYnyLL64qtiaFRd5jYs+ePcbHx8f4+PiY3LlzG29vb/Paa6+Zd99919hsNtO4cWPTv39/kz9/fpM8eXLzyy+/xGt8rjyPH3GnQpzVnlRsfeS/RVdvb+8oL4jnap42xq4oqrv7L02sHOPvv//e2Gw20759+ydenOHevXumRYsWxm63J8iYu+IL7owZM5oOHTpE2Pbmm28am81mKlWqZMaPH2/69etncubMaQIDA83BgwfjnIezYvJ+HJOLvtlsNjNjxgzz9ttvm0yZMhm73W4aNGhgdu3aZVnssRUfr1vRSYhiqytyfuWVV0xwcLC5dOlStO3++ecfkzJlykjngCu5aowbN25s3nzzTfPWW29F+gImoTHGcc933LhxTn1e3r59u/Hy8jKDBw+OeeBuJqrPmt99953jwpzFixc3uXLlMjabzRQtWtTcuHEj3mLjPI57zn///bf5+++/zfr1602GDBlMpkyZzMaNGx3bkXBYs9VNhIaGysfHx3H/0b+jWqswWbJkCg8Pf2qft2/fdvqCOenSpdPt27edjNYaV65c0TPPPOO4X6xYMRljlCdPnkhty5cvr9WrVzvdd+7cuZU7d261b99eknT16lX99NNP+umnn7Rt2zZNnz5dISEhKlSoULwsev7It99+6/h/vnHjhmw2m1asWKGjR49Gamuz2SJd+Opx7r4Gj9X5StL8+fMtj9OdjR49WunSpdOOHTuUNWtW/fXXX2rYsKEWLVqk7t27a86cOZL+XRi9ePHiGjt2rNasWRNv8bniPN67d2+E+xcvXpQk/fnnn5H2PVKiRImYhB0ry5cvV5MmTZxqGxYWptdee80xPrGVMWNGzZgxQ4MHD9bEiRP1wQcfxKm/2HDla7U78vf3V/78+Z/aztvbWwMHDnSqzwEDBujEiRPatGmTatSo8cR23333nZo3b65BgwbF62udlWNcuXJlDR8+XOPHj9e2bdvUrl07lShRQilSpNDNmze1e/duLVy4UGfPnlWvXr0SZI3eq1evRljvrly5cpL+b73z/8qcObOuX7/uVJ/FixePsG369OmqXbu21q9f79g2btw4lSxZUsOGDdOqVatimYH78PPzU/fu3dWzZ0/NmTNHkydPVtmyZVWrVi2NGDFCFSpUSJC4XPG61aBBA6ePf+TIEafbWsUVOQ8YMECLFi3SCy+8oM8//zzKawEcOHBAbdq0UUhIiNNrf1rBFfmuXLlS+/bt05IlSyRJBQsW1KpVq9SwYUPrAo8Dxjju+fbu3Vvz589XrVq1NHHiRHXo0EG+vr6O/ffu3dO8efM0fPhwZcyYMdIa864yduxY1atXL9Ln2Rs3bmjJkiX6888/lTJlSr344osqWbJknI83cOBABQcHa/v27cqVK5ckacOGDWrSpIlGjx6t6dOnx/kYzuA8jnvO2bNnlyS1b99eY8aMkc1m01tvvaUtW7ZYGzhijGKrm4vLhWEKFy6s5cuXq1evXrLb7U9sZ/7/QtvxfeXZdOnS6Z9//nHcz5Iliz788EOVKlUqUtsbN24obdq0sT5WmjRplC9fPl28eFHnzp3TyZMnderUKR0+fDjWfcbGsmXLtGzZsgjbPvrooyjbOlt8dGeuyNddC8yu+pC0e/dude3a1XEBl5w5c6pv377q0aOHRo8e7Wjn7++v1q1ba8aMGVak4zRXnMelSpWK8rWvT58+kbYZY2Sz2eLlis/NmzfX22+/rV69ekXb7sqVK3r55Ze1ffv2OBdbH/lv0TW+uWKMH/8Qb9WXMe5q7dq1evXVV6MttEpSjRo11LNnT3344YfxFNm/rB7jsWPHqmjRoho1apTGjx/vOJ+NMZL+/UJ0/vz5CfZ67oovuNOlS6ebN2867l+/fl1Xr15VmzZtIrTz9/dXmzZtNGXKlNiGHytffPGFDh06JEm6c+eObDab5syZow0bNkRoF9vPRb6+vurTp0+EomulSpVUtWpVjRgxQlWrVo1rCjHiitetAwcOxOhz+X//mI4Prsg5Z86cWrJkiVq3bq3ChQurfPnyKlmyZIQvT3bu3KlkyZJpwYIFevbZZy3NKTquyDdVqlSaO3eu44I0c+fOTZALCD8JYxz3fIOCgrRx40Y1adJEr7zyinr37q0CBQo48j169KgePHigvHnz6quvvorySzhXGD16tDJmzBjh74i//vpLlStX1tmzZx3bRo4cqY4dOz7x7yln3L9/X7/++qtGjBjhKLRK0osvvqh27drpq6++irdiK+exNTnPmzdPN27cUPfu3WWM0Xvvvad58+apa9eulsaOmLGZR598kaDsdruKFi3q+GB2//59ffvttypTpozSp08foe2pU6d08ODBpxYXVq9erZdffllFihRR+/btVaBAAQUFBcnPz0+hoaG6du2adu3apeXLl+vQoUNasWKF6tev77IcH9ekSRMdP37cqZmlFStWlLe3t77//nun+g4PD9fevXv1ww8/aPv27frxxx916dIlGWOUIUMGVa5cWVWqVFGVKlXi7UPDyZMnY/yYR99UPUn16tXVr18/1atXL8L2nTt3aubMmY4CX506ddSzZ0/5+/vHOIbYckW+7sxut2vOnDnq3r27Y1tUH5JsNluMPiSlTJlSo0ePjlBoqlevnsLCwiLMlpKkKVOmaMSIEY4raccHV5zHCxcujHEc8VG06dy5sxYuXKi+fftq6tSpUbY5cOCAGjZsqFOnTmn8+PEaMmRItH0uXLhQFSpUUN68eV0RsiVcMcbRfQEYlfgqqEty+irFj9hstqeez/7+/nrrrbei/MLgcTNmzNDQoUN19+7dGMURF658Pz5x4oR+++033bx5U0FBQcqfP7/y5csX15Dj5PHX67Nnzypbtmxau3at6tSpE6Ht3Llz9eqrrz71+deyZUsdO3bMMfv+4cOH8vf31+eff66mTZtGaPvmm29q4sSJunXrloVZPZmrzreo3vceuX//vqPoeuHCBVWoUEHDhw9X7dq1YxRLbLnyOe2uXJnzyZMnNXnyZK1Zs0ZnzpxxbM+YMaPq1q2rAQMGOPWLACsxxtFjjKMXHh6u5cuXa/Xq1Tpy5EiE96i6deuqRYsWSpYsWVxTcFpUr6cNGzbUpk2b9MEHH6hBgwa6ceOGZs+erSlTpjj15f+T+j5//rwyZ86sJUuWqHnz5hHazpw5U4MGDYq3vyU4j6PnbM6HDx9WSEiIo1i/d+9e+fj4xOuXI4hCfK9bgKhlz57d5MiRI0Y3Z3zzzTemTJkyUa5r+egCSmXKlDHffPONizOMbMWKFaZUqVLmwIED0bZ7tP7b+PHjn9rnqFGjTI0aNUxgYKAjxyxZspjWrVubuXPnmqNHj1oVvltw5zV4XMVdF3mPaiwaNGhg/Pz8zCeffGJu3LhhTp06ZQYNGmTsdruZOXOmU/2WKVPG1KtXz3F/586dxm63m2bNmkVq+8ILL5gCBQrELZEYcsV57M7GjRtn7Ha7adGihXnw4EGEfcuWLTOBgYEmKCjIrFy5MoEitJ4rxvjROlIxucWXgIAAExgY6LgFBAQYu91u/Pz8Imx/dAsKCnpqn6VLlzbPP/+8CQsLi7ZdeHi4qVq1qildurRV6TjF087jR2vj1q9f39SvX9+88MILxmazmbJlyzq2PboVK1bMqbU3d+zYYex2u3nttdcca9WWKFEi0v/V7du3Td68eU316tVdklt8etqa08b8uz7v22+/bTJmzBiva5h62nPamPjL+datW+bs2bPm5s2bsXq8VRjjJ2OME5/HX09DQ0NN8uTJTc+ePSO1bdy4cYwu3GWz2czbb79trly5Yq5cuWKOHDlibDabWbp0aaS2w4cPN5kzZ45dErHgSWP8iCfm7KkotnqIS5cume+//94sXbrUfPLJJ2bp0qVm69at5uLFiwkd2lOdOXPG7Nu3z9y6deupbW02m8mePbtp3769+eijj8wff/wRDxFa4/jx42b37t2RCjjRieoPnZIlS5pMmTKZP//807Ft/fr1xt/f3/Tt29eyeOMqNvka474FZld9SPrss8+MzWYzzZs3NyNGjDDp06c3+fLlM5UqVTJdu3Y1R44cMb///rvp1auXsdvtZty4cZblZKWYnMfRCQ8PNxcuXDC3b9+2KLKY+/TTT42vr6+pUqWK4/k1cuRIY7fbTY4cOZ764SmpsmqM3c3p06fjfPGuVatWGS8vL/Pcc8+Z6dOnm3Xr1pkffvjB7N692+zcudOsX7/ejB071hQrVsx4eXmZ1atXW5iBdZwZ4zFjxpg9e/ZE2n79+nUzZ84cM2DAADN+/Hize/duV4YaLVd9wT1z5kxjt9tNrly5zKhRo0zXrl1N/vz5zZ49e8y2bdvMe++9ZwoWLGh8fHzMTz/95OIsXc+ZYusj9+7dMzNmzHBxRDFn1evWpUuXTLVq1czOnTstisx1ktL7sTNc9d50+fJl8/3331vap1UYY2vcvn3bnDx50tI+n+Tx19MLFy4Ym81mFi1aFKnt1KlTja+vb4z6jupCwlFN/ChevLh5/vnnY5eEC8V0jMPCwsznn39u2rRpY4oVK2YyZsxoUqVKZTJlymSKFi1qWrZsaRYvXmwePnzo4shjz8rn9cOHDxNF3SepodiKBPOkP8jiwp2vuLdr1y5Tvnx5M3bs2AjbL1++7Jh9bLfbTbp06cyKFSuc6vPxN+Z79+4Zu91uRo0aFaltjx49TPbs2eOQQcy4Il9j3LfA7MoPSSNGjDB+fn7GZrOZQoUKmaNHj5oTJ06YXLlyRZipXrt27RgXr+PKFefxk4wePdqkSZPG2O124+XlZZ5//vl4O/bjtm3bZoKDg02hQoVMw4YNjc1mM5UrV37q1XzjYvfu3aZTp04u6/9J4nOMoxKff+xEJbort8eEO//SxMoxjuo1+sSJEyZr1qyOPB/l37lzZ0uO6U5++uknU7NmTePl5RUh10fjXLp0abN9+/aEDtMSVatWjfKXJu4gvl+3Hr1OrFu3Lt6O+ThPez9O6PemR1cQj0+MsTXOnz9v3nrrLdOzZ08zduzYJ35JPmfOnHgbY5vNZj744APH/XPnzhmbzRbl30gTJkwwadOmdbrv0aNHR3l7/PPGunXr4n0mpSvG+NKlS6ZEiRLGZrOZZ555xtSuXdu0bt3adOjQwbRu3drUrl3bpE2b1vErlwsXLlh6/Kdx1fN6586dpk2bNqZRo0ZmwYIFju1DhgwxPj4+ji+D169fb/mxETWKrYnArVu3zM8//2wOHDhgwsPDY9XH8ePHzYoVK8ycOXPMjBkzzNy5c81XX32VoD+rj8mMiKSgf//+JlmyZJGKBm3btjU2m820adPGvPPOO6ZSpUrGx8fHHD9+/Kl9Pv5/+OiN+YsvvojUdsaMGcbHxyfuiTjJFfka474FZld+SDLm3zwf/0by7t27ZtGiRWbixIkJ9sbpivM4Y8aMpkOHDhG2vfnmm8Zms5lKlSqZ8ePHm379+pmcOXOawMBAc/DgQUuP76zff//d5M6d29jtdtO+fXuXF7oT4g87Y1z3Wu2Of+xExapi6yPu+EsTK8fYVUuqJDbXrl1zjPOnn35q1q5da06dOpUgsVSrVi1GN09Z4iCmgoKCnngLDAw0NpvN+Pv7O7alSJHC0uM/jae9Hyf03xEJ8Z7MGMfdyZMnTfr06SN9+dehQ4dIM3jju9iaNWtWU6RIEVOkSBHz7LPPGrvdHuFvi0fq1KljihcvHi9xuZorxrhjx44mVapU5ttvv4223bfffmvSpEljOnbsaOnxn8YVOe/Zs8f4+PgYHx8fkzt3buPt7W1ee+018+677xqbzWYaN25s+vfvb/Lnz2+SJ09ufvnlF0uPj6i5z2XZPFxYWJj69++vL774QlmzZtX06dNVqVIlffLJJ3rttdd0584dSf8uUj5z5sxIF1t4kq+++krDhg3TsWPHHFcB/i+bzaZcuXJp/PjxatGihaU5JaQ1a9ZoxYoVOnz4sK5fvy4fHx+lT59exYsX18svv6zy5cvHe0zfffedqlevHuHqtLdu3dKXX36pMmXKaPHixZKknj176tlnn9WoUaP06aefPrXfu3fv6urVq5L+vfqxFPUVlS9fvhxvV9SUXJfv465fvy5jjAoVKhRpX8GCBbVgwYJY5xBTY8eO1TvvvCPp34X3bTabLl26FKndtm3blC1bthj17evrK19f3wjb/Pz81K5du9gH7KauXr2q4sWLR9g2ffp01a5dO8JFwcaNG6eSJUtq2LBhWrVqVXyHqXz58unnn39Ww4YNtXv3bp0/fz7G4/roYjrOOHHiRExDdFunTp1S6dKlI5wfo0ePVrt27TR79mwFBAQkYHSulTZtWlWuXDmhw4g3YWFh2rBhgzp37qy2bdtKklKkSKGJEyfq2LFj+vDDD52+yEd8uH37tn777Tf5+fmpcOHCMbr6/H+lSpXKbcb54sWLTuVx4cIFXb58OdY5J3WpU6fW6dOnVbNmTWXJkiXCvrt37+rLL79UhQoVlDVr1gSK0HqJ5f3YKkWKFHH6+X/z5k0XRxM/PG2Mhw0bprt372rNmjWqXr26zpw5ozlz5mjWrFk6ePCg1q1bpwwZMsR7XJUrV4703EuXLl2kv98PHz6sb775Rq+//np8hpeorF27Vq+++qpq1KgRbbsaNWqoZ8+e+vDDD+MpMtcZPXq00qVLpx07dihr1qz666+/1LBhQy1atEjdu3fXnDlzJP37t2rx4sU1duxYrVmzJoGjTvootrqJmTNnatasWSpfvrx8fHxUr149rV27Vt26dVOjRo1Uo0YNXbt2TXPnzlWrVq2UJk0aVa9ePdo+v/jiC7Vq1UqVKlXSwIEDVbBgQQUHBytZsmR6+PChrly5ol9++UXLli1T69atZYxRy5Yt4ylj1wgJCVGjRo20ceNGpUqVSoULF1aePHm0c+dOXblyRTt37tT06dNVs2ZNffLJJ0qfPn28xXbixAk1btw4wradO3fqwYMHqlu3rmObt7e3mjZtqo8//tipfvv376/+/fs77ttsNp07dy5Su7Vr1ypXrlyxjD7mXJWv5J4F5oT4kHT58mU1b95cEyZMUNmyZePcn7tIly5dhD9irl+/rqtXr6pNmzYR2vn7+6tNmzaaMmVKvMQVFBQU5fMsPDxcd+/eVd68eZU8eXLHdpvNphs3bkTbZ6lSpZz+w84Yk2SKIO76x05CSKrn8SNXrlzRw4cPVaFChUj7KlasGOGP+fjiqi+4o3LlyhUtWrRIx48fV2BgoGrVqqWaNWtalYpTDh06FO3+33//XW+99ZY+//xzBQYG6pVXXnGq3+rVq6tfv36qV69ehO07d+7UzJkz9eeffyplypSqU6eOevbsKX9//1jn4A727NmjJk2a6NChQxo9enSE5/SZM2f05Zdfqm/fvhE+4yR27vp+7Cq1atXSe++9pwcPHujZZ5+N9jNkeHh4PEbmOp42xj/88IO6deuml156SZKUN29eTZs2TfXq1VOTJk1UvXp1bd26VenSpYvXuLZu3epUu5QpU2rz5s1RTjJ5krFjx8put6tx48YecXX627dvOz1+6dKl0+3bt10ckevt3r1bXbt2dXzZlzNnTvXt21c9evTQ6NGjHe38/f3VunVrzZgxI2EC9TAUW93EwoUL9cILL+ibb76RJPXp00fNmzdX06ZNHTMAJenVV19V4cKF9dZbbz212Dpx4kTVrFlTGzdujHJ/3rx5Va5cOfXq1Uu1a9fWpEmT4r3YevXqVZ06dcrp9v+dJRmVcePGaePGjXrvvffUtWtXeXl5SZImTJig5cuX69SpU1q+fLn69eunGjVqaNeuXZFmC7rKvXv3lCZNmgjb9u3bJ5vNptKlS0fYnjFjRl25cuWpfY4aNSrK7QUKFIhwf/369dq3b5/GjRsXw6hjzxX5PuKOBWZXfkh6kpCQEG3dutVReE4oVp/Hzz//vFasWKERI0ZIkgICAuTl5RXluWqz2eLtD54mTZpYXuzcv3+/2rVrp4MHD6p3794qVqzYE9vu2rVL77//vqXHd5bVY+yuf+xEx1WF7qR4Hv/3/+rR+RkUFBSp3cOHDxUYGBiDKK3hii+43333XQUHB6tVq1aObUeOHFHVqlV16dIleXt7KzQ0VNOmTVPz5s312WefJfiXJwcPHtT48eP11VdfKSgoSEOGDFGfPn2UOnVqpx6/devWSJ8bN2/erDp16shms6lQoUI6ceKE3njjDS1cuFA//PCDUqRI4YpUomT161batGn13XffqXfv3qpevbreeecddevWTZLrXh9iylPejx+xOt9p06YpT548ev311zVx4sRoC+fLli1LkF8FMsbRe1q+ly5dUs6cOSNtr1atmr7//nvVqlXL8RnEHWXNmjXGs+dHjx4tLy8vjR49Wo0bN9bw4cOj/bwZ36we48KFC2v58uXq1auX7Hb7E9sZY7Rq1SoVLlzY6WNbxeqc79y5E+m9e/ny5apRo4YyZswYYbu/v7/u3bvnfLCIvQRcwgD/ERgYaKZPn+64v2fPHmOz2cynn34aqe2wYcOcWgfK19fX6bXQZsyYEaML9ljhSVdGjO72NDlz5jRt27aNtH3Xrl3Gbrc7LqD1999/mxQpUpgJEyZYnteTZM6c2QwcODDCtqZNmxq73W7OnTsXYfvIkSNNYGBgvMXmCq7K150XeXeFxLBGnNXn8Y4dO4zdbjevvfaaCQkJMcYYU6JEiUhjefv2bZM3b95Ev77gpk2bjN1uf+qaoAm5ZqvVY+zv729mzZoV5b4DBw6YjBkzmsKFC5vLly/H+5qtgYGBTz3P/ntz5pzzpPM4Maw7V7RoUVOrVi3H/d69e5tMmTKZNm3aRGh369Ytkz17dlOjRo2n9unr62tGjhwZYVuVKlVMmjRpHOvG3bp1y8yePdt4e3ubSZMmWZBJ7Pz888+mQYMGjotUvvXWW+bmzZsx7sddL1j5KDarX7f+a968ecbHx8e88sor5uHDh5av7RwbnvZ+7KoxPnLkiFNjuWzZMmOz2axIxWmMcdzzLVCggOndu/cT9x8+fNhkyJDBlCtXzkybNi1eP39cvXrVjBo1ytSvX9/Ur1/fTJw40dy5cyfO/dpsNjNjxgzz9ttvm0yZMhm73W4aNGhgdu3aZUHUcY/N6jFetWqV8fLyMs8995yZPn26Wbdunfnhhx/M7t27zc6dO8369evN2LFjTbFixYyXl5dZvXp1PGT6f1yRc5kyZUy9evUc93fu3Gnsdrtp1qxZpLYvvPCCKVCggKU5IWrMbHUT5rGfGmfKlElZsmSJ9E2E9O8364+3j0qqVKl09uxZp45//vx5pUyZ0rlgLdSkSRNLfzb5zz//qESJEpG2BwcHyxijEydOKHv27MqePbtat26tpUuXavDgwZYdPzrVq1fXp59+qiFDhihVqlS6ePGivvnmGxUrVizSOG/fvl3Zs2ePl7hcxVX5Pmk27+Pq1KkT79++X7t2TTNnznSsw1mxYkX973//i9NPJxPDGnFWn8flypXT22+/rb59+2r9+vVq166dSpQooU8++UR16tTRnTt3dOjQIb3zzjs6deqUFi5caNmxrXLnzh2tWrVKrVu3fmrbfPny6Zlnnnnq8yQgIOCp32y7itVj/Mwzz+jPP/+Mcl+RIkUcaz7Xq1dPzZo1s+y4znDFDGZPOo8Tw7pzJ06cUMeOHR3327dvr1mzZkWaxRYYGKi2bds61uKOTvLkyZU2bVrH/dDQUP3www8aM2aMY924wMBAvfrqq9q9e7fmzZungQMHWpOQk7Zt26bx48fru+++U8aMGTV16lT17NlTfn5+lvR///59/frrrxoxYkSEX5W8+OKLateunb766itNnz7dkmM5w+rXrf/q0qWLnn32WTVp0kSHDx92m59ketr7sSvGOFeuXNqyZYuKFCkSbbsqVapoy5Ytlh7bGYxx3BQtWtTxWmSPYtZjoUKFtGnTJlWtWlVjxoyx7LhPc/78eVWoUEF///23/Pz8lDx5cn399dd67733tGnTJuXLly9O/fv5+al79+7q2bOn5syZo8mTJ6ts2bKqVauWRowYEeVSP/HF6jFu0KCB1q1bpxEjRuiNN96ItAzXo88jpUuX1rp161SrVi3Lju0sq3Pu06eP2rRpoxYtWih//vyaO3eu8uTJo/Pnz6tbt27q37+/7Ha7Zs+ere+++y5en9seLSErvfg/JUuWjDSj4kk6d+5snn322ae269GjhwkMDDTLli0z4eHhUbZ58OCBWbZsmUmRIoXp0aNHjGKOK1dciS9fvnymffv2kbavWbPG2O12s2fPHse2WbNmmaCgIEuPH53ff//dBAQEmGzZspnmzZubLFmyGLvdbpYsWRKh3bFjx4yXl5fp2bNnjPo/efKkWbdunfnss8/M8uXLzQ8//BCrmSpWcXW+7ubcuXMmZ86cjhlqqVKlMjabzTzzzDPm999/j3W/ly5dMpUrVzaZMmUyP/74Y4R9p0+fdouZNK66GvBPP/1katasaby8vCJcMdZutxubzWZKly5ttm/f7pJjx8bdu3fN0qVLTZMmTYy/v3+CzEJ1BVeMcfPmzU22bNlMWFjYE9scOHDApEmTxqRIkSLR/1968nn8JKdPnzZbt241Fy9ejNfjGmNMQEBAhF8T/fPPPyZr1qzmu+++i9T27bffduqzQokSJUzXrl0d969evWpsNpv56quvIrWdPn268fHxiWX0MbdhwwZTqVIlY7PZTI4cOcz777/vmMEWF48/b86dO2dsNpv54osvIrWdMWNGvOYcX8/ps2fPmnLlypng4GCnfqHgSp72fpwQr1sJjTGOu61bt5qWLVuanTt3Rttu+/btxs/PL94+f/To0cP4+vqapUuXOv5u37p1q0mfPr0pX758nPqO6v/x3r17EWa6Vq9e3WzZsiVOx7EqNitdunTJfP/992bp0qXmk08+MUuXLk2wzx6PuCrnESNGGD8/P2Oz2UyhQoXM0aNHzYkTJ0yuXLkinMu1a9c2Dx48sPz4iIyZrW6iZs2aeuedd3TlypVoF2O/c+eOVqxY4VjnLjqTJk3S4cOH1axZM6VKlUp58+ZVUFCQ/Pz8FBoaqmvXrungwYO6f/++KlSooEmTJlmZUoLo1q2bBg8erKJFi6pbt25KkSKFDh06pDfeeEPp06eP8C319evXFRoaGm+x5cuXTxs3blTv3r21evVqPfPMM3r//fcjrff02muvKTw83KkZcdK/ax/26dNH+/btkxTxIjp2u13Vq1fXqFGj4v0bS1fl+1+nTp3S4cOHdf36dfn4+Ch9+vQqVqxYlGsDutro0aN17tw5ffHFF2ratKlsNpu+//57NW/eXB07dtRPP/0Uq34TwxpxrlK+fHlt2rRJ169f14EDB3ThwgU9fPhQqVKlUpEiRZQtW7aEDlEhISFat26dvvjiC61du1Z3796Vn5+f6tatqyZNmiR0eG7r1Vdfld1u165du574zX6RIkW0evVqvfDCC/EcnfU8+Tx+ktisO2eVAgUKaM+ePY77mTJl0unTp6Nse/DgQadmlDdt2lTjx49X//79VaBAAaVOnVopU6aMcj3y/fv3R/nLJVepU6eOMmXKpClTpqhdu3by9vZ+6gVBHl9z/Unc8YKV8Slz5sz6/vvvNX/+fN2/fz/JXnwmMbwfI248ZYyrVKmiKlWqPLVdxYoVdffu3XiI6F8bNmxQq1atIvyap0qVKpowYYK6deumv/76K8q1ZmPL19dXffr0iTDTtUaNGqpQoYKGDx+u2rVrW3ashJQ2bVpVrlw5ocOIF2PHjtXQoUN169atCNc8OHTokJYtW6Z//vlHxYoV04svvpiAUXqYhK724l8XL140X3/9tTl//ny07VavXm0aNWpkvv/+e6f7XrdunXn99ddNtWrVTKFChUzu3LlNoUKFTNWqVc1rr71m1q1bF9fwY8UV3+qEh4ebtm3bOr6N9fHxMXa73fj5+ZkNGzZEaPvFF188cc3AhHT9+nVz/fp1p9r++OOPJlmyZCZv3rxm+vTpZvPmzebnn382L774oqlRo4Z56623TNGiRY2Xl5fT6/fGt5jk+8i2bdtMiRIlInxL9+jf3t7eplatWpFmj7la9uzZTadOnSJt/+ijj4zdbjcnTpyI8zHcdY04T5tZ8uDBA7N69WrTpk0bx6xLX19fY7fbzahRo8zdu3cTOkRLeeIYP/LHH3+Yt956y7Rp08a8+OKLpl27dmbatGnm9OnTse6T8zjhDRo0yPj7+5vLly9H2+727dsmderUUa4F/7h79+6ZQoUKmeDgYPPBBx+YkJAQ88orr5jWrVtHaDdv3jzj7e1thg4dGqccYuLxWWpWrW35pD6j+rxRvHhx8/zzz1udWrSxedJz2hjPy9kV+Y4ZM8aMGzfOHDp0yNJ+rcIYJ13Jkyc377zzTqTtx44dMzabLcpfXjjLmf/HRzNdM2bMGK+/JkroMf7999/NwoUL4/WYCZ0z4g8zW91EunTpnJqtWr9+fdWvXz9GfdepU0d16tSJbWguU6VKFWXOnNnSPm02mz755BN17txZ33zzja5du6bs2bOrVatWkb4NbN68uaXHtsqjtXNHjBihcePGRdt2zJgxypYtm3799VcFBAQ4tr/yyitq3bq1Vq9erYEDB2rUqFHq16+fypYt67L1y2IrJvlK0k8//aQaNWooR44cmjp1qp577jkFBARo1KhRevjwoWrUqKElS5aocuXKmj59unr16uXqFCRJ586di3K94EqVKskYY8k30u64RpwrzuPo3L59W7/99pv8/PxUuHDheJ0VuH79ei1dulSrVq3S9evXFRgYqPr166tZs2bKkyePihYtqqJFi1qy/uHChQs1ffp0HT9+XIGBgapdu7befPPNBFmzNb7H2B2EhobqjTfe0OzZsxUWFiY/Pz+lTp1a169f1+LFizV48GC98cYbGjdunLy8vGLUtyecxytWrNDSpUuVLVs2xy9LDh48qNdff127du2Sn5+fatasqTfffFN58uSx7LjO6t+/vypVqvTUX7ds3rxZVapUccxCjo6vr6+2bt2q9u3bq0ePHvrf//6nzJkz69y5c6pfv77u3r2r3377TRcuXFDdunWdXn/cCq461pP6LVCgQIT769ev1759+5x6j7eKq163wsPDtXTpUn399dc6dOiQLly4oPv378vPz0/p0qVToUKFVK9ePbVo0ULe3vH7Z5YnvR9LrsnX3a/antDvx8eOHdPOnTvVvn37eDmeq/INCwvTli1bdOfOHVWrVk0pUqSQJO3bt0+zZs3ShQsXVLZsWfXp08exz9UePnyo5MmTR9r+aD3/+/fvu/T4/53pOnfuXJce679cNcbffvutpkyZolOnTilz5sxq3LixunbtKl9f3wjttmzZoldffTXentOS689jd/rlp8dL6Gov/lWtWjWzZs2ahA4DbuDBgwemdevWTn2rGBgYaMaNGxdp+6NvQf97lcly5cpFeUXChBaTfI0xplatWiZXrlzm9u3bEbavWrXKBAQEmDt37pjQ0FAzbNgw4+Xl9dQ1mazypG8pXTFrzZ3WiHOF0NBQ07t3b5MxY0ZTqlQps23bNmOMMYsWLTJBQUGO2VOZM2c2X375ZbzFZbPZTIoUKUyrVq3MihUrzP379x37/vjjD2Oz2czy5ctj1GfZsmUjXRV33rx5xmazmdy5c5uuXbuaJk2amJQpU5qMGTOakydPWpGKWwgNDTWbNm0yK1euNDdu3HBs//XXX02nTp1M3bp1zZgxYyLsiy+tWrUyNpvNdO7c2Rw8eDDCvsOHD5sePXoYm80W5Rrhzkqq5/HKlSuNzWYzvr6+xs/Pz+TPn9+cPXvWZMiQweTMmdO0bt3a1K1b1yRPntwEBwebP/74I6FDttz27dtN3759Hb8oyps3ryldurTp3Llzkhnn2Lh3715ChxAnly5dMiVKlHCsx167dm3TunVr06FDB9O6dWtTu3ZtkzZtWmOz2UyxYsXMhQsXEjrkOHPX92NXceertrvSpk2bTK1atUyBAgVM9erVzTvvvBPl+TpnzpxEv4b6jRs3TKlSpRy/jEuVKpXZuHGj2bNnj/H393e8N9lsNvPss8/G22cQm81mmjVrZqZNmxbhNnLkSGO3203Xrl0jbP/vuuNPU7VqVY+qM2zfvt14eXmZoKAgU6ZMGZMlSxZjs9lMwYIFzeHDhyO0TQrP6Ufc8Zefno6ZrW5i69atatmypeP+hQsXVL58eX388ceqWrVqrPq02+2y2+1q2rSphg8frsKFC1sUrTU6d+4co/Y2m00fffSRZceP729nJemXX37RpEmTdPToUaVIkULPP/+8+vXrp0yZMkn6d72zRo0aadu2berXr99T+7Pb7QoPD4+0/cGDB5L+XeP3kfr162v27NkWZeIcq/OV/p3ZOmjQoAgzeSWpYMGCjplDpUqVclxxedq0aVq6dKnluUXl22+/jbQO3o0bN2Sz2bRixQodPXrUsd1ms6lv376xOo47rRHnivN45syZmjVrlsqXLy8fHx/Vq1dPa9euVbdu3dSoUSPVqFFD165d09y5c9WqVSulSZNG1atXj0saTgsNDdWDBw/08OFDhYWFxbm/vXv3RpplP2rUKJUtW1bff/+9Y5bDuXPnVKZMGQ0dOlSLFy+O83FjwhVjfPPmTdWoUUN79+6VMUYpU6bU0qVLFRwc7JhxGBQU5JhN/NNPP8Xb7JI1a9ZoyZIlGj9+vIYOHRppf6FChTRnzhzlypVLQ4YMUbNmzVSvXr0YHyepnsdvv/22ChYsqB07dih58uRq3LixXn75ZeXOnVvfffedY1bJgQMHVLFiRY0cOVKffvppnHOIierVq6tfv36xGjdnVKxYURUrVnRJ37Hx4MEDvf7668qePbuGDRsm6d/336jed319ffXee+851W/v3r01c+ZMp9qeOXNGDRs2jLBWriu54nVrwIABOnHihDZt2qQaNWo8sd13332n5s2ba9CgQZo/f36M4ogLT3s/dtXfEe581XZX5Pzjjz/qxRdflL+/vwoWLKjff/9dvXr10nvvvadly5apUKFCcQk5TlyR74QJE7R//35NnDhR5cuX1+bNm9WuXTtlyZJFuXPn1ubNm5U2bVpt3rxZ9erV05gxYzRt2rS4pOG0ZcuWadmyZVHuezyvmPwdsWXLljjH5iquGONRo0bpmWee0Y4dO5QhQwZJ/36269Gjh55//nl9/fXXSe48dtdffnq8hK724l+Pz4qzYjbco6vNPfpmtnHjxmbPnj1WhGsJm83m+NYpMDDwqTdnrgZsjPt+O7tz507j4+NjfHx8TPHixU3JkiWNj4+PyZIli/njjz/MiRMnTIECBUyyZMnMnDlznOqzQYMGJk+ePObWrVsRtvfv39/Y7fYI6wrOnDnTJE+e3NKcouOKfI0xJkWKFGbMmDGRth/6f+3dd1RTydsH8O8NSK/SESkWFhVRLCgrIgoo2Dt2rNgrdkXBgnVtq65lXdvPrqsiAi52xS72soqKYAEUQQVEEOb9w5esEQIJpEGezzmcozeTe+fJ3Ll3Mpk7c/8+4ziOnT17lr9t4cKFzNLSUiLxlKRgTjxR/yrKr6jSqMdOTk6sdevW/P+PHz+eWVhYsL59+wqk+/z5M7OxsWGenp4Sj6sod+7cYTNnzmQ1atRgHMcxbW1t1r17d7Z//352586dUo1sNTMzY4sXL+b//9OnT4zjOLZly5ZCaYOCgpiRkVGZ4xCXNMp4+vTprFKlSmzp0qXswoULLCQkhJmZmbEGDRqwunXrsnfv3jHGGDt16hTT1NRkkyZNknaYfN27d2c1atTgrwZcnNq1a7POnTvLIFfSJckyNjIyYgsWLOD///Tp04zH47HNmzcXSjtmzBhmamoqlZiK83ObKykpidnZ2UltJeb8/HyWnJxc6IkMWdm5cyfj8XgsLCyMv62gnampqSlQtjwej0VHR4u0X47jWPfu3QVG+RclJiaGmZmZMXV19TLFIQ5pXLdMTExEnmt35syZzMTEpKxhiEWZ7seMSSdeRV61vSB/ko7Z09OT2dnZCawbEhYWxiwsLJihoaHASDhZf3eSRrz29vZs4MCBAttWrFjBeDyewHcIxhgbMWIEs7W1lWhMwsTHx4v9VxFI61pd1HfFpKQk5uzszPT09Njly5cZY/IZ2SqNmBX1yU9lR52tCkJana0bN24s1Eho166dQlSwpk2bMo7jmJeXFwsPD5fIPhX5sQFfX19mZmbGHj9+zN/2/Plz1qBBA9agQQNmamrK9PX12YkTJ0Te54MHD5iuri6rUqUKmzhxIlu0aBHr1q0b4ziO9e7dWyDtsGHDmLm5ucTiKYk04mVMcTuYpd1IunTpEtu7dy87f/48y83NLTLNrl27mJ2dnSTCEZk06rGOjo7A41E3b95kHMexXbt2FUo7a9YspqenJ5HjiuPGjRts6tSpzM7OjnEcx9TU1BiPx2OzZs0Sq0OlQ4cOzN3dnf//rKwsxuPxiuy0XbJkCdPU1JRI/sUhjTJW1C87jDFmZ2dXaGoHYQIDA5mNjY1Ej//+/XuxFsGUBEmWsbq6usDik/Hx8YzjuCIfYVy9erVczmlptLmECQ4OZpUrV2Y8Ho+pqKgwNzc3mf/w3aVLF/bLL78IbCsq5vz8fGZra1uobgqzbds2pqamxtzc3NiHDx+KTFOwIJyxsbFMz2tpXLc0NTXZypUrRUq7cuVKmZ/bynY/lka8xS1c8/P3KTc3t0KL70qbNGJW5I4pacSrpaXFVq1aJbBtwoQJha6RjH2vx7L6HnHu3Dn29u1bmRyrOLJug0ijjA0MDAqVcYFPnz6xZs2aMX19fRYbGyuXzlZpXavL+9SCFRF1tioIaXa2Fvi5keDt7c2fe0lerl+/zvr168fU1dWZvb09W7duXZlW8lbkX2eNjY3Z1KlTC20/fvw44ziO2dnZlWr101u3bjF3d3emoqLCH203fPhwua+ILq14FbWDWVqNpIyMDNaiRQuBuXcsLCzY1q1bC6WV17xDkq7H2traAl/u3rx5w6ysrIpciXXlypUij3qXlqtXr7JJkyYxa2tr/kixjh07irS6aUREBOM4ji1ZsoS/zcHBgS1btkwg3bdv31jDhg1ZkyZNJJ5/UUi6jBX1yw5j388/YY30n61atUriHSoHDx4s1/XY3t6ejR07lv//vLw8FhcXV+S+Jk2axKpXr16mfJeGNNpc5ubmzN/fX2DbwoULGcdxrHnz5mzBggVs0qRJzM7Ojuno6BSaC1iaqlatysaMGSOwTVjMY8aMYTVq1BB536dOnWKGhoasVq1aAj8i5uXlsbFjxzIej8ccHR3ZixcvyhRDaUj6utW4cWPm5ubG8vLyik2Xn5/PPDw8WOPGjUt9rNJStvuxpONV5FXbC0g6ZkXvmJJ0vDVr1mQTJkzg/z89PZ0ZGhoWukYyxtjo0aOZmZlZqY8lDh6PpxAr1MujDSLpMnZ2di52Tv3Pnz+zpk2bMjMzM/5gHVmTdMyK+uSnsqPOVgUhi87WAj93urZo0YKdPHmy1MeRhLdv37K5c+fyO0WnTJlSqoVgFPnXWRUVFbZ+/fpC2wsW1vnx8T5xnT59mi1YsIDNnz9f7h3oBaQZryJ2MEurkTRr1izG4/HY5MmTWUREBNu4cSN/Yv9hw4YJfOmT9yTvkqrHDRs2LPSIojCDBw9mderUEfsY0hITE8PGjx/PqlSpInJZTJ48mXEcx5o1a8a2bt3KJkyYwBo2bMhSU1NZQkICi4iIYC1btmQ8Hk9iv4CXlqTKWFG/7DAm2pftAtKoc/LqbC1Q1jIePHgwMzc3Zzk5OcWmy8/PZ9WrV2ft27cva5bFJo02l5qaWqEOCyMjI+bj4yOwLTMzkzk4OLCOHTuW+lilydvSpUsFtgmL+bfffmMaGhpi7f/hw4fM1taWmZubs9jYWPbhwwfm6enJOI5jHTp0KPQkiqxJ6rp19OhRpqKiwurXr89WrFjBIiIi2IULF9iNGzfYlStXWGRkJJs3bx6rV68eU1FRKVM7p6yU7X4sqXjFuf5/+fJF5B/mpEFSMZeHjinGJBfvpEmTmL6+PouIiGDPnj1jHTt2ZBoaGqxatWrs4sWL/HTXrl1jWlpahQZySIs45540ybMNIqkyHjVqFNPV1WVpaWlC06SnpzMnJye5T+0mqZgV9clPZUedrQqC4zjWqlUrNnbsWDZ27Fg2ePBgxuPxWIcOHfjbfvwbN26cSPss7qL9c6erq6urJEMqlZycHLZ+/Xqmrq7OKlWqxP7++2+x3q/Iv86WtFp9RESE2PvMyclhnTp1KrTqYNeuXeU2N1wBacT7I0XrYJZWI8nBwYH16dNHYFt+fj4LCQlhPB6P9ejRg9/hKu/O1gJlrcfTpk1jWlpa7P3798Wmy8jIYIaGhqxfv35lya7UXLhwQeS0e/fuZfb29vw6/OMKohzHMUtLS7Zv3z4p5lY8ZS1jRf2yw9j3ujx79mx28+bNEv9mzpwpUp1zdHRkdevWFenPxsamXNfj27dvs8mTJ7MnT54Um27Pnj3MwMCgyFH60iaNztYqVaqwefPm8f+flpbGOI5jO3fuLJR2/vz5Mn3cWl9fv1DbKDc3l924caPQStvLly8vVd6Sk5OZi4sL09XV5U+vMm3aNJHmPpaVsl63GGPsxIkTzMXFpdB1uuD/HMcxFxcXsadIkhZlux+XNd7yuGp7WWMuTx1TjJU93rS0NObg4CBQb9evX8/27t3LVFRUmIuLC3N1dWVqamrMwMCgxHuZpEizs7W8tUHKWsbPnz9nGzZsYI8ePSo2XVJSEvP19WUeHh5lya5ElDXmH5/8nDRpksI8+ansVOW9QBf5z5kzZwqtFhgeHl5kWo7jRF4BVhgNDQ1MmDBBYLVNeXr58iXWrFnDX22vT58+aNCggVj7sLOzQ2xsbJGv6erqIioqCt7e3vD19UW/fv3KnGdx7du3D/fv3xfYlpmZCY7j8McffyAyMpK/XZQyXrZsGcLCwjBy5Ej0798fmpqaOHLkCEJDQzFixAjs3LlTKnGIStLxAkBubi569OiBY8eOAQAYY+A4Dp07d8aOHTugra0t2SDkLCEhASNHjhTYxnEc5syZA1tbWwwdOhT9+vWT+WrewkiiHgcGBvJXpC/O6dOn0aJFCwwbNqzU+RWHnp5eiWnU1NRgb2+P0aNHo2/fviWm//r1K9TV1eHn5wc/Pz/cvXsXt2/fRnJyMnJzc2FgYIC6devC1dUVqqqKccuWRBkHBQUhIiKCvxo8Ywzr1q1D5cqV0aJFCzRs2BAqKiq4efMmtLS0EBISIvE4ihMaGorQ0NAS0xVcf0rSunVrrF+/Hjk5OahTpw6MjIyEps3Pzxcrr9JQljKuV68e6tWrV2K6Xr16oVevXmXKZ1n8eH8quC9t2LABUVFRhdKKcn9yc3PD4cOHERQUBADQ1taGiooKNDQ0ityfLMu5evXquHv3rsA2VVVVNGzYsFDaa9euoVq1amIfw9TUFOfOnUOfPn1w5MgRzJgxAwsXLix1niVNEtct4Htdbt26Nd6/f4+HDx8iOTmZfx03NTVF7dq1YWJiIunsl0pFvh8XRRLxKvKq7UWRRMyTJ0+Gk5MTkpKSYGBgUGQafX19/PPPPxg0aBC+fPlS1myXmiTiNTAwQGxsLPbt24d3797B3d0dTZo0AQDo6Ohg5cqVeP36Nbp06YK5c+eiZs2aEo9DGFHaE6VRntogkuoPGD58eInpzMzMEBERUap8SpIkYq5duzbOnz+P8ePHY/Xq1cjPz4eWlhYCAgKwcuVKgbSbNm2SWN5JCeTc2UukSNxfyL58+SLF3Ah38eJF1q1bN6aiosLMzc1ZcHAwS05OLtW+FPnXWWmsVu/o6Mh8fX0LbZ89ezZTUVFhqamp0ghFJNKIl7H/5sAbNWoUu3z5Mrt9+zYLDg5mampqch1RwXEc27Rpk8T3a2try4KCgoS+vm/fPqaqqsoCAgLkOrJVkvVYUXEcx1xdXdnAgQOF/vn5+bHatWszHo8n0og9PT091r9/fxYeHl7iY9fyJukyzsrKYlu3bmVLly4VWLQxPDyceXp6MgcHB+bn58cePnwoieyLbNu2bWL/iWL9+vWMx+OVOHrywIEDVI+lTBr3p8uXLzMej8dGjx7Nvn79yhhjrEGDBmzBggUC6TIyMljNmjVZq1atpBJbUYKCgpi2tnaJCzM+ePCAqampsWnTpom036JGe1+/fp15eXkxXV1dtmfPnkKvy5qynNM/UraYlS1expQvZmWIl+M4ZmxszOzs7ET6q1atmlj7V/Q2iDKU8c+kFXNGRgZ78+aN0AWViexQZ2s58OXLFxYdHc0OHTokVgU8e/Yse/PmjVjHkaVdu3axxo0bMx6Pxxo2bMi2b9/O/4JSWuXxsYGy0NTUFFi4oMC9e/cYx3EK8Xi9pClqB7O0Gkk+Pj6sYcOGxab53//+x3g8Hv+xKFmSRj1WVOL8gNW2bVvm5ORUYrrevXszPT09xuPxmKGhIRs0aBCLiopi3759K2t2JUaZyliaHj16JNKj6gcPHmQcx8koV99JsoxbtmzJPD092R9//KF058nq1asZj8dj1apVY3PnzmVDhw5lv/zyC7t58yY7f/48W79+PatVqxZTV1dnly5dklm+3r9/z0xMTJi9vb3QDs+YmBhmY2PDDA0NRW47/vwY/c+P0/+8TZb3J2lft16+fMkiIiLY7t272aFDh9iFCxfYp0+fJLb/0lC2a7U04uU4jqmoqDA/Pz+ZLmInKipj6cebm5vLUlJSpHqMonAcx5ycnFj79u1F/hOHorZBpFXGCQkJbMOGDQLxZmZmsjlz5jBfX1/WtWtXtn79ernUH2Wrx8qMY4wxeY+uJd8fW+nbty/8/f2xaNEi/vb4+Hh4eHggMTERjDGoq6tj2bJlGDNmTIn7LBhGLorExER07twZN2/eLHUM4uLxeFBRUUG7du3g7u5eYnqO4zBx4kQZ5Ez2srOzcfHiRXz69Alubm4wNTUV6X08Hg8bNmxAQECAwPZXr17B2toax48fh6+vrzSyXCaljRcAtLS0sHDhwkLnwv379+Hk5IRz586hefPmks5yiXg8HurWrQtra2uR31MwFUJxtmzZgtmzZ+PgwYNo1qyZ0HRr167FuHHjwHEc8vLyRM5DWUmjHvN4PPB4PHTv3h2zZ8+Go6OjpLJbJsLqW1F+//13TJs2DVlZWSWm/fr1KyIjI3HgwAGEh4cjIyMDhoaG6NatG3r27ImWLVuCx+NJIoRSkce1+tu3b0hLS5PL47jVqlVDaGioSI+47969G7Nnz8bz589LTJuTk4PLly+jbt26qFy5stB079+/x4MHD9CiRQux8l0WkixjHo8HW1tbxMfHw9LSElOnTkVAQECRj9MrorLcnwDg8uXLmDNnDs6cOcN/HLPg0VDGGBo1aoSVK1cWez2XhitXrqBDhw748OEDmjZtisaNG0NPTw/p6em4cuUKbt68CT09PRw6dAitWrUSaZ/bt28XOx/+/v5iv6c0pHXdunDhAiZMmIDbt2+j4CtUQfnyeDy0atUKc+fOxa+//lq2AEpBme7HgPTibd26Ne7evYvk5GR06tQJs2fPLtWUE9IgrfM6MTERERERqFq1Ktq2bQsAyMrKwpIlS3D9+nVoamrCy8sLQ4YMgZqaWpnjEJW04r169Sp+//13ZGZmonPnzvzr0syZM7FixQrk5ubC1tYW69atg4+PT5njEIU4bczSUNQ2iDTK+O7du2jevDk+f/4MjuPQqVMn7N+/H66uroiNjYW5uTk+f/6MjIwMNGrUCGfOnJHpNHTSiLlVq1aYNGkSf3quAleuXMHq1avx7Nkz6Ovrw9fXFyNGjICWllaZYiAikmtXL+EbOXIk09TUZO/evRPY3qFDB8ZxHJs5cyY7duwY69WrF1NRUWGxsbEl7pPjONa9e3eWnZ1dbLqYmBhmbm7O1NXVyxSDuKT1mLmiOn36NLOwsGDTp08X2P7ixQv+ZOQcxzENDQ32+++/i7RPYQu5REZGMo7j2Jo1a+T2CJ804mVM+OjCxMREiSy8VVqKsIrotWvXZL4yrjTqMcdxrE2bNvzF+7p06SKXx09/9vjxY5aeni5S2pCQEGZoaCj2MbKzs9mRI0dY3759mb6+PuPxeMzU1JSNGjWKnT17Vuz9SYK0rtVXrlxhffv2ZZ07dxZ4FH/GjBlMXV2dP0IwMjJSWqEVieM4NmLECHbo0KES/4YPHy6Ve9Pt27clvs/iSLKMC66Ft27dYp06dWIcxzEzMzO2bNkyuS/cWEBa96cfpaWlsXPnzrH9+/ezXbt2sePHj7OEhAT+6/KYuunt27dswoQJzMbGRqA8q1atykaPHi2Qv/JOGtetmJgYVqlSJVazZk22YsUKdvr0aXb16lXm4+PDPD09WWhoKHNycmIqKips9erVMohSkDLdjxmTXrwbN24stIhwu3btBKa7kRdpxHznzh2mp6fHT9+lSxeWm5vLGjVqxHg8HrO0tGS6urqM4zjWuHFjmV7HpRHvzZs3mbq6OlNXV2fVq1dnqqqqbPTo0Wzt2rWM4zjWpUsXFhgYyH755RempqbGrl27JoNIFeN7RAFZtkGkUcY9e/ZkRkZG7MSJE+zs2bPM3t6eDRo0iNnY2PCnpsrPz+dPrTBlyhRphylAmteuH506dYqpqakxdXV15uzszKpVq8Y47vsI6p8XxyTSQSNbFYSjoyNq1qyJw4cP87elpqbC3NwcHh4eiI6O5m9v2LAhrKyscPTo0WL3uX37dgQEBMDFxQVhYWEwNDQslGbLli0YPXo0dHV1cejQIZF+XZGUly9fiv0eGxubYl9X5F/gR40ahW3btiEhIQHGxsb87R07dkR4eDhmzJgBV1dX7Nq1CwcOHMD169fh7Oxc7D55PJ7QydTZTwu3FPxfVqMepREv8D3mWbNmoUuXLgLbU1JS0LZtW6xevbrQiCFZjEiQ9i/SooiMjESvXr3w8eNHmR1TWvV4w4YNGDBgAH/xvuTkZPj6+iIoKIi/kIEiCw0Nxbt37wpNSi+OnJwcREVF4cCBAzh27Bg+f/4Mc3NzvH79WoI5LZk0yjg2NpY/8svKygovX77E8OHDUatWLYwdOxadO3dGtWrVEB4ejhcvXuDixYto3LhxqfIvroLrqqjNI0lfV8t7Pf75Wnjr1i2EhIQgLCwMRkZGmDhxIsaOHQtdXd0y5bkspHF/UvSniX6WkZGBT58+QU9PDzo6OnLLh7RI47rVpk0bxMXF4e7duwKjoMLCwtCnTx+kpKRAXV0dc+fOxeLFixETEyPT+5Wy3Y+lGW/B9Ss7O1sgbk9PTwQFBcnlCSpAOjH7+fnh1KlT2L17N9TV1REQEIBmzZrh9OnTiIyMRK1atcAYw4YNGzBmzBgEBgbKbFFlacTbsWNH3Lp1C5cvX4aVlRVevHiBTp06IT4+Hn369MGGDRsAfB/Z6+zsDHt7e5GeRCsrRfgeAci+DSKNMra0tET//v2xZMkSAMChQ4fQs2dPLF26FIGBgQJpe/bsiRs3boj0hJKkyOLaBQCNGjXCmzdvcPHiRf7Cl1FRUejWrRuGDx+OFStWiJ0PIiZ59vSS/+jo6LDQ0FCBbeHh4YzjOLZ48WKB7XPnzmXGxsYi7ffUqVPM0NCQ1apVS2BhhLy8PDZ27FjG4/GYo6Mje/HiRZljUASK/At8nTp1WOfOnQW2vX//nqmqqjIvLy+B7Q0aNGAdO3YscZ/SWshFEqQRL2PC54j7cU44ecwRJ81fpL99+8bCwsLY0qVL2YYNG4qcR2zDhg2sUqVK7JdffpFKHmTp58/y5xEm3t7eFXI+YmHevXvHVq9ezQwNDcv9CP8CHTp0YFZWViwxMZEx9n2+7bp16zJdXV02fPhwfrrMzExmb28v9txkZcFxHAsMDGRnz54t8W/ixIkil4mo9VhVVbVc12Nh18LY2Fj+SFdDQ0M2d+5c9uHDBznkUDr3J44T/WkiMzMzmT9NRMpOR0eHzZ8/v9D2J0+eMI7j2PXr1/nbmjZtynr06CHL7EmFst2PhV2/fo67RYsW7OTJk3LIoeRZWFiwqVOn8v9/8OBBxuPx2PLlywul7dGjB7Ozs5Nl9iTOwsKi0MKzf/31F6tUqRJ7+/atwPbg4GBmYGAgk3xJe2SrsrRBGGNMTU2NrV27lv//p0+fMo7j2JEjRwqlXbFiRYW4Hxd1rebxeGzu3LmF0g4fPpzZ2NjILnNKTFXenb3ku5ycnEIjC27fvg2O49CoUSOB7UZGRiL/2tSqVSvExMSgbdu2aNq0KSIiImBra4sePXrg9OnTaN++PXbv3l2hRjV07doVR44c4f8S3bhxY4X5Bb5v374C265cuYK8vDx4eXkJbO/QoQPWrVtX4j5lNfdZaUgjXgDYunWrxPJYHnz8+BHe3t64efOmwPxwAwYMwObNm6GqqoqpU6di+fLlaN68ucDo+IpCQ0MDEyZMwIgRI/j12sPDA82bN0dQUBA8PT3lnUWJS0lJwd9//40DBw7gwoULyMvLg6WlJQYOHCjvrEnEjRs3MHToUFhZWQEA7OzsMHHiRAwfPhzBwcH8dFpaWujTpw9WrVols7zZ2NigYcOGIs1X9vr1a5HmaaZ6DDg7O+PIkSP8ka7z5s3DypUrMWrUKIG56mVBGvenrVu3IiAgAF5eXiI9TfTPP/+ULQgiczwejz8P749ycnIAAJmZmfxt4rRryhNlvB8DRcfdunVrNGnSBJcuXZJ39sokNTVV4D5Wr149MMZQo0aNQmldXV0RFhYmy+xJXGZmZqHr86FDh+Dp6Qlzc3OB7VpaWvjy5YtM8lXUtUVSlK0NYmJigjdv3vD/X6VKFWzevLlQnwrw/bP58QmXiiI9PR2MMdSuXbvQa7Vq1cK2bdtknyklRJ2tCsLS0hJxcXEC265cuQKO4wo9uvb27VuxJjWuVasWrl69ig4dOqBFixYwNjZGfHw8pk6dikWLFgl9DF3aBg8eLFZ6juOwZcsWkdIW1Sj69ddf5fr4j7Q61BWVtOJV1A5maTWS5s6di9jYWMyePRvdunUDj8fD0aNHERoaisqVKyMxMREHDx5E//798eeff6JSpUpSyYcw0qzHP6vIX3YAIDk5GYcOHcKBAwdw8eJF5OXlwcLCAiNGjICfn5/MF9QpII0yVtQvOwDw4sULkdP26dMHffr0KTEd1eP//NzpunTpUpl3tkrj/uTv74+qVauie/fuaNasGSIjI/mP/eXn52PChAlYt24dateujWPHjsHW1lZi8ZDCpHFOe3h4YOfOnZg0aZLA+bN161ZwHIfq1avzt+no6OD9+/fiZbqMlO1+LMt4CxQVtyxJI2ZF7piSRrwODg44ffo0f8Ghq1evIjIyEt26dSuUNjo6GnZ2dmLlQREpchtEGmXcpEkTHDt2DAsXLgQAaGpqYsiQIUWmjY6OFrh2y4K0rl1ZWVn48OEDgO+drQXv/dn79+9hZGQkVh5IKcl1XC3hGzFiBKtcuTL/cf5///2XaWhoMDc3t0JpXV1dmbOzs9jH+PLlC+vSpQvjuO8LbsmbtrY209HRKfFPQ0OD/1h4SRT58R9bW1s2btw4gW3t27dnKioqLDU1VWD7jBkzmL6+vgxzJ3nKFq+02NnZsV69ehXa/vvvv/OnSSjqsUZZkWU9/tmP9bq8evPmDfv999+Zu7s7U1VV5S8mVLAgVn5+vryzKJUydnFxEZga4MqVK4zH4xX52K23tzdzcHCQaEyypkz1WNxHIW/duiWBCMQjzfvTw4cPma2tLTM3N2exsbHsw4cPzNPTk3Ecxzp06MA+f/4siRBICaRx3Xrw4AHT1dVlVapUYRMnTmSLFi1i3bp1YxzHsd69ewukHTZsGDM3N5dWeEVStvuxPOMtIOuF7qQRc9euXVndunVFOv6vv/7K3N3dyxqGyKQR7+7duxnHcaxnz54sKCiImZqaMnt7e9a8eXM2dOhQ9ujRI/bvv/+ycePGyf3eLCmK3AaRRhkfOXKENWrUiN29e7fYdOfOnWMcx7EFCxZIKhyRSOvaVdQUe0Ut1ujs7FxkHxORPOpsVRBv3rxhJiYmTEdHh7m4uDBtbW3G4/EKdQjevHmTcRzHJk2aVOI+f16F/ubNm+z69evMy8uL6erqsj179shtpXpRfPjwgc2ZM4cZGhoyFRUV1rNnzxLfU1Ij6edOV1dXV0lmuViy6FBXJMoWr7Soq6uzFStWFNp+584dxnEc++233+SQK9FJox7/TB6rektKQWPI2NiYBQQEsFOnTrG8vDx5Z0sspSljZfuyo0z1eNq0aQrXnviZtO9PycnJzMXFhenq6jI7OzvGcRybNm2aQvx4Qr4rzXWLse8/Dri7uzMVFRXGcRzT1tZmw4cPZ1lZWVLOcdkp2/24NPGePXuWvXnzRuRjyHLVdlGUJmZF75gqTmnrcVBQENPU1GQcx7HatWuzx48fs+fPn7Nq1aoJrPvQpk0blpOTI+UopK88t0FKW8aiePXqFbt9+za7ePGixPYpCaWJOTg4uMi/EydOCKSLiIhQuHpckVFnqwJ5+vQp6969O3NwcGCtW7dmUVFRAq/n5+czT09PZmBgINJIEGG/cHAcJ9eFhEqSkpLCpk6dyvT09JiamhobMGAAe/z4sUjvVeRf4KXRoa7IlC1eaRF2Tr969YpxHMeOHz8uh1yVrCz1uLx/2RHHkCFD2IkTJ9i3b9/knRWxlaWMGVOuLzvKVI/t7OzY4cOHZZPBUpLF/UnRniYi35X1usUYY6dPn2YLFixg8+fPLxeLQynb/VgSZSyKiIgIpqenJ/H9loYsYlakjilJxPvlyxeWkpIisC0rK4vt2LGDLV68mEVGRjLGFOOcLqvy2Aaheiy9mBmrGOd1eUCdrRWYIq9UX5RXr16xcePGMS0tLaaurs4CAgLY8+fPxdqHov8CL+kOdUWnbPFKA8dxbPbs2YVGoUdGRjKO49iaNWsUaoS6JOqxOBSpkaQsJFnGyvJlR5nqcaNGjRiPx2NeXl7swYMHUs5p6Un6/lTenyaq6CRx3crJyWGdOnUqNFCha9euLCMjQ0o5Lz1lux9LKt7ytGo7lXHFjldSylMbhOqx9M/ryMjICnFelwfU2Urk7vnz52zYsGFMXV2daWlpsXHjxrFXr16Val/l8Rd4QoojbIR6UaPU5TlCXZL1mLHy1UhSFpIuY1FVhC87ylaPt23bxqpUqcIqVarExo0bx9LS0iSfWQVTXp8mqugked1auHAh4ziOjRo1il2+fJndvn2bBQcHMzU1NdavXz8J57z0lO1+LMl409PTWePGjQvV3YEDB7Lc3FzGGGNTpkxhHMcxd3f3QnM8ywqVccWOV9LKQxuE6jGd1xWRqrwX6CLK69GjRwgNDcXevXuhpaWFCRMmIDAwECYmJqXeZ4sWLUROGxkZiV69eom0yjAh8rJ161Z5Z6FY0qjHHz9+hLe3N27evAnGGIDvq2kOGDAAmzdvhqqqKqZOnYrly5ejefPmOHz4sKTCIUWQRhkDQF5eHiIiIvD48WPo6emhWbNmcHR0FEizceNGjBkzRuYrxUqastVjf39/9OjRA4sWLcKKFSuwe/duLFiwAAEBAUWujFsRKHoZKxtpXLf27NkDHx8frFu3jr+tXr16+PbtGxYtWoTVq1ejcuXKksh+qSjb/Vga8Sryqu0AlXFFj1daFPn+RPWYzusKTa5dvUSp8Xg8ZmRkxIKDgyU66oV+1SFEdqRRj8ePH89UVFTYnDlz2J07d9i9e/fYggULmJaWFps0aRLr0aMH4ziODRgwoNzP5VkeSKOMy8uoA2UhrfsxY4wlJCSwjh07Mh6Px7S0tJiurq7AX3kftUwUkzTOaU1NzSIXmbl37x7jOE7u87cq2/1YGvEq8qrtjFEZS4Iix6uMqB6nSWSfdF4rJo6x/+/6JkTGeDwe1NXVRf41ieO4EkehluZXHXmOQiCkvJNGPa5WrRqaNGmCPXv2CGxfu3Ytxo0bB47jEBISgtmzZ5c630R00ijjCRMmYO3atZg1a1ahUQcjRoyQ+6gDZSONMgaA/Px8rF+/HsHBwUhPT0f79u1haGhYKJ0ij7oh5ZM0zmkej4cNGzYgICBAYPurV69gbW2N48ePw9fXt9R5Litlux9LI14NDQ0sWrQIEydOFNh+9+5d1K9fH8uXL8ekSZNKneeyojIuXnmPVxlRPS4ZndflF00jQOTG399f4vtU9McGCKlopFGP37x5AxcXl0Lb3d3dAQDLli2TayNJ2UijjMPCwtCjRw+EhITwtzk6OkJfX5/fKJw3bx41CmVEGmV8/PhxTJkyBY8fP4a7uztWrlwJZ2dniR+HkKJI45wGgMTERMTGxgpsS0lJAQDExcUVeq1BgwZSyUdRlO1+LI14c3JyoK2tXWi7kZERAMDBwUHixxQHlXHZKXK8yojqsWTQea2YqLOVyI00RrLQF3hCZEsa9VjRG0nKRhplTI1CxSLJMr5//z4mTZqEU6dOwdbWFgcOHEC3bt0ktn9CRCGt0dKhoaEIDQ0t8rUJEybw/80YA8dxyMvLk0o+iqJs92NplXFxHerPnj2Ta4c6lXHZKXK8yojqsWTQea2YqLOVKIRnz55h//79ePDgAVJTU2FiYoL69evDz88PVapUEXk/9AWeEPmRVD0GFLuRpMwkVcbUKFRcZS3j+vXrQ1tbG6GhoZg4cSLU1NRkkGtChJPUdas8TXehbPdjScZbXIf6+PHj+Yv8yaND/UdUxhU7XmVE9ZjO64qG5mwlcvXt2zdMnjwZ69atQ15eHjQ1NWFoaIj09HRkZWVBVVUVkydPxvz586GiolLi/oTNp/X69WtUrVoV4eHhaNu2rbTCIUQpSaMeC1ux/Me5mAv+L89GkrKQRhnPmjULXbp0EdiekpKCtm3bYvXq1WjWrJnAa9QolC5JlXFAQAAWLFgAU1NTGeaekMIkfd0qD5TtfizpeLdv3y52HqQ1ZYUwVMYVO15lRPWYzuuKika2ErkaMGAA9u7di0GDBmHixIlwdHTkv/bw4UOsWbMGixcvxuvXr0W+cNKvOoTIlqTrcXkaPaQspHGtLi+jDpSFpMp406ZNssguISWSxnVL0Snb/VjS8cq6w6U0qIwrdrzKiOoxndcVFiNETsLCwhjHcWzhwoXFpluyZAnj8Xjs2LFjJe6T4zjG4/GK/OM4TuD1gn8TQkpPGvWYKBZplPG2bdvE/iPSQ/WYVDTKeE4rW8zKFi9jyhezssWrjJSxjJUxZmVF0wgQuenRowdu376NJ0+eCB32XqBOnTqwt7fH4cOHi01XHh4bIKQikUY9JoqFyrjiozImFY0yntPKFrOyxQsoX8zKFq8yUsYyVsaYlRVNI0Dk5ubNm+jYsWOJFxkA8PX1xcGDB0tMRx2nhMiWNOoxUSxUxhUflTGpaJTxnFa2mJUtXkD5Yla2eJWRMpaxMsasrHjyzgBRXikpKbCzsxMpbdWqVfnzrhJCFAfV44qPyrjiozImFY0yntPKFrOyxQsoX8zKFq8yUsYyVsaYlRV1thK5ycrKgqampkhpNTQ08PXrVynniBAiLqrHFR+VccVHZUwqGmU8p5UtZmWLF1C+mJUtXmWkjGWsjDErK5pGgMhVYmIiYmNjS0yXkJAgg9wQQkqD6nHFR2Vc8VEZk4pGGc9pZYtZ2eIFlC9mZYtXGSljGStjzMqIFsgicsPj8USaqwQAGGPgOA55eXlSzhUhRBxUjys+KuOKj8qYVDTKeE4rW8zKFi+gfDErW7zKSBnLWBljVlY0spXIzdatW+WdBUJIGVE9rviojCs+KmNS0SjjOa1sMStbvIDyxaxs8SojZSxjZYxZWdHIVkIIIYQQQgghhBBCCJEAWiCLEEIIIYQQQgghhBBCJIA6WwkhhBBCCCGEEEIIIUQCqLOVEEIIIYQQQgghhBBCJIA6WwkhhBBCCCGEEEIIIUQCqLOVEEIIIUSJxcfHg+M4+Pj4yDsrhBBCCCGElHvU2UoIIYQQIkd37tzByJEjUatWLejo6EBdXR02Njbo1q0bdu7ciezsbH7a4OBgcByHvXv3Ct2fKGlkTUNDA7a2tvLORrGSkpLAcRw8PDzknRWxbNu2DRzHYeDAgTI9bnkoU0IIIYQQeVCVdwYIIYQQQpQRYwzTpk3DsmXLwOPx0KxZM7Rs2RL5+fmIi4vD8ePH8ffff2Pp0qW4d++evLNLFJSjoyMCAwPRuHFjeWeFEEIIIYSAOlsJIYQQQuQiKCgIy5YtQ82aNXHgwAHUq1dP4PUPHz5gxYoV2L9/v5xySMqDRo0aoVGjRvLOBiGEEEII+X80jQAhhBBCiIzFxcVh8eLF0NHRwYkTJwp1tAJA5cqVsWDBAkRHR8ssX/fv30e/fv1gZmYGdXV1NGzYEP/880+RaWNjY9GlSxcYGxtDXV0d9vb2CAoKQmZmJj9NwXywX79+xcuXL8FxHP/vx8feY2Ji0KlTJ5iYmKBSpUqwtLREp06dsHPnzkL78vHxwY0bN9C1a1cYGRlBU1MTbm5uiImJKTKfp06dQuvWrWFgYABNTU04OTnht99+w7dv3/hpzp49CwsLCwDAuXPnBPIZHBxc7GcWGhoKjuOwePHiIl//3//+B47jMGfOHABAamoqevToAUdHRxgbG0NVVRVGRkbw8vJCWFhYofcPHDgQHMchJiYGmzZtQoMGDaCurg4tLS1+3jmOw4gRIwTet3jxYri6usLKygrq6urQ1taGo6Mj5s6di6ysLIG0P+7j3Llz6NKlCypXrgxNTU20aNECN2/eLFQOJZUpIYQQQoiyopGthBBCCCEytn37duTl5WHIkCGws7MrNq2NjY1M8nTt2jU4OzujVq1a6N69O969e4eDBw+iQ4cOePr0KaytrflpIyIi0KVLF3AcB29vb1hYWODq1atYsGABIiMjcf78eWhpaUFfXx+BgYFYtWoVtLW1MWzYMP4+Ch57j4qKQvv27aGtrY127drByMgIz58/x8mTJ3Hq1Cn0799fIJ8XLlyAi4sLGjZsCD8/P7x8+RJRUVHw8vLCpUuX4OzszE/7559/IiAgADo6OvD29oaBgQHOnz+PyZMn48yZMwgLCwOPx0PVqlUxcuRI/PHHH7CysoKfnx9/H7/++muxn9vAgQMxZ84cbN++HdOnTy/0esGcqoMHDwYApKWl4fjx43BxcUHjxo2hp6eH5ORkREdHo1OnTti1axf69OlTaD9+fn748OED2rZti+bNmyMpKanYfEVERODjx49wc3ODqakpsrOzcfnyZcybNw9Xr15FVFRUoffs378fGzduhKurK/r27Ytnz54hMjISPj4+eP78OXR1dUUqU0IIIYQQpcYIIYQQQohMtWzZkgFgUVFRYr1v7ty5DABr3749CwwMLPLP1dWVAWB79uwRaZ8vXrxgAJiBgQHbtm2bwGuhoaEMAJs+fTp/W2ZmJjMxMWF6enrs+vXr/O35+fksMDCQAWBTpkwR2I+6ujqzsbEp8vhubm4MALt//77A9vT0dLZkyZJC+TQ1NWVHjx4VSLtlyxYGgPn6+vK3JSYmMnV1dWZtbc3i4uL4279+/cp69erFALB169bxt799+5YBYC1atBDySQnXvn17BoBdvnxZYHtCQgLj8XjM29ubvy03N5d9/fq10D7i4uIYj8djzs7OAtv9/f0ZAObu7s6eP39e6H1nzpxhANjw4cMFtmdmZhZKm5eXxz/37ty5U2gfVapUYeHh4QLvCQgIYADYhg0bBLYXV6aEEEIIIcqMRrYSQgghhMhYwajEH0eLiiM8PBzh4eGSzBKaNGkCf39/gW2+vr6YOXMmnjx5wt925MgRvHv3DjNmzBCYK7TgUfqdO3fizz//xOLFi8HjlTxj1cePHwEABgYGAtv19fUxderUQumdnZ3RsWNHgW0DBw5ESEgIoqOjkZ2dDQ0NDezYsQNfv37FrFmzUL16dX5aNTU1rF27FgcPHsSmTZswatSoEvNYkiFDhiA8PBzbtm1D06ZN+dt37NiB/Px8DB06lL9NVfV78zszMxNXr15FXFwcPn78iE+fPkFLSwtxcXFFHmPJkiUljoL+UcE0A3fu3MG9e/fw/v17fPr0CV+/fgXwfSoLJycngfe0b98e7dq1E9jm6+uLTZs2CZwDhBBCCCFEOOpsJYQQQggpZ/bs2YNevXoV+VpwcDBCQkIkchwjIyMA3xfrKhAbGwsAaNasWaH0qqqqcHFxQXh4OF68eCHQySlMhw4dcO/ePTRu3Bj9+vWDm5sbGjRoACsrK5HzyePxULt2bSQkJOD58+eoXbt2sfk0MjKCg4MD7t27h9zcXFSqVEnkYxWlffv2MDc3x759+7Bq1SpoaGgA+N7ZamxsjM6dO/PT5uTkYObMmfjjjz8KzZ0KAOrq6mXKS4Hw8HBMnDhRaOdtTk6OSPsp6hwghBBCCCHC0QJZhBBCCCEyZmpqCgBISEiQc06KV9AJyRjjb0tPTwcA6OnpFfmegu0F6Uoyb948fufwsmXL0KlTJ1StWhUNGjTAxYsXRc5rwcjYL1++iJzP/Px8fPr0SeRjCKOqqgp/f3+kp6fjyJEjAIBLly7hyZMn6N+/P9TU1Phpx48fj99++w1WVlZYv349bt68ieTkZOTk5Ehsft5Lly6hc+fOSExMxJQpU3Dq1CnEx8cjMzOTv1CXqIo6BwghhBBCiHDU2UoIIYQQImNNmjQBABw/flzOORGfvr4+AODt27dFvv7mzRsAhacFEEZFRQVz5szB69evcf/+ffz111/o1asX7t69C19fX6HHEXZcY2NjkfPJ4/GEdsaKa8iQIQC+L4gFfF8EDYDAFAK5ubn466+/YGRkhGvXrmHkyJFo0KABTE1Nyzy69kebN29GXl4e/vzzTyxduhStWrWCjY0NtLS0wHGcxI5DCCGEEEIKo85WQgghhBAZ69u3LziOw5YtW/Ds2bNi06ampsooV6Jp2LAhACAmJqbQaxkZGbhz5w4MDQ0F5hdVVVUt8bF1juNQp04dDBo0CHv27MGoUaOQkZGBS5culZinjx8/4vbt27CwsOCPDi0unwkJCUhMTISjoyO/k7NgLlVRH6//Wc2aNeHu7o7o6Gg8e/YM+/btg6urK2rXrs1Pk5aWhpycHNSpU4ffGSwNycnJAIqeQkFSRClTQgghhBBlRJ2thBBCCCEy5uTkhFGjRiErKwtt2rTBjRs3CqX58uULVq1aJTDfpyLo1KkTjIyMsHnzZly7do2/PT8/HzNmzEBaWhoGDx4ssDhWlSpV8PbtWzx8+LDQ/gYNGlTk4kuvX78G8N9I1QJFdfAFBQXh06dPGDZsGH9bv379oKamhqVLl+Lp06f87V+/fsXEiRORl5cnkN7Q0BBaWlq4fft2qTu4hw4divz8fPTu3RsfP34UGNUKfJ8+wsDAAPfv30daWhp/e15eHpYsWcIfnVtW9vb2AIALFy4IbL906RJ27dolkWMUV6aEEEIIIcqMFsgihBBCCJGDVatWISsrC1u3bkXjxo3RtGlT1K9fHzweD/Hx8bhw4QI+f/4Md3d3eWdVgLa2NrZt24auXbuiefPmaN26NSwsLHD16lXcvXsXzs7OhRbo8vPzw/z58+Hp6Yl27dohMzMTTk5OmDFjBg4fPozt27fDxcWFH/+1a9dw8+ZN+Pj4FIr/zJkzsLe3R7NmzaCjo4OYmBjcunULTZs2xYwZM/jprK2tsWbNGowcORINGzaEt7c3DAwMcP78ecTFxcHHxwcjR47kp1dRUUG3bt2wc+dOuLi4wNPTEx8+fEC7du0waNAgkT6b7t27Y+zYsbh+/Tp0dXXh5+dXKM306dMxffp0ODk5wdfXFwBw6tQp/rQGkjBu3Dhs3rwZQ4cORVhYGCwsLPDo0SOcOXNGYtMmFFemhBBCCCHKjEa2EkIIIYTIgaqqKv766y+cPn0avXv3RnJyMrZt24atW7fi33//RadOnXDgwAGcPHlS3lktpH379rh06RJ8fX1x6dIlbN++HVlZWZg5cybOnz8PbW1tgfRBQUGYM2cONDU1sXPnTly8eJGf5u+//8aAAQOQkpKCrVu3Yt++feDxePjjjz9w9OjRQnOM1qtXD61atcKZM2ewadMmpKWlYebMmTh58iQ0NDQE0g4fPhwnTpxAkyZNcPLkSezatQvq6upYunQpwsLCoKKiIpB+3bp1GDNmDHJzc7Fjxw7ExsZCV1dX5M9FU1MTffr0AQD06tWr0OcAAFOnTsXq1auhqamJHTt24MSJE/D09MSDBw9gbm4u8rGKU61aNZw/fx7NmjXDiRMnsHv3bnAch4MHD2L8+PESOUZxZUoIIYQQosw4RkuLEkIIIYQQBRcfHw87Ozu0adMGUVFR8s4OIYQQQgghRaKRrYQQQgghhBBCCCGEECIB1NlKCCGEEEIIIYQQQgghEkCdrYQQQgghhBBCCCGEECIBNGcrIYQQQgghhBBCCCGESACNbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJoM5WQgghhBBCCCGEEEIIkQDqbCWEEEIIIYQQQgghhBAJUJV3BoTJy8tDbm6uvLNBCCGEEEIIIYQQQghRUpUqVYKKiorI6RWus5UxhqSkJKSnp8s7K4QQQgghhBBCCCGEECVnYGAAc3NzcBxXYlqF62wt6Gg1NTWFlpaWSEEQQgghhBBCCCGEEEKIJDHGkJWVhZSUFACAhYVFie9RqM7WvLw8fkerkZGRvLNDCCGEEEIIIYQQQghRYpqamgCAlJQUmJqaljilgEItkFUwR6uWlpacc0IIIYQQQgghhBBCCCH/9VWKsr6UQnW2FqCpAwghhBBCCCGEEEIIIYpAnL5KhexsJYQQQgghhBBCCCGEkPJGoeZsFabqzFiZHSsxtIHMjlUUDw8PeHh4IDg4uFwfg0jetMMJMjnOki7WMjlOcageEEWTXcNYJsfRiHsvk+OIKz4+HnZ2dnjx4gVsbW1Fes/AgQMBANu2bZNavmRxDCJZ39YslclxVMdNlclxxEV1iVQEWyz0ZXasIW8/yuxY8hQcHIyzZ8/i7Nmz5foYRPJurVgik+M4T5omk+MoAqpvRBZoZGs5M2/ePNSoUUNi+3v58iVUVVVx7tw5ie1T0aiqqmLHjh3yzgaRIKoH5ce5c+egqqqKly9fyjsrRMHs2LEDqqqS/c23ol/va9SogXnz5sk7G0TBUF0qP6i9IT+2trbl+scEabSnKvo9xdPTE0OGDJF3NpQS1bfCKnp9kydFbbOUi5Gt5D9z5szBnDlzJLY/GxsbfPv2TWL7U0QVPT5lRPWg/GjRogV9tqRIAwYMwIABAyS6z4p+rsXFxck7C0QBUV0qP6i9QUpLGu2pin5POXXqlLyzQMopqm/li6LeV2lkaxl4eHigT58+6NatG4yMjFClShVs3bqV/3peXh6WLFmCGjVqQE9PD25ubrh69WqJ+3306BE6d+4MIyMj2NnZYfPmzfzXgoOD4eHhIXIeAGDPnj2oXbs2tLS0UL9+fTx58oT/Wnx8PDiOQ3x8PIDvj49ZWloiJCSEn+9evXrhy5cv/Pc8e/YMbdu2hZ6eHkxMTFC3bl2RH4X70cWLF+Hu7g59fX1YWlqiV69e+PTpE54+fYratWtDX18f2tracHJywqFDh/jv27ZtG6pUqYKFCxfC3t4eampqePToEQYOHIi2bdti2LBhsLS0hKWlJYDvkxj/OHw/PDwcLi4u0NXVRY0aNbBmzZoS933+/Hk0a9YMurq6MDU1ha+vLxITE8WOuSKielD6elBw3NDQUDRo0AA6OjpwdXUVuBm/fv0aPXr0gImJCSwtLTFs2DB8/PjfI3WixC7Mli1bULduXWhra6NmzZqYPXs2AGDfvn2ws7ODlpYWDAwM0KpVK9y7d4//PmF1zdbWFuPHj0fbtm1hYGCA3r174+zZswITiX/79g3Lli2Dg4MDdHR00LBhQ5w+fbrEfW/atAm1a9eGpqYmrK2tMXjwYLE+64ro2LFjMDAwQH5+PgDg2rVrAudxeno6VFRU8PTpUwBAYmIievXqBXNzcxgbG6NPnz54//6/aQvc3d1hbGwMDQ0N2NjYICQkROix09LS0LhxY/Tr1w95eXlC071+/Rr9+/eHubk5LC0tMX/+fP5r27ZtE6gzAwcORJs2bTBo0CBYWFjAxMQECxcuFNjfyZMn0bBhQ2hra+OXX37BxYsXBV7/8XpfcO6tXr0atWvXho6ODtq0aYN3797x06ekpKB3796oXLkyDA0N0aBBg1It0nn//n20bdsWRkZGMDExQceOHZGQkIBPnz6hfv36MDQ0hKamJuzt7bFhwwb++wryuH79etStWxfq6uqIiopCcHAwGjVqhMmTJ8PW1hZqamrIzMwsNFLk0qVLaNGiBfT19WFjY4OgoCB+g1PYvu/du4fWrVvDwMAAlStXhoeHB+7cuSN2zBUJ1SXFqUscx2HmzJlwc3ODrq4u6tati+vXrwt8XkOHDoWlpSVMTEzQs2dPvHnzRqzYhTl69CgaNWoEHR0d2NjY8EfEnT9/HjVr1oSuri50dXXRtGlTgZGpwuqrh4cH+vfvDz8/PxgbG+PXX38t1N4AgK1bt8LJyQk6OjqoU6cO9u/fX+K+Dx8+jAYNGkBbWxuWlpbo3r07MjMzxf68Kxphn0unTp3w8uVLDBkyBKqqqvz6IqzN8ejRI7Rp0wYGBgawsbHBtGnT8PXrV/5xbG1tMXr0aLRp0wb6+vqoXr06IiIi+K8zxrB69WpUr14dWlpaqFu3LvT09Eoc6Zeeno7Ro0fD2toaJiYmGDduHH/F65/bU8HBwXB2dsaECRNgbW0NQ0NDjB49GowxfprY2Fi4u7tDR0cHtra2At+nCuIoyJMobdKMjAyMGjUKZmZm0NPTg4uLS6HzWRQvX75Ez549YW5uDkNDQ3h5eeH27dsAir9+FuRx2bJlcHFxgaamJjZu3Cj0+9vP04U9fPgQ7dq1g5GRESwsLDBq1Ch+vRG274SEBHTt2hXGxsbQ19dH06ZNqRP3/1F9Kx/1raTPLzs7G1OnToWNjQ0MDAzg4+ODx48fixW7MMre30OdrWV0+/ZtjBw5Eq9evcK8efMwYsQIpKamAgBWrVqFTZs24dChQ0hOTsaAAQPg5eWFV69eFbvPp0+fYuLEiXj16hXWrVuHUaNGITZW+Ly1xeXhxIkTGDFiBFasWIH09HSEhYXByMio2OOnpqZCR0cHV65cwd27d3H+/Hn8+eefAIAvX76gTZs2qFGjBhISEhAXF4fWrVuL85EBAP/L3rBhw/D27VvExMSgRo0ayMzMhKmpKfbt24fk5GR8/vwZK1asQO/evfkxAeA38K9du4aMjAzUrFkTwPeLnJeXF+Lj44v89ejUqVPo06cPgoOD8f79e+zatQshISE4evSo0H1Xq1YN7du3x4ABA5CUlIQbN27A19cXWVlZYsddUVE9KF09KBAXF4eDBw/i7du30NXVxbRp3+dMysvLQ/v27aGnp4cnT54gNjYWb9684c+tJ0rswmzcuBFz587FH3/8gdTUVOzfv5/fwGjatClOnz6NT58+ITk5GR4eHvD39xd4v7C6FhMTgzlz5uDDhw/YsmVLoeOGhIRg+/btOHDgAFJTUzFu3Dh07NhR4Mvyz/u+fPkyJkyYgL/++gtpaWmIjIyEhYWFWJ9xReTh4YHMzEx+R9mBAwegq6uLffv2AQAuX76MqlWrombNmsjOzoanpydMTU3x+PFjPH78GJmZmRg2bBh/f7/99huePXuG7OxsnDlzBn/++SfCwsIKHTctLQ3e3t6oV68eduzYARUVFaF5fPbsGXr16oVnz57h8OHDWL58OQ4fPiw0/YMHD9ChQwfExcVh9+7dmD17Nr+j//79++jYsSMmTZqE1NRUnD9/HnZ2diV+TsnJyTh58iTi4+Px9u1bLF68GMD3xnmXLl3w7ds3PH78GK9evUK/fv1K3N/PkpKS4O7uDg8PD8THx+P27dto0aIFUlNToaGhgT///BOJiYnIysrCgQMHEBgYiLt37wrs4+nTpzh16hSysrLQsmVLAMC///6L6tWr4+HDh/j48SO0tLQE3vPo0SN4e3tjyJAhSEpKQlRUFPbu3Yt169YVu+/u3bujSZMmePnyJR4+fIhBgwYhIyND7LgrEqpLilGXCty7dw8bNmxASkoKGjdujBEjRvBfGzBgAN6+fYvY2Fg8efIEurq66NixI7+jvKTYhTlx4gT69++PkJAQvH//HidOnICmpiYA4JdffkF4eDhSU1ORlpaG4cOHo1u3bgJfNIXV18uXL2Pw4MH8z+5nW7ZswezZs/n34t9++w3+/v64deuW0H1/+PABPXv2xPz58/H+/XtcuHAB9erVQ05OTuk+8AoiMTFR6Ody9OhR2NjYYMuWLfj27ZtAZ8XPbY7Pnz/Dy8sLTZo0watXr3D69GmcPn0a06dPFzjerVu3MG/ePCQnJ6Nfv37w9/fnnxObNm3C0qVLsXPnTnz8+BG7du2Curp6iTG8fv0abm5uuH//Pi5cuICjR49i7dq1QtO/ePECTk5OuHfvHs6dO4ctW7YgPDwcwPf66uXlhU6dOiE5ORm3bt1CgwYlr00irE0KAMOGDcODBw9w9epVvHv3DlOmTClxfz/LysqCp6cnLCws8PDhQzx9+hS9e/dGcnIyANGun7du3cKBAweQmZnJb58K+25YICUlBe7u7nB3d8fLly9x9epVxMbG8gcaCNv3sGHDYGhoiMePH+P58+eYOnUq/bABqm/lpb4VKO7zmzJlCs6ePYvTp0/j1atXcHFxgbe3t0DbsLjYhaH+HupsLbOePXvCy8sLmpqa6N27N3Jycvgj5jZv3oxp06ahXr160NTUREBAABwdHbFr165i99mxY0e0aNECmpqaaNu2Ldzc3ARODnHysHTpUgwdOhQ+Pj5QU1ODtbV1iZ1MFhYWCAwMhLGxMWxtbfkXIQA4ePAgPn36hBUrVsDAwAD6+vqoW7euOB8ZAGDDhg1o27Yt+vfvDy0tLdjZ2WHBggWwsLCAlpYWIiMj0bJlS1hZWaFLly7Izc3Fo0eP+O+3tLTErFmzYGBgADU1Nf5cZT4+PvDz84OamlqhL6bA946/cePGoW3btlBXV0eTJk3Qu3dv/P3330L3nZubi5ycHLx9+xaZmZmwtrbGuHHj8Msvv4gdd0VF9aB09aDAnDlzUK1aNejq6qJz587841y/fh3379/HmjVrYGhoCHNzc6xYsQJHjhzhN0pLil2YNWvWYPr06XBzc4OGhgacnZ2xZMn3Cfg1NTWxcuVKNGjQAFWrVkVoaGihziFhdW3MmDFo2rQpeDxekXVw9erVWLVqFX+knb+/P2xtbXHixAmh+87IyEBeXh7evn2LvLw81KlTR+SRShXZzyOsDh06hBUrVmDv3r0Avnd8e3l5AQCOHz+O7OxsrFq1CgYGBjA2NkZQUBCOHTvG72QvGGliY2MDZ2dnvH79ulC5p6Wl8RvFmzdvBo9XfDPC3d0d7dq1g7a2Npo0aYKuXbviyJEjQtN7eXmha9eu0NbWhre3N0xMTPj1YeXKlfD19UXfvn2hoaEBMzMzVKlSpcTPKTQ0FJaWljA2NoaPjw9/fxcvXsSVK1ewadMmmJqaQltbW6SG8c927tyJatWqYerUqdDV1UWVKlUQGBgIZ2dnVKpUCbGxsWjfvj2qVq2K5s2bIysrq9DnunLlSpiamkJFRYX/BaFhw4YYOXIktLS0oKmpWWiU4Pr169GlSxcMGDAAmpqaqFWrFoYPHy5wPytq3xkZGUhJScGnT59gbm4Of39/NGvWTOy4KxKqS4pRlwoEBgbC0dERmpqa6N69O/84SUlJCA8Px8qVK/kj4tasWYM7d+4IjH4tLnZh1qxZg4CAALRr1w4aGhpwcHDgf+nW1tbGzp074erqCisrK4wePRqpqal4/fo1//3C6mu/fv3Qpk0bqKioCG2XhoSEoFmzZlBXV4ePjw88PT0FyvbnfX/58gXfvn3D69ev8fXrV1SvXh1BQUEwNDQs9WdeEZT2c/m5zXH8+HFwHIeQkBDo6OigevXqWLBgATZu3CjQwR4QEIAmTZpAQ0MDfn5+eP/+PZKSkgB8b3dOmzYNv/76KypVqgQnJydoa2uXGEOdOnXQu3dv6OnpwcHBAYMHDy62ntevXx+DBw+Gvr4+nJyc4OTkxD/XN23aBDs7OwQGBkJbWxuGhoYi/agirE2amJiIvXv34vfff4etrS3U1dXRuHHjEvf3s+PHjyMjIwMrVqxA5cqVYWxsjCFDhqBNmzYARLt+hoaGwsbGBjweDxoaGgCEfzcssHPnTjg4OGDatGnQ0dGBtbU1Jk+eXOie+fO+MzIy+D+0GBkZoWvXrujYsaPYcVc0VN/KR30rIOzzy8/Px5YtW7BgwQJUr14dOjo6CA4ORn5+Po4fPy5S7MJQfw91tkpUQWEXDHtPSEiAjY2NQBobGxuxJ1o2NzfnX0zEzcOrV69gb28v1vGK2mfB/hITE1G9evUyL8Tw7NkzoQscBQYG4o8//sD06dNx+/ZtpKWlAZDMXBxxcXFYsmQJNDQ0+H+bNm0SGFX3Mx0dHRw6dAjR0dGwtrZGlSpVMHHiRIHHG8h/qB6UzY/HSUhIgLGxsUCDoeCzFPb5/Rx7wSM8BX8Fk4cLq4OMMbRu3Rr37t3Dpk2b8OTJE0RGRhb7eKuoCn69LPgyW/D377//Cnxp/Zm3tzcWLlyImTNnwsDAAHXq1BF4FFuZeXt749y5c7h+/TrMzc0xaNAgvHnzBv/++y9iYmLg7e0N4Pu179WrV9DS0uJ/7m5ubgC+d2D8/fffGDBgANq3b48LFy7wR4D8fN3t1q0bHjx4gNGjR5fqEWFx6jEgWB+kUY9NTU3L3EFR3P1s5cqVmDlzJoYMGYJr167h/fv3EpuzMS4uDnv37hWoS7NmzcLbt2+Lfd/Bgwfx/PlzODg4wMTEBIMHD+bfZ5UZ1SXxSKMuCTtOwYjNhIQEABBoU2hra8PIyKjYNsWPeZ03b57APbFgsZLi6nGfPn0QHh6OpUuX4sGDB/xRspKqx6NGjRKox//880+x7VJ7e3ts27YNGzZsgImJCapXr4758+eL9EhnRSapzyUhIQFVq1YVqJc2Njb48uWLwNQZP/q57ZWYmCiRQRmKVs8BlDmuZ8+ewdbWtsiR/KJeP0uj4EmpH+ta3759S7xnFjyl1bhxY+jr66N79+40lRyovhXkQ9Hrm7DjAN8/v/fv3+PLly8C91Uej4eqVauKfF8tWKSz4K9gCh7q76HOVqmytrbGixcvBLbFx8cX6ngqybNnz8R+TwELCwuxLholMTc3F3rhE4epqanQSaKjo6MxefJkdOrUCaamphLt0DI1NcWcOXOQnZ3N/8vJyUF0dHSx72vXrh1iYmKQkZGB/fv3Y/fu3SLPi6nsqB6UnrW1Nd69eyfwGEfBoziifhYFj/AU/BUsoiKsDqakpODOnTtYu3YtmjZtCgMDg1J1BBRFX18fampqOHnypEAdzM3NLfQY188mT56MR48eIS0tDdOmTcPIkSMF5hNSVt7e3rhw4QL279+PXr16QUVFBd27d8f//vc/3LhxA56engC+l7ednZ3A556dnY1v376hatWqiI6ORrt27TB27FhYW1sLffwqJCQEY8aMgbe3N549eyZ2fhWtHqenp/NHI5ZWSfezQYMGoX///rC0tISamlqZjvXzcf39/QXK8+vXryWOand1dUV0dDQ+f/6M6OhoXL58mT+qXZlRXSo9SdWlklhbWwOAQJuiYNSZqJ/FnDlzBO6JBYttllSPFy5cCE9PTxgZGRU73YO4TE1NsWXLlkLt0h/nqS+Kv78/YmNjkZGRgbVr12LhwoUCT4goq+I+FxUVFYHpJoSxtrbGy5cvBdLGx8dDU1MTJiYmIuVDUu1ERavnAMocl6mpKeLj44vsVBH1+lna43p4eBS6Z5Y0/YaDgwOOHDmCtLQ0XLt2DSkpKWV6nLsiofr2H0WtbyUxNjaGpqamwH01Pz8fiYmJIn8WAwYMELivFvxAQf091NkqVcOGDcOSJUtw+/ZtfPnyBZs2bcKDBw/Qt2/fYt8XFxeHDx8+ICsrC2vWrMHTp08xdOjQUuWha9eu+Ouvv/Dw4UN8/vwZe/fuxYMHD0q1L+D7sO3Xr19j27ZtyM7OxsOHDwUm8hfVmDFjcOzYMezcuRNfvnzBmzdvEBoaioyMDDg4OODMmTPIzMxESkoKJk6cWOr8/mzUqFFYsWIFDh8+jMzMTHz8+BExMTHFzjny4sULTJ48md+xY2ZmBlVV1RIfQyffUT0ovcaNG6Nu3boYO3Ys0tLSkJSUhMDAQHTu3BlmZmZl2veECROwePFiXLp0Cbm5uXjw4AGWL18OY2NjGBkZISoqCt++fcOjR48k9si+hoYGBg8ejHHjxuHatWvIzc1FUlISwsLCBOan+9mRI0ewdu1avHr1Cpqamvw6qKenJ5F8lWcuLi749u0bNm/ejB49egAAevXqhdWrV6NmzZowNjYGAHTp0gVZWVmYMGECXr9+jdzcXDx58oQ/v6eDgwNu3bqF169f48uXL9i4cSNu3rxZ6HjNmzfH8uXL0a5dO5HmXk5MTERSUhK+fv2KAwcOIDw8HOPHjy9VrF27dsXff/+Ny5cvIysrCxEREbhw4UKp9gWA/2PC4sWL8eXLF7x48aJUjaohQ4bgyZMnWLp0Kb/jZ/369YiLi4ODgwMuXbqEDx8+4OPHjwgNDS12FLc4AgICsHfvXmzZsgUfP35EZmYmbt68yX/0vSg5OTkYNmwYbt++jW/fvsHMzAzq6up0PwPVJUWoSyUxNzdH+/btMXHiRLx9+xZpaWkYP3486tWrV6bHK4Hv98RNmzYhIiICX79+xfPnz/mjXh0cHBAdHY2cnBwkJCRg5syZkggHwPd26ezZs3H69GlkZ2cjNTUVJ0+eFFiA62eXL1/GggUL8OLFC6iqqsLc3Bw8Hg+VK1eWWL7Ko5I+F1tbW0RGRiIzM7PYkYnt2rUDx3GYM2cOMjIy8Pz5cwQFBWH48OEi//jcpUsXrFmzBq9fv8anT5+wY8cOkTpN3r17h/j4eOTm5uLUqVPYtGkTJk+eLNoHUEQeLl68iPDwcGRnZ+PChQtFzh0tKjs7O9SvXx/z589HRkYG3rx5gz/++EPs/XTr1g2VKlVCYGAgPnz4gM+fP2PXrl24cuWKyNfP0hgwYACuX7+OxYsX4/3798jOzsb9+/fx119/Ffu+UaNG4fLly8jOzoaJiQm0tLTongmqb0XlQRHrW0l4PB6GDBmC2bNn49mzZ8jIyEBwcDB4PB7atWtXpn1Tfw91tkrVhAkTMGLECHTr1g2mpqbYsWMHoqOjYWVlVez7YmNjUbduXRgbG2Pv3r34559/YGpqWqo8jBw5Et27d4eHhwdq1qyJkydPlmlhGUtLSxw8eBBLly6Fvr4++vXrBxMTE4HROvn5+TAyMuIPIS+Ki4sLvwPF1NQUjRo1wr///gs1NTWsWrUKKSkp/JVbJbkQjp+fH3bs2IHly5fDysoKNjY2mDJlSrG/vOnr6+PVq1fw8vLir8A7btw4/pcxUjyqB8LrQUlUVFRw7NgxfP78GTVr1oSzszMsLS0l8iV2woQJmDRpEn/+nS5duiAvLw8qKirYt28fNm7cCF1dXfj7+6NevXplPl6BNWvWoH///hg8eDAMDAxQr149bN68mT/nVlGqVq2Kw4cPo379+tDT08PMmTNx+PBh/gqUykxFRQUtW7aEs7Mz/5x2c3ODgYEBf45JADAwMMCVK1eQmpoKV1dX6OnpoV27dvzV1UeOHIlmzZrhl19+ga2tLe7cuYPq1asLPe6GDRvg6uoKT09PgfmDfxYfHw9XV1cYGBhgwYIFOHToEOrUqVOqWDt27IgZM2agR48esLa2xrZt20o9AgH4/gjUkSNHcPToURgaGsLHxwc6OjqoVKmSQLr69evzRzUWxcbGBtHR0YiKioKVlRUcHBxw6tQp6OvrY86cOahcuTKsrKzg6OiIz58/8zvtyqpZs2Y4ceIE9u7dixo1asDc3BzDhg0rduGOgpEmnTt35q9s27p1a4k2cssrqkvyr0ui2LFjB6pUqYIGDRqgZs2a+PTpE8LCwkqc87YkXbt2xbp16zBjxgxUrlwZLVu2RHp6OgBg69atOHfuHPT09NC2bdsyPyr6o6lTpyIkJASTJ0+Gqakp7O3tsWjRomJH+ZiZmeHq1atwdXWFjo4O+vXrh3Xr1sHFxUVi+SqPSvpcFi1ahEePHsHQ0BDdunUTuh9dXV2cPHkS169fh5WVFTw8PNCqVSv+gnCiWLBgARwcHODo6AhbW1tcu3YNurq6JT7dkJaWBh8fH+jp6WHUqFFYs2ZNqRdgrVevHv7880+MHz8eZmZmCA0NFWkOSWE4jsO+ffvw+PFjmJqaomnTpvxHxn+Mq3PnzqhZs6bQ71b6+vqIjo7GixcvYG9vD2tra+zcuROVK1cW+/opjurVq+PSpUu4cuUK6tatCyMjI/j5+RV73S3I74ABA1C5cmXUrl0bVlZWYp0LFRXVN0GKWt9EsWzZMrRs2ZI/f+rVq1cRHR0NHR2dUu8ToP4eAOCYAk3wk52djRcvXsDOzq7YL95EsaxYsQKRkZH8odlXrlzBr7/+igcPHqBWrVpyzh0hskH1gJDy7++//8a0adP4HWdJSUmwtLREREQEfHx85Jw7QsoPqkuE/IcxBgMDA0RERFSoBQljY2Px66+/IisrCzweD7m5uTAyMsLSpUsxYsQIeWePKCmqb0SaxOmzpJGtRGyRkZGIi4tDbm4u7t27hw0bNsDPz4//elRUFNq1a0cdTKRCo3pASPl39uxZPHjwADk5OXj+/DmWLVtWqB47OjpS5xAhJaC6RMh/EhMTcerUKaSnp+Pz589YuHAhjI2N0ahRI3lnrUyuXbuG2NhYZGdnIykpCfPmzUP37t35I8tjYmKgoaEBf39/OeeUKBOqb1TfFBV1thKxhYeHo1GjRtDR0UG3bt0wZMgQDB48mP96VFQUTRxOKjyqB4SUf5cvX0aLFi2go6MDT09PNG/eXGCxtqioqFLP30WIMqG6RMh/Pn78iGHDhsHMzAx2dna4ePEijh07JtEFn+Th0aNHaNu2LfT09NCoUSOYmJjg999/578eFRWFMWPGQFNTU465JMqG6hvVN0VF0wgQQgghhBBCCCGEEEKIEDSNACGEEEIIIYQQQgghhMiYQna2KtBgW0IIIYQQQgghhBBCiBITp69SoTpbK1WqBADIysqSc04IIYQQQgghhBBCCCHkv77Kgr7L4qhKOzPiUFFRgYGBAVJSUgAAWlpa4DhOzrkihBBCCCGEEEIIIYQoG8YYsrKykJKSAgMDA6ioqJT4HoVaIAv4HkRSUhLS09PlnRVCCCGEEEIIIYQQQoiSMzAwgLm5uUiDQhWus7VAXl4ecnNz5Z0NQgghhBBCCCGEEEKIkqpUqZJII1oLKGxnKyGEEEIIIYQQQgghhJQnCrVAFiGEEEIIIYQQQgghhJRX1NlKCCGEEEIIIYQQQgghEkCdrYQQQgghhBBCCCGEECIB1NlKCCGEEEIIIYQQQgghEkCdrYQQQgghhBBCCCGEECIB1NlKCCGEEEIIIYQQQgghEkCdrYQQQgghhBBCCCGEECIB/weg8Sxt6j4TaQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 1600x400 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# Plotting\n",
    "colors1 = ['#1c79cd', '#f72313', '#9c180e']\n",
    "colors2 = ['#76aee1', '#fb9189', '#ce8d88']\n",
    "\n",
    "plt.figure(figsize = (16, 4))\n",
    "\n",
    "for i, var in enumerate(order_all):\n",
    "\n",
    "    plt.bar(2*i, stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['no binding'], color=colors1[0], width=bar_width, label='no binding, carriers' if i == 0 else None)\n",
    "    plt.bar(2*i+0.5, stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['no binding'], color=colors2[0], width=bar_width, label='no binding, non-carriers' if i == 0 else None)\n",
    "    \n",
    "    plt.bar(2*i, stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['weak binding'], color=colors1[1], width=bar_width, label='weak binding, carriers' if i == 0 else None, bottom=stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['no binding'])\n",
    "    plt.bar(2*i+0.5, stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['weak binding'], color=colors2[1], width=bar_width, label='weak binding, non-carriers' if i == 0 else None, bottom=stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['no binding'])\n",
    "    \n",
    "    plt.bar(2*i, stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['strong binding'], color=colors1[2], width=bar_width, label='strong binding, carriers' if i == 0 else None, bottom=stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['no binding']+stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['weak binding'])\n",
    "    plt.bar(2*i+0.5, stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['strong binding'], color=colors2[2], width=bar_width, label='strong binding, non-carriers' if i == 0 else None, bottom=stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['no binding']+stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['weak binding'])\n",
    "\n",
    "plt.xlim(-1, 2*i+2)\n",
    "order_all2 = [item for item in order_all for _ in range(2)]\n",
    "tick_positions = [tick + 0.25 if i % 2 == 0 else tick for i, tick in enumerate(range(len(order_all) * 2))]\n",
    "tick_labels = [f'{var}' if i % 2 == 0 else '' for i, var in enumerate(order_all2)]\n",
    "plt.xticks(tick_positions, tick_labels, rotation=90, fontsize = 12)\n",
    "plt.yticks(fontsize = 13)\n",
    "plt.xlabel('CH hotspot variant', fontsize = 14)\n",
    "plt.ylabel('Percentage (%)', fontsize = 14)\n",
    "plt.title('MHC II', fontsize = 16)\n",
    "plt.legend(bbox_to_anchor=(0.5, -0.4), loc='upper center', ncol = 6, fontsize = 9.5)\n",
    "plt.tick_params(left = True, bottom = False) \n",
    "plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/8b_UKBB_{param}_percentages_by_threshold_mhcii.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at VAF scores for cases where you have both binding and no binding \n",
    "ch_all_el_rank_log_labels = ch_hla_merge_sub_all[ch_hla_merge_sub_all['gene_var2'].isin(order)]\n",
    "\n",
    "# let's add to everyone group based on score\n",
    "conditions = [\n",
    "    (ch_all_el_rank_log_labels['log_score'] < -1 * np.log10(5)),\n",
    "    ((ch_all_el_rank_log_labels['log_score'] >= -1 * np.log10(5)) & (ch_all_el_rank_log_labels['log_score'] < -1 * np.log10(1))),\n",
    "    (ch_all_el_rank_log_labels['log_score'] >= -1 * np.log10(1))]\n",
    "\n",
    "values = ['no binding', 'weak binding', 'strong binding']\n",
    "ch_all_el_rank_log_labels['binding'] = np.select(conditions, values)\n",
    "ch_all_el_rank_log_labels['binding']= pd.Categorical(ch_all_el_rank_log_labels['binding'], categories=['no binding', 'weak binding', 'strong binding'], ordered=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### split by binding / non-binding to have more variants to look at "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# okay now binding / no binding\n",
    "# look at VAF scores for cases where you have both binding and no binding \n",
    "ch_all_el_rank_log_labels = ch_all_el_rank_log_labels[ch_all_el_rank_log_labels['gene_var2'].isin(order)]\n",
    "\n",
    "# let's add to everyone group based on score\n",
    "conditions = [\n",
    "    (ch_all_el_rank_log_labels['log_score'] < -1 * np.log10(5)),\n",
    "    (ch_all_el_rank_log_labels['log_score'] >= -1 * np.log10(5))\n",
    "]\n",
    "values = ['no binding', 'binding']\n",
    "\n",
    "ch_all_el_rank_log_labels['binding_any'] = np.select(conditions, values)\n",
    "ch_all_el_rank_log_labels['binding_any']= pd.Categorical(ch_all_el_rank_log_labels['binding_any'], categories=['no binding', 'binding'], ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# okay now repeat this but binding / no binding \n",
    "\n",
    "# define function to plot cumulative probability\n",
    "def plot_cumulative_mhcii(data, x, y, hue, colors, xlabel, ylabel, var, score, name):\n",
    "\n",
    "    var_name = var.split('_')[0:2]\n",
    "    var_name = ' '.join(var_name)\n",
    "\n",
    "    nr_cases = data.shape[0]\n",
    "\n",
    "    plt.figure(figsize=(4,4)) \n",
    "    sns.lineplot(x=x, y=y, data=data, hue=hue, palette=colors, legend = True)\n",
    "    sns.scatterplot(x=x, y=y, data=data, hue=hue, palette=colors, legend = False, alpha = 0.3, linewidth=0)\n",
    "\n",
    "    # specify title and axes labels \n",
    "    plt.title(f'{var_name} (n={nr_cases})', fontsize = 12)\n",
    "    plt.xlabel(xlabel, fontsize = 10)\n",
    "    plt.ylabel(ylabel, fontsize = 10)\n",
    "\n",
    "    # convert both axes to log scale \n",
    "    plt.gca().set_xscale('log')\n",
    "    plt.gca().set_yscale('log')\n",
    "\n",
    "    # adjust legend \n",
    "    if var == 'DNMT3A_R326C':\n",
    "        legend = plt.legend()\n",
    "        legend.prop.set_size(10) \n",
    "        plt.legend(title = 'MHC II presentation', loc = 'lower left', fontsize = 10, title_fontsize = 11)\n",
    "    else:\n",
    "        plt.legend().remove()\n",
    "\n",
    "    # specify ticks on the x axis \n",
    "    x_ticks = [1, 3, 10, 30]\n",
    "    plt.xticks(x_ticks, x_ticks, fontsize = 9)\n",
    "\n",
    "    y_ticks = [1, 0.3, 0.1, 0.03]\n",
    "    plt.yticks(y_ticks, y_ticks, fontsize = 9)\n",
    "    \n",
    "    plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_VAF_{hue}_{name}_{score}_{var}_threshold_mhcii.pdf', bbox_inches='tight')\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# okay now do the cumulative plots for this \n",
    "\n",
    "# plots scores for each group on the same plot \n",
    "# note: an alternative is to split by thresholds for binding  \n",
    "# script to do this is included in the complete analysis, but I am not using this way due to issues described there \n",
    "\n",
    "# for extreme groups in each split \n",
    "dfs_list = ['ch_all_el_rank_log_labels']\n",
    "\n",
    "for name in dfs_list:\n",
    "\n",
    "    df = globals()[name]\n",
    "    score = 'min_el_rank'\n",
    "    \n",
    "    df['VAF_percent'] = df['VAF'] * 100\n",
    "    colors = colors1\n",
    "\n",
    "    variants = df['gene_var'].unique()\n",
    "\n",
    "    # plot for each variant separately \n",
    "    for var in variants: \n",
    "\n",
    "        data = df[df['gene_var'] == var]        \n",
    "        data = data.sort_values(by='VAF')\n",
    "        N = data.shape[0]\n",
    "        \n",
    "        # add number of people in each category \n",
    "        data['group_size'] = data.groupby('binding_any')['binding_any'].transform('count')\n",
    "        data['index_vaf_group'] = data.groupby('binding_any')['VAF'].rank(ascending=False) # index by VAF\n",
    "        data['fraction_index_presenter_group'] = data['index_vaf_group'] / data['group_size']\n",
    "    \n",
    "        plot_cumulative_mhcii(data, 'VAF_percent', 'fraction_index_presenter_group', 'binding_any', colors, 'VAF (%)', f'Cumulative probability', var, score, f'groups_thesholds_allon1')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
