{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MHC class I genotype vs CH risk"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- note: this is the script to reproduce the figures from the report\n",
    "- more complete analysis is provided in additional scripts "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n",
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n"
     ]
    }
   ],
   "source": [
    "# imported packages\n",
    "import random\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import matplotlib.ticker as plticker\n",
    "from matplotlib.ticker import LinearLocator, FormatStrFormatter\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "import scipy.special\n",
    "import scipy.integrate as it\n",
    "from scipy import integrate\n",
    "from scipy.interpolate import interp1d\n",
    "from scipy.stats import kde\n",
    "import copy\n",
    "import glob, os\n",
    "import re\n",
    "# from sklearn import datasets, linear_model\n",
    "import pandas as pd\n",
    "from decimal import *\n",
    "from operator import itemgetter    \n",
    "from collections import OrderedDict\n",
    "import timeit\n",
    "import time \n",
    "import csv\n",
    "import seaborn as sns \n",
    "import scipy as sp\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from matplotlib.ticker import LogLocator\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy.stats import kruskal\n",
    "from myfunctions import transform_format\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# specify font size for plotting \n",
    "plt.rcParams.update({'font.sans-serif':'Verdana'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "timestr = time.strftime(\"%Y%m%d\") # get current date "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import batch data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Examined file: batch_21_annotated_data\n",
      "Examined file: batch_18_annotated_data\n",
      "Examined file: batch_15_annotated_data\n",
      "Examined file: batch_11_annotated_data\n",
      "Examined file: batch_16_annotated_data\n",
      "Examined file: batch_12_annotated_data\n",
      "Examined file: batch_13_annotated_data\n",
      "Examined file: batch_17_annotated_data\n",
      "Examined file: batch_14_annotated_data\n",
      "Examined file: batch_19_annotated_data\n",
      "Examined file: batch_20_annotated_data\n",
      "Number of samples called with variants: 76718\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# path to folder with files (from Hamish)\n",
    "folder_path = '/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/ukb_hotspot_calls_annotated'  # Path to ukb files from Hamish \n",
    "\n",
    "# Get all files in the folder \n",
    "files = glob.glob(os.path.join(folder_path, '*.txt'))  \n",
    "\n",
    "# Initialize an empty dictionary \n",
    "dataframes = {}\n",
    "\n",
    "# Iterate through the CSV files and read each one with pandas\n",
    "for csv_file in files:\n",
    "    \n",
    "    df = pd.read_csv(csv_file, sep = '\\t')\n",
    "    dataframes[csv_file] = df\n",
    "\n",
    "# Extract dataframes \n",
    "for file_name, df in dataframes.items():\n",
    "    \n",
    "    variable_name = file_name.split('/')[8].split('.')[0] + '_data'  # Removing the file extension\n",
    "    print('Examined file:', variable_name)\n",
    "    \n",
    "    df['batch'] = variable_name # add column to indicate source \n",
    "    globals()[variable_name] = df  # assign dataframe to the variable \n",
    "\n",
    "# Concat all into one df\n",
    "dfs_to_concat = [v for k, v in globals().items() if k.endswith('_data') and isinstance(v, pd.DataFrame)]\n",
    "\n",
    "# Row bind all batch dataframes into a single one \n",
    "batch_all = pd.concat(dfs_to_concat, ignore_index=True)\n",
    "batch_all = batch_all.dropna(subset=['batch']) # remove rows which are read incorrectly (w/o batch number)\n",
    "\n",
    "print('Number of samples called with variants:', batch_all.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples carrying more than a single read with the variant sequence: 2823\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>end_position</th>\n",
       "      <th>VAF</th>\n",
       "      <th>var_depth</th>\n",
       "      <th>depth</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Codons</th>\n",
       "      <th>batch</th>\n",
       "      <th>alt_variant</th>\n",
       "      <th>ref_variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>2100538</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7674894</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>2</td>\n",
       "      <td>55</td>\n",
       "      <td>R/*</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>*</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>2100679</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25234373</td>\n",
       "      <td>0.048780</td>\n",
       "      <td>4</td>\n",
       "      <td>82</td>\n",
       "      <td>R/H</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>cGc/cAc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>H</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2100706</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25234307</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>2</td>\n",
       "      <td>103</td>\n",
       "      <td>P/L</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>cCg/cTg</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>L</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>2100777</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240313</td>\n",
       "      <td>0.020833</td>\n",
       "      <td>2</td>\n",
       "      <td>96</td>\n",
       "      <td>R/*</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>*</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>2100788</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240420</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>8</td>\n",
       "      <td>61</td>\n",
       "      <td>Y/C</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>tAc/tGc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>C</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>2101006</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240420</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>2</td>\n",
       "      <td>72</td>\n",
       "      <td>Y/S</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>tAc/tCc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>2101361</td>\n",
       "      <td>chr17</td>\n",
       "      <td>76736877</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>2</td>\n",
       "      <td>78</td>\n",
       "      <td>P/H</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>cCc/cAc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>H</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>2101637</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240313</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>2</td>\n",
       "      <td>92</td>\n",
       "      <td>R/*</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>*</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>2101708</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7675088</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>2</td>\n",
       "      <td>59</td>\n",
       "      <td>R/H</td>\n",
       "      <td>TP53</td>\n",
       "      <td>cGc/cAc</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>H</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>2102011</td>\n",
       "      <td>chr15</td>\n",
       "      <td>90088702</td>\n",
       "      <td>0.097561</td>\n",
       "      <td>4</td>\n",
       "      <td>41</td>\n",
       "      <td>R/Q</td>\n",
       "      <td>IDH2</td>\n",
       "      <td>cGg/cAg</td>\n",
       "      <td>batch_21_annotated_data</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_ID chromosome  end_position       VAF  var_depth  depth  \\\n",
       "41     2100538      chr17       7674894  0.036364          2     55   \n",
       "56     2100679       chr2      25234373  0.048780          4     82   \n",
       "59     2100706       chr2      25234307  0.019417          2    103   \n",
       "64     2100777       chr2      25240313  0.020833          2     96   \n",
       "65     2100788       chr2      25240420  0.131148          8     61   \n",
       "83     2101006       chr2      25240420  0.027778          2     72   \n",
       "109    2101361      chr17      76736877  0.025641          2     78   \n",
       "126    2101637       chr2      25240313  0.021739          2     92   \n",
       "130    2101708      chr17       7675088  0.033898          2     59   \n",
       "154    2102011      chr15      90088702  0.097561          4     41   \n",
       "\n",
       "    Amino_acids  SYMBOL   Codons                    batch alt_variant  \\\n",
       "41          R/*    TP53  Cga/Tga  batch_21_annotated_data           *   \n",
       "56          R/H  DNMT3A  cGc/cAc  batch_21_annotated_data           H   \n",
       "59          P/L  DNMT3A  cCg/cTg  batch_21_annotated_data           L   \n",
       "64          R/*  DNMT3A  Cga/Tga  batch_21_annotated_data           *   \n",
       "65          Y/C  DNMT3A  tAc/tGc  batch_21_annotated_data           C   \n",
       "83          Y/S  DNMT3A  tAc/tCc  batch_21_annotated_data           S   \n",
       "109         P/H   SRSF2  cCc/cAc  batch_21_annotated_data           H   \n",
       "126         R/*  DNMT3A  Cga/Tga  batch_21_annotated_data           *   \n",
       "130         R/H    TP53  cGc/cAc  batch_21_annotated_data           H   \n",
       "154         R/Q    IDH2  cGg/cAg  batch_21_annotated_data           Q   \n",
       "\n",
       "    ref_variant  \n",
       "41            R  \n",
       "56            R  \n",
       "59            P  \n",
       "64            R  \n",
       "65            Y  \n",
       "83            Y  \n",
       "109           P  \n",
       "126           R  \n",
       "130           R  \n",
       "154           R  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Change numerical variables to integers\n",
    "batch_all['end_position'] = batch_all['end_position'].astype(int)\n",
    "batch_all['position'] = batch_all['position'].astype(int)\n",
    "batch_all['sample_ID'] = batch_all['sample_ID'].astype(int)\n",
    "\n",
    "# Subset and create new useful columns\n",
    "batch_all = batch_all[['sample_ID', 'chromosome', 'end_position', 'VAF', 'var_depth', 'depth', 'Amino_acids', 'SYMBOL', 'Codons', 'batch']]\n",
    "batch_all['alt_variant'] = batch_all['Amino_acids'].str.split('/', expand = True)[1] # alternative (CH) variant\n",
    "batch_all['ref_variant'] = batch_all['Amino_acids'].str.split('/', expand = True)[0] # reference variant \n",
    "\n",
    "# There are some cases where there is no change in amino acids, for now save as NaN \n",
    "batch_all['alt_variant'].fillna(batch_all['Amino_acids'], inplace=True)\n",
    "batch_all['ref_variant'].fillna(batch_all['Amino_acids'], inplace=True)\n",
    "\n",
    "# Decided to only use calls with at least two variants based on previous investigation \n",
    "batch_all_ns = batch_all[batch_all['var_depth'] >= 2]\n",
    "\n",
    "# Number of samples with two variant reads or more \n",
    "print('Number of samples carrying more than a single read with the variant sequence:', batch_all_ns.shape[0])\n",
    "batch_all_ns.head(n = 10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Import indexes tested in each batch \n",
    "folder_path = '/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/ukb_hotspot_calls/batch_ids'  # Path to ukb files from Hamish \n",
    "\n",
    "# Get all files in the folder \n",
    "files_ids = glob.glob(os.path.join(folder_path, '*.tsv'))  \n",
    "\n",
    "# Read each file one by one \n",
    "\n",
    "indexes = {}\n",
    "\n",
    "# Iterate through the CSV files and read each one with pandas\n",
    "for file in files_ids:\n",
    "    \n",
    "    id = pd.read_csv(file, sep = '\\t')\n",
    "    id = id.rename(columns={'batch ID': 'sample_ID'})\n",
    "    id['sample_ID'] = id['sample_ID'].str.split('_', n = 1).str[0]\n",
    "    indexes[file] = id\n",
    "\n",
    "for file_name, df in indexes.items():\n",
    "    \n",
    "    variable_name = file_name.split('/')[9].split('.')[0]    # Remove file extension\n",
    "    globals()[variable_name] = df  # Assign the DataFrame to a variable with the file name\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate samples with genetic variants carried "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples which have been correctly annotated: 2818\n",
      "Number of variants identified in batches analysed: 89\n",
      "Number of sites identified in batches analysed: 41\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Find out how many gene_variants were called:\n",
    "\n",
    "# don't show warnings \n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# identify variants called \n",
    "batch_all_ns['variant_coord'] = batch_all_ns['chromosome'].astype(str) + \"_\" + batch_all_ns['end_position'].astype(str) # specific position in the genome \n",
    "batch_all_ns['variant_coord'] = batch_all_ns['variant_coord'].astype('category')\n",
    "batch_all_ns['variant_coord_pos'] = batch_all_ns['variant_coord'].astype(str) +  \"_\" + batch_all_ns['SYMBOL'].astype(str) + \"_\" + batch_all_ns['ref_variant'].astype(str) + \"_\" + batch_all_ns['alt_variant'].astype(str) # change to a specific aa\n",
    "batch_all_ns['variant_coord_pos'] = batch_all_ns['variant_coord_pos'].astype('category')\n",
    "\n",
    "# remove samples that have not been annotated (you can tell from the coordinate what is likely but these could be different mutations)\n",
    "batch_all_ns = batch_all_ns.dropna(subset=['SYMBOL']) # remove column where gene is not known \n",
    "print('Number of samples which have been correctly annotated:', batch_all_ns.shape[0]) # but at each site, you are getting reads modified to sth else \n",
    "# ok so sth worked wrong with annotation in only 5 cases > that looks good \n",
    "\n",
    "# identify the number of variants in a specific position\n",
    "num_variants = pd.DataFrame(batch_all_ns['variant_coord_pos'].value_counts())\n",
    "num_variants = num_variants[num_variants['count']!=0]\n",
    "num_variants = num_variants.sort_values(by = 'count')\n",
    "num_variants['variant_coord_pos'] = num_variants.index\n",
    "num_variants = num_variants.reset_index(drop=True)\n",
    "num_variants = num_variants.sort_values(by = 'variant_coord_pos')\n",
    "print('Number of variants identified in batches analysed:', num_variants.shape[0]) # but at each site, you are getting reads modiifed to sth else \n",
    "\n",
    "# identify the number of positions we looked at\n",
    "num_sites = pd.DataFrame(batch_all_ns['variant_coord'].value_counts())\n",
    "num_sites = num_sites[num_sites['count']!=0]\n",
    "num_sites = num_sites.sort_values(by = 'count')\n",
    "num_sites['variant_coord'] = num_sites.index\n",
    "num_sites = num_sites.reset_index(drop=True)\n",
    "num_sites = num_sites.sort_values(by = 'variant_coord')\n",
    "print('Number of sites identified in batches analysed:', num_sites.shape[0]) # okay see so you only found 37 sites \n",
    "\n",
    "# we have more variants than sites because for each site, we look at any change from the reference sequence (any possible variant)\n",
    "# at the same time, because we have this data, we can check if we have more variants that \"random\" mutations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants which have been investigated: 42\n",
      "Number of sites which have been investigated: 36\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Annotation (using df with variant names and genomic coordinates)\n",
    "\n",
    "# read in the df with coordinates\n",
    "coord_gene_var = pd.read_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/ch_variants_coordinates_tp53_added_nts.csv')\n",
    "\n",
    "# all coordinates identified in the batches  \n",
    "coord_out = num_sites['variant_coord'].tolist()\n",
    "\n",
    "# intersection (annotate)\n",
    "coord_gene_var['variant_coord'] = coord_gene_var['chromosome'] + \"_\" + coord_gene_var['end_position'].astype(str) # find cariant coordinates\n",
    "coord_gene_var['SYMBOL'] = coord_gene_var['gene_var'].str.split('_').str[0] # find gene analysed\n",
    "coord_gene_var['ref_variant'] = coord_gene_var['gene_var'].str.split('_').str[1].str[0] # reference sequence variant \n",
    "coord_gene_var['alt_variant'] = coord_gene_var['gene_var'].str.split('_').str[1].str[-1] # alternative sequence (CH / mutation) variant\n",
    "coord_gene_var['variant_coord_pos'] = coord_gene_var['variant_coord'].astype(str) + \"_\" + coord_gene_var['SYMBOL'].astype(str) + \"_\" + coord_gene_var['ref_variant'].astype(str) + \"_\" + coord_gene_var['alt_variant'].astype(str) # specific mutation \n",
    "\n",
    "print('Number of variants which have been investigated:', len(coord_gene_var.variant_coord_pos.unique()))\n",
    "print('Number of sites which have been investigated:',len(coord_gene_var.variant_coord.unique()))\n",
    "\n",
    "# NOTE: \n",
    "# TP53_R249S was searched for but not found in any of the batches so far \n",
    "# there have been 5 samples identified (in DNMT3A, at 2 different sites) with variants that I did not originally search for \n",
    "# in addition, some of the variants were offset by one base (could have been a deletion / insertion)\n",
    "# this is why we have differences in the number of sites in the two datasets \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants identified with annotations: 32\n",
      "Number of samples with a mutation in a CH-relevant position (with >2 reads): 2818\n",
      "Number of patients with a mutation in a CH-relevant position (with >2 reads): 2761\n",
      "Number of samples with a CH hotspot mutation (with > 2 reads): 2249\n",
      "Number of patients with a CH hotspot mutation (with > 2 reads): 2214\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# identify CH variants typed in the dataset \n",
    "gene_vars = coord_gene_var[['variant_coord_pos', 'gene_var']]\n",
    "\n",
    "# subset batch df to only include variants which were successfully identified \n",
    "batch_gene_vars = pd.merge(batch_all_ns, gene_vars, on = 'variant_coord_pos', how = 'inner')\n",
    "batch_gene_vars['gene_var'] = batch_gene_vars['gene_var'].astype('category')\n",
    "batch_gene_vars['gene_var'].value_counts()\n",
    "# rename sample ID to person ID column \n",
    "batch_gene_vars = batch_gene_vars.rename(columns={'sample_ID': 'Person_ID'})\n",
    "\n",
    "gene_vars_count = pd.DataFrame(batch_gene_vars['gene_var'].value_counts())\n",
    "gene_vars_count['gene_var'] = gene_vars_count.index\n",
    "gene_vars_count = gene_vars_count.reset_index(drop=True)\n",
    "gene_vars_sorted = gene_vars_count.sort_values(by = 'count', ascending=False)\n",
    "\n",
    "print('Number of variants identified with annotations:', gene_vars_sorted.shape[0])\n",
    "# so we will have 30 variants to look at bc the rest was not mapped correctly \n",
    "\n",
    "print('Number of samples with a mutation in a CH-relevant position (with >2 reads):', batch_all_ns.shape[0])\n",
    "print('Number of patients with a mutation in a CH-relevant position (with >2 reads):', len(batch_all_ns.sample_ID.unique()))\n",
    "print('Number of samples with a CH hotspot mutation (with > 2 reads):', batch_gene_vars.shape[0])\n",
    "print('Number of patients with a CH hotspot mutation (with > 2 reads):', len(batch_gene_vars.Person_ID.unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add age data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals for whom age data is available: 459385\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# load the age dataset \n",
    "# NOTE: age here is only available for healthy (ie cancer-free), therefore we also exclude cancer patients from analysis \n",
    "age_data = pd.read_csv('~/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/2022-05-20_healthy_pheno.tsv', sep = '\\t')\n",
    "\n",
    "age_df = age_data[['ID_v0', 'Age.when.attended.assessment.centre_v0']]\n",
    "age_df.columns.values[0] = 'Person_ID'\n",
    "age_df.columns.values[1] = 'age'\n",
    "\n",
    "print('Number of individuals for whom age data is available:', age_df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sample with available CH variant: 2249\n",
      "Number of sample with available CH variant and age: 1986\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# add age data to the rest of the data \n",
    "\n",
    "batch_gene_age = pd.merge(batch_gene_vars, age_df, on = 'Person_ID') \n",
    "print(\"Number of sample with available CH variant:\", batch_gene_vars.shape[0])\n",
    "print(\"Number of sample with available CH variant and age:\", batch_gene_age.shape[0])\n",
    "# note: this is throwing away patients who have been diagnosed with cancer (this dataset is for healthy individuals only)\n",
    "# I think it is valid and makes sense to not include patients who are diagnosed with cancer "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load MHC class I genotyping data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with imputed HLA: 488377\n",
      "Number of samples with confidently imputed MHC I: 443620\n",
      "Number of samples with confidently imputed MHC II: 378317\n",
      "Number of UKBB participants with correctly genotyped MHC class I and II: 346968\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# add MHC genotype data to CH cases\n",
    "\n",
    "# path to file\n",
    "file_hla = \"/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/data/UKBB/ukb_typed_hla_with_ids_07_threshold_processed.txt\"\n",
    "\n",
    "# get the header \n",
    "header = pd.read_csv(file_hla, sep='\\t', nrows=1, header=None).values.tolist()\n",
    "head = [item for sublist in header for item in sublist]\n",
    "\n",
    "# load the dataframe \n",
    "df = pd.read_csv(file_hla, skiprows = 1, sep = ' ', header = None)\n",
    "\n",
    "# add columns \n",
    "df.columns = head \n",
    "\n",
    "# subset data for HLA-I and HLA-II class alleles \n",
    "# we are only interested in HLA-I for the momnet \n",
    "df_hla1 = df.filter(regex='^(Person_|A_|B_|C_)') # 488377 cases \n",
    "df_hla2 = df.filter(regex='^(Person_|D)') # 488377 cases \n",
    "\n",
    "# clean up the dataset (filtering according to Krishna et al., 2024 Science: remove people in whom not all alleles were genotyped with sufficient confidence)\n",
    "df_clean_hla1 = df_hla1[~df_hla1.isin([0.5]).any(axis=1)] # exclude people for whom we lacked a genotype\n",
    "df_clean_hla1 = df_clean_hla1[~df_clean_hla1.isin([1.5]).any(axis=1)] # exclude people for whom we lacked a genotype\n",
    "df_clean_hla2 = df_hla2[~df_hla2.isin([0.5]).any(axis=1)] # exclude people for whom we lacked a genotype\n",
    "df_clean_hla2 = df_clean_hla2[~df_clean_hla2.isin([1.5]).any(axis=1)] # exclude people for whom we lacked a genotype\n",
    "\n",
    "print(\"Number of samples with imputed HLA:\", df_hla1.shape[0])\n",
    "print(\"Number of samples with confidently imputed MHC I:\", df_clean_hla1.shape[0])\n",
    "print(\"Number of samples with confidently imputed MHC II:\", df_clean_hla2.shape[0])\n",
    "\n",
    "# identify people who have correctly genotyped MHC class I and class II \n",
    "ids_hla1 = df_clean_hla1.Person_ID.tolist()\n",
    "ids_hla2 = df_clean_hla2.Person_ID.tolist()\n",
    "ids_correct_genotype = list(set(ids_hla1).intersection(ids_hla2))\n",
    "print('Number of UKBB participants with correctly genotyped MHC class I and II:', len(ids_correct_genotype))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with available CH variant: 2249\n",
      "Number of samples with available CH variant (healthy): 1986\n",
      "Number of samples with annotated variants CH variant and age and HLA genotype: 1798\n",
      "Number of HLA alleles which have been identified in the UK BioBank: 214\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# now you need to merge the data that you want to use with the allele data to get predictions\n",
    "# most comprehensive dataset we have is batch_gene_age (includes genetic variant someone carries and age)\n",
    "# Find IDs of CH-affected persons and match to their HLA genotype \n",
    "\n",
    "batch_gene_vars.rename(columns={'sample_ID': 'Person_ID'}, inplace=True)\n",
    "batch_gene_age.rename(columns={'sample_ID': 'Person_ID'}, inplace=True)\n",
    "ids_batch = batch_gene_age['Person_ID']\n",
    "hla_batch_ids = df_clean_hla1[df_clean_hla1['Person_ID'].isin(ids_batch)] \n",
    "\n",
    "# add HLA genotype data to samples with annotated CH variant and age\n",
    "batch_gene_age_hla = pd.merge(batch_gene_age, hla_batch_ids, on='Person_ID')\n",
    "print(\"Number of samples with available CH variant:\", batch_gene_vars.shape[0])\n",
    "print(\"Number of samples with available CH variant (healthy):\", batch_gene_age.shape[0])\n",
    "print('Number of samples with annotated variants CH variant and age and HLA genotype:', batch_gene_age_hla.shape[0]) # 1 HLA genotype missing \n",
    "\n",
    "# first, filter the df such that we only have MHC alleles for which predictions are available \n",
    "hla_ukbb = df_clean_hla1.filter(regex='\\d').columns # HLA from all UKBB\n",
    "print('Number of HLA alleles which have been identified in the UK BioBank:', len(hla_ukbb))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants to examine (top 20 most common ones): 20\n",
      "Number of samples with a CH hotspot mutation (with > 2 reads): 2249\n",
      "Number of patients with a CH hotspot mutation (with > 2 reads): 2214\n",
      "Number of samples with a CH hotspot mutation (with > 2 reads) (common variants): 1737\n",
      "Number of patients with a CH hotspot mutation (with > 2 reads) (common variants): 1710\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Filtering 2: we will only be looking at variants for which we have more than 10 samples \n",
    "gene_vars_count = pd.DataFrame(batch_gene_vars['gene_var'].value_counts())\n",
    "gene_vars_count['gene_var'] = gene_vars_count.index\n",
    "gene_vars_count = gene_vars_count.reset_index(drop=True)\n",
    "gene_vars_sorted = gene_vars_count.sort_values(by = 'count', ascending=False)\n",
    "\n",
    "variants_to_examine = gene_vars_sorted[gene_vars_sorted['count'] >= 15].gene_var.tolist()\n",
    "print('Number of variants to examine (top 20 most common ones):', len(variants_to_examine))\n",
    "\n",
    "# filter the dataframe to only include CH variants with minimum 10 samples \n",
    "# we will be using this dataset going forward \n",
    "batch_gene_vars_hla_10 = batch_gene_age_hla[batch_gene_age_hla['gene_var'].isin(variants_to_examine)]\n",
    "print('Number of samples with a CH hotspot mutation (with > 2 reads):', batch_gene_vars.shape[0])\n",
    "print('Number of patients with a CH hotspot mutation (with > 2 reads):', len(batch_gene_vars.Person_ID.unique()))\n",
    "print('Number of samples with a CH hotspot mutation (with > 2 reads) (common variants):', batch_gene_vars_hla_10.shape[0])\n",
    "print('Number of patients with a CH hotspot mutation (with > 2 reads) (common variants):', len(batch_gene_vars_hla_10.Person_ID.unique()))\n",
    "\n",
    "# make sure you only have correct levels of things \n",
    "batch_gene_vars_hla_10.gene_var = batch_gene_vars_hla_10.gene_var.astype(str)\n",
    "batch_gene_vars_hla_10.gene_var = batch_gene_vars_hla_10.gene_var.astype('category')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load MHC binding affinity prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# need to change format of HLA allele name to align with HLA genotyping file\n",
    "\n",
    "import re\n",
    "\n",
    "# transform format (from NetMHCpan data to UKBB format)\n",
    "def transform_format_netmhc(input_string):\n",
    "    # a regular expression pattern to match the input format\n",
    "    pattern = re.compile(r'HLA-(\\w\\S*)(\\d{2}):(\\d{2})')\n",
    "\n",
    "    # check if there is a match\n",
    "    match = pattern.match(input_string)\n",
    "\n",
    "    # if match, apply transformation\n",
    "    if match:\n",
    "        group1 = match.group(1).replace('*', '') # remove the star \n",
    "        group2 = int(match.group(2)) # remove zeros at the start \n",
    "        group3 = match.group(3) # leave as it is \n",
    "\n",
    "        # Format the output string\n",
    "        output_string = f'{group1}_{group2}{group3}' # stitch back \n",
    "        return output_string # return transformed string \n",
    "\n",
    "    # if no much, return original string \n",
    "    return input_string\n",
    "\n",
    "# transform format (from NetMHCpan data to UKBB format) - VERY annoyingly running the same pipeline I can get different formats of HLA allele names\n",
    "def transform_format_netmhc_2(input_string):\n",
    "    # a regular expression pattern to match the input format\n",
    "    pattern = re.compile(r'HLA-(\\w)(\\d{2}):(\\d{2})')\n",
    "\n",
    "    # check if there is a match\n",
    "    match = pattern.match(input_string)\n",
    "\n",
    "    # if match, apply transformation\n",
    "    if match:\n",
    "        group1 = match.group(1) # remove the star \n",
    "        group2 = int(match.group(2)) # remove zeros at the start \n",
    "        group3 = match.group(3) # leave as it is \n",
    "\n",
    "        # Format the output string\n",
    "        output_string = f'{group1}_{group2}{group3}' # stitch back \n",
    "        return output_string # return transformed string \n",
    "\n",
    "    # if no much, return original string \n",
    "    return input_string\n",
    "\n",
    "# transform format (from PRIME format to UKBB format)\n",
    "def transform_format_prime(input_string):\n",
    "    # regular expression pattern to match the input format\n",
    "    # Nb PRIME is almost the same as NetMHC but there is no star after \"HLA-A*\"\n",
    "    pattern = re.compile(r'HLA-(\\w)(\\d{2}):(\\d{2})')\n",
    "\n",
    "    # check if there is a match\n",
    "    match = pattern.match(input_string)\n",
    "\n",
    "    # if match, apply transformation\n",
    "    if match:\n",
    "        group1 = match.group(1).replace('*', '') # remove the star \n",
    "        group2 = int(match.group(2)) # remove zeros at the start \n",
    "        group3 = match.group(3) # leave as it is \n",
    "\n",
    "        # Format the output string\n",
    "        output_string = f'{group1}_{group2}{group3}' # stitch back \n",
    "        return output_string # return transformed string \n",
    "\n",
    "    # if no much, return original string \n",
    "    return input_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of alleles for which predictions are available (NetMHC): 194\n",
      "Number of unique variants (NetMHC): 42\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Import NetMHC scores  \n",
    "\n",
    "pred_file_net = '/Users/barbarawalkowiak/Desktop/msc_thesis/task1_predict_binding_to_HLA/netMHC_out/scores/20240210_NetMHC_HLA_UKBB_with_affinities_bestscores.csv' # EL scores, BA scores, EL rank, BA rank, affinity prediction\n",
    "pred_filename_net = pred_file_net.split('/')[2].split('.')[0] # identify file name \n",
    "pred_method_net = pred_file_net.split('_out')[0] # identify method used to obtain predictions\n",
    "\n",
    "# load the csv\n",
    "netmhc = pd.read_csv(pred_file_net) # load the csv \n",
    "netmhc = netmhc.rename(columns={'Aff(nM)': 'Aff_nM'})\n",
    "\n",
    "# specify HLA allele naming format \n",
    "NET_col = netmhc.HLA\n",
    "NET_formatted = NET_col.apply(transform_format_netmhc)\n",
    "netmhc = pd.concat([netmhc, NET_formatted.rename('HLA_formatted')], axis = 1)\n",
    "\n",
    "# select required columns and sor values \n",
    "netmhc = netmhc[['HLA_formatted', 'Peptide', '%Rank_EL', 'Score_EL', '%Rank_BA', 'Score_BA', 'Aff_nM', 'gene', 'variant', 'genotype']]\n",
    "netmhc = netmhc.sort_values(by=['HLA_formatted', 'gene', 'variant', 'genotype'])\n",
    "\n",
    "# identify gene variants \n",
    "netmhc['gene_var_gt'] = netmhc['gene'] + '_' + netmhc['variant'] + '_' + netmhc['genotype'] # add complete genotype data\n",
    "netmhc['gene_var'] = netmhc['gene'] + '_' + netmhc['variant']\n",
    "netmhc = netmhc.rename(columns={'Aff(nM)': 'Aff_nM'}) # rename affinity column \n",
    "scores_netmhc = netmhc[['HLA_formatted', 'Score_EL', '%Rank_EL', 'Score_BA', '%Rank_BA', 'Aff_nM', 'gene_var', 'gene_var_gt']] # select columns of interest\n",
    "\n",
    "# Print the results \n",
    "print('Number of alleles for which predictions are available (NetMHC):', len(netmhc.HLA_formatted.unique()))\n",
    "print('Number of unique variants (NetMHC):',  len(netmhc.gene_var.unique()))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define the function to find best score for variant carried by someone\n",
    "def find_best_score_for_variant_carried(row, df, param):\n",
    "\n",
    "    '''\n",
    "    this function assigns predictions from a prediction dataset (df)\n",
    "    based on the specified parameter of interest \n",
    "    in this script, I am only considering %Rank_EL\n",
    "    '''\n",
    "    \n",
    "    row_values = pd.to_numeric(row[1:-2], errors='coerce')\n",
    "    \n",
    "    # find alleles which are present \n",
    "    hla = row.index[1:-2][row_values >= 1] \n",
    "    \n",
    "    # find values corresponding to these alleles \n",
    "    vals = df.loc[df['gene_var'] == row['gene_var'], hla].values.flatten() \n",
    "    # print(vals)\n",
    "    # print(type(vals))\n",
    "    # it can be that nothing was found e.g., bc there were no predictions made for this variant\n",
    "    if vals.size == 0: \n",
    "        value = None\n",
    "    \n",
    "    else: \n",
    "    \n",
    "        if param == \"Aff_nM\":\n",
    "            value = min(vals) # highest affinity corresponds to lowest value (see plots above)\n",
    "        elif param == \"%Rank_EL\":\n",
    "            value = min(vals) # the best rank is the lowest one (indicates peptide in top x % of binders)\n",
    "        elif param == \"Score_EL\":\n",
    "            value = max(vals) # the best score is the highest one \n",
    "        elif param == \"%Rank_BA\":\n",
    "            value = min(vals) # the best rank is the lowest one (indicates peptide in top x % of binders)\n",
    "        elif param == \"Score_BA\":\n",
    "            value = max(vals) # the best score is the highest one\n",
    "        else:\n",
    "            print('Incorrect parameter provided') \n",
    "    \n",
    "    return value   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# get predictions\n",
    "\n",
    "# define a function to obtain dataframes with predictions for different scores \n",
    "def pred_to_scores(df, param):\n",
    "\n",
    "    '''\n",
    "    Allowed parameters are:\n",
    "    Aff_nM - affinity (raw number)\n",
    "    Score_BA - binding affinity score\n",
    "    Score_EL - elution score\n",
    "    %Rank_BA - %Rank of binding affinity cf a set of random peptides\n",
    "    %Rank_EL - %Rank of elution cf a set of random peptides\n",
    "    '''\n",
    "\n",
    "    # for a given parameter, prepare a dataframe where you store the genetic variant and best score for each HLA allele\n",
    "    pred_sub = df[['HLA_formatted', 'gene_var_gt', param]]\n",
    "    pred_sub_wide = pd.pivot(pred_sub, index='gene_var_gt', columns='HLA_formatted', values=param)\n",
    "    pred_sub_wide = pred_sub_wide.reset_index() # this is to make sure that you have the gene_var column in there too \n",
    "    \n",
    "    # create a list of HLA alleles genotyped in the UKBB for which predictions are available\n",
    "    hla_intersect = pred_sub_wide.columns[pred_sub_wide.columns.isin(hla_ukbb)] # HLA in the UKBB which I have predictions for \n",
    "    hla_intersect_list = hla_intersect.tolist() \n",
    "\n",
    "    # subset predictions to only include scores for available \n",
    "    pred_sub = pred_sub_wide[hla_intersect_list + pred_sub_wide.columns[pred_sub_wide.columns.str.contains('gene_var')].tolist()]\n",
    "\n",
    "    # prepare gene variants names to match names in the Patient file \n",
    "    pred_sub = pred_sub[pred_sub['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "    pred_sub['gene_var'] = pred_sub['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "    pred_sub['gene_var'] = pred_sub['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "\n",
    "    # subset batch_gene_age_hla file \n",
    "    ch_hla_sub = batch_gene_age_hla[hla_intersect_list + batch_gene_age_hla.columns[batch_gene_age_hla.columns.str.contains('gene_var')].tolist()]\n",
    "    ch_hla_sub = pd.concat([batch_gene_age_hla[\"Person_ID\"], ch_hla_sub], axis=1) # add CH cases \n",
    "    ch_hla_sub['score'] = ch_hla_sub.apply(find_best_score_for_variant_carried, df=pred_sub, param=param, axis=1) # add score for the parameter\n",
    "    ch_hla_scores = ch_hla_sub.dropna() # remove NA (incorrectly annotated cases)\n",
    "\n",
    "    # merge scores with VAF and age\n",
    "    age_vaf = batch_gene_age_hla[['Person_ID', 'VAF', 'var_depth', 'age', 'gene_var']]\n",
    "    ch_hla_merge = pd.merge(ch_hla_scores, age_vaf, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "    # now add the columns with VAF and age \n",
    "    col_to_select = ['Person_ID', 'gene_var', 'score', 'age', 'var_depth', 'VAF'] # subset the data \n",
    "    ch_hla_merge_sub = ch_hla_merge[col_to_select]\n",
    "\n",
    "    return ch_hla_merge_sub\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# run the function \n",
    "param = '%Rank_EL' # specify parameter of interest \n",
    "\n",
    "# get individual dataframes\n",
    "ch_best_el_rank = pred_to_scores(netmhc, param) # generate dataframe with scores for all CH cases\n",
    "ch_best_el_rank['score_type'] = '%Rank_EL_NetMHC' # add so you know what score you are using \n",
    "\n",
    "# for affinity and ranking, convert to negative log (higher -1log(rank) = tighter binding / more immunogenic)\n",
    "ch_best_el_rank['log_score'] = -1*np.log10(ch_best_el_rank['score']) # convert score to -log10(score)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define the function to find best scores \n",
    "\n",
    "def find_best_score_for_all_variants(row, df, param):\n",
    "\n",
    "    '''\n",
    "    Allowed parameters are:\n",
    "    Aff_nM - affinity (raw number)\n",
    "    Score_BA - binding affinity score\n",
    "    Score_EL - elution score\n",
    "    %Rank_BA - %Rank of binding affinity cf a set of random peptides\n",
    "    %Rank_EL - %Rank of elution cf a set of random peptides\n",
    "    '''\n",
    "\n",
    "    hlas = row.index[1:-1][row[1:-1] >= 1] # select alleles which each Person (row) carries\n",
    "   \n",
    "    # it may be that we don't have this column in case we are doing this for non-CH cases\n",
    "   \n",
    "    variants = df['gene_var'] # get out variants which are present \n",
    "   \n",
    "    scores = {} # initialise empy dictionaries\n",
    "\n",
    "    # depending on the parameter, pick the minimum of maximum value \n",
    "    if param == \"Aff_nM\":\n",
    "        for var in variants:\n",
    "            # Find the minimum value for each variant in the category that is present\n",
    "            best_value = min(df.loc[df['gene_var'] == var, hlas].values[0])\n",
    "            # Update the dictionary with the minimum value for the corresponding variant\n",
    "            scores[f'score_{var}'] = best_value\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Score_BA\":\n",
    "        for var in variants:\n",
    "            \n",
    "            best_value = max(df.loc[df['gene_var'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "        \n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Score_EL\":\n",
    "        for var in variants:\n",
    "           \n",
    "            best_value = max(df.loc[df['gene_var'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"%Rank_BA\":\n",
    "        for var in variants:\n",
    "\n",
    "            best_value = min(df.loc[df['gene_var'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"%Rank_EL\":\n",
    "        for var in variants:\n",
    "            \n",
    "            best_value = min(df.loc[df['gene_var'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define a function to find best scores for all examined CH variants\n",
    "def pred_to_scores_all_variants(df, param):\n",
    "\n",
    "    '''\n",
    "    Allowed parameters are:\n",
    "    Aff_nM - affinity (raw number)\n",
    "    Score_BA - binding affinity score\n",
    "    Score_EL - elution score\n",
    "    %Rank_BA - %Rank of binding affinity cf a set of random peptides\n",
    "    %Rank_EL - %Rank of elution cf a set of random peptides\n",
    "\n",
    "    For df, you can supply the dataframe with predictions either for NetMHC or PRIME\n",
    "    '''\n",
    "\n",
    "    # for a given parameter, prepare a dataframe where you store the genetic variant and best score for each HLA allele\n",
    "    pred_sub = df[['HLA_formatted', 'gene_var_gt', param]]\n",
    "    pred_sub_wide = pd.pivot(pred_sub, index='gene_var_gt', columns='HLA_formatted', values=param)\n",
    "    pred_sub_wide = pred_sub_wide.reset_index() # this is to make sure that you have the gene_var column in there too \n",
    "    pred_sub_wide.columns = map(transform_format, pred_sub_wide.columns)\n",
    "\n",
    "    # create a list of HLA alleles genotyped in the UKBB for which predictions are available\n",
    "    hla_intersect = pred_sub_wide.columns[pred_sub_wide.columns.isin(hla_ukbb)] # HLA in the UKBB which I have predictions for \n",
    "    hla_intersect_list = hla_intersect.tolist() \n",
    "\n",
    "    # create the prediction dataset \n",
    "    pred_sub = pred_sub_wide[hla_intersect_list + pred_sub_wide.columns[pred_sub_wide.columns.str.contains('gene_var')].tolist()]\n",
    "    ch_hla_sub = batch_gene_age_hla[hla_intersect_list + batch_gene_age_hla.columns[batch_gene_age_hla.columns.str.contains('gene_var')].tolist()]\n",
    "    ch_hla_sub = pd.concat([batch_gene_age_hla[\"Person_ID\"], ch_hla_sub], axis=1)\n",
    "    ch_hla_sub = batch_gene_age_hla[hla_intersect_list + batch_gene_age_hla.columns[batch_gene_age_hla.columns.str.contains('gene_var')].tolist()]\n",
    "    ch_hla_sub = pd.concat([batch_gene_age_hla[\"Person_ID\"], ch_hla_sub], axis=1) # add CH cases \n",
    "    \n",
    "    # prepare gene variants names to match names in the Patient file \n",
    "    pred_sub = pred_sub[pred_sub['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "    pred_sub['gene_var'] = pred_sub['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "    pred_sub['gene_var'] = pred_sub['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "\n",
    "    # apply the function to find, for each participant, the best score for each examined CH variant \n",
    "    ch_scores_all_var = pd.concat([ch_hla_sub, ch_hla_sub.apply(find_best_score_for_all_variants, df=pred_sub, param=param, axis=1)], axis=1)\n",
    "\n",
    "    # merge scores with VAF and age\n",
    "    age_vaf = batch_gene_age_hla[['Person_ID', 'VAF', 'var_depth', 'age', 'gene_var']]\n",
    "    ch_hla_merge = pd.merge(ch_scores_all_var, age_vaf, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "    return ch_hla_merge\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# run the function \n",
    "params = '%Rank_EL' # here we are only interested in %Rank_EL\n",
    "ch_all_el_rank = pred_to_scores_all_variants(netmhc, param)\n",
    "\n",
    "# add a column to indicate the name of the scoring system that you used\n",
    "ch_all_el_rank['score_type'] = '%Rank_EL_NetMHC'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# now let's also get best log_scores \n",
    "def find_best_score_all_variants_log(row, df, param):\n",
    "    \n",
    "    '''\n",
    "    The same function but log scores \n",
    "    the only allowed parameters are %Rank_EL and %Rank_BA\n",
    "    we want -1 * log(score) so the highest score if the most \"immunogenic\" (best binding) one\n",
    "    '''\n",
    "    \n",
    "    hlas = row.index[1:-1][row[1:-1] >= 1] # select alleles which each Person (row) carries\n",
    "\n",
    "    variants = df['gene_var']\n",
    "    \n",
    "    scores = {} # initialise empty dictionaries\n",
    "\n",
    "    if param == \"%Rank_EL\":\n",
    "        for var in variants:\n",
    "\n",
    "            # take maximum of the negative log score \n",
    "            best_value = max(-1 * np.log10(df.loc[df['gene_var'] == var, hlas].values[0]))\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"%Rank_BA\":\n",
    "        for var in variants:\n",
    "\n",
    "            # take maxium of the negative log score \n",
    "            best_value = max(-1 * np.log10(df.loc[df['gene_var'] == var, hlas].values[0]))\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Aff_nM\":\n",
    "        for var in variants:\n",
    "\n",
    "            # take maxium of the negative log score (low aff = more immunogenic)\n",
    "            best_value = max(-1 * np.log10(df.loc[df['gene_var'] == var, hlas].values[0]))\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "\n",
    "# define a function to find best scores for all examined CH variants\n",
    "def pred_to_scores_all_variants_log(df, param):\n",
    "\n",
    "    '''\n",
    "    Allowed parameters are:\n",
    "    Aff_nM - affinity (raw number)\n",
    "    %Rank_BA - %Rank of binding affinity cf a set of random peptides\n",
    "    %Rank_EL - %Rank of elution cf a set of random peptides\n",
    "    '''\n",
    "\n",
    "    # for a given parameter, prepare a dataframe where you store the genetic variant and best score for each HLA allele\n",
    "    pred_sub = df[['HLA_formatted', 'gene_var_gt', param]]\n",
    "    pred_sub_wide = pd.pivot(pred_sub, index='gene_var_gt', columns='HLA_formatted', values=param)\n",
    "    pred_sub_wide = pred_sub_wide.reset_index() # this is to make sure that you have the gene_var column in there too \n",
    "    pred_sub_wide.columns = map(transform_format, pred_sub_wide.columns)\n",
    "\n",
    "    # create a list of HLA alleles genotyped in the UKBB for which predictions are available\n",
    "    hla_intersect = pred_sub_wide.columns[pred_sub_wide.columns.isin(hla_ukbb)] # HLA in the UKBB which I have predictions for \n",
    "    hla_intersect_list = hla_intersect.tolist() \n",
    "\n",
    "    # create the prediction dataset \n",
    "    pred_sub = pred_sub_wide[hla_intersect_list + pred_sub_wide.columns[pred_sub_wide.columns.str.contains('gene_var')].tolist()]\n",
    "    ch_hla_sub = batch_gene_age_hla[hla_intersect_list + batch_gene_age_hla.columns[batch_gene_age_hla.columns.str.contains('gene_var')].tolist()]\n",
    "    ch_hla_sub = pd.concat([batch_gene_age_hla[\"Person_ID\"], ch_hla_sub], axis=1)\n",
    "    ch_hla_sub = batch_gene_age_hla[hla_intersect_list + batch_gene_age_hla.columns[batch_gene_age_hla.columns.str.contains('gene_var')].tolist()]\n",
    "    ch_hla_sub = pd.concat([batch_gene_age_hla[\"Person_ID\"], ch_hla_sub], axis=1) # add CH cases \n",
    "    \n",
    "    # prepare gene variants names to match names in the Patient file \n",
    "    pred_sub = pred_sub[pred_sub['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "    pred_sub['gene_var'] = pred_sub['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "    pred_sub['gene_var'] = pred_sub['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "\n",
    "    # apply the function to find, for each participant, the best score for each examined CH variant \n",
    "    ch_scores_all_var = pd.concat([ch_hla_sub, ch_hla_sub.apply(find_best_score_all_variants_log, df=pred_sub, param=param, axis=1)], axis=1)\n",
    "\n",
    "    # merge scores with VAF and age\n",
    "    age_vaf = batch_gene_age_hla[['Person_ID', 'VAF', 'var_depth', 'age', 'gene_var']]\n",
    "    ch_hla_merge = pd.merge(ch_scores_all_var, age_vaf, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "    return ch_hla_merge\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# run the function \n",
    "param = '%Rank_EL' # allowed parameters (scores we are interested in)\n",
    "ch_all_el_rank_log = pred_to_scores_all_variants_log(netmhc, param)\n",
    "ch_all_el_rank_log['score_type'] = '%Rank_EL_log_NetMHC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# melt the dataset \n",
    "df_all_names = ['ch_all_el_rank', 'ch_all_el_rank_log']\n",
    "dict_melted = {}\n",
    "\n",
    "for name in df_all_names:\n",
    "\n",
    "    df = globals()[name] \n",
    "    key = f'df_melted_{name}'\n",
    "\n",
    "    score_columns = [col for col in df.columns if col.startswith('score_')]\n",
    "    score_columns.remove('score_type') # get rid of this column, we don't want that \n",
    "    col_to_select = ['Person_ID', 'gene_var'] + score_columns\n",
    "    df = df[col_to_select]\n",
    "\n",
    "    df_melted = pd.melt(df, id_vars=['Person_ID', 'gene_var'])\n",
    "    dict_melted[key] = df_melted\n",
    "\n",
    "# get individual dataframes\n",
    "ch_all_el_rank_melted = dict_melted['df_melted_ch_all_el_rank']\n",
    "ch_all_el_rank_log_melted = dict_melted['df_melted_ch_all_el_rank_log']\n",
    "\n",
    "# add a column to indicate the name of the scoring system that you used\n",
    "ch_all_el_rank_melted['score_type'] = '%Rank_EL_NetMHC'\n",
    "ch_all_el_rank_log_melted['score_type'] = '%Rank_EL_log_NetMHC'\n",
    "\n",
    "# In this dataset, for everyone you have the variant they carry (gene_var)\n",
    "# and the score for each possible CH variant (variable)\n",
    "# to find the score for the variant they have, need to find all cases where gene_var == log_score_gene_var_ch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_names_melted = ['ch_all_el_rank_melted', 'ch_all_el_rank_log_melted']\n",
    "dict_melt_add = {}\n",
    "\n",
    "for name in df_names_melted:\n",
    "\n",
    "    df = globals()[name] \n",
    "    df['is_variant_present'] = df.apply(lambda row: row['gene_var'] in row['variable'], axis=1)\n",
    "\n",
    "    levels = [True, False] # set order of levels for plotting\n",
    "    df['is_variant_present'] = pd.Categorical(df['is_variant_present'].astype(bool), categories=levels, ordered=True) # convert to categorical\n",
    "    df['gene_var'] = df['gene_var'].astype('category') # convert to categorical\n",
    "\n",
    "    # NOTE: I want to sort by log score for present variants\n",
    "    df['gene_var'] = df['gene_var'].astype(str) # convert gene_var to string (not category)\n",
    "    df['variable'] = df['variable'].astype(str)\n",
    "    df['mean_variant_score'] = df.groupby('variable')['value'].transform('mean')\n",
    "    df_sort = df.sort_values(by='mean_variant_score', ascending = False)\n",
    "\n",
    "    key = f'df_melted_{name}'\n",
    "    dict_melt_add[key] = df_sort\n",
    "\n",
    "ch_all_el_rank_melted_sort = dict_melt_add['df_melted_ch_all_el_rank_melted']\n",
    "ch_all_el_rank_log_melted_sort = dict_melt_add['df_melted_ch_all_el_rank_log_melted']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batches that were examined for variants: {11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21}\n",
      "Number of samples examined for variants: 99597\n"
     ]
    }
   ],
   "source": [
    "# subset the dataframe to only include individuals who were actually screened for CH\n",
    "batch_all['batch_number'] = batch_all['batch'].apply(lambda x: [int(num) for num in re.findall(r'\\d+', x)])\n",
    "batch_numbers_examined = set(batch_all['batch_number'].sum()) \n",
    "print('Batches that were examined for variants:', batch_numbers_examined)\n",
    "\n",
    "all_dataframes = [f'batch_{i}_ids' for i in range(11, 61)]\n",
    "selected_ids = []\n",
    "\n",
    "# Iterate through all df name \n",
    "for df_name in all_dataframes:\n",
    "    number = int(re.search(r'\\d+', df_name).group())\n",
    "    # check if number matches numbers of batches examined \n",
    "    if number in batch_numbers_examined:        \n",
    "        selected_ids.append(globals()[df_name])\n",
    "\n",
    "ids_examined = pd.concat(selected_ids, ignore_index=True)\n",
    "print('Number of samples examined for variants:', ids_examined.shape[0])\n",
    "\n",
    "ids_examined_list = ids_examined['sample_ID'].tolist()\n",
    "ids_examined_list = [int(id) for id in ids_examined_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of healthy (non-CH) individuals in the UKBB: 88462\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# select a random set of 2000 non-carriers and compare to scores in carriers\n",
    "\n",
    "# identify UKBB participants who are not diagnosed with CH \n",
    "ch_sampleids = batch_gene_vars['Person_ID']\n",
    "df_clean_hla1_screened = df_clean_hla1[df_clean_hla1['Person_ID'].isin(ids_examined_list)]\n",
    "ukbb_no_ch = df_clean_hla1_screened[~df_clean_hla1_screened['Person_ID'].isin(ch_sampleids)]\n",
    "hla_ukbb = ukbb_no_ch.filter(regex='\\d').columns # identify HLA alleles carried by non-CH-individuals in the UKBB data\n",
    "\n",
    "# prepare dataset with predictions (now just from NetMHC)\n",
    "scores_netmhc_elrank = scores_netmhc[['HLA_formatted', '%Rank_EL', 'gene_var_gt']]\n",
    "scores_netmhc_elrank_wide = pd.pivot(scores_netmhc_elrank, index='gene_var_gt', columns='HLA_formatted', values='%Rank_EL')\n",
    "\n",
    "hla_intersect = scores_netmhc_elrank_wide.columns[scores_netmhc_elrank_wide.columns.isin(hla_ukbb)]\n",
    "hla_intersect_list = hla_intersect.tolist() # select predictions for alleles which are available in NetMHC / PRIME\n",
    "\n",
    "pred_sub_wide = scores_netmhc_elrank_wide[hla_intersect_list + scores_netmhc_elrank_wide.columns[scores_netmhc_elrank_wide.columns.str.contains('gene_var_gt')].tolist()]\n",
    "\n",
    "ukbb_no_ch_hla = ukbb_no_ch[hla_intersect]\n",
    "ukbb_no_ch = pd.concat([ukbb_no_ch['Person_ID'], ukbb_no_ch_hla], axis = 1)\n",
    "ukbb_no_ch.head()\n",
    "\n",
    "ids_non_carriers_healthy = ukbb_no_ch.Person_ID.unique().tolist()\n",
    "print('Number of healthy (non-CH) individuals in the UKBB:', len(ids_non_carriers_healthy))\n",
    "\n",
    "pred_sub_wide['gene_var_gt'] = pred_sub_wide.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define new functions in case you want to get predictions for both CH and refseq\n",
    "\n",
    "def find_best_score_for_all_variants_ref_ch(row, df, param):\n",
    "\n",
    "    '''\n",
    "    Allowed parameters are:\n",
    "    Aff_nM - affinity (raw number)\n",
    "    Score_BA - binding affinity score\n",
    "    Score_EL - elution score\n",
    "    %Rank_BA - %Rank of binding affinity cf a set of random peptides\n",
    "    %Rank_EL - %Rank of elution cf a set of random peptides\n",
    "    '''\n",
    "\n",
    "    hlas = row.index[1:-1][row[1:-1] >= 1] # select alleles which each Person (row) carries\n",
    "   \n",
    "    # it may be that we don't have this column in case we are doing this for non-CH cases\n",
    "   \n",
    "    variants = df['gene_var_gt'] # get out variants which are present \n",
    "   \n",
    "    scores = {} # initialise empy dictionaries\n",
    "\n",
    "    # depending on the parameter, pick the minimum of maximum value \n",
    "    if param == \"Aff_nM\":\n",
    "        for var in variants:\n",
    "            # Find the minimum value for each variant in the category that is present\n",
    "            best_value = min(df.loc[df['gene_var_gt'] == var, hlas].values[0])\n",
    "            # Update the dictionary with the minimum value for the corresponding variant\n",
    "            scores[f'score_{var}'] = best_value\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Score_BA\":\n",
    "        for var in variants:\n",
    "            \n",
    "            best_value = max(df.loc[df['gene_var_gt'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "        \n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Score_EL\":\n",
    "        for var in variants:\n",
    "           \n",
    "            best_value = max(df.loc[df['gene_var_gt'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"%Rank_BA\":\n",
    "        for var in variants:\n",
    "\n",
    "            best_value = min(df.loc[df['gene_var_gt'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"%Rank_EL\":\n",
    "        for var in variants:\n",
    "            \n",
    "            best_value = min(df.loc[df['gene_var_gt'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "\n",
    "def find_best_score_all_variants_log_ref_ch(row, df, param):\n",
    "    \n",
    "    '''\n",
    "    The same function but log scores \n",
    "    the only allowed parameters are %Rank_EL and %Rank_BA\n",
    "    we want -1 * log(score) so the highest score if the most \"immunogenic\" (best binding) one\n",
    "    '''\n",
    "    \n",
    "    hlas = row.index[1:-1][row[1:-1] >= 1] # select alleles which each Person (row) carries\n",
    "\n",
    "    variants = df['gene_var_gt']\n",
    "    \n",
    "    scores = {} # initialise empty dictionaries\n",
    "\n",
    "    if param == \"%Rank_EL\":\n",
    "        for var in variants:\n",
    "\n",
    "            # take maximum of the negative log score \n",
    "            best_value = max(-1 * np.log10(df.loc[df['gene_var_gt'] == var, hlas].values[0]))\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"%Rank_BA\":\n",
    "        for var in variants:\n",
    "\n",
    "            # take maxium of the negative log score \n",
    "            best_value = max(-1 * np.log10(df.loc[df['gene_var_gt'] == var, hlas].values[0]))\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Aff_nM\":\n",
    "        for var in variants:\n",
    "\n",
    "            # take maxium of the negative log score (low aff = more immunogenic)\n",
    "            best_value = max(-1 * np.log10(df.loc[df['gene_var_gt'] == var, hlas].values[0]))\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# select only HLAs for which we have predictions\n",
    "df_clean_hla1_hla = df_clean_hla1[hla_intersect]\n",
    "df_clean_hla1_hla_sel = pd.concat([df_clean_hla1['Person_ID'], df_clean_hla1_hla], axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define necessary functions\n",
    "def get_score(row):\n",
    "    \n",
    "    category = row['gene_var']\n",
    "    return row[f'score_{category}']\n",
    "\n",
    "def assign_labels(group):\n",
    "\n",
    "    total_samples = group.shape[0]\n",
    "    ordered_group = group.sort_values('score', ascending = True)\n",
    "    # Assign \"bottom\" to the first half of the entries and \"top\" to the rest\n",
    "    ordered_group['label'] = ['bottom half'] * (total_samples // 2) + ['top half'] * (total_samples - (total_samples // 2))\n",
    "    \n",
    "    return ordered_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# assign labels using the function within each category\n",
    "\n",
    "df_names_to_label = ['ch_all_el_rank_log']\n",
    "dict_names_to_label = {}\n",
    "\n",
    "for name in df_names_to_label:\n",
    "\n",
    "    # get score for each dataframe\n",
    "    df = globals()[name] \n",
    "    df['score'] = df.apply(get_score, axis = 1)\n",
    "\n",
    "    # find median score for each variant \n",
    "    median_score = df.groupby('gene_var')['score'].transform('median')\n",
    "    df['median_score_for_group'] = median_score\n",
    "\n",
    "    # first, add an extra column, group ('top half', 'bottom halfs' for each variant)\n",
    "    df['group_score'] = 'bottom half'  # everyone gets the low score\n",
    "\n",
    "    # Note: TOP people will have a HIGHER score than median because we have done negative log transform \n",
    "    df.loc[df['score'] >= median_score, 'group_score'] = 'top half'  # top 50% presenters if index is higher than median \n",
    "\n",
    "    # make sure ordering is correct\n",
    "    df['gene_var'] = df['gene_var'].astype('category') \n",
    "\n",
    "    # add labels\n",
    "    df = df.groupby('gene_var', group_keys=False).apply(assign_labels)\n",
    "\n",
    "    # add the df to the dictionary \n",
    "    key = f'df_labels_{name}'\n",
    "    dict_names_to_label[key] = df\n",
    "\n",
    "ch_all_el_rank_log_labels = dict_names_to_label['df_labels_ch_all_el_rank_log']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# new way of doing this where you do not sample 5x but you only sample ONE reference from EVERYONE who does not have this variant (CH with different variant or healthy)\n",
    "# we use 2k for the reference to have a really good idea of how the distribution looks like \n",
    "# we also do the plot for all variants together \n",
    "\n",
    "# full script to look across variants  \n",
    "\n",
    "np.random.seed(1) # ensure this is all reproducible \n",
    "\n",
    "# specify list of variants to look at \n",
    "variants = batch_gene_vars.gene_var.unique()\n",
    "colors = ['#f00071', '#0422ed']\n",
    "cols_ch = ['Person_ID', 'gene_var', 'score']\n",
    "\n",
    "# initialize empty dataframe to store results in \n",
    "df_compare_to_ref = pd.DataFrame()\n",
    "\n",
    "for var in variants:\n",
    "\n",
    "    var_name = var.replace('_', ' ')\n",
    "    \n",
    "    # we will compare to %Rank EL (log) derived from NetMHC now\n",
    "    ch_scores = ch_all_el_rank_log[ch_all_el_rank_log['gene_var'] == var]\n",
    "    ch_scores = ch_scores[cols_ch]\n",
    "\n",
    "    # identify number of variant carriers \n",
    "    n =  ch_scores.shape[0]\n",
    "\n",
    "    if n > 10: # we have filtered the dataset before so all variants should fulfill this condition \n",
    "\n",
    "        # find all the dataset that you can sample from\n",
    "        ids_non_carriers_ch = ch_all_el_rank_log[ch_all_el_rank_log['gene_var'] != var].Person_ID.unique().tolist()\n",
    "        ids_non_carriers = ids_non_carriers_healthy + ids_non_carriers_ch\n",
    "        ukbb_not_carrier = df_clean_hla1_hla_sel[df_clean_hla1_hla_sel.Person_ID.isin(ids_non_carriers)]\n",
    "        ref = ukbb_not_carrier.sample(n=2000, replace=False) # no duplicates\n",
    "            \n",
    "        # specify parameter \n",
    "        param = '%Rank_EL'\n",
    "\n",
    "        # add scores \n",
    "        ref_scores = pd.concat([ref, ref.apply(find_best_score_all_variants_log_ref_ch, df=pred_sub_wide, param=param, axis=1)], axis=1)\n",
    "        ref_scores['gene_var'] = 'reference set'\n",
    "        ref_scores.rename(columns={f'score_{var}_ch': 'score'}, inplace=True)\n",
    "        ref_scores_var = ref_scores[cols_ch]\n",
    "        \n",
    "        # concatenate all dataframes\n",
    "        scores_compare_all = pd.concat([ch_scores, ref_scores_var], axis=0)\n",
    "        scores_compare_all[\"CH_variant_carrier\"] = scores_compare_all['gene_var'] == var \n",
    "        scores_compare_all[\"CH_variant_carrier\"] = scores_compare_all[\"CH_variant_carrier\"].map({True: 'Carrier', False: 'Non-carrier'})\n",
    "        \n",
    "        # add rows to dataframe with data for all variants\n",
    "        df_compare_to_ref = pd.concat([df_compare_to_ref, scores_compare_all], axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add mutation type \n",
    "def add_mut_type(string):\n",
    "    if string[-1] == '*':\n",
    "        return 'stop codon'\n",
    "    else:\n",
    "        return 'substitution'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# okay we need to add a colum with the variant for everyone (because you have reference sets to a variant but the gene_var column does not say which variant this is reference for)\n",
    "labels = []\n",
    "current_label = None\n",
    "\n",
    "# Iterate through the DataFrame\n",
    "for idx, row in df_compare_to_ref.iterrows():\n",
    "    # If the category is 'ch', update the current label\n",
    "    if row['CH_variant_carrier'] == 'Carrier':\n",
    "        current_label = row['gene_var']\n",
    "    # Append the current label to the list\n",
    "    labels.append(current_label)\n",
    "\n",
    "# Add the labels list as a new column to the DataFrame\n",
    "df_compare_to_ref['CH variant'] = labels\n",
    "\n",
    "# Add column based on the last character of strings in the 'var' column\n",
    "df_compare_to_ref['mutation_type'] = df_compare_to_ref['CH variant'].apply(add_mut_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# figure out desired order of variants on the plot \n",
    "ch_all_el_rank_log_melted_sort['median_score'] = ch_all_el_rank_log_melted_sort[ch_all_el_rank_log_melted_sort['is_variant_present']==True].groupby('gene_var')['value'].transform('median')\n",
    "ch_all_el_rank_log_melted_sort_sort = ch_all_el_rank_log_melted_sort.sort_values(by='median_score', ascending = False)\n",
    "\n",
    "# change names on the x axis for plotting \n",
    "df_compare_to_ref['CH variant2'] = df_compare_to_ref['CH variant'].str.replace('_', '\\n')\n",
    "df_compare_to_ref['gene_var2'] = df_compare_to_ref['gene_var'].str.replace('_', '\\n')\n",
    "ch_all_el_rank_log_melted_sort_sort['gene_var2'] = ch_all_el_rank_log_melted_sort_sort['gene_var'].str.replace('_', '\\n')\n",
    "order = ch_all_el_rank_log_melted_sort_sort.gene_var2.unique()\n",
    "variants_examined = df_compare_to_ref['CH variant2'].unique().tolist()\n",
    "order_2 = [var for var in order if var in variants_examined]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>gene_var</th>\n",
       "      <th>score</th>\n",
       "      <th>CH_variant_carrier</th>\n",
       "      <th>CH variant</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>CH variant2</th>\n",
       "      <th>gene_var2</th>\n",
       "      <th>nr_carriers</th>\n",
       "      <th>CH variant3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2100679</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>0.189096</td>\n",
       "      <td>Carrier</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>substitution</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>336</td>\n",
       "      <td>DNMT3A\\nR882H\\n(n=336)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2110983</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>0.970616</td>\n",
       "      <td>Carrier</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>substitution</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>336</td>\n",
       "      <td>DNMT3A\\nR882H\\n(n=336)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2113574</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>0.835647</td>\n",
       "      <td>Carrier</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>substitution</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>336</td>\n",
       "      <td>DNMT3A\\nR882H\\n(n=336)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2116112</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>0.970616</td>\n",
       "      <td>Carrier</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>substitution</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>336</td>\n",
       "      <td>DNMT3A\\nR882H\\n(n=336)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2118726</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>0.759451</td>\n",
       "      <td>Carrier</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>substitution</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>DNMT3A\\nR882H</td>\n",
       "      <td>336</td>\n",
       "      <td>DNMT3A\\nR882H\\n(n=336)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41732</th>\n",
       "      <td>1546550</td>\n",
       "      <td>reference set</td>\n",
       "      <td>1.008774</td>\n",
       "      <td>Non-carrier</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>substitution</td>\n",
       "      <td>SF3B1\\nK666N</td>\n",
       "      <td>reference set</td>\n",
       "      <td>14</td>\n",
       "      <td>SF3B1\\nK666N\\n(n=14)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41733</th>\n",
       "      <td>1288100</td>\n",
       "      <td>reference set</td>\n",
       "      <td>1.008774</td>\n",
       "      <td>Non-carrier</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>substitution</td>\n",
       "      <td>SF3B1\\nK666N</td>\n",
       "      <td>reference set</td>\n",
       "      <td>14</td>\n",
       "      <td>SF3B1\\nK666N\\n(n=14)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41734</th>\n",
       "      <td>1410949</td>\n",
       "      <td>reference set</td>\n",
       "      <td>0.910095</td>\n",
       "      <td>Non-carrier</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>substitution</td>\n",
       "      <td>SF3B1\\nK666N</td>\n",
       "      <td>reference set</td>\n",
       "      <td>14</td>\n",
       "      <td>SF3B1\\nK666N\\n(n=14)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41735</th>\n",
       "      <td>2040272</td>\n",
       "      <td>reference set</td>\n",
       "      <td>1.008774</td>\n",
       "      <td>Non-carrier</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>substitution</td>\n",
       "      <td>SF3B1\\nK666N</td>\n",
       "      <td>reference set</td>\n",
       "      <td>14</td>\n",
       "      <td>SF3B1\\nK666N\\n(n=14)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41736</th>\n",
       "      <td>1562365</td>\n",
       "      <td>reference set</td>\n",
       "      <td>-0.028571</td>\n",
       "      <td>Non-carrier</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "      <td>substitution</td>\n",
       "      <td>SF3B1\\nK666N</td>\n",
       "      <td>reference set</td>\n",
       "      <td>14</td>\n",
       "      <td>SF3B1\\nK666N\\n(n=14)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41737 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Person_ID       gene_var     score CH_variant_carrier    CH variant  \\\n",
       "0        2100679   DNMT3A_R882H  0.189096            Carrier  DNMT3A_R882H   \n",
       "1        2110983   DNMT3A_R882H  0.970616            Carrier  DNMT3A_R882H   \n",
       "2        2113574   DNMT3A_R882H  0.835647            Carrier  DNMT3A_R882H   \n",
       "3        2116112   DNMT3A_R882H  0.970616            Carrier  DNMT3A_R882H   \n",
       "4        2118726   DNMT3A_R882H  0.759451            Carrier  DNMT3A_R882H   \n",
       "...          ...            ...       ...                ...           ...   \n",
       "41732    1546550  reference set  1.008774        Non-carrier   SF3B1_K666N   \n",
       "41733    1288100  reference set  1.008774        Non-carrier   SF3B1_K666N   \n",
       "41734    1410949  reference set  0.910095        Non-carrier   SF3B1_K666N   \n",
       "41735    2040272  reference set  1.008774        Non-carrier   SF3B1_K666N   \n",
       "41736    1562365  reference set -0.028571        Non-carrier   SF3B1_K666N   \n",
       "\n",
       "      mutation_type    CH variant2      gene_var2 nr_carriers  \\\n",
       "0      substitution  DNMT3A\\nR882H  DNMT3A\\nR882H         336   \n",
       "1      substitution  DNMT3A\\nR882H  DNMT3A\\nR882H         336   \n",
       "2      substitution  DNMT3A\\nR882H  DNMT3A\\nR882H         336   \n",
       "3      substitution  DNMT3A\\nR882H  DNMT3A\\nR882H         336   \n",
       "4      substitution  DNMT3A\\nR882H  DNMT3A\\nR882H         336   \n",
       "...             ...            ...            ...         ...   \n",
       "41732  substitution   SF3B1\\nK666N  reference set          14   \n",
       "41733  substitution   SF3B1\\nK666N  reference set          14   \n",
       "41734  substitution   SF3B1\\nK666N  reference set          14   \n",
       "41735  substitution   SF3B1\\nK666N  reference set          14   \n",
       "41736  substitution   SF3B1\\nK666N  reference set          14   \n",
       "\n",
       "                  CH variant3  \n",
       "0      DNMT3A\\nR882H\\n(n=336)  \n",
       "1      DNMT3A\\nR882H\\n(n=336)  \n",
       "2      DNMT3A\\nR882H\\n(n=336)  \n",
       "3      DNMT3A\\nR882H\\n(n=336)  \n",
       "4      DNMT3A\\nR882H\\n(n=336)  \n",
       "...                       ...  \n",
       "41732    SF3B1\\nK666N\\n(n=14)  \n",
       "41733    SF3B1\\nK666N\\n(n=14)  \n",
       "41734    SF3B1\\nK666N\\n(n=14)  \n",
       "41735    SF3B1\\nK666N\\n(n=14)  \n",
       "41736    SF3B1\\nK666N\\n(n=14)  \n",
       "\n",
       "[41737 rows x 10 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# maybe add the number of observations (ie carriers) to the gene (n = )\n",
    "# however am wondering if this is going to be misleading given that the number of non-carriers is 2000\n",
    "df_carriers_counts = df_compare_to_ref[df_compare_to_ref['CH_variant_carrier']=='Carrier']['CH variant'].value_counts().reset_index()\n",
    "df_carriers_counts.columns.values[1] = 'nr_carriers'\n",
    "\n",
    "df_compare_to_ref2 = pd.merge(df_compare_to_ref, df_carriers_counts, on = 'CH variant')\n",
    "df_compare_to_ref2['nr_carriers'] = df_compare_to_ref2['nr_carriers'].astype(str)\n",
    "df_compare_to_ref2['CH variant3'] = df_compare_to_ref2['CH variant2'] + '\\n(n=' + df_compare_to_ref2['nr_carriers'] + ')'\n",
    "df_compare_to_ref2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change names on the x axis for plotting \n",
    "order = [var for var in order if var in variants_examined]\n",
    "order_3 = df_compare_to_ref2['CH variant3'].unique().tolist()\n",
    "\n",
    "ch_all_el_rank_log_melted_sort['gene_var2'] = ch_all_el_rank_log_melted_sort['gene_var'].str.replace('_', '\\n')\n",
    "median_scores_df = ch_all_el_rank_log_melted_sort[ch_all_el_rank_log_melted_sort['is_variant_present']==True][['gene_var2', 'median_score']]\n",
    "median_scores_df.columns.values[0] = 'CH variant2'\n",
    "df_compare_to_ref2_merged = pd.merge(df_compare_to_ref2, median_scores_df, on = 'CH variant2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Okay, first, we need to get a score column so that will be a log score which matches the gene var someone has \n",
    "def get_score(row):\n",
    "    \n",
    "    category = row['gene_var']\n",
    "    return row[f'score_{category}']\n",
    "\n",
    "# split into a desired number of groups\n",
    "def split_into_equal_groups(df, num_groups):\n",
    "\n",
    "    # first, assign a rank to everyone based on score\n",
    "    # the highest score = the lowest rank (ie highest score means you get rank 1)\n",
    "    # if two people have the same score, assign consecutive ranks \n",
    "    df['rank'] = df['score'].rank(method='first', ascending=False)\n",
    "    \n",
    "    # assign groups based on rank \n",
    "\n",
    "    # first, determine the number of samples\n",
    "    total_samples = len(df)\n",
    "\n",
    "    # now, determine the number of samples in each group\n",
    "    samples_per_group = total_samples // num_groups\n",
    "    remainder = total_samples % num_groups\n",
    "    group_sizes = [samples_per_group + 1 if i < remainder else samples_per_group for i in range(num_groups)] # add one person if there is a reminder \n",
    "    \n",
    "    # now, assign the group based on the rank \n",
    "    df_sort = df.sort_values(by = 'rank')\n",
    "    group_assignments = []\n",
    "    group_number = 1\n",
    "    start = 0\n",
    "    for size in group_sizes:\n",
    "        end = start + size\n",
    "        group_assignments.extend([group_number] * (end - start))\n",
    "        start = end\n",
    "        group_number += 1\n",
    "\n",
    "    # Add a new column 'group' to the DataFrame indicating the group assignment for each row\n",
    "    df_sort['group'] = group_assignments\n",
    "    \n",
    "    return df_sort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define function to determine if a given case presents the variant well (top 50%)\n",
    "def presents_variant(row, param):\n",
    "    scores = {}\n",
    "    if row[param] == 'top half': # indicates strong binding\n",
    "       scores[f'presents_variant_{param}'] = True\n",
    "    elif row[param] == 'top':\n",
    "        scores[f'presents_variant_{param}'] = True\n",
    "    else:\n",
    "        scores[f'presents_variant_{param}'] = False\n",
    "    return pd.Series(scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define function to plot\n",
    "def plot_ranks(data, x, y, hue, colors, order, plt_type, nr, hue_order = None):\n",
    "\n",
    "    score = data.score_type[0]\n",
    "\n",
    "    plt.figure(figsize = (16, 4))\n",
    "\n",
    "    if hue_order is None:\n",
    "        if plt_type == 'box':\n",
    "            p = sns.boxplot(x=x, y=y, hue=hue, data=data, showfliers=False, palette=colors, order=order)\n",
    "            plt.xlabel(f'CH hotspot variant', fontsize = 14)\n",
    "            plt.ylabel(r'$\\log_{10}$(min %EL rank)', fontsize = 14)\n",
    "            plt.xticks(fontsize = 12, rotation = 90)\n",
    "            plt.yticks(fontsize = 13)\n",
    "            plt.title('MHC I', fontsize = 16)\n",
    "            plt.ylim(-2.5, 3.5)\n",
    "            plt.legend(title = 'Variant presentation', loc = 'upper right', fontsize = 13, title_fontsize = 14)\n",
    "            plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_dist_{hue}_{y}_{plt_type}_{score}_{nr}.pdf', bbox_inches='tight')\n",
    "\n",
    "        if plt_type == 'jitter':\n",
    "            p = sns.stripplot(x=x, y=y, hue=hue, data=data, palette=colors, order=order, size = 5, alpha = 0.3)\n",
    "            plt.xlabel(f'CH hotspot variant', fontsize = 14)\n",
    "            plt.ylabel(r'$\\log_{10}$(min %EL rank)', fontsize = 14)\n",
    "            plt.xticks(fontsize = 12, rotation = 90)\n",
    "            plt.yticks(fontsize = 13)\n",
    "            plt.title('MHC I', fontsize = 16)\n",
    "            plt.ylim(-2.5, 3.5)\n",
    "            plt.legend(title = 'Variant presentation', markerscale = 2, loc = 'upper right', fontsize = 13, title_fontsize = 14)\n",
    "            legend = plt.gca().get_legend()\n",
    "            for handle, color in zip(legend.legendHandles, colors):\n",
    "                handle.set_color(color)\n",
    "            for lh in legend.legendHandles:\n",
    "                lh.set_alpha(1)\n",
    "            plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_dist_{hue}_{y}_{plt_type}_{score}_{nr}.pdf', bbox_inches='tight')\n",
    "\n",
    "    else:\n",
    "        if plt_type == 'box':\n",
    "            p = sns.boxplot(x=x, y=y, hue=hue, data=data, showfliers=False, palette=colors, order=order, hue_order=hue_order)\n",
    "            plt.xlabel(f'CH hotspot variant', fontsize = 14)\n",
    "            plt.ylabel(r'$\\log_{10}$(min %EL rank)', fontsize = 14)\n",
    "            plt.xticks(fontsize = 12, rotation = 90)\n",
    "            plt.yticks(fontsize = 13)\n",
    "            plt.title('MHC I', fontsize = 16)\n",
    "            plt.ylim(-2.5, 3.5)\n",
    "            plt.legend(title = 'Variant presentation', loc = 'upper right', fontsize = 13, title_fontsize = 14)\n",
    "            plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_dist_{hue}_{y}_{plt_type}_{score}_{nr}.pdf', bbox_inches='tight')\n",
    "\n",
    "        if plt_type == 'jitter':\n",
    "            p = sns.stripplot(x=x, y=y, hue=hue, data=data, palette=colors, order=order, hue_order=hue_order, size = 5, alpha = 0.3)\n",
    "            plt.xlabel(f'CH hotspot variant', fontsize = 14)\n",
    "            plt.ylabel(r'$\\log_{10}$(min %EL rank)', fontsize = 14)\n",
    "            plt.xticks(fontsize = 12, rotation = 90)\n",
    "            plt.yticks(fontsize = 13)\n",
    "            plt.title('MHC I', fontsize = 16)\n",
    "            plt.ylim(-2.5, 3.5)\n",
    "            plt.legend(title = 'Variant presentation', loc = 'upper right', markerscale = 2, fontsize = 13, title_fontsize = 14)\n",
    "            legend = plt.gca().get_legend()\n",
    "            for handle, color in zip(legend.legendHandles, colors):\n",
    "                handle.set_color(color)\n",
    "            for lh in legend.legendHandles:\n",
    "                lh.set_alpha(1)\n",
    "            plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_dist_{hue}_{y}_{plt_type}_{score}_{nr}.pdf', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# specify colors for plotting\n",
    "col0 = '#910e0e' # this is used to plot the top group (highest score = most immunogenic / strongest binding)\n",
    "col1 = '#ed5d5d'\n",
    "col1b = '#fab9b2' # this is used to plot the middle group \n",
    "col2 = '#65e6f5'\n",
    "col3 = '#1091af' # this is used to plot the bottom group "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# create a column to use for plotting (nicer gene names)\n",
    "ch_all_el_rank_log_labels['gene_var2'] = ch_all_el_rank_log_labels.gene_var.str.replace('_', '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# split df into equal groups\n",
    "df_names_to_label = ['ch_all_el_rank_log_labels']\n",
    "\n",
    "for name in df_names_to_label:\n",
    "\n",
    "    df = globals()[name]\n",
    "    \n",
    "    # apply function separately to each category \n",
    "    df_groups2_eqsize = df.groupby('gene_var2').apply(lambda x: split_into_equal_groups(x, num_groups=2))\n",
    "    df_groups3_eqsize = df.groupby('gene_var2').apply(lambda x: split_into_equal_groups(x, num_groups=3))\n",
    "    df_groups4_eqsize = df.groupby('gene_var2').apply(lambda x: split_into_equal_groups(x, num_groups=4))\n",
    "    df_groups5_eqsize = df.groupby('gene_var2').apply(lambda x: split_into_equal_groups(x, num_groups=5))\n",
    "\n",
    "    # add a more descriptive name for the group\n",
    "    df_groups2_eqsize['group_name'] =  df_groups2_eqsize['group'].map({1: 'top', 2: 'bottom'})\n",
    "    df_groups3_eqsize['group_name'] =  df_groups3_eqsize['group'].map({1: 'top', 2: 'mid', 3: 'bottom'})\n",
    "    df_groups4_eqsize['group_name'] =  df_groups4_eqsize['group'].map({1: 'top', 2: 'mid-high', 3: 'mid-low', 4: 'bottom'})\n",
    "    df_groups5_eqsize['group_name'] =  df_groups5_eqsize['group'].map({1: 'top', 2: 'mid-high', 3: 'mid', 4: 'mid-low', 5: 'bottom'})\n",
    "\n",
    "    # make sure you only include variants examined\n",
    "    df_groups2_eqsize = df_groups2_eqsize[df_groups2_eqsize['gene_var2'].isin(variants_examined)]\n",
    "    df_groups3_eqsize = df_groups3_eqsize[df_groups3_eqsize['gene_var2'].isin(variants_examined)]\n",
    "    df_groups4_eqsize = df_groups4_eqsize[df_groups4_eqsize['gene_var2'].isin(variants_examined)]\n",
    "    df_groups5_eqsize = df_groups5_eqsize[df_groups5_eqsize['gene_var2'].isin(variants_examined)]\n",
    "\n",
    "    colors2 = [col0, col3]\n",
    "    hue_order2 = ['top', 'bottom']\n",
    "    scores_groups2_eqsize = df_groups2_eqsize.sort_values(by = 'score', ascending = False)\n",
    "    \n",
    "    colors3 = [col0, col1b, col3]\n",
    "    hue_order3 = ['top','mid', 'bottom']\n",
    "    scores_groups3_eqsize = df_groups3_eqsize.sort_values(by = 'score', ascending = False)\n",
    "    \n",
    "    colors4 = [col0, col1, col2, col3]\n",
    "    hue_order4 = ['top','mid-high', 'mid-low', 'bottom']\n",
    "    scores_groups4_eqsize = df_groups4_eqsize.sort_values(by = 'score', ascending = False)\n",
    "    \n",
    "    colors5 = [col0, col1, col1b, col2, col3]\n",
    "    hue_order5 = ['top', 'mid-high', 'mid', 'mid-low', 'bottom']\n",
    "    scores_groups5_eqsize = df_groups5_eqsize.sort_values(by = 'score', ascending = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# specify order of levels in group descriptors\n",
    "cat_type2 = pd.CategoricalDtype(categories=['top', 'bottom'], ordered=True)\n",
    "df_groups2_eqsize['group_name'] = df_groups2_eqsize['group_name'].astype(cat_type2)\n",
    "df_groups2_eqsize['group_name'] = df_groups2_eqsize['group_name'].astype('category')\n",
    "\n",
    "cat_type3 = pd.CategoricalDtype(categories=['top', 'mid', 'bottom'], ordered=True)\n",
    "df_groups3_eqsize['group_name'] = df_groups3_eqsize['group_name'].astype(cat_type3)\n",
    "df_groups3_eqsize['group_name'] = df_groups3_eqsize['group_name'].astype('category')\n",
    "\n",
    "cat_type4 = pd.CategoricalDtype(categories=['top', 'mid-high', 'mid-low', 'bottom'], ordered=True)\n",
    "df_groups4_eqsize['group_name'] = df_groups4_eqsize['group_name'].astype(cat_type4)\n",
    "df_groups4_eqsize['group_name'] = df_groups4_eqsize['group_name'].astype('category')\n",
    "\n",
    "cat_type5 = pd.CategoricalDtype(categories=['top', 'mid-high', 'mid', 'mid-low', 'bottom'], ordered=True)\n",
    "df_groups5_eqsize['group_name'] = df_groups5_eqsize['group_name'].astype(cat_type5)\n",
    "df_groups5_eqsize['group_name'] = df_groups5_eqsize['group_name'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# define a new function just to do it for one variant \n",
    "def find_best_score_one_variant_log_ref_ch(row, df, param, variant):\n",
    "    \n",
    "    '''\n",
    "    The same function but log scores \n",
    "    the only allowed parameters are %Rank_EL and %Rank_BA\n",
    "    we want -1 * log(score) so the highest score if the most \"immunogenic\" (best binding) one\n",
    "    '''\n",
    "    \n",
    "    hlas = row.index[1:-1][row[1:-1] >= 1] # select alleles which each Person (row) carrie\n",
    "    \n",
    "    scores = {} # initialise empty dictionaries\n",
    "\n",
    "    if param == \"%Rank_EL\":\n",
    "\n",
    "        # take maximum of the negative log score \n",
    "        best_value = max(-1 * np.log10(df.loc[df['gene_var_gt'] == variant, hlas].values[0]))\n",
    "        scores[f'score_{variant}'] = best_value\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"%Rank_BA\":\n",
    "        \n",
    "        # take maxium of the negative log score \n",
    "        best_value = max(-1 * np.log10(df.loc[df['gene_var_gt'] == variant, hlas].values[0]))\n",
    "        scores[f'score_{variant}'] = best_value\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Aff_nM\":\n",
    "        \n",
    "        # take maxium of the negative log score (low aff = more immunogenic)\n",
    "        best_value = max(-1 * np.log10(df.loc[df['gene_var_gt'] == variant, hlas].values[0]))\n",
    "        scores[f'score_{variant}'] = best_value\n",
    "        return pd.Series(scores)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define a function to split people into groups based on MHC presentation score \n",
    "\n",
    "def split_equal_groups_by_varscore(df, num_groups, var):\n",
    "\n",
    "    '''\n",
    "    first, assign a rank to everyone based on score\n",
    "    the highest score = the lowest rank (ie highest score means you get rank 1)\n",
    "    if two people have the same score, assign consecutive ranks \n",
    "    '''\n",
    "    \n",
    "    # first, determine the number of samples\n",
    "    total_samples = len(df)\n",
    "\n",
    "    # now, determine the number of samples in each group\n",
    "    samples_per_group = total_samples // num_groups\n",
    "    remainder = total_samples % num_groups\n",
    "    group_sizes = [samples_per_group + 1 if i < remainder else samples_per_group for i in range(num_groups)] # add one person if there is a reminder \n",
    "    \n",
    "    # now, assign the group based on the rank \n",
    "    group_assignments = []\n",
    "    group_number = 1\n",
    "    start = 0\n",
    "    for size in group_sizes:\n",
    "        end = start + size\n",
    "        group_assignments.extend([group_number] * (end - start))\n",
    "        start = end\n",
    "        group_number += 1\n",
    "\n",
    "    # Add a new column 'group' to the DataFrame indicating the group assignment for each row\n",
    "    return group_assignments\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of healthy (non-CH) individuals examined: 441910\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# cases who have CH with a genotyped variant (from one of the ones we were looking at) are in batch_gene_vars \n",
    "\n",
    "ch_sampleids = batch_gene_vars_hla_10['Person_ID']\n",
    "ukbb_no_ch = df_clean_hla1[~df_clean_hla1['Person_ID'].isin(ch_sampleids)]\n",
    "hla_ukbb = ukbb_no_ch.filter(regex='\\d').columns # identify HLA alleles carried by non-CH-individuals in the UKBB data\n",
    "\n",
    "# prepare dataset with predictions (now just from NetMHC)\n",
    "scores_netmhc_elrank = scores_netmhc[['HLA_formatted', '%Rank_EL', 'gene_var_gt']]\n",
    "scores_netmhc_elrank_wide = pd.pivot(scores_netmhc_elrank, index='gene_var_gt', columns='HLA_formatted', values='%Rank_EL')\n",
    "\n",
    "hla_intersect = scores_netmhc_elrank_wide.columns[scores_netmhc_elrank_wide.columns.isin(hla_ukbb)]\n",
    "hla_intersect_list = hla_intersect.tolist() # select predictions for alleles which are available in NetMHC / PRIME\n",
    "\n",
    "pred_sub_wide = scores_netmhc_elrank_wide[hla_intersect_list + scores_netmhc_elrank_wide.columns[scores_netmhc_elrank_wide.columns.str.contains('gene_var_gt')].tolist()]\n",
    "\n",
    "ukbb_no_ch_hla = ukbb_no_ch[hla_intersect]\n",
    "ukbb_no_ch = pd.concat([ukbb_no_ch['Person_ID'], ukbb_no_ch_hla], axis = 1)\n",
    "ukbb_no_ch.head()\n",
    "\n",
    "ids_non_carriers_healthy = ukbb_no_ch.Person_ID.unique().tolist()\n",
    "print('Number of healthy (non-CH) individuals examined:', len(ids_non_carriers_healthy))\n",
    "\n",
    "# okay so now we have a dataset for everyone who does NOT have CH\n",
    "# and for all of them, we have their alleles for which we can make predictions\n",
    "\n",
    "pred_sub_wide['gene_var_gt'] = pred_sub_wide.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select only HLAs for which we have predictions\n",
    "df_clean_hla1_hla = df_clean_hla1[hla_intersect]\n",
    "df_clean_hla1_hla_sel = pd.concat([df_clean_hla1['Person_ID'], df_clean_hla1_hla], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batches that were examined for variants: {11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21}\n",
      "Number of samples examined for variants: 99597\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# subset the dataframe to only include individuals who were actually screened for CH\n",
    "batch_all['batch_number'] = batch_all['batch'].apply(lambda x: [int(num) for num in re.findall(r'\\d+', x)])\n",
    "batch_numbers_examined = set(batch_all['batch_number'].sum()) \n",
    "print('Batches that were examined for variants:', batch_numbers_examined)\n",
    "\n",
    "all_dataframes = [f'batch_{i}_ids' for i in range(11, 61)]\n",
    "selected_ids = []\n",
    "\n",
    "# Iterate through all df name \n",
    "for df_name in all_dataframes:\n",
    "    number = int(re.search(r'\\d+', df_name).group())\n",
    "    # check if number matches numbers of batches examined \n",
    "    if number in batch_numbers_examined:        \n",
    "        selected_ids.append(globals()[df_name])\n",
    "\n",
    "ids_examined = pd.concat(selected_ids, ignore_index=True)\n",
    "print('Number of samples examined for variants:', ids_examined.shape[0])\n",
    "\n",
    "ids_examined_list = ids_examined['sample_ID'].tolist()\n",
    "ids_examined_list = [int(id) for id in ids_examined_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>A_101</th>\n",
       "      <th>A_102</th>\n",
       "      <th>A_103</th>\n",
       "      <th>A_1101</th>\n",
       "      <th>A_1102</th>\n",
       "      <th>A_1103</th>\n",
       "      <th>A_201</th>\n",
       "      <th>A_202</th>\n",
       "      <th>A_203</th>\n",
       "      <th>...</th>\n",
       "      <th>C_403</th>\n",
       "      <th>C_407</th>\n",
       "      <th>C_501</th>\n",
       "      <th>C_602</th>\n",
       "      <th>C_701</th>\n",
       "      <th>C_702</th>\n",
       "      <th>C_704</th>\n",
       "      <th>C_801</th>\n",
       "      <th>C_802</th>\n",
       "      <th>C_804</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1118855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1734232</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1539760</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1396487</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1759529</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 195 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Person_ID  A_101  A_102  A_103  A_1101  A_1102  A_1103  A_201  A_202  \\\n",
       "4     1118855    0.0    0.0    0.0     1.0     0.0     0.0    1.0    0.0   \n",
       "9     1734232    0.0    0.0    0.0     0.0     0.0     0.0    2.0    0.0   \n",
       "14    1539760    1.0    0.0    0.0     0.0     0.0     0.0    1.0    0.0   \n",
       "15    1396487    0.0    0.0    0.0     0.0     0.0     0.0    2.0    0.0   \n",
       "19    1759529    1.0    0.0    0.0     0.0     0.0     0.0    0.0    0.0   \n",
       "\n",
       "    A_203  ...  C_403  C_407  C_501  C_602  C_701  C_702  C_704  C_801  C_802  \\\n",
       "4     0.0  ...    0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0   \n",
       "9     0.0  ...    0.0    0.0    1.0    0.0    1.0    0.0    0.0    0.0    0.0   \n",
       "14    0.0  ...    0.0    0.0    0.0    0.0    0.0    2.0    0.0    0.0    0.0   \n",
       "15    0.0  ...    0.0    0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0   \n",
       "19    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "\n",
       "    C_804  \n",
       "4     0.0  \n",
       "9     0.0  \n",
       "14    0.0  \n",
       "15    0.0  \n",
       "19    0.0  \n",
       "\n",
       "[5 rows x 195 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df_clean_hla1_hla_sel = df_clean_hla1_hla_sel[df_clean_hla1_hla_sel['Person_ID'].isin(ids_examined_list)]\n",
    "df_clean_hla1_hla_sel.head() # note that this "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# define functions to compare groups\n",
    "\n",
    "def percentage_difference(group, nr_groups):\n",
    "\n",
    "    x_a = group.loc[group[f'groups_{nr_groups}'] == 1, 'CH_variant_positive'].iloc[0]  # first group \n",
    "    x_b = group.loc[group[f'groups_{nr_groups}'] == nr_groups, 'CH_variant_positive'].iloc[0]  # last group \n",
    "    \n",
    "    result = (x_b - x_a) / x_a * 100\n",
    "    return result\n",
    "\n",
    "def normalized_difference(group, nr_groups):\n",
    "\n",
    "    x_a = group.loc[group[f'groups_{nr_groups}'] == 1, 'CH_variant_positive'].iloc[0]  # first group \n",
    "    x_b = group.loc[group[f'groups_{nr_groups}'] == nr_groups, 'CH_variant_positive'].iloc[0]  # last group \n",
    "    num_observations = group['count']   # Number of observations in this category\n",
    "    \n",
    "    result = (x_b - x_a) / num_observations * 100\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loop for all variants\n",
    "\n",
    "scores_2groups = {}\n",
    "scores_for_each_var_ukbb = {}\n",
    "\n",
    "for var in variants_to_examine:\n",
    "    \n",
    "    # for everyone in the UKBB, add binding score for each variant based on their MHC genotype\n",
    "    # note this is for everyone, carrier or not\n",
    "    df_all_ukbb_scores = pd.concat([df_clean_hla1_hla_sel, df_clean_hla1_hla_sel.apply(find_best_score_one_variant_log_ref_ch, df=pred_sub_wide, param=param, variant = f'{var}_ch', axis=1)], axis=1)\n",
    "    scores_for_each_var_ukbb[var] = df_all_ukbb_scores[f'score_{var}_ch'] # add this so we can later correlate with immunogenicity\n",
    "\n",
    "    # order the df by rank\n",
    "    df_all_ukbb_scores['rank'] = df_all_ukbb_scores[f'score_{var}_ch'].rank(method='first', ascending=False)\n",
    "    df_all_ukbb_scores_sort = df_all_ukbb_scores.sort_values(by = 'rank')\n",
    "\n",
    "    # split people into 2 groups for each variant (how strongly they bind)\n",
    "    df_all_ukbb_scores_sort['groups_2'] = split_equal_groups_by_varscore(df_all_ukbb_scores_sort, num_groups = 2, var = var)\n",
    "\n",
    "    # okay now we need to add a column to indicate if that person has CH driven by this variant\n",
    "    # identify IDs of people with the variant\n",
    "    ids_var = batch_gene_vars[batch_gene_vars['gene_var']==var].Person_ID.tolist()\n",
    "    df_all_ukbb_scores_sort[f'{var}_positive'] = df_all_ukbb_scores_sort.Person_ID.isin(ids_var) # added True if someone has the variant, False otherwise \n",
    "\n",
    "    # calculate % of people in each of the 5 or 10 groups (as % of total CH positive cases)\n",
    "    counts2 = df_all_ukbb_scores_sort.groupby('groups_2')[f'{var}_positive'].sum().reset_index()\n",
    "    counts2['percentage_all_cases'] = counts2[f'{var}_positive'] / counts2[f'{var}_positive'].sum() * 100\n",
    "    \n",
    "    # add data to the dictionary \n",
    "    scores_2groups[var] = counts2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get dfs\n",
    "scores_2groups_df = pd.DataFrame()\n",
    "\n",
    "for key, value in scores_2groups.items():\n",
    "    temp_df = scores_2groups[key]\n",
    "    temp_df['gene_var'] = key\n",
    "    temp_df.rename(columns = {f'{key}_positive':'CH_variant_positive'}, inplace = True)\n",
    "    scores_2groups_df = pd.concat([scores_2groups_df, temp_df], axis=0)\n",
    "\n",
    "# I think one way we can analyse this is for each variant, plot the difference between top and bottom group, normalized by the total nr of people\n",
    "genevar_counts = pd.DataFrame(batch_gene_vars_hla_10.gene_var.value_counts().reset_index())\n",
    "scores_2groups_counts = pd.merge(scores_2groups_df, genevar_counts, on = 'gene_var')\n",
    "\n",
    "# Group by Category and apply the custom calculation function\n",
    "result_2groups1 = pd.DataFrame(scores_2groups_counts.groupby('gene_var').apply(percentage_difference, nr_groups = 2))\n",
    "result_2groups2 = pd.DataFrame(scores_2groups_counts.groupby('gene_var').apply(normalized_difference, nr_groups = 2))\n",
    "result_2groups = pd.merge(result_2groups1, result_2groups2, on = 'gene_var')\n",
    "result_2groups = result_2groups.groupby('gene_var').first().reset_index()\n",
    "result_2groups.columns.values[1] = 'percentage_diff'\n",
    "result_2groups.columns.values[2] = 'norm_diff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# each group had 45237 observations (we have 90474 people in total and we split them into 2)\n",
    "scores_2groups_df['CH_variant_negative'] = ( df_clean_hla1_hla_sel.shape[0]/2 - scores_2groups_df['CH_variant_positive']).astype(int)\n",
    "scores_2groups_df['total'] = int(df_clean_hla1_hla_sel.shape[0]/2 )\n",
    "scores_2groups_df['Group'] = scores_2groups_df['groups_2'].map({1: 'better presenters', 2: 'worse presenters'})\n",
    "scores_2groups_df['gene_var2'] = scores_2groups_df['gene_var'].str.replace('_', '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simpler plot \n",
    "order_percent2 = result_2groups.sort_values(by = 'percentage_diff').gene_var.tolist()\n",
    "order_norm2 = result_2groups.sort_values(by = 'norm_diff').gene_var.tolist()\n",
    "order_percent2 = [x.replace('_', '\\n') for x in order_percent2]\n",
    "order_norm2 = [x.replace('_', '\\n') for x in order_norm2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "order_by_total = scores_2groups_df.groupby('gene_var2')['CH_variant_positive'].sum().reset_index()\n",
    "order_by_total = order_by_total.sort_values(by = 'CH_variant_positive', ascending = False).gene_var2.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test 4: split by absolute threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>score</th>\n",
       "      <th>gene_var</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>0.419075</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>0.982967</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>0.970616</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>0.189096</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>0.970616</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90469</th>\n",
       "      <td>488357</td>\n",
       "      <td>-0.028571</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90470</th>\n",
       "      <td>488359</td>\n",
       "      <td>0.463442</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90471</th>\n",
       "      <td>488361</td>\n",
       "      <td>-0.250420</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90472</th>\n",
       "      <td>488369</td>\n",
       "      <td>0.463442</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90473</th>\n",
       "      <td>488376</td>\n",
       "      <td>0.463442</td>\n",
       "      <td>SF3B1_K666N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1809480 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index     score      gene_var\n",
       "0           4  0.419075  DNMT3A_R882H\n",
       "1           9  0.982967  DNMT3A_R882H\n",
       "2          14  0.970616  DNMT3A_R882H\n",
       "3          15  0.189096  DNMT3A_R882H\n",
       "4          19  0.970616  DNMT3A_R882H\n",
       "...       ...       ...           ...\n",
       "90469  488357 -0.028571   SF3B1_K666N\n",
       "90470  488359  0.463442   SF3B1_K666N\n",
       "90471  488361 -0.250420   SF3B1_K666N\n",
       "90472  488369  0.463442   SF3B1_K666N\n",
       "90473  488376  0.463442   SF3B1_K666N\n",
       "\n",
       "[1809480 rows x 3 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a dataframe with scores for everyone in the UKBB \n",
    "scores_each_var_df = pd.DataFrame()\n",
    "\n",
    "for key, value in scores_for_each_var_ukbb.items():\n",
    "    temp_df = scores_for_each_var_ukbb[key].reset_index()\n",
    "    temp_df['gene_var'] = key\n",
    "    temp_df.rename(columns = {f'score_{key}_ch':'score'}, inplace = True)\n",
    "    scores_each_var_df = pd.concat([scores_each_var_df, temp_df], axis=0)\n",
    "\n",
    "scores_each_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop rows with non-numeric score columns\n",
    "scores_each_var_df['score'] = pd.to_numeric(scores_each_var_df['score'], errors='coerce')\n",
    "scores_each_var_df = scores_each_var_df.dropna(subset=['score'])\n",
    "scores_each_var_df['gene_var2'] = scores_each_var_df['gene_var'].str.replace('_', '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# specify order (for all)\n",
    "scores_each_var_df['median_score'] = scores_each_var_df.groupby('gene_var2').score.transform('median')\n",
    "order_all = scores_each_var_df.sort_values(by = 'median_score', ascending = False).gene_var2.unique().tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_var2</th>\n",
       "      <th>no binding</th>\n",
       "      <th>weak binding</th>\n",
       "      <th>strong binding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNMT3A\\nP904L</td>\n",
       "      <td>22</td>\n",
       "      <td>11388</td>\n",
       "      <td>79064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNMT3A\\nR320*</td>\n",
       "      <td>90474</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNMT3A\\nR326C</td>\n",
       "      <td>96</td>\n",
       "      <td>50875</td>\n",
       "      <td>39503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNMT3A\\nR598*</td>\n",
       "      <td>75093</td>\n",
       "      <td>15381</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNMT3A\\nR729W</td>\n",
       "      <td>268</td>\n",
       "      <td>8632</td>\n",
       "      <td>81574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_var2  no binding  weak binding  strong binding\n",
       "0  DNMT3A\\nP904L          22         11388           79064\n",
       "1  DNMT3A\\nR320*       90474             0               0\n",
       "2  DNMT3A\\nR326C          96         50875           39503\n",
       "3  DNMT3A\\nR598*       75093         15381               0\n",
       "4  DNMT3A\\nR729W         268          8632           81574"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# okay now for each variant, let's first calculate people at each category \n",
    "bins = [-float('inf'), -1*np.log10(2), -1*np.log10(0.5), float('inf')]\n",
    "labels = ['no binding', 'weak binding', 'strong binding']\n",
    "\n",
    "# Categorize scores into bins\n",
    "scores_each_var_df['score_bin'] = pd.cut(scores_each_var_df['score'], bins=bins, labels=labels)\n",
    "\n",
    "# Group by category and score bins, then count observations\n",
    "grouped_df = scores_each_var_df.groupby(['gene_var2', 'score_bin']).size().unstack(fill_value=0)\n",
    "\n",
    "# Create a new DataFrame with desired columns\n",
    "scores_each_var_df_bins = grouped_df.reset_index()\n",
    "scores_each_var_df_bins.columns.name = None  # Remove the name of columns index\n",
    "scores_each_var_df_bins.columns = ['gene_var2', 'no binding', 'weak binding', 'strong binding']\n",
    "\n",
    "scores_each_var_df_bins.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_var2</th>\n",
       "      <th>no binding_population</th>\n",
       "      <th>weak binding_population</th>\n",
       "      <th>strong binding_population</th>\n",
       "      <th>no binding_carriers</th>\n",
       "      <th>weak binding_carriers</th>\n",
       "      <th>strong binding_carriers</th>\n",
       "      <th>no binding_noncarriers</th>\n",
       "      <th>weak binding_noncarriers</th>\n",
       "      <th>strong binding_noncarriers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNMT3A\\nP904L</td>\n",
       "      <td>22</td>\n",
       "      <td>11388</td>\n",
       "      <td>79064</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>86</td>\n",
       "      <td>22</td>\n",
       "      <td>11379</td>\n",
       "      <td>78978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNMT3A\\nR320*</td>\n",
       "      <td>90474</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>90405</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNMT3A\\nR326C</td>\n",
       "      <td>96</td>\n",
       "      <td>50875</td>\n",
       "      <td>39503</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>45</td>\n",
       "      <td>96</td>\n",
       "      <td>50835</td>\n",
       "      <td>39458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNMT3A\\nR598*</td>\n",
       "      <td>75093</td>\n",
       "      <td>15381</td>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>75020</td>\n",
       "      <td>15370</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNMT3A\\nR729W</td>\n",
       "      <td>268</td>\n",
       "      <td>8632</td>\n",
       "      <td>81574</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>52</td>\n",
       "      <td>268</td>\n",
       "      <td>8627</td>\n",
       "      <td>81522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_var2  no binding_population  weak binding_population  \\\n",
       "0  DNMT3A\\nP904L                     22                    11388   \n",
       "1  DNMT3A\\nR320*                  90474                        0   \n",
       "2  DNMT3A\\nR326C                     96                    50875   \n",
       "3  DNMT3A\\nR598*                  75093                    15381   \n",
       "4  DNMT3A\\nR729W                    268                     8632   \n",
       "\n",
       "   strong binding_population  no binding_carriers  weak binding_carriers  \\\n",
       "0                      79064                    0                      9   \n",
       "1                          0                   69                      0   \n",
       "2                      39503                    0                     40   \n",
       "3                          0                   73                     11   \n",
       "4                      81574                    0                      5   \n",
       "\n",
       "   strong binding_carriers  no binding_noncarriers  weak binding_noncarriers  \\\n",
       "0                       86                      22                     11379   \n",
       "1                        0                   90405                         0   \n",
       "2                       45                      96                     50835   \n",
       "3                        0                   75020                     15370   \n",
       "4                       52                     268                      8627   \n",
       "\n",
       "   strong binding_noncarriers  \n",
       "0                       78978  \n",
       "1                           0  \n",
       "2                       39458  \n",
       "3                           0  \n",
       "4                       81522  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# okay, now we are doing the same thing for carriers \n",
    "scores_each_var_df_carriers = ch_all_el_rank_log_melted_sort[ch_all_el_rank_log_melted_sort['is_variant_present']==True]\n",
    "scores_each_var_df_carriers = scores_each_var_df_carriers[['Person_ID', 'gene_var2', 'value']]\n",
    "scores_each_var_df_carriers.rename(columns = {'value':'score'}, inplace = True)\n",
    "\n",
    "# add score bins\n",
    "bins = [-float('inf'), -1*np.log10(2), -1*np.log10(0.5), float('inf')]\n",
    "labels = ['no binding', 'weak binding', 'strong binding']\n",
    "scores_each_var_df_carriers['score_bin'] = pd.cut(scores_each_var_df_carriers['score'], bins=bins, labels=labels)\n",
    "\n",
    "grouped_df_carriers = scores_each_var_df_carriers.groupby(['gene_var2', 'score_bin']).size().unstack(fill_value=0)\n",
    "scores_each_var_df_bins_carriers = grouped_df_carriers.reset_index()\n",
    "scores_each_var_df_bins_carriers.columns.name = None  # Remove the name of columns index\n",
    "scores_each_var_df_bins_carriers.columns = ['gene_var2', 'no binding', 'weak binding', 'strong binding']\n",
    "\n",
    "# merge all together and calculate values for non-carriers\n",
    "merged_df = scores_each_var_df_bins.merge(scores_each_var_df_bins_carriers, on='gene_var2', suffixes=('_population', '_carriers'))\n",
    "merged_df['no binding_noncarriers'] = merged_df['no binding_population'] - merged_df['no binding_carriers']\n",
    "merged_df['weak binding_noncarriers'] = merged_df['weak binding_population'] - merged_df['weak binding_carriers']\n",
    "merged_df['strong binding_noncarriers'] = merged_df['strong binding_population'] - merged_df['strong binding_carriers']\n",
    "\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# do mann-whitney test to compare population (non-carriers) and carriers scores \n",
    "pvalues_mwu = {}\n",
    "cat_series = pd.Series(['Carrier', 'Non-carrier'])\n",
    "df_scores_all_plot = pd.DataFrame()\n",
    "\n",
    "for var in order_all:\n",
    "    carrier_scores = scores_each_var_df_carriers[scores_each_var_df_carriers['gene_var2']==var].score.tolist()\n",
    "    all_scores = scores_each_var_df[scores_each_var_df['gene_var2']==var].score.tolist()\n",
    "\n",
    "    # Remove elements from x only once if they appear once in y\n",
    "    from collections import Counter \n",
    "    counts = Counter(carrier_scores)\n",
    "\n",
    "    noncarrier_scores = []\n",
    "    for x in all_scores:\n",
    "        if counts[x]:\n",
    "            counts[x] -= 1\n",
    "        else:\n",
    "            noncarrier_scores.append(x)\n",
    "\n",
    "    scores_series = pd.Series(carrier_scores + noncarrier_scores).reset_index(drop=True)\n",
    "    repetitions = [len(carrier_scores), len(noncarrier_scores)]\n",
    "    cat_series_rep = pd.Series(cat_series.repeat(repetitions)).reset_index(drop=True)\n",
    "    df_x = pd.DataFrame({'score': scores_series, 'CH status': cat_series_rep})\n",
    "    df_x['gene_var2'] = var\n",
    "\n",
    "    df_scores_all_plot = pd.concat([df_scores_all_plot, df_x], axis = 0)\n",
    "\n",
    "    statistic, p_value = mannwhitneyu(carrier_scores, noncarrier_scores)\n",
    "    pvalues_mwu[var] = p_value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do a nice visualisation to compare these \n",
    "# the easiest I can think about is doing % of people in each group\n",
    "# and maybe having carriers / non-carriers side by side or sth\n",
    "merged_df['no binding_carriers_percent'] = merged_df['no binding_carriers'] / (merged_df['no binding_carriers'] + merged_df['weak binding_carriers'] + merged_df['strong binding_carriers']) * 100\n",
    "merged_df['weak binding_carriers_percent'] = merged_df['weak binding_carriers'] / (merged_df['no binding_carriers'] + merged_df['weak binding_carriers'] + merged_df['strong binding_carriers']) * 100\n",
    "merged_df['strong binding_carriers_percent'] = merged_df['strong binding_carriers'] / (merged_df['no binding_carriers'] + merged_df['weak binding_carriers'] + merged_df['strong binding_carriers']) * 100\n",
    "\n",
    "merged_df['no binding_noncarriers_percent'] = merged_df['no binding_noncarriers'] / (merged_df['no binding_noncarriers'] + merged_df['weak binding_noncarriers'] + merged_df['strong binding_noncarriers']) * 100\n",
    "merged_df['weak binding_noncarriers_percent'] = merged_df['weak binding_noncarriers'] / (merged_df['no binding_noncarriers'] + merged_df['weak binding_noncarriers'] + merged_df['strong binding_noncarriers']) * 100\n",
    "merged_df['strong binding_noncarriers_percent'] = merged_df['strong binding_noncarriers'] / (merged_df['no binding_noncarriers'] + merged_df['weak binding_noncarriers'] + merged_df['strong binding_noncarriers']) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_percents = merged_df[['gene_var2', 'no binding_carriers_percent', 'weak binding_carriers_percent', 'strong binding_carriers_percent',\n",
    "                                'no binding_noncarriers_percent', 'weak binding_noncarriers_percent', 'strong binding_noncarriers_percent']]\n",
    "\n",
    "merged_df_percents_melt = pd.melt(merged_df_percents, id_vars = 'gene_var2')\n",
    "merged_df_percents_melt[['binding level', 'group', 'drop']] = merged_df_percents_melt['variable'].str.split('_', expand = True)\n",
    "merged_df_percents_melt['CH status'] = merged_df_percents_melt['group'].map({'carriers' : 'Carrier', 'noncarriers': 'Non-carrier'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABTAAAAHJCAYAAACyvxagAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/OQEPoAAAACXBIWXMAAA9hAAAPYQGoP6dpAAC5FUlEQVR4nOzdd3gU1fv38c9sEpKQhBY6IoQOUqQ36QiKgCBVkF4tX6p06b0XG1gBRRFBBKQIUgQUlCJVEBSlSe81kOQ8f/hkf8YUNsnuZmHfr+va62Jnzs7cN7Mzu7n3zDmWMcYIAAAAAAAAADyQLaUDAAAAAAAAAID4UMAEAAAAAAAA4LEoYAIAAAAAAADwWBQwAQAAAAAAAHgsCpgAAAAAAAAAPBYFTAAAAAAAAAAeiwImAAAAAAAAAI9FARMAAAAAAACAx6KACQAAAAAAAMBjUcAEAACAy+TOnVuWZcmyLK1atSrBtk2bNrW3HTFiRIx1mzZtkmVZypcv3wP3Gb2NU6dOxdvm6NGj6tevn8qUKaPQ0FD5+fkpffr0Klu2rHr06KH169fLGONQjlLMPP/9+G8eAAAASDzflA4AAAAA3mHs2LGqV69enOt+++03LV261OUxREZGauDAgZo+fboiIyOVNm1alShRQkFBQbp48aL27dunnTt36s0331Tx4sW1d+9eh7Zbp04dnT9/Xj/++KMuXLigkiVL6vHHH1ehQoVcnBEAAMCjjwImAAAAXC5//vz68ccftWHDBtWsWTPW+kmTJikqKkr58+fX0aNHXRZHu3bttGDBAqVJk0YzZ85UmzZt5OPjY19/584dffXVVxozZoyOHz/u8Hbfe+89SVLt2rW1fv169ejRQ+3bt3d2+AAAAF6JW8gBAADgcv3795ckjR49Ota606dP69NPP1XFihVVtWpVl8Uwb948LViwQP7+/tqwYYPat28fo3gpSYGBgWrdurX27Nmjvn37uiwWAAAAOI4CJgAAAFyudu3aKleunDZt2qQffvghxrqpU6fq3r17GjJkiEtjGDNmjCTp9ddfV+nSpRNs6+/vr6FDh7o0HgAAADiGAiYAAADcYtCgQZJi9sK8fPmy3n//fT355JN67rnnXLbvgwcP6vfff5ckdevWzWX7AQAAgPMxBiYAAADc4vnnn9cTTzyhb7/9Vjt37lSZMmX05ptv6ubNmxo8eLBD2zh79qwaNWqU6H3v2LFDkvT4448rZ86ciX49AAAAUg4FTAAAALiFZVkaOHCg2rRpozFjxuizzz7Tm2++qYIFC6pJkyYObePWrVtatmxZovd9/vx5SVLWrFkT/VoAAACkLG4hBwAAgNu8+OKLCgsL0/Lly9WjRw9dunRJgwYNks3m2NfSvHnzyhiT4CMhlmU5Iw0AAAC4EQVMAAAAuI2Pj4/69+8vY4w+/PBD5c6dW61bt3b5fjNmzCjpn1vQAQAA8HChgAkAAAC36tChg7JlyyZJ6t+/v3x9XT+qUYkSJSRJx48f16lTp1y+PwAAADgPBUwAAAC4lb+/v9588029+uqr6tixo1v2WbJkSWXJkkWS9M4777hlnwAAAHAOCpgAAABwuyZNmuitt96Sv7+/W/Zns9nUr18/SdK0adP0888/J9g+IiJCn3zyiTtCAwAAwANQwAQAAIBX6Nmzp2rWrKnw8HDVqlVL7777rsLDw2O0CQ8P15IlS1SmTBlNnTo1hSIFAADAv1HABAAAgFfw9fXVsmXL1KJFC928eVOvvPKKQkNDVaVKFT333HMqX7680qdPr6ZNm2rv3r0KCwtzeNtdu3ZVo0aNtG/fPknSrFmz1KhRIy1cuNBV6QAAAHgN14+YDgAAAHiI4OBgLVy4UD179tTHH3+sH374Qfv27dPt27eVPn16lSxZUlWqVNELL7ygcuXKObzdtWvX6vjx4/bnv/zyi3755Rc9+eSTLsgCAADAu1jGGJPSQQAAAAAAAABAXLiFHAAAAAAAAIDH8vgC5qBBgxQcHBzv+iNHjqhp06bKmDGjAgIC9MQTT2jmzJmKioqK1TY8PFwjRoxQvnz55O/vr6xZs6pDhw46c+aMK1MAAAAAAAAAkEQeeQv51atXdeLECW3YsEEDBw6Ur6+vbt68Gavd8ePHVbp0aUlSt27dlClTJq1du1arV69W//79NXHixBjtmzVrpsWLF+vFF19U+fLldfz4cc2ZM0fZs2fX7t27FRIS4pb8AAAAAAAAADjGIwuYI0aM0MiRI+3Pg4KC4ixgdu7cWR9//LF2796tEiVK2Je/8MILWrFihY4dO6acOXNKkrZu3aoqVaqoT58+mjp1qr3t8uXL9fzzz2vcuHEaNGiQC7MCAAAAAAAAkFgeeQt5q1attGLFCq1YsUJFixaNs40xRl9//bUqVKgQo3gpSd27d1dERISWLVtmX7Z06VL7un9r2LChsmfPriVLljg5CwAAAAAAAADJ5ZvSAcSlQIECKlCggCRpypQp+vPPP2O1OX36tC5duqQiRYrEWlesWDFJ0t69e+3L9u7dKz8/P+XLly9W+6JFi+r7779XVFSUbLa4a7rh4eEKDw+3P4+KitLly5cVGhoqy7ISlyAAAAAAAADg5YwxunHjhrJnzx5vTU7y0AKmI6In3kmbNm2sdaGhoTHaRP87TZo0cRYbQ0NDFR4ersuXLytjxoxx7m/8+PExbmsHAAAAAAAAkHwnT57UY489Fu/6h7aAeffuXUlS6tSpY63z8/OTJN25cydG+7jaxtf+vwYNGqQ+ffrYn1+7dk2PP/64Tp48qTRp0iQ+AQfMzx//gYtL26On7P/e+9b0RL22xGu9E9XeFbwtX8n7cva2fCXvy9nb8pW8L2dvy1fyvpy9LV/J+3L2tnwl78vZ2/KVvC9nb8tXku4+mTtR7QP2/GX/d8TsGYl6rW/3Xolq7wrelq8kFR65J1HtDw1/0v7vYStOJuq1oxrkTFR7V/C2fONz/fp15cyZ84ETaz+0BcyAgABJUmRkZKx19+/flyQFBgbGaB9d9HSk/X/5+/vL398/1vI0adK4rIAZaEvcren/jiP4////JOW1KcXb8pW8L2dvy1fyvpy9LV/J+3L2tnwl78vZ2/KVvC9nb8tX8r6cvS1fyfty9rZ8JSlVAreWxiXgX3FHJDJnXw/I2dvylSSbf3Ci2v/7vemfOuHiV0KvTSnelu+DPGh4Ro+cxMcRWbJkkSRduXIl1rqLFy/GaBP976tXryquSdcvXryoVKlSKX369C6KFgAAAAAAAEBSPLQFzJw5cyp9+vTavn17rHXRy4oXL25fVrx4cd2+fVv79++P0TYqKko7duxQkSJF5OPj49qgAQAAAAAAACTKQ1vAtCxLDRs21J49e2IVJRcsWCAfHx81bNjQvqxRo0aSpPnz58dou3LlSl29elWNGzd2ecwAAAAAAAAAEscjx8D89ddf9euvv0qSLly4oMjISC1evFiSlDt3bpUpU0aSNGzYMC1btkxPP/20unTpoowZM2rdunVauXKlevXqpbCwMPs2q1evroYNG2ratGm6cOGCSpcurRMnTmj27NnKnTu3evTo4f5EAQAAAAAAACTIIwuYixYt0siRI2Msa9asmSSpXbt2mjt3riQpT548+uGHHzRkyBC9/fbbunXrlvLmzaupU6eqV69esba7cOFCjR07Vp999pk+//xzpU+fXk2aNNG4ceOULl06F2cFAAAAAAAAILE8soA5YsQIjRgxwqG2RYoU0dKlSx1qGxgYqDFjxmjMmDHJiA4AAAAAAACAuzy0Y2ACAAAAAAAAePRRwAQAAAAAAADgsShgAgAAAAAAAPBYFDABAAAAAAAAeCwKmAAAAAAAAAA8FgVMAAAAAAAAAB6LAiYAAAAAAAAAj0UBEwAAAAAAAIDHooAJAAAAAAAAwGNRwAQAAAAAAADgsShgAgAAAAAAAPBYFDABAAAAAAAAeCwKmAAAAAAAAAA8FgVMAAAAAAAAAB6LAiYAAAAAAAAAj+Wb0gEAAAAAAJBSivlaKR0CAOAB6IEJAAAAAAAAwGPRAxMAAAAA8FBrHeSX0iEAAFyIHpgAAAAAAAAAPBY9MAEAAOAUjCMHAAAAV6CACQAAADtuwwQAAICn4RZyAAAAAAAAAB6LAiYAAAAAAAAAj0UBEwAAAAAAAIDHooAJAAAAAAAAwGNRwAQAAAAAAADgsShgAgAAAAAAAPBYFDABAAAAAAAAeCwKmAAAAAAAAAA8FgVMAAAAAAAAAB6LAiYAAAAAAAAAj+Wb0gEAAAAAAAC4ypCcLRPVfqKL4gCQdBQwAQAAAAAAPEj+5msT1f6ki+IAPAUFTAAAAAAAHiKtg/xSOgQAcCsKmAAAAAAAeIlivlZKhwAAiUYBEwAAAAAAeDRuqQa8GwVMAAAAAIAkeucBADyTLaUDAAAAAAAAAID4UMAEAAAAAAAA4LEoYAIAAAAAAADwWBQwAQAAAAAAAHgsCpgAAAAAAAAAPBYFTAAAAAAAAAAeiwImAAAAAAAAAI/lm9IBAAAAAAAAwDmG5GyZqPYTXRQH4Ez0wAQAAAAAAADgsShgAgAAAAAAAPBYFDABAAAAAAAAeKyHvoB5+/ZtTZw4UcWLF1dwcLCyZs2qWrVqaeXKlbHaHjlyRE2bNlXGjBkVEBCgJ554QjNnzlRUVFQKRA4AAAAAAADgQR7qSXzCw8NVtWpV7dmzRy1atFD79u117do1ff7556pfv74mTJigAQMGSJKOHz+uSpUqSZK6deumTJkyae3aterVq5f+/vtvTZzIsLUAAAAAAACAp3moC5iLFi3Srl27NHXqVPXp08e+vH///ipevLjGjBmj119/XT4+Pho9erSuXLmi3bt3q0SJEpKkXr166YUXXtC0adP02muvKWfOnCmVCgAAAAAAAIA4PNS3kJ85c0aSVLRo0RjLg4KCVLp0ad2+fVsREREyxujrr79WhQoV7MXLaN27d1dERISWLVvmtrgBAAAAAAAAOOahLmDWqlVLlmVp4MCB+uOPP+zLb926pR9++EENGjSQv7+/Tp8+rUuXLqlIkSKxtlGsWDFJ0t69exPcV3h4uK5fvx7jAQAAAAAAAMC1HuoCZunSpfXJJ5/o2LFjKliwoBo1aqSPP/5Y9erVU758+fThhx9K+r+emmnTpo21jdDQ0Bht4jN+/HilTZvW/uB2cwAAAAAAAMD1HuoCpiRlzJhRGTNmVNu2bXX58mV17txZW7ZsUYUKFZQ6dWpJ0t27dyXJ/vzf/Pz8JEl37txJcD+DBg3StWvX7I+TJ086ORMAAAAAAAAA//VQT+KzceNG1a9fX2+//ba6du0qSTp9+rQmT56syZMn6+eff9aGDRsUEBAgSYqMjIy1jfv370uSAgMDE9yXv7+//P39nZwBAAAAAAAAgIQ81AXM8ePHK126dOrSpYt9WY4cOTRjxgwZYzRr1ixt2bJFuXPnliRduXIl1jYuXrwoScqSJYtbYgYAAAAAV2od5JfSIQAA4FQP9S3kJ0+elM1mk2VZsdblzZtXknT27FnlzJlT6dOn1/bt22O1i15WvHhx1wYLAAAAAAAAINEe6gJmqVKldP78eX3yyScxlkdERGjRokXy8fFR2bJlZVmWGjZsqD179mj//v0x2i5YsEA+Pj5q2LChO0MHAAAAAAAA4ICH+hbyUaNGad26dWrXrp2WL1+uChUq6MaNG/ryyy/166+/auzYsQoLC5MkDRs2TMuWLdPTTz+tLl26KGPGjFq3bp1WrlypXr162dsBAAAAAAAA8BwPdQEzb9682r17t0aPHq01a9bo66+/VurUqVW2bFlNmjRJzz33nL1tnjx59MMPP2jIkCF6++23devWLeXNm1dTp05Vr169Ui4JAAAAAAAAAPF6qAuYkvTYY49pzpw5DrUtUqSIli5d6uKIAAAAAAAAADjLQz0GJgAAAAAAAIBHGwVMAAAAAAAAAB6LAiYAAAAAAAAAj0UBEwAAAAAAAIDHooAJAAAAAAAAwGNRwAQAAAAAAADgsShgAgAAAAAAAPBYFDABAAAAAAAAeCwKmAAAAAAAAAA8lm9yN3Dp0iVduHBBV69eVbp06ZQ5c2ZlyJDBGbEBAAAAAAAA8HKJLmCGh4dr4cKFWrVqlb777jtdvXo1Vpt06dLp6aefVr169dSiRQv5+/s7I1YAAAAAAAAAXsbhAmZkZKTeeustTZo0SWfOnFFAQICqVKmikiVLKkeOHAoKCtLNmzd16tQp/fLLL/r666/15ZdfatCgQRowYIBeffVV+fj4uDIXAAAAAAAAAI8YhwqYv//+u1q1aqWdO3eqRIkSmjhxopo0aaLAwMB4X3Pz5k19+umnmjJlinr37q0FCxbos88+U968eZ0WPAAAAAAAAIBHm0MFzJIlS8rHx0fvv/++OnbsKMuyHvia4OBgde/eXR06dNC4ceM0adIklSpVSteuXUt20AAAAIAnKOb74O/FAAAASB6HZiEPDQ3V9u3b1alTJ4eKl//m7++vkSNHauvWrQoODk5SkAAAAAAAAAC8k0M9MLdv366sWbMma0elS5fWli1bkrUNAAAAAAAAAN7FoR6YyS1eRsuTJ49TtgMAAAAAAADAOzhUwAQAAAAAAACAlODQLeQJiYiI0Lx587Rp0yadP39egYGBKlq0qNq3b698+fI5I0YAAAAAAAAAXipZBcyIiAhVq1ZN27dvlzHGvnz58uWaOnWqVq9ererVqyc3RgAAAAAAAABeKlm3kH/55Zfatm2bunTpoj/++EN3797VjRs3tGPHDpUoUUL9+/d3VpwAAAAAAAAAvFCyCpgHDx6UZVmaPHmywsLClCpVKgUFBal06dLq06eP9u/f76w4AQAAAAAAAHghhwuYd+7cibUsbdq0kqQjR47EWnf06FGlSZMmGaEBAAAAAAAA8HYOj4GZN29eDR8+XJ07d5aPj48kqWHDhhoyZIjq1KmjBg0aKEeOHIqIiNDhw4e1Zs0aderUyWWBAwAAAAAAAHj0OVzArFOnjl599VVNnTpVY8aMUfPmzVWwYEEtXLhQPXr00Pz58/9vo76+atmypaZOneqSoAEAAAAkTesgv5QOAQAAIFEcvoV87ty52rNnjwoVKqSWLVuqTJkyWrdunV544QWdPHlSv/32m7Zs2aIdO3bo0qVLmj9/vgIDA10ZOwAAAAAAAIBHXKIm8SlatKiWL1+urVu3KnXq1HrmmWdUq1Yt7dy5U/nz51flypVVunRphYSEuCpeAAAAAAAAAF4kSbOQV6pUSZs3b9by5ct18eJFVahQQc2aNYtzMh8AAAAAAAAASKokFTCjPffcc9qzZ4/mzZun3bt3q2jRouratatOnz7trPgAAAAAAAAAeDGHJ/GRpGvXrumzzz7T77//rjt37qhgwYJq0aKFXnrpJbVo0ULvvvuuxo0bpwULFui1117TwIEDlT59elfFDgAAAMCNivlaKR0CAADwQg4XMPft26caNWro6tWrMsbYl48YMUJbtmxR0aJF1aNHD3Xq1ElTpkzRtGnT9P777+vy5csuCRwAAAAAAADAo8/hW8j79Omj8PBwffjhh9q/f7/279+vjz76SBEREerTp4+9XVBQkIYPH67ff/9d7dq1c0nQAAAAAAAAALyDwz0wd+/erRdffFHt27e3L3viiSe0detWLV26NFb7TJkyafr06U4JEgAAAAAAAIB3crgHZlBQUJy3g1+9elWBgYFODQoAAAAAAAAApET0wKxdu7bmz5+vzp07q3Tp0pKkXbt26euvv9aLL77osgABAAAAAAAAeC+HC5jTpk3TgQMH9NFHH+mjjz6yLy9evLimTZvmkuAAAAAAAAAAeDeHC5jp06fXjh079Msvv+jIkSOSpHz58tl7YwIAAAAAAACAszlcwIxWsmRJlSxZ0hWxAAAAAAAAAEAMDk/iAwAAAAAAAADu5lABc9asWU7Z2ZIlS5yyHQAAAAAAAADewaECZq9evTRgwABFRUUleUeTJ09Wy5Ytk/x6AAAAAAAAAN7HoQJm7dq1NXnyZFWoUEE//PBDonawa9culStXTgMGDFDNmjWTFCQAAAAAAAAA7+RQAXPt2rWaPHmyDhw4oKpVq6py5cqaM2eOjh8/Hmf7CxcuaPbs2XrqqadUrlw5HThwQFOmTNG3337r1OABAAAAAAAAPNocnoW8b9++atWqlSZMmKAPPvhA27Ztk2VZSp8+vXLlyqXg4GDdvn1bZ86c0ZkzZ2SMUWBgoP73v/9pwIABypYtmyvzAAAAAAAAAPAIcriAKUnZsmXTzJkzNX78eG3YsEHr1q3T77//rgsXLujMmTNKly6dnnzySTVt2lRPP/20atSoodSpU7sqdgAAAAAAAACPuEQVMKOlTp1a9evXV/369Z0dDwAAAAAAAADYOTQGJgAAAAAAAACkBAqYAAAAAAAAADzWI1PAXLNmjZo2baocOXIoVapUSpcunXbs2BGjzZEjR9S0aVNlzJhRAQEBeuKJJzRz5kxFRUWlUNQAAAAAAAAAEpKkMTA9SUREhDp16qT58+erVKlS6tKli9KnT69jx47Jz8/P3u748eOqVKmSJKlbt27KlCmT1q5dq169eunvv//WxIkTUyoFAADwiCrma6V0CAAAAMBD76EvYI4cOVLz58/XiBEjNHz48HjbjR49WleuXNHu3btVokQJSVKvXr30wgsvaNq0aXrttdeUM2dOd4UNAAAAAAAAwAEP9S3kFy5c0KRJk1S1alUNGzYs3nbGGH399deqUKGCvXgZrXv37oqIiNCyZctcHS4AAAAAAACARHqoe2AuWbJE9+7d0//+9z9ZlqV79+7p1q1bSp8+fYx2p0+f1qVLl1SkSJFY2yhWrJgkae/evQnuKzw8XOHh4fbn169fd0IGAAAAAAAAABLyUBcwd+7cKUn65ZdfNGzYMP3222+KiopSSEiImjVrpilTpih9+vQ6c+aMJClt2rSxthEaGipJ9jbxGT9+vEaOHOnkDAAAgKdrHeT34EYAAAAAXCZZBczjx49r/fr1Onv2rMqVK6fatWtL+qc347Vr11S+fHn5+/s7JdC4RBcdv/vuO7Vr10558uTRnTt39M033+ijjz7Snj17tH37dt29e1eSlDp16ljbiJ7o586dOwnua9CgQerTp4/9+fXr1xkzEwAAAAAAAHCxJBcwBwwYoGnTpikyMlKWZalv3772AuYvv/yiTp06ad68eXrppZecFux/3bt3T5K0fft2Wdb/zfLZtm1bvfLKK3r33Xe1atUqZc+eXZIUGRkZaxv379+XJAUGBia4L39/f5cWYwEAAAAAAADElqRJfD744ANNnjxZVatW1RdffCFjTIz1L730kkJDQ7Vo0SKnBBmfDBkySJJu3boVa12DBg0kSYcOHVKWLFkkSVeuXInV7uLFi5JkbwMAAADv0jrIL1EPAAAAuFeSCpjvvvuuwsLCtGbNGjVr1izWel9fX9WoUUM///xzsgNMyJNPPilJ2rdvX6x1ly9fliSFhIQoZ86cSp8+vbZv3x6rXfSy4sWLuy5QAAAAAAAAAEmSpALm4cOHVb16dfv4kXHJnDmzvYjoKi1atJCPj49GjBihiIgI+/KoqCh9/PHH8vX11bPPPivLstSwYUPt2bNH+/fvj7GNBQsWyMfHRw0bNnRprAAAAAAAAAASL0ljYPr4+MjHxyfBNn/99ZfSpEmTpKAclSdPHo0aNUpDhgxRqVKl1KJFC0nS119/rZ07d2rSpEnKkyePJGnYsGFatmyZnn76aXXp0kUZM2bUunXrtHLlSvXq1UthYWEujRUAAAAAAABA4iWpgFm0aNE4b8eOdurUKW3cuFEVK1ZMcmCOGjx4sPLkyaNp06Zp7NixstlsKl26tL755hs999xz9nZ58uTRDz/8oCFDhujtt9/WrVu3lDdvXk2dOlW9evVyeZwAAAAAAAAAEi9JBcwuXbqoU6dOGjdunAYPHhxj3Z9//qlWrVrpzp07at++vTNifKCWLVuqZcuWD2xXpEgRLV261A0RAQDwaGICEwAAAADulqQxMDt06KCXXnpJb7zxhvLlyyfLsvTll1+qZMmSKlCggH766Sc1btxYrVu3dna8AAAAAAAAALxIkgqYkjR//ny9//77CgkJkSQdP35ce/fuVVhYmKZMmaJFixY5LUgAAAAAAAAA3ilJt5BH69Spkzp16qQ7d+7oypUrCgkJsRc0AQAAAOBhV8zXSukQAADweskqYEYLDAxUYGCgMzYFAAAAAAAAAHZJKmCeOHHigW1sNpuCg4OVLl26pOwCAAAAAAAAAJJWwMydO7csy7FbKUJDQ9WwYUONGjVK2bNnT8ruAAAAAAAAAHipJBUwCxcurEOHDqlatWoKCAiItf7AgQO6dOmSqlatqr///lsfffSR1qxZox07dihbtmzJDhoAAAAAAACAd0hSATNv3ry6ffu2Nm7cGOf6zZs3q3r16urRo4fq1aun1atXq2HDhho3bpzefPPNZAUMAAAAAAAAwHvYkvKibdu2qVq1avGur1q1qqpUqaLJkydLkp599lk988wzWr16ddKiBAAAAAAAAOCVklTAvHbtmtKmTZtgm+LFi2vXrl3253nz5tXp06eTsjsAAAAAAAAAXipJBcxcuXJpy5YtCbY5fvy4oqKi7M+vXLkiHx+fpOwOAAAAAAAAgJdKUgHzpZde0p49e9SrVy/dvHkz1vqlS5dqzZo1Klu2rH3ZsWPHlClTpqRHCgAAAAAAAMDrJGkSnwEDBujbb7/VrFmz9Mknn6hs2bLKkSOHrl+/roMHD+q3336Tv7+/JkyYIEm6deuWdu7cqbp16zo1eAAAAAAAAACPtiQVMAMCArRx40aNHj1a7777rtauXWtfZ1mWatSooYkTJ6p06dKSpFSpUunAgQNKly6dU4IGAAAAAAAA4B2SVMCUJH9/f40ZM0ajR4/WkSNHdOnSJQUEBCh//vwKCQmJ0dbPz0958+ZNdrAAAAAAAAAAvEuSC5jRLMtSwYIFnRELAAAAAAAAAMSQ7ALm1atXde/evXjXZ86cObm7AAAAAAAAAOClklzAnDhxoqZMmaLLly/H28ayLEVERCR1FwAAAAAAAAC8XJIKmHPnztWgQYOUPn16Zc+eXadPn1a1atXs63fv3i1jjH0SHwAAAAAAAABICltSXvTBBx8oR44c+v3339WxY0dZlqV169Zp48aN2rhxo6ZPn667d+9q5syZzo4XAAAAAAAAgBdJUgHz4MGDqlmzptKnT6+0adNKki5dumRf37FjRxUuXFiTJ092TpQAAAAAAAAAvFKSCpi3bt1ShgwZJMlewDxx4kSMNpUrV9amTZuSFx0AAAAAAAAAr5akAmb69Ol15coVSVLBggVljNEPP/wQo014eHiCE/wAAAAAAAAAwIMkqYCZLVs2HT16VJJUvnx5pU+fXlOnTrX3wvzrr7+0bNky5cuXz3mRAgAAAAAAAPA6SSpgFi1aVDt37tTdu3fl5+enfv366fTp0ypQoIDCwsJUsGBBXb58WYMHD3Z2vAAAAAAAAAC8SJIKmJ988olu376tgIAASdLAgQM1fvx4Zc+eXefPn1eRIkW0cOFCtWjRwqnBAgAAAAAAAPAuvkl5kWVZ8vHxibFswIABGjBggFOCAgAAAAAAAAApiT0wO3bsqC+//DLBNhMnTtRbb72VpKAAAAAAAAAAQEpiAXPu3LnasWNHgm327t2ryZMnJykoAAAAAAAAAJCSWMB0RFBQkM6fP++qzQMAAAAAAADwAkkaA/NBDh8+rJUrVypHjhyu2DwAAAAAOKx1kF9KhwAAAJLB4QJmnjx5Yjx///33tXjx4ljt7t27pzNnzkiShg0blszwAAAAAAAAAHgzhwuYf//9tyzLkvTPLOQ3b97U3bt3Y2/Q11f58+dXixYt9MYbbzgvUgAAAAAAAABex+EC5r+LlTabTb1799akSZNcEhQAAAAAAAAASEkcA/PPP/9U2rRpnR0LAMSrmK+V0iEAAAAAAIAUkKQCZq5cuZwdBwAAAAAAAADEkuRZyG/duqXPPvtMR44c0eXLl2WMidXGsix9+OGHyQoQAAAAAAAAgPdKUgFz9+7devbZZ3Xx4sU4C5fRKGACAAAAAAAASI4kFTD79OmjCxcuqGvXrnr22WcVGhoqPz8/Z8cGAAAAAAAAwMslqYD5888/q169epo9e7az4wEAAAAAAAAAO1tSXhQYGKi8efM6OxYAAAAAAAAAiCFJPTCrVq2qXbt2OTsWAADwCCvma6V0CAAAAAAeQknqgdm7d29t27ZNa9ascXY8AAAAAAAAAGCXpB6YCxcuVLp06dS1a1fVr18/3naWZentt99OcnAAAAAAAAAAvFuSCpjRk/dcuXIlwYl8KGACAAAAAAAASI4kFTA3btzo7DgAAAAAAAAAIJYkFTCrVavm7DgAAAAAAAAAIJYkTeLj6caOHSvLsvTaa6/FWvfzzz/rmWeeUbp06ZQ6dWqVKVNGn376aQpECQAAAAAAAOBBklzAPHnypDp37qzHH39cqVKlUv/+/e3rpk2bppo1a+rAgQNOCTIxPvvsMw0dOjTOdTt37lS1atV06NAh9enTR6NHj1aqVKnUpk0bvfvuu26OFAAAAAAAAMCDJKmA+ddff6lMmTL6+OOPFRISooiIiBjrGzRooO+//97tPRu///57dejQQe3bt49z/cCBA+Xn56cff/xRw4YNU9++fbVlyxaVLVtWQ4YM0e3bt90aLwAAAAAAAICEJamAOWzYMF26dElr1qzRwYMHY63Pnz+/ypUrp1WrViU7QEcdPnxYjRs3VqtWrTRs2LBY669cuaJNmzapQYMGypEjh325j4+POnfurCtXrmjDhg1uixcAAAAAAADAgyWpgPndd9/p2Wef1dNPPx1vmyeffFJ//PFHkgNLjPPnz6tevXoqVaqU3nvvvTjbHDx4UJGRkSpSpEisdcWKFZMk7d27N959hIeH6/r16zEeAAAAAAAAAFwrSQXMS5cuKSwsLME2d+/elTEmSUElxu3bt9WgQQMFBQVpyZIl8vPzi7PdmTNnJElp06aNtS40NDRGm7iMHz9eadOmtT9y5szphOgBAAAAAAAAJCRJBcwsWbLo9OnTCbbZuXOny4t8UVFRat26tU6ePKmVK1fGWZyMdvfuXUlS6tSpY62LLnreuXMn3tcPGjRI165dsz9OnjyZzOgBAAAAAAAAPIhvUl5Up04dff755zp9+nSM8SSjvf/++zp48KBee+21ZAeYkC+//FIrV67U4sWLZbPZdOrUKUnS2bNnJUm3bt3SqVOnlCZNGgUEBEiSIiMjY23n/v37kqTAwMB49+Xv7y9/f39npwAAAAAAAAAgAUnqgTl06FD5+/uratWq+uSTTyRJp0+f1rJly9ShQwe98sorSpMmjfr27evUYP/rzp07un//vp5//nnlzJnT/qhYsaIkae7cucqZM6emTZumLFmySPpnMp//unjxoiTZ2wAAAAAAAADwDEnqgZkrVy6tXbtWLVq0ULt27SRJCxcu1MKFC2WMUZYsWfTFF18oV65cTg32v2rXrq0VK1bEWn7+/Hl16tRJzz33nLp3764CBQooY8aMstls2r59e6z20cuKFy/u0ngBAAAAAAAAJE6SCpiSVKZMGf32229auXKltm3bpsuXLyskJESlS5dW48aNE7wd21kee+wxPfbYY7GW//XXX5Kk3Llzq379+vblVapU0bp163T27FllzZpV0j+3lC9cuFAhISGqVauWy2MGAAAAAAAA4LgkFzAlydfXV88//7yef/55Z8XjUhMmTFC1atVUuXJldejQQalTp9bSpUu1Y8cOzZgxQ8HBwSkdIgAAAAAAAIB/SdIYmN9//73y5Mmj/v37x7l+0qRJKlWqlHbt2pWs4JytQoUK2rhxo/LkyaNJkyZpyJAhunXrlubOnauePXumdHgAAAAAAAAA/iNJPTA/+ugj/f333+rVq1ec69u2bathw4Zpzpw5eu+995ITX5Lkzp1bxpg411WqVEnr1q1zc0QAAAAAAAAAkiJJPTA3bNigmjVrKnv27HGuz5o1q2rVqqXNmzcnKzgAAAAAAAAA3i1JBczz588rf/78CbbJmzevTpw4kaSgAAAAAAAAAEBKYgEzQ4YMOnPmTIJtzp07x6Q4AAAAAAAAAJIlSQXMatWqaeXKlTpy5Eic648ePapvvvlGFSpUSFZwAAAAAAAAALxbkgqYgwcPVmRkpCpWrKiJEyfqp59+0okTJ7Rjxw5NmjRJlSpVUnh4uPr27evseAEAAAAAAAB4kSTNQl68eHF9/vnnatu2rQYPHhxjnTFGfn5+evvtt1WtWjWnBAkAAAAAAADAOyWpgClJjRs31rFjxzR37lzt2LFDV69eVZo0aVSyZEm1bt1auXPndmKYAAAAAAAAALxRkgqY27ZtU0hIiIoWLap+/fo5OyYAAAAAAAAAkJSMSXymT5/u7FgAAAAAAAAAIIYkFTD9/f0VEhLi7FgAAAAAAAAAIIYkFTArVaqkAwcOODsWAAAAAAAAAIghSQXM/v37a9OmTdq8ebOz4wEAAAAAAAAAuyRN4rNr1y5ly5ZN7dq108svv5xg2/79+ycpMAAAAAAAAABIUgFz4MCBcf77vyzLooAJAAAAAAAAIMmSVMD8+OOPnR0HAAAAAAAAAMSSpAJmu3btnB0HAAAPpdZBfikdAgAAAAA80pI0iQ8AAAAAAAAAuEOSC5g3btzQmDFjVLVqVRUoUEATJkywr3v//ffVsWNH/fnnn04JEgAAAAAAAIB3StIt5BcvXlTlypV19OhRBQUF6fbt27p8+bJ9fZkyZdStWzeFhYVp6NChTgsWAAAAAAAAgHdJUg/M4cOH6+jRo3r//fd1/fp1GWNirC9ZsqSKFy+u5cuXOyVIAAAAAAAAAN4pSQXMb775RjVq1FCnTp1kWVacbcqWLavDhw8nKzgAAAAAAAAA3i1JBcyzZ8+qcOHCCbaxLEv3799PUlAAAAAAAAAAICWxgJkhQwZdunQpwTZ79+5VtmzZkhQUAAAAAAAAAEhJLGBWq1ZN3377ra5duxbn+hUrVmjHjh2qUaNGsoIDAAAAAAAA4N2SVMB84403dPfuXdWpU0dbtmyRJN28eVN79+7VyJEj1bJlS6VKlUp9+/Z1arAAAAAAAAAAvItvUl5UtGhRffnll3rppZdUvXp1SdKcOXM0Z84cGWMUEBCgefPm6YknnnBmrAAAAAAAAAC8TJIKmJL03HPP6dixY5o/f762bdumy5cvKyQkRKVLl1b79u2VPXt2Z8YJAAAAAAAAwAslqYB548YNhYeHK2PGjOrZs6d69uzp7LgAAAAAAAAAIHFjYO7cuVOlS5dWunTplCVLFuXJk0effvqpq2IDAAAAAAAA4OUc7oF58uRJ1a5dW9evX1dQUJB8fX31119/qV27drp165a6devmyjgBAAAAAAAAeCGHe2DOnj1b169f15gxY3TlyhVduXJF27ZtU9asWdWvXz9dvXrVhWECAAAAAAAA8EYOFzC//vprFSpUSIMHD5av7z8dN8uXL69p06bp5s2b+vbbb10WJAAAAAAAAADv5PAt5H/99ZfatGkTa3mtWrUkSUePHnVeVAAAAAAAIE75m69NVPuT//r3kJwtE/XaiYlqDQCu4XAPzDt37igkJCTW8tDQUEnSzZs3nRcVAAAAAAAAACgRPTAlybKseJdFRUU5JyIAj6zWQX4pHQIAAAAAAHjIJKqAuWTJEh04cMDhdZZlaeXKlUmPDgAAAAAAAIBXS1QB888//9Sff/7p8Lq4emwCAAAAAAAAgKMcLmDGV7gEAAAAAAAAAFdxuICZK1cuV8YBAIDXKebLnQoAAAAA8CAOz0IOAAAAAAAAAO5GARMAAAAAAACAx6KACQAAAAAAAMBjUcAEAAAAAAAA4LEoYAIAAAAAAADwWBQwAQAAAAAAAHgsCpgAAAAAAAAAPBYFTAAAAAAAAAAeiwImAAAAAAAAAI/lm9IBJMelS5c0ffp0bdq0SX/++aeuXLminDlzqm7duhoyZIiyZMkSo/3PP/+sYcOGafv27bp3756KFCmiXr166aWXXkqhDAAAAAAAKWlIzpaJaj/RRXEAAOL3UBcw9+/fr/Hjx6tmzZrq1q2bQkJCtHv3br3zzjv66quvtGfPHmXMmFGStHPnTlWrVk2ZM2dWnz59FBQUpCVLlqhNmza6ceOGXn755RTOBgAAAAAAAMB/PdQFzLx58+rw4cPKnz9/jOWlSpVSnz599MEHH2jgwIGSpIEDB8rPz08//vijcuTIIUnq1auXKlasqCFDhqhdu3ZKnTq123NwlWK+VkqHAAAAAAAAACTbQz0GZs6cOWMVLyWpTp06kqS//vpLknTlyhVt2rRJDRo0sBcvJcnHx0edO3fWlStXtGHDBrfEDAAAAAAAAMBxD3UBMz4nTpyQJHux8uDBg4qMjFSRIkVitS1WrJgkae/evQluMzw8XNevX4/xAAAAAAAAAOBaD/Ut5PGZOXOmLMtS06ZNJUlnzpyRJKVNmzZW29DQ0Bht4jN+/HiNHDnSyZECAABPl7/52kS1P+miOAAAAABv9cj1wJw8ebK+/fZb9ezZU4ULF5Yk3b17V5LiHOPSz89PknTnzp0Etzto0CBdu3bN/jh5kj9PAAAAAAAAAFd7pHpgzpkzRwMGDFCDBg00efJk+/KAgABJUmRkZKzX3L9/X5IUGBiY4Lb9/f3l7+/vxGgBAAAAAAAAPMgjU8CcOnWq+vXrpwYNGmjRokXy9f2/1LJkySLpn8l8/uvixYsx2gDwTENytkxU+4kuigMAAAAAALjXQ38LuTFGffv21euvv66OHTvqq6++itVTsmjRorLZbNq+fXus10cvK168uFviBQAAAAAAAOC4h7qAefv2bTVp0kTTp0/XyJEj9cEHH8jHxydWuwwZMqhKlSpat26dzp49a18eGRmphQsXKiQkRLVq1XJn6AAAAAAAAAAc8FDfQt6vXz8tXbpUderUUebMmTV79uxYbRo2bKjs2bNrwoQJqlatmipXrqwOHTooderUWrp0qXbs2KEZM2YoODg4BTIAAAAAAAAAkJCHuoB569YtSdLatWu1du3aONsUKlRI2bNnV4UKFbRx40YNHz5ckyZN0v3791W4cGHNnTtX7dq1c2fYAAAAAAAAABz0UBcw586dq7lz5zrcvlKlSlq3bp3rAgIAAAAAAADgVA/1GJgAAAAAAAAAHm0UMAEAAAAAAAB4LAqYAAAAAAAAADwWBUwAAAAAAAAAHuuhnsQHAADAkw3J2TJR7Se6KA4AAADgYUYBEwAAAADwUMvffG2i2p90URwAANeggOnBWgf5pXQIAAAAAAAAQIpiDEwAAAAAAAAAHosCJgAAAAAAAACPRQETAAAAAAAAgMeigAkAAAAAAADAY1HABAAAAAAAAOCxmIUcgNvkb742Ue1PuigOAAAAAADw8KCACQDwCMV8rZQOAQAAAADggbiFHAAAAAAAAIDHogcmkIJaB/mldAgAAAAAAAAejR6YAAAAAAAAADwWBUwAAAAAAAAAHosCJgAAAAAAAACPRQETAAAAAAAAgMeigAkAAAAAAADAYzELOfCQKuZrpXQIAAAAAAAALkcBEwA8FEVqAAAAAAAoYAIAnKh1kF9KhwAAAAAAeMRQwAQAAIBd/uZrE9X+pIviAAAAAKJRwAQAF6JHIgDA01CkBgAADxtmIQcAAAAAAADgseiBCQAAAKcYkrNlotpPdFEcAAAAeLTQAxMAAAAAAACAx6KACQAAAAAAAMBjcQs5AABwGJN/AAAAAHA3emACAAAAAAAA8Fj0wMQjoZivldIhAAAAAAAAwAUoYAIAALdghmoAAAAAScEt5AAAAAAAAAA8FgVMAAAAAAAAAB6LW8gBAACAJPK2oRG8LV8AAOAZKGDCY7QO8kvpEAAAAACvRpEaAOCJuIUcAAAAAAAAgMeiByYAAAAAPELyN1+bqPYnXRQHAADOQg9MAAAAAAAAAB6LAiYAAAAAAAAAj0UBEwAAAAAAAIDHooAJAAAAAAAAwGNRwAQAAAAAAADgsShgAgAAAAAAAPBYFDABAAAAAAAAeCzflA7A3aKiovTmm2/q/fff1++//66goCDVqFFD48aNU4ECBVI6PADAQyZ/87WJan/SRXEAAAAAwKPK63pgDhgwQL169dLjjz+uCRMmqHv37tq0aZMqV66skyf5sxIAAAAAAADwJF7VA/P48eOaPn26mjRposWLF9uXN2vWTKVKldKYMWM0Z86cFIzQeYbkbJmo9hNdFAcAAAAAAACQHF5VwFy+fLkiIyPVrVu3GMuffPJJVahQQV999dUjU8DEw4FbTwEAAAAAABLmVQXMvXv3SpKKFCkSa12xYsW0bds2nT59Wjly5Ii1Pjw8XOHh4fbn165dkyRdv37dRdFKd6OiEtX+3r9iCb99I1GvdWUejkpOvhF37ybqtb4ekK8kRYXfTFT7fx+nYaHPJeq1ozwg5+Tk+zC+p6Xkva8fxmPMefxgnMdJe21K8racvS1fyfu+c3GMH+xh/3zyxmPsbTl7W76S9+XsbflK3pezt+Ubn+jYjDEJtrPMg1o8Qp577jmtWrVKN27cUHBwcIx1gwcP1vjx47Vjxw6VKVMm1mtHjBihkSNHuitUAAAAAAAAwCucPHlSjz32WLzrvaoH5t3//+tp6tSpY63z8/OTJN25cyfO1w4aNEh9+vSxP4+KitLly5cVGhoqy7JcEC0AAAAAAADw6DLG6MaNG8qePXuC7byqgBkQECBJioyMlM0WcwL2+/fvS5ICAwPjfK2/v7/8/f1jLEuXLp3zgwQAAAAAAAC8RNq0aR/YxvbAFo+QLFmySJKuXLkSa93FixdjtAEAAAAAAACQ8ryqgFm8eHFJ0vbt22Ot2759u9KnT5/g/fYAAAAAAAAA3MurCpgNGzaUzWbTJ598EmP5vn37dODAATVq1IjxLAEAAAAAAAAP4lWzkEtSz549NWvWLDVs2FA1a9bUhQsX9N577ykiIkK7du1SWFhYSocIAAAAAAAA4P/zugJmZGSkZsyYoQ8//FB//PGHgoKCVL16dY0dO1aFCxdO6fAAAAAAAAAA/IvXFTABAAAAAAAAPDy8agxMAAAAAAAAAA8XCpgAAAAAAAAAPBYFTAAAAAAAAAAeiwImAAAAAAAAAI9FARMPjfDw8JQOwe28MWdv423H2NvyBR5FnMePPm87xt6Wr+SdOcO7bd68WWfPnk3pMJzK285jb8tX8s6cE0IBEx4tIiJC33zzjdq0aaPMmTOndDhu4Y05J9fNmzd19uxZRUZGJthuyZIl+uabb9wUVfy87Rh7W77RDh06pFGjRqV0GHChs2fPavPmzSkdhlt463mcVHfu3NGBAwe0fft2HTt2TPfu3Yu37dtvv605c+a4Mbq4edsx9rZ8Je/MWfKuazXiV6NGDS1fvtz+/M6dO5o/f77+/vvvFIwq8bztPPa2fCXvzNlhBg+dnTt3mg4dOjywXe/evc22bdvsz69du2Z69+5tDh06FKvtjz/+aHr37u3UOJMqIiLCrF692rRv396kT5/e2Gw2kypVKlO3bt2UDs1lvDHn5Pr1119Nx44dTY4cOYzNZjM2m834+PiYXLlymf79+5s///wz1mu6du1qMmXK5P5gjfcdY2/LNy6LFy82NpstpcNAEhw4cMB07drVPPPMM6Zjx45mxYoVcbabPXv2I32MOY8Tb926daZmzZomVapU9s8mm81m/P39TfPmzc1PP/0U6zW9evUy6dOnT4Fove8Ye1u+xjzaOXOtRmJYlmXmzJljf37q1Cljs9nMypUrUzAqxzzK53FcvC1fY7wz56TwTekCKhLvr7/+0rx58/TRRx8l2G7GjBkqVKiQKlSoIEm6ceOGZs6cqdq1a6tQoUIx2u7bt08zZ87UtGnTXBZ3QqKiorRhwwYtWrRIS5cu1eXLlyVJTz31lDp16qSGDRsqXbp0KRKbq3hjzs4yf/58devWTREREapSpYqaNm2qkJAQ3bhxQ/v379e0adM0e/ZszZ8/X88//7z9dXny5NHVq1fdFqe3HWNvyDc6J0fcuHHDoXbXr19X6tSp5evrG2NZcHCwbLaYN0pERETo9u3bSpMmjcNxuMuuXbv09ttvP/CzSfrnXEwMy7L0xx9/JDW0RDlw4IAqVKig27dvK1OmTPruu+80d+5cVatWTQsWLFC2bNncEkdK8Ybz2FXGjh2rYcOGKW3atGrdurWKFy9u/2zat2+fli1bpqVLl2rGjBl65ZVX7K/LmDGjbt265bY4ve0Ye1u+knfk7O3Xamfx5M9jdzDGpHQI8fKG8/jfvC1fyTtzTraUrqDiH7t27XL4MWnSJId+RYzrVybLsuL8lSmlfpnctGmTefnll03mzJmNzWYzAQEB5vnnnzcTJ040lmWZJUuWuD0mV/PGnJ1p9+7dxtfX11SsWNH88ccfcbY5duyYqVSpkgkMDDRHjhwxxhhz5MgRU7hwYVOwYEGXx+htx9ib8rUsK0avKkceD2Kz2RzuEeDJvUgS0+P0ueeeM/Xr13/gI2vWrPb/c3dp2LChyZw5szlw4IAxxpjbt2+bWbNmmeDgYJM7d27z22+/2dt68vFILG86j11hw4YNxrIs07hxY3P16tU421y/ft288MILxtfX1/z888/m3r17Zu3atSZHjhymVKlSLo/R246xt+VrjHfl7K3Xamf77+dx7dq1jWVZply5cvF+Nj+sEvO3cUrypvPYGO/L1xjvzNlZ6IHpIcqUKSPLshxqa4xxuK0ny549u86dO6dUqVKpbt26atasmRo2bKiQkBD98ccfGjhwYEqH6HTemLOzTZ48WVmyZNHatWsVHBwcZ5uwsDCtWbNGhQsXVuXKleXn56ezZ8/Kx8dHixYtcml83naMvS3fZcuW6X//+59Onjypli1bqkCBAvG2PXjwoJYsWfLAbZo4fv2Pa1lK2L17t8Ntjx075nDbB41Fu27dOo0ZM0bnzp1Trly51L9/f4e3nVw7d+5Uly5d9MQTT0iSAgMD9b///U+1a9fWM888oxo1auj7779Xvnz53BaTq3nbeewKM2bMUP78+bVo0aIYvan/LSQkRAsXLtQTTzyhypUr28dtTpMmjWbOnOnS+LztGHtbvpL35eyN12pX+O/n8alTp/T4449r+PDhqlevXgpF5b287Tz2tnwl78zZmShgeoi9e/eqTZs22r9/v3r27KkSJUrE23bHjh1699133Rida5w9e1aZM2fWtGnT1KxZM/n5+aV0SC7njTk726ZNm/TSSy/FW7yMFhISotatW2vq1Klq2LCh8ufPr65duypv3rwujc/bjrG35dugQQNJUqNGjdS6desEv9wvWbLEoQKmJ3P3j2srVqzQ2LFjtWPHDuXLl08ffvih2rRpE29ByBWuX7+uLFmyxFpeuHBh/fDDD6pdu7Zq1Kjh0gkh9u7dm+D3AGfztvPYFbZt26Zu3bo98L3q5+enli1baty4cerZs6fy58+vVq1aKW3atC6Nz9uOsbflK3lfzp5wrX4UPQqdZBJy+/Zt+226V65ckfTPkD/xDRGUIUMGt8Umed957G35St6ZszMxC7mHKFasmKZMmSJJql27ttq1axfvo0aNGikcrXP07NlTfn5+9tm12rdvr1WrVun+/fspHZrLeGPOznbp0iU9/vjjDrXNmTOnbDabvvrqK02cONHlxUvJ+ce4d+/e+vrrr3Xp0iUnR+oc3vieLly4sMM9JD2lJ2VS7d27V8WKFZP0z7H+6KOP4n38e0y/xFq0aJGefPJJPf/887p165Y+/fRTHTp0SB06dHBr8VKScufOrYMHD8a57rHHHtOGDRsUGBiounXr6vz5807f/+rVq1W1alWnbzch3ngeO1t8xZS4ZM6cWT4+Ppo2bZpefvlllxcvJe87xt6Wr+R9Oaf0tdrdli5dqhdffFH9+/e357N//35Vq1ZNqVOnVmhoqFq0aKHff/89hSN1rsjISJ07d05RUVHxttm0aZM6duzo0Pb69u2rTJkyKVOmTPYfClu1amVf9t+Hu3nbeext+UrembNTpeT964jp+PHjJnfu3Gbjxo0Jtlu9erXJnTv3A7dnWZZ544037GNnrl692liWZWbNmhVrXM3Bgwen2NgwW7ZsMa+99prJnj27sSzLpE+f3jRo0MDYbDbz1VdfpUhMruZtOUdERJizZ8+ayMjIeNts3LjRdOjQ4YHbyp49u+nfv79D++3fv7/JkSOHw3E6k7OOcUBAgH0sxSJFipju3bubzz77zJw6dcqF0Seet72nncnTxytet26dQ7N0JnbW9cjISDNv3jxTqFAhY1mWKV26tEe8V9q1a2cyZsxo7ty5E2+bEydOmJw5c9pnmnZERESEWb58uZk0aZKZPXu22b9/f6w2s2fPNr6+vm4Zqzcu3nIef/XVV6Zly5amX79+5ty5c8YYY/bt22eqVq1qAgMDTYYMGUzz5s3N0aNHHd5mrly5TI8ePRxq26NHD4e+x7mCtxzjaN6WrzHek7MrrtVhYWGJeuTJk8eZKcXr66+/NpZlmYCAABMYGGgKFixoTp8+bbJkyWLCwsJMq1atTL169UyqVKlMaGio+f3335O8L08aF3Lo0KEmODjYPhtzkyZNzK+//hqrnaPfjUaMGJHoR0rxlvM4mrfla4x35uwMFDAfYXFNNhHfBBTuniQhLlFRUWbjxo3m5ZdfNlmyZDGWZZm0adOal156ySxdujTBLygPK2/I2dlfPl588UXz2GOPmVu3biXY7saNGyZHjhymdevWSY7dGZJ7jO/du2e2bdtmJk2aZBo1amQyZcpkP1/DwsJMu3btzIcffpioP7RdyRve084WXwFz1apVsdqmRAHT2T+uGfNPHmFhYcayLFO5cmWzevVqJ0TqHHv27DEDBw40u3fvTrDdoUOHTOHChU1YWNgDt3n16lVTtmxZ++dt9Dncvn17c//+fWOMMf369TOWZZmqVauaS5cuOSWXpHqUz2NXFQI6d+5sQkND7QXR+Jw9e9ZkyJDBdO3a1RnpJJmzjvGpU6fM8ePHE2xz4sSJFP/R7VF+T8cnOTnfvn3bdO7c2fTq1ctERUXF227YsGGma9euJjw83BUpJMgV12pPndCmWrVqpkiRIubatWvmzp075plnnjHly5c3lSpVinEc9+7da4KDg02rVq2SvC9PKWB+/PHHxrIsU758edO/f3/z4osvmqCgIJM6dWrz6aefxmj7KE/S5G3XLm/L1xjvzDk5LGMe8vvbEK958+Yl+jXt2rVzQSSJFxUVpU2bNmnhwoX6+uuvdfHiRQUFBenGjRsJvu7EiROJ3pejtyO7WlJyzpMnT6L2YVmW/vjjj+SG6rC5c+eqY8eOKleunKpVq6aTJ09q+fLlMsbovffeU+vWre1t58yZo1deecU+qUF8fvnlF5UrV06VKlXSp59+qpw5c8Zqc+rUKb300kvatm2bduzYoeLFizs9t6RI6vv6v44cOaKtW7dq69at+vHHH3XkyBFZlqWsWbOqatWq+vzzz12UQeI4K9+HyZUrV+Tr66uQkBCHX2Oz2VS8eHH7teju3bv67rvvVK5cOWXOnDlG2xMnTmj//v0PPE88nc1mU7Zs2dSnTx+Hh0UpVaqUi6NynV69eumtt97SkCFD1KRJE9lsNi1btkzjxo1T9+7ddfLkSS1evFht2rTRBx984FHjISX1PD579qyqVKmi/Pnza+XKlfGOq9ahQwf9+OOP2rp1q1tu16tevbouXLigbdu2KVWqVGrcuLGuXLkiHx8frV+/XgEBAZKkffv2qXLlymrYsKEWLFjwwO3+8ccfKlGihHLlyqUPPvhAFStWjNVm+/bt6tSpk06ePKm9e/cqLCzM6fklRVKP8Z49e1S6dGnVr19fy5Yti7ddjx499Pbbb2vfvn32CVdS0qP2nnZEYnN+//331b17d3366ad68cUX493u5s2bVb16dc2dO1dt27Z1VfgpJnpCm2+++SZFJ7TJmDGjevfurSFDhkiSNm7cqNq1a2vOnDnq3LlzjLb/+9//tGjRIp07d86hbf93sr7z58+rXr16mjlzpipXrhzna9zxeVy6dGkFBQXFGMP07Nmzatu2rdavX69p06apZ8+ekhz/G+Jh523fq5OTb3h4uPz9/WP929N5Y86JlsIFVDjgxo0bZvfu3Wbfvn0J/gr6qIqIiDBr1qwxnTp1emDb+HqYJvTwRI7m7Km/FEcrVaqUqVKlSoxlZ86cMU8//bSx2WxmxowZ9uWJ+fX0448/Nv7+/sbPz8/UqlXL9O7d2wwbNsz07t3b1K5d26RKlcr4+/ubTz75xKn5OFNi3tcJuXnzpvn8889NyZIlPaIndXycle9/vffee6ZGjRpO3aYjzp8/b+7du2d/PnfuXPPYY4/Zryvly5c3P/30k0PbypUrl8mdO3eiHg+7f/dCfNDD3e/rDh06mE2bNjl1m2FhYaZly5axlr/55pv2/EaPHu3UfbpCYs7jUaNGGZvNZnbu3Jlguz///NP4+/ubSZMmOSvMBIWGhpoxY8bYn2/YsMHYbDbz/vvvx2r72muvmcyZMzu87TVr1pjQ0FBjs9lM3rx5TePGjU3btm1N48aNTb58+YzNZjOhoaHmu+++c0ourpCYY/zaa6+ZoKAgc+bMmQTbXb9+3WTMmNG8/vrrzgrTaR6F93RiOZJz3bp1TbFixRzaXpUqVdz+/dIY11yr/8tTeiP6+/ubWbNm2Z//9ddfxrIss2LFilhtZ86caQIDAx3etqfesRcSEmImT54ca3lkZKTp1KmTsdls5s033zTGOP43RFhYmMmbN68ZMGCAOX/+vNNjdqfEXLvCw8PNxYsX7c8vXryYIr2mkyMx+Y4ePdp07tzZ/rxz584PxXes//LGnB1BD0wPcu/ePe3du1fZsmXTY489JkkaOXKkJk2apLt370qSsmbNqrfffluNGjVKwUg915QpU2L8In716lWNHTtWHTt2VOHCheN8Td++fd0Vnst5yi/F0dKkSaNhw4bp9ddfj7E8KipKXbt21ccff6yZM2fqtddeS/Svp4cPH9aUKVO0evVqnTlzxr48S5YseuaZZ9S/f/94j/nD7MyZM/rhhx/0ww8/aOvWrdq7d68iIyOVN29ePfXUU6pSpYo6dOiQ0mG6zaBBgzRp0iS3/ep+5swZNWjQQL/88ouCgoI0efJk5ciRQw0bNlT58uVVvXp13b59W1999ZUuX76sn376SUWLFnVLbO528+ZNHT16VL6+vipatGiiZi715DsEbDabZs+era5du0qSLl68qObNm2v8+PEqX758krYZEBCg8ePHq3fv3jGW79u3T08++aSmTJmiPn36JDt2T1KhQgWFhIRo3bp1D2zbpEkTnTt3Tlu3bnV5XAEBAZo8ebL+97//SZKOHz+usLAwLV++XPXr14/RdtasWRo4cKBu377t8PYvXryo2bNna/Xq1Tp8+LCuX7+ukJAQFSxYUM8884xeeeUVj+mVl1xFihRRiRIlHOr137lzZ/3888/at2+fGyJzDU99T7tC5syZ1alTJ40fP/6BbYcPH6733nsvxncxd3DFtfq/Tp8+rZw5c6b49+qCBQuqbt26mjVrlqR/vkf/+eefyp49uwIDA2O07du3r5YtW+bwZD6e+nmcLVs29erVSwMGDIhzfc+ePfXWW2/pww8/VHh4uEN/Q9hsNpUuXVoHDx6UzWZT165d1a9fP2XLls0VKXiMUaNGae/evVqyZIkk6YUXXlDJkiU1dOjQFI7MNS5evKiiRYtq3bp1unfvnp577jkdOnRI6dOnT+nQXMabcnbv1J6I12+//aZatWrpzJkzsixLAwYMUMGCBTVy5Eg1a9ZMNWvW1J07dzR//nw1a9ZMmzZtirdb/79FRUVp0aJF+uabb3TgwAGdO3dOd+/eVWBgoDJlyqQiRYqofv36atGihdtnepWkkydPatWqVcqZM6f9i8Ht27c1ceJE7dixQ4GBgapdu7Y6deqkVKlSPXB7/y2UnT59WmPHjtULL7zgEQU9yfk5/1tiCgjuEBQUFOeXCZvNpg8++EBBQUHq2bOngoODE73tQoUK6YMPPpAk3bhxw/5HYpo0aZIdd3I58xgfPHjQXqz84Ycf9Oeff8rX11dPPvmkqlSpokGDBumpp56KdauxO7nyPe1pRo4cqV9++UXt27eXzWZTnz59lC1bNtWsWVPfffedvd2wYcP05JNPavTo0friiy9SMOLkc8WPa54yXIkjwsPDtWnTJl26dCnJ27h3756CgoJiLQ8NDZX0z/UspTn7PD506JD69evn0L7Lli2riRMnJit+R+XKlUtHjx61P8+ZM6eOHj2q7Nmzx2p7/PjxOJcnJGPGjHrjjTf0xhtvJDtWZ3P2Mf7zzz8dnvm3SJEibr8West7+t+clfO1a9eUI0cOh/aZKVMmXb582SnxJ4czrtWe6qmnntKXX36pqVOnys/PTzabTXnz5o3VzhijZcuWJeoHfE/9PM6bN682b94cbwFz5syZun//vrp27apnn33W4e126dJFDRs21IQJEzR79mzNnj1bHTt21IABA+IcliolOPva1bdvXxUtWlRbtmzRvXv3tGfPHoeGRnEXZ+ebMWNGTZw4UT179tT9+/c1ZcoUjyvkeWPOTpPCPUDx/7300ksmMDDQjBo1yowZM8ZkyJDBZMqUybzwwgsx2t2+fdsUKFDAPPfccw/c5oULF0ypUqWMZVnm8ccfN3Xr1jWtWrUy7dq1M61atTJ169Y1GTNmNJZlmRIlSjxw4Hln27t3r0mTJo39doTGjRub+/fvmzJlyhibzWayZ89uQkJCjGVZpmzZsubmzZuJ3oen3PoRzdU5e1q+lStXNvXq1Uuwzcsvv2z8/PxMw4YNPfb258Rw9jGO3k7BggXNqFGjzPr16x84gZE7ueM8TsjAgQPd+r7JmTOnadu2rf35O++8E+/tp3379jVZs2Z94DZHjhxpRo8ebQ4cOODUWJ3h8OHDJkeOHMZmsxkfHx8zePBgM2/ePGNZlmnevLmZPXu2mT59uilZsqTx9fU1W7duTfY+o6KizLlz55z+XnFUYmaFT8w233jjDbNr164Yj9WrVxvLssysWbNirXMnV5zHAQEB9tv7HmTWrFnG398/uWk4pGPHjiZr1qwxhoCIS1RUlMmbN6/Tbo29f/9+it6y6IpjHBwcHGMomIRMnTrVBAUFJTcNh3nTezqaM3POli2bGTp0qEP7HTRokMmSJYuz0nCYK67V/+Up36v37NljXn/9dXPkyJEE233++ecmXbp05uOPP3ZPYC40fPhw4+PjYw4dOpRgu1atWjl8a/t/3zNnzpwxPXv2NIGBgcbf39906dLF/PHHH8mOPTlc9b36m2++MSVLljRFixY1a9ascXEWjnN2vtWrVzc1atQw1atXNzabzfj6+poaNWrYH57AG3N2JgqYHiJLlizm5Zdftj9fsGCBsSzLzJs3L1bbQYMGmYwZMz5wm+3btzfp0qV74FhL3333ncmQIYNp37594gNPhubNm5vQ0FDz7bffmk2bNpkCBQqYDh06mFy5ctlnqY6KirIXCPr165fofXjKF49ors7Z0/J1xZcPY/4Z/+bzzz83rVu3NiVKlDBZs2Y16dKlM9myZTPFixc3LVu2NJ9++ql9dl93cvYxbty4sX1GuuzZs5tmzZqZmTNnmp07d5qIiAh3pJQgd5zHCXF3ATNVqlQx/mA/evRoguNQ+fn5PXCblmUZX19f4+PjY5o2bWr27Nnj1JiTwxU/riVkxIgRJkOGDPaC6VNPPeX2Yp6rCpgJjfH57/UpMZatK87jfPnymZ49ezq0/5dfftltY7u6shCwfft207p1a9OoUSMzd+5c+/JBgwYZf39/Y7PZTJ48eczq1auTGn6SueIYlyxZ0uHZjps2bWqKFi2arBwSw5ve09GcmXPTpk1NoUKFHvg94969eyZPnjzJvvYnhSuu1Yn5kSklfmxyhRo1asT5HWbbtm2mZcuWpmzZsqZ27dpm6tSpbv0B/datW+bw4cPm0qVLCba7f/++mThxohkxYsQDt/nf90y0fxcy/fz8TLt27R74GeEqzr525c6d24SFhZncuXPbi1vRz8PCwtyRUoKcne+mTZvMpk2bzKeffmrSpElj0qVLZz7//HP7ck/gjTk7EwVMD+Hn5xdjcOY//vjDWJZlvvnmm1ht33zzTePr6/vAbWbKlMkMHjzYof0PHjzYZMqUyfGAnSBbtmymf//+9ueLFy82NpvNTJkyJVbbZs2aJeki62kFPVfn7Gn5uuLLh6f3LHbVMf7999/NvHnzTNeuXU2RIkWMzWYzQUFBpkaNGuaNN94wq1evNlevXnVaHo5yx3mcEHcXMBPzB5Ojg8pblmVmzJhhpk+fbrJly2ZsNptp2LCh2bFjh1NjTwpX/LhmjDFZs2Y17dq1i7Fs7NixxrIsU6VKFTNmzBjTp08fExYWZoKDg83+/fuTlUdiuOKP4rlz5yb64U6uOI9ffvllExoaai5cuJBgu7///tukTZs21vvhYbNr1y7j7+9v/P39Td68eY2vr6959dVXzVtvvWUsyzKNGzc2ffv2NQULFjSpUqUyP//8s1vjc8UxHj16tEO5bN261fj4+JiBAwcmPvAk8sb3tDNz/v77741lWaZt27bxTvZx584d06JFC2Oz2VLke6e7fmzyhAltXCmuot769evtE2KWLFnS5MmTx1iWZYoXL26uXbuWQpEmX3wFzGj/LmT6+vrGOfmeqzn72vXXX3+Zv/76y6xevdpkyZLFZMuWzaxdu9a+PKW56u+Ixo0bm7Fjx5px48bF+tE9pXljzs7EGJgeIiIiIsZU99H/jmtMQz8/P0VFRT1wmzdv3nR4oPhMmTLp5s2bDkbrHJcuXdLjjz9uf16iRAkZY5QvX75YbStWrKjly5e7MzyXcHbOu3fvjvH8/PnzkqQ//vgj1rpopUqVSmzYSZY6dWoVLFjwge18fX3Vv39/h7bZr18/HTt2TOvWrVOtWrXibbd+/Xo1b95cAwYM0Mcff+xwzMnlqvd13rx5lTdvXrVt21aSdPnyZf3444/68ccftXnzZk2bNk3h4eEqUqSIWydJcEW+DRs2dHj/hw4dcritJwsMDFTXrl3VvXt3zZ49W5MmTVL58uVVp04dDR06VJUqVUqRuC5fvhxjLK0KFSpI+r+xG/8te/bsunr1qsPbLVmyZIxl06ZNU926dbV69Wr7stGjR6t06dIaMmSIli1bloQMkua7776zfyZeu3ZNlmVp6dKlOnz4cKy2lmXFmpznvzx1jLForjiP+/Xrp/nz5+vpp5/W559/Huc4n/v27VPr1q0VHh7u8NiCnmrEiBHKlCmTtm3bpscee0x//vmnnn/+ec2fP19du3bV7NmzJf0zmULJkiU1atQorVixwm3xueIY9+zZUx9//LHq1KmjCRMmqF27dgoICLCvv3Pnjj744AO98cYbypo1a6xxyl3JG9/Tzsy5atWqeuONNzRmzBht3rxZbdq0UalSpZQmTRpdv35dO3fu1Lx583T69Gn16NEjxcaZd/a12p3fFxNj1KhRql+/fqzv8NeuXdPChQv1xx9/KG3atHrmmWdUunTpZO+vf//+Cg0N1datW5UnTx5J0po1a9SkSRONGDFC06ZNS/Y+HmTJkiVq0qSJQ20jIyP16quv2q+zSZU1a1bNmDFDAwcO1IQJE/Tee+8la3tJ4exrV65cuSRJbdu21ciRI2VZlsaNG6eNGzc6N/AkcsW1+uuvv9aePXu0cOFCSVLhwoW1bNkyPf/8884LPBm8MWdnooDp4ZIzKUvRokW1ZMkS9ejRQzabLd525v8P+OzumXIzZcqkv//+2/48R44cev/991WmTJlYba9du6aMGTM+cJv//UB1xpcZZ3J2zmXKlInzPdKrV69Yy4wxsizLbbM1u8rKlSv1yiuvJFi8lKRatWqpe/fuev/9990U2T9c8b6OS4YMGVSgQAGdP39eZ86c0fHjx3XixAkdPHgwybEnhSvy3bdvX6Kuff/+EuAOX3zxhQ4cOCBJunXrlizL0uzZs7VmzZoY7ZJyLAICAtSrV68YhcwqVaqoevXqGjp0qKpXr+6MFBzmih/XpH/eN9evX7c/v3r1qi5fvqzWrVvHaJc6dWq1bt1akydPTkr4SbZ48WItXrw4xrIPP/wwzrbu/hxxBVecx2FhYVq4cKFatWqlokWLqmLFiipdunSMAsj27dvl5+enuXPn6oknnnBqTvFxVSFg586d6ty5s32iq7CwMPXu3VvdunXTiBEj7O1Sp06tVq1aacaMGc5Ix2GuOMYhISFau3atmjRpopdfflk9e/ZUoUKF7Mf48OHDunfvnvLnz6+vvvoqzh8+XMWb3tPRnJ3zqFGjVLx4cQ0fPlxjxoyxX/eNMZL++WH1448/TtEfaJx9rfbUH5tGjBihrFmzxrhu/fnnn6patapOnz5tXzZs2DC1b98+3v8DR9y9e1e//PKLhg4dai9eStIzzzyjNm3a6KuvvnJLAbN58+aaPn26evTokWC7S5cu6YUXXtDWrVuTXcCM9u9Cpru54tr1wQcf6Nq1a+ratauMMXrnnXf0wQcfqHPnzk6NPSlckW+6dOk0Z84c++Q3c+bMSZHJiuPjjTk7k2WiP4WQomw2m4oXL27/Q/zu3bv67rvvVK5cuVizC584cUL79+9/YCFq+fLleuGFF1SsWDG1bdtWhQoVUkhIiAIDAxUREaErV65ox44dWrJkiQ4cOKClS5eqQYMGLsvxv5o0aaKjR4861FuscuXK8vX11ffff59gu4QKtXFxd0HP2TnPmzcv0TG488uZo7OTRrMs64FfulKnTq1x48bFWaT9rxkzZmjw4MG6fft2ouJIDle8ryUpKipKu3fv1pYtW+wzkl+4cEHGGGXJkkVVq1ZVtWrVVK1aNbf+0eSqfD2VK64xNptNs2fPVteuXWOtu3v3rr2Qee7cOVWqVElvvPGG6tatm6g4kuq/sZ0+fVo5c+bUypUrY836OWfOHL3yyisOXVNbtmypI0eO2HuK379/X6lTp9bnn3+upk2bxmg7duxYTZgwQTdu3HBSVgk7fvx4ol8T3cMhPjVr1lSfPn1Uv379GMu3b9+umTNn2otmzz77rLp3767UqVMnOobkcOV5fPz4cU2aNEkrVqzQqVOn7MuzZs2qevXqqV+/fg711HeWuM63uAoBlmUlqhCQNm1ajRgxIkaBpH79+oqMjIzRq1iSJk+erKFDh+ru3bvJzMZxrjzGUVFRWrJkiZYvX65Dhw7p+vXrCgkJUcGCBVWvXj21aNFCfn5+yU0hUbzpPR3NlTkfO3ZMv/76a4xjW6BAgeSGnCyuuFZ7qriuW88//7zWrVun9957Tw0bNtS1a9f09ttva/LkyQ4V/uLb9tmzZ5U9e3YtXLhQzZs3j9F25syZGjBggFuuXR07dtS8efPUu3dvTZkyJc42+/bt0/PPP68TJ05ozJgxGjRoUILbnDdvnipVqqT8+fO7ImSncMV5fPDgQYWHh9sL4Lt375a/v7/bf2SJi7f9HSF5Z85O5e571hG3XLlymdy5cyfq4Yhvv/3WlCtXLs4xXKInDihXrpz59ttvXZxhbEuXLjVlypQx+/btS7Bd9Dg8Y8aMeeA2o8fzSMzDnVyRsycLCgoywcHB9kdQUJCx2WwmMDAwxvLoR0hIyAO3WbZsWfPUU0+ZyMjIBNtFRUWZ6tWrm7JlyzorHYc4+xgPHz7c1KpVywQHB9vP2xw5cphWrVqZOXPmmMOHDzsz/ETztve0KzxoTCZj/hlrbPr06SZr1qxuH/OzRIkSpkGDBqZBgwbm6aefNpZlmfLly9uXRT9KlCjhcGzbtm0zNpvNvPrqq/bx1UqVKhXr/XHz5k2TP39+U7NmTafn5k6ePsaYu87jGzdumNOnT5vr168n6fXOENexaNiwoQkMDDSffPKJuXbtmjlx4oQZMGCAsdlsZubMmQ5tt1y5cjFmLN++fbux2WymWbNmsdo+/fTTplChQslLJJG87VrtTe/paN52jF3BUye0+e91KyIiwqRKlcp07949VtvGjRsnasIsy7LM9OnTzaVLl8ylS5fMoUOHjGVZZtGiRbHavvHGGyZ79uxJSyIJRo8ebWw2m2nRooW5d+9ejHWLFy+2/+3w9ddfuy0mV/O289jb8jXGO3N2JgqYXuLChQvm+++/N4sWLTKffPKJWbRokdm0aZM5f/58Sof2QKdOnTJ79uwxN27cSOlQ3MYZOUdFRZlz586ZmzdvOjGy5Dl58mSyB1hftmyZ8fHxMU8++aSZNm2aWbVqldmyZYvZuXOn2b59u1m9erUZNWqUKVGihPHx8THLly93YgbO4+gxtizL5MqVy7Rt29Z8+OGH5vfff3dThM7lrPP4woULpkaNGmb79u1OiizlOVLAjHbnzp0Ys6C7mqt+XDPmn1nao2dlHj58uOncubMpWLCg2bVrl9m8ebN55513TOHChY2/v7/58ccfXZhl4hw9etTs3Lkz1h9TCYnrGJcuXdpky5bN/PHHH/Zlq1evNqlTpza9e/d2WrzO5KzzOCU/n1xVCPjss8+MZVmmefPmZujQoSZz5symQIECpkqVKqZz587m0KFD5rfffjM9evQwNpvNjB492mk5OZOrvnPdvHnTHD9+3KnbdIZH4T2dWI7kPHLkyDhn2L569aqZPXu26devnxkzZozZuXOnK0NNFmddqz3hx6b/xnXu3DljWZaZP39+rLZTpkwxAQEBidp2XJMUxfXjTcmSJc1TTz2VtCSSaMGCBSYgIMBUq1bN/v89bNgwY7PZTO7cuR9YBHpUOfNaff/+fY+vCbjqs+nixYvm+++/d+o2nSWxOUdGRprPP//ctG7d2pQoUcJkzZrVpEuXzmTLls0UL17ctGzZ0nz66afm/v37Lo7ctShgIsXE9+XIXVLiy7S7ch4xYoTJkCGDsdlsxsfHxzz11FMp+n8dzVmzpHtyz2JnH2NPmCEwIe4+j6PfQ6tWrXLbPl2tevXqcfb48AY//vijqV27tvHx8bGfu/8+l8uWLWu2bt3q1ph27NhhKlasaEaNGhVj+cWLF+3XHZvNZjJlymSWLl3q0Db/+8fnnTt3jM1mM8OHD4/Vtlu3biZXrlzJyCDx3Hkep/TnkysLAUOHDjWBgYHGsixTpEgRc/jwYXPs2DGTJ0+eGO/runXrJqqo4gyuOsZnz54148aNM927dzejRo2Kt5gwe/Zst/Ye96b3dDRn5hxXIe/YsWPmscces1+ro6/XHTt2dMo+E8sd12pjPOPHJsuyzHvvvWd/fubMGWNZVpx5jR8/3mTMmNHhbY8YMSLOx3+/Q69atSrFeoNt3rzZhIaGmiJFipjnn3/eWJZlqlatai5cuOCyfe7cudN06NDBZduPj6uuXdu3bzetW7c2jRo1MnPnzrUvHzRokPH397f/oLx69Wqn7zshKV0PiJ4B3J1ckfOFCxdMqVKljGVZ5vHHHzd169Y1rVq1Mu3atTOtWrUydevWNRkzZrTfWXXu3Dmn7t+dKGA+BG7cuGF++ukns2/fPhMVFZWkbRw9etQsXbrUzJ4928yYMcPMmTPHfPXVVyl6+2lieh0lhqd+mTbG+TlnzZrVtGvXLsaysWPHGsuyTJUqVcyYMWNMnz59TFhYmAkODjb79+932r6TwlkFzGie2LPYVe9rT+WKfENCQuJ9BAcHG8uyTOrUqe3L0qRJ49T9J6RGjRqJejzstz67y5UrV+zn8oIFC8zKlSvNiRMnUiSWvn37Gj8/v1g/cL300kvGsizTunVr8+abb5oqVaoYf39/c/To0Qdu87/nSfQfn1988UWstjNmzDD+/v7JTyQRXHEee+rnkysLAcb8U5z+72fQ7du3zfz5882ECRPc/sdhNFcc4+PHj5vMmTPHKmi1a9cuVk9Ed3/n8qb3dDRn5uyqoRacyR3Xak/5scmyLPPYY4+ZYsWKmWLFipknnnjC2Gy2GNeyaM8++6wpWbKkW+Jyp99++83kzZvX2Gw207ZtW5f/CJQShS1jXHPt2rVrl/H39zf+/v4mb968xtfX17z66qvmrbfeMpZlmcaNG5u+ffuaggULmlSpUpmff/7ZqftPSEr/3ZQSx9kVObdv396kS5fOfPfddwm2++6770yGDBlM+/btnbp/d3o0pyZ6CEVGRqpv37764osv9Nhjj2natGmqUqWKPvnkE7366qu6deuWpH8GCJ85c2asSQ7i89VXX2nIkCE6cuSIfdbAf7MsS3ny5NGYMWPUokULp+aUEk6cOKGyZcvqwoUL9mUjRoxQmzZt9PbbbysoKCgFo3O+y5cvq2TJkjGWTZs2TXXr1o0xacDo0aNVunRpDRkyRMuWLXN3mC6TMWNGVa1aNaXDcIsVK1Zo6dKlOnjwoK5evSp/f39lzpxZJUuW1AsvvKCKFSumdIhOkz59ep08eVK1a9dWjhw5Yqy7ffu2vvzyS1WqVMk+2687nT9/3qEZ0s+dO6eLFy8majb1h83Nmzf166+/KjAwUEWLFk1WrunSpfOYc3n9+vWqWbNmjNntb9y4oS+//FLlypXTp59+Kknq3r27nnjiCQ0fPlwLFix44HZv376ty5cvS/pn1nUp7tncL1686NbZml3Fkz+fRo0apTfffFPSPxPQWJYV43tDtM2bNytnzpyJ2nZAQIACAgJiLAsMDFSbNm2SHrCHGjJkiG7fvq0VK1aoZs2aOnXqlGbPnq1Zs2Zp//79WrVqlbJkyZLSYTqNJ7+nXS0yMlJr1qxRx44d9dJLL0mS0qRJowkTJujIkSN6//33HZ40xllcda3+t6tXr8oYoyJFisRaV7hwYc2dOzdZOTiqatWqsT4vMmXKFOtvu4MHD+rbb7/Va6+95pa43KlAgQL66aef9Pzzz2vnzp06e/Zsoq/P0RMHOuLYsWOJDdFjjRgxQpkyZdK2bdv02GOP6c8//9Tzzz+v+fPnq2vXrvbZ20eNGqWSJUtq1KhRWrFiRQpHnXTFihVz+Dvp9evXXRyNe6xcuVKvvPKKatWqlWC7WrVqqXv37nr//ffdFJnzUcD0EDNnztSsWbNUsWJF+fv7q379+lq5cqW6dOmiRo0aqVatWrpy5YrmzJmjF198URkyZFDNmjUT3OYXX3yhF198UVWqVFH//v1VuHBhhYaGys/PT/fv39elS5f0888/a/HixWrVqpWMMWrZsqWbMnYNb/synSlTphgX3qtXr+ry5ctq3bp1jHapU6dW69atNXnyZHeH6FYXL15U8+bNNX78eJUvXz6lw3GK8PBwNWrUSGvXrlW6dOlUtGhR5cuXT9u3b9elS5e0fft2TZs2TbVr19Ynn3yizJkzp3TIybZr1y41adJEBw4c0IgRI1SpUiX7ulOnTunLL79U7969Va9ePbfHduDAgQTX//bbbxo3bpw+//xzBQcH6+WXX37gNj15hmpX/bgWn0uXLmn+/Pk6evSogoODVadOHdWuXdsZqTjs2LFjaty4cYxl27dv171792K853x9fdW0aVN99NFHDm23b9++6tu3r/25ZVk6c+ZMrHYrV65Unjx5khi95/DUz6eUKAQ8ip9NkrRlyxZ16dJFzz33nCQpf/78mjp1qurXr68mTZqoZs2a2rRpkzJlypTCkTqHp76n3eHSpUu6f/9+jM/jaJUrV45RwHUXV12rPfHHpk2bNjnULm3atNqwYUOcBdf4jBo1SjabTY0bN/aIWamjhYSExPn/HhUVpdu3byt//vxKlSqVfbllWbp27VqC2yxTpozDhS1jzCPzI/TOnTvVuXNn+w//YWFh6t27t7p166YRI0bY26VOnVqtWrXSjBkzUiZQJ6lTp47eeecd3bt3T0888USC52lUVJQbI3OdmzdvOvxZmylTJt28edPFEbkOBUwPMW/ePD399NP69ttvJUm9evVS8+bN1bRpU/sviJL0yiuvqGjRoho3btwDC5gTJkxQ7dq1tXbt2jjX58+fXxUqVFCPHj1Ut25dTZw40e0FzMuXL+vEiRMOt//3r6xxeRi+TDsz56eeekpLly7V0KFDJUlBQUHy8fGJ1ftD+ueD3VMu0q76QhAeHq5NmzbZv3imFGce49GjR2vt2rV655131LlzZ/n4+EiSxo8fryVLlujEiRNasmSJ+vTpo1q1amnHjh1xHn9XcvZ5nDFjRq1fv149e/ZUzZo19eabb6pLly6SXPfeSa79+/drzJgx+uqrrxQSEqJBgwapV69eSp8+/QNfu2nTpljX3g0bNujZZ5+VZVkqUqSIjh07ptdff13z5s3Tli1blCZNGlelEoMrflyTpLfeekuhoaF68cUX7csOHTqk6tWr68KFC/L19VVERISmTp2q5s2b67PPPnPbsb9z544yZMgQY9mePXtkWZbKli0bY3nWrFl16dKlB25z+PDhcS4vVKhQjOerV6/Wnj17NHr06ERGnXzOPo899fPJlYWA+DyKn02SdOHCBYWFhcVaXqNGDX3//feqU6eO/XtXSvCW9/S/OTPnf19zo3MJCQmJ1e7+/fsKDg5ORJTO4YprtfRw/9j02GOPJfrulBEjRsjHx0cjRoxQ48aN9cYbb6hEiRIuitBxTZo0cfrn/t69e9WmTRvt379fPXv2TDDPHTt26N1333Xq/h3l7GvXrVu3Yn0fXbJkiWrVqqWsWbPGWJ46dWrduXPH8WCdwNn5Tp06Vfny5dNrr72mCRMmJNjhYfHixSlyF6qzcy5atKiWLFmiHj16yGazxdvOGKNly5apaNGiDu/b46Tg7ev4l+DgYDNt2jT78127dhnLssyCBQtitR0yZIhDY74FBAQ4PCbNjBkzEjVQvTPEN+tdQo8HSZ06tZk1a1ac6/bt22eyZs1qihYtai5evJhiY2A6M+dt27YZm81mXn31VRMeHm6MMaZUqVKxBti+efOmyZ8/v9vH4wsODn7g+IX/fjjyvvbk8RGNcf4xDgsLMy+99FKs5Tt27DA2m80+yc9ff/1l0qRJY8aPH++SvOLjivP43z744APj7+9vXn75ZXP//n2nj6OaXD/99JNp2LChfaKAcePGmevXrydqG546aYAxxhQvXtzUqVPH/rxnz54mW7ZspnXr1jHa3bhxw+TKlcvUqlXLoe0GBASYYcOGxVhWrVo1kyFDBvv4PTdu3DBvv/228fX1NRMnTkxmJo7Lnj276d+/f4xlTZs2NTabzZw5cybG8mHDhpng4GC3xeYqrjiPPf3zydm87bPJGGMKFSpkevbsGe/6gwcPmixZspgKFSqYqVOnun0MTG97Tzsz54dhzEVXXKs9eUKby5cvm+HDh5sGDRqYBg0amAkTJphbt24le7uWZZkZM2aY6dOnm2zZshmbzWYaNmxoduzY4YSoPc+6deuMzWZ74PfIlBwD09nXrnLlypn69evbn2/fvt3YbDbTrFmzWG2ffvppU6hQIafmlBBX/R1x6NAhh/5eWLx4sbEsyxmpOMwVOS9btsz4+PiYJ5980kybNs2sWrXKbNmyxezcudNs377drF692owaNcqUKFHC+Pj4mOXLl7shU9egB6aHMP+5dSlbtmzKkSNHrF9FpH96J/23fVzSpUun06dPO7T/s2fPKm3atI4F60RNmjRx6u1Ujz/+uP7444841xUrVsw+Xk79+vXVrFkzp+03MZyZc4UKFTR9+nT17t1bq1evVps2bVSqVCl98sknevbZZ3Xr1i0dOHBAb775pk6cOKF58+Y5Zb+OcsWvp548PmI0Zx7jv//+W6VKlYq1PDQ0VMYYHTt2TLly5VKuXLnUqlUrLVq0SAMHDnTKvh3l7PP43zp16qQnnnhCTZo00cGDBz3mtpbNmzdrzJgxWr9+vbJmzaopU6aoe/fuCgwMTPa27969q19++UVDhw6N0bvjmWeeUZs2bfTVV19p2rRpyd6PI44dO6b27dvbn7dt21azZs2K9Wt2cHCwXnrpJfu4gg+SKlUqZcyY0f48IiJCW7Zs0ciRI+3j9wQHB+uVV17Rzp079cEHH6h///7JT8gBNWvW1IIFCzRo0CClS5dO58+f17fffqsSJUrE+kzeunWrcuXK5Za4XM3Z57Enfz5duXJFM2fOtI+HVrlyZf3vf/9L1vAM3vbZJEnFixe3X49scfT4KFKkiNatW6fq1atr5MiRTtuvo7zpPR3NWTk/DGMuuuJaHV9v+f969tln3drD9uzZs6pUqZL++usvBQYGKlWqVPrmm2/0zjvvaN26dSpQoECyth8YGKiuXbuqe/fumj17tiZNmqTy5curTp06Gjp0aJxDB3iSW7duadmyZWrVqtUD2xYoUECPP/74A6/3QUFBD+z15irOvnb16tVLrVu3VosWLVSwYEHNmTNH+fLl09mzZ9WlSxf17dtXNptNb7/9ttavX+/267Ur/o7IkyePNm7cqGLFiiXYrlq1atq4caNT9+0IZ+fcsGFDrVq1SkOHDtXrr78eawiE6Gt32bJltWrVKtWpU8dp+3a7lKye4v+ULl06Vo+W+HTs2NE88cQTD2zXrVs3ExwcbBYvXhzv7OX37t0zixcvNmnSpDHdunVLVMzJ5YoZuJo3b25y5sxpIiMj422zb98+kyFDBpMmTRq3/7LmqpnWfvzxR1O7dm3j4+MTYyZQm81mLMsyZcuWNVu3bnX6flPChQsXTNWqVU22bNnMDz/8EGPdyZMnU7x3nrOPcYECBUzbtm1jLV+xYoWx2Wxm165d9mWzZs0yISEhTtu3I9w1e+Dp06dNhQoVTGhoqEO/nLvKmjVrTJUqVYxlWSZ37tzm3XfftffESSpPnqE6KCgoxt0Bf//9t3nsscfM+vXrY7WdPn26w++/UqVKmc6dO9ufX7582ViWZb766qtYbadNm+bWnH/77TcTFBRkcubMaZo3b25y5MhhbDabWbhwYYx2R44cMT4+PqZ79+6J2v7x48fNqlWrzGeffWaWLFlitmzZkuheu87myvPY0z6fzpw5Y8LCwuw9ItOlS2csyzKPP/64+e2335K8XW/7bDLGmE2bNpmWLVua7du3J9hu69atJjAw0O09ML3lPR0tJWbzPXnypNm0aZM5f/68W/drjOuv1Z6kW7duJiAgwCxatMj+N92mTZtM5syZTcWKFZO17bjeN3fu3InRI7NmzZpm48aNydqPs92+fdssWrTINGnSxKROnTpFeku6gqvO46FDh5rAwEBjWZYpUqSIOXz4sDl27JjJkydPjOtX3bp1XT7D+7+l9CzkKcHVOV+4cMF8//33ZtGiReaTTz4xixYtSrHrtCvQA9ND1K5dW2+++aYuXbqU4ECzt27d0tKlS+1jPCZk4sSJOnjwoJo1a6Z06dIpf/78CgkJUWBgoCIiInTlyhXt379fd+/eVaVKlTRx4kRnppQiXnnlFdlsNu3YsSPeXzWKFSum5cuX6+mnn3ZzdK5TsWJFrVu3TlevXtW+fft07tw53b9/X+nSpVOxYsUSPUufJ3sYx0dMji5dumjgwIEqXry4unTpojRp0ujAgQN6/fXXlTlz5hi/LF69elUREREpGK3rZM+eXd9//70+/vhj3b17N8UGmn/22WeVLVs2TZ48WW3atJGvr+8DB8L+7xhdcfHESQOkf8Zo3LVrl/15tmzZdPLkyTjb7t+/3+HeCk2bNtWYMWPUt29fFSpUSOnTp1fatGnjHKNs7969cd6N4CoFChTQ2rVr1bNnTy1fvlyPP/643n333VhjJL366quKiopyqMeH9M8Yzb169dKePXskxZwgwGazqWbNmho+fLjH93RJLE/7fBoxYoTOnDmjL774Qk2bNpVlWfr+++/VvHlztW/fXj/++GOStuttn03SPz1XqlWr9sB2lStX1u3bt90QkXt42ns6JSVlzEVncdW1OtqJEyd08OBBXb16Vf7+/sqcObNKlCgR5zigrrZmzRq9+OKLMe4eq1atmsaPH68uXbrozz//jHM82qQKCAhQr169YvTIrFWrlipVqqQ33nhDdevWddq+EiM8PFyrVq3SF198oZUrV+r27dsKDAxUvXr11KRJkxSJ6WExatQoDR48WDdu3IgxF8SBAwe0ePFi/f333ypRooSeeeaZFIwSzpAxY0ZVrVo1pcNwnZSuoOIf58+fN9988405e/Zsgu2WL19uGjVqZL7//nuHt71q1Srz2muvmRo1apgiRYqYvHnzmiJFipjq1aubV1991axatSq54ScJv7h4l99//92MGzfOtG7d2jzzzDOmTZs2ZurUqebkyZNJ3qYnjo/o7GMcFRVlXnrpJXsPD39/f2Oz2UxgYKBZs2ZNjLZffPFFvGPAuoq3vaf/29vGGePWxLe9uMYwLlmypHnqqadckVqcBgwYYFKnTm0uXryYYLubN2+a9OnTxzlea1zu3LljihQpYkJDQ817771nwsPDzcsvv2xatWoVo90HH3xgfH19zeDBg5Ocg6tcvXrVXL161aG2P/zwg/Hz8zP58+c306ZNMxs2bDA//fSTeeaZZ0ytWrXMuHHjTPHixY2Pj4/DY1c7kzedx7ly5TIdOnSItfzDDz80NpvNHDt2LNn78IbPJk/nbfka4505OyIx12pjjNm8ebMpVapUjF5p0f/29fU1derUidXL2tVSpUpl3nzzzVjLjxw5YizLivOuCEc58r6J7pGZNWtWt/d0vHfvnlm+fLlp3bq1/e65gIAAY7PZzPDhw83t27fdGo+redt57Ip8R44caUaPHm0OHDjg1O06S0of499++83MmzcvxfafXPTA9BCZMmVyqFdlgwYN1KBBg0Rt+9lnn9Wzzz6b1NBcplq1asqePXtKh+FW7sz55s2b+vXXXxUYGKiiRYumWA+QiIgIvf7663r77bcVGRmpwMBApU+fXlevXtWnn36qgQMH6vXXX9fo0aPtM2w7yhPHR3T2MbYsS5988ok6duyob7/9VleuXFGuXLn04osvxvq1vXnz5k7br6Nc9Z6OiorSokWL9M033+jAgQM6d+6c7t69q8DAQGXKlElFihRR/fr11aJFC/n6uu+jzNHxsZyxTU+Yobpv376qUqXKA3v2btiwQdWqVbP3OHuQgIAAbdq0SW3btlW3bt30v//9T9mzZ9eZM2fUoEED3b59W7/++qvOnTunevXqueT/Pbmix40eOnToA4/JyJEjlTNnTv3yyy8KCgqyL3/55ZfVqlUrLV++XP3799fw4cPVp08flS9f3mXjysbF3Z/HKfn5dObMmTjHFa5SpYqMMU7pyeQNn03RIiMjtXHjRt26dUs1atRQmjRpJP0zG/SsWbN07tw5lS9fXr169bKvc4eU/o555MgRbd++XW3btnXbPp2d89KlS7Vo0SLlzJnTftfH/v379dprr2nHjh0KDAxU7dq1NXbsWOXLl89p+3W2xFyrf/zxR9WqVUu5c+fWlClT9OSTTyooKEjDhw/X/fv3VatWLS1cuFBVq1bVtGnT1KNHD3ekoPv37ytVqlSxlkeP43j37l2X7v/fPTLnzJnj0n1FW716tRYtWqRly5bp6tWrCg4OVoMGDdSsWTPly5dPxYsXV/HixZ0y9vi8efM0bdo0HT16VMHBwapbt67Gjh2bImNguvra5Uk9iyXX5DtixAj5+PhoxIgRaty4sd54440EZ5x3N1cd4++++06TJ0/WiRMnlD17djVu3FidO3dWQEBAjHYbN27UK6+84tbPJ6dK6Qoq/lGjRg2zYsWKlA7jkRAREWHWrVtnvv76a3Pt2jX78l9++cV06NDB1KtXz4wcOTLGuodVRESE6dmzp8maNaspU6aM2bx5szHGmPnz55uQkBD7L8bZs2c3X375ZYrE+OKLLxrLskzHjh3N/v37Y6w7ePCg6datm7EsK85xHh3lKeMjwjkuXLhgSpUqZR+Xrm7duqZVq1amXbt2plWrVqZu3bomY8aMxrIsU6JECXPu3LmUDtmt7ty5k9IhONXWrVtN79697XcJ5M+f35QtW9Z07NjRo8/le/fumVatWjnUGyU4ONiMHj061vLo3jP/nu21QoUKcc4M+rDx1M+n+Ho+uKKX5KP+2XTt2jVTpkwZey+1dOnSmbVr15pdu3aZ1KlTm1SpUpnQ0FBjWZZ54oknHonvXevWrTN16tQxhQoVMjVr1jRvvvlmnNfk2bNnP9Rj8n399dfGsiwTEBBgAgMDTcGCBc3p06dNlixZTFhYmGnVqpWpV6+e/Rj//vvvKR1yvBJzra5Tp47JkyePuXnzZozly5YtM0FBQebWrVsmIiLCDBkyxPj4+Dxw/FdnsSzLNGvWzEydOjXGY9iwYcZms5nOnTvHWP7vsasfpHr16h75N6hlWSZNmjTmxRdfNEuXLjV37961r/v999+NZVlmyZIlidpm+fLlTc+ePWMs++CDD4xlWSZv3rymc+fOpkmTJiZt2rQma9as5vjx485IxSN4Ys9iV7Esy8yYMSPGOK4NGzaM8V3rUbN161bj4+NjQkJCTLly5UyOHDmMZVmmcOHC5uDBgzHaPuyfT/TA9BCbNm1Sy5Yt7c/PnTunihUr6qOPPlL16tWTtE2bzSabzaamTZvqjTfeUNGiRZ0UrXN07NgxUe0ty9KHH36YYJvr16+rVq1a2r17t4wxSps2rRYtWqTQ0FB7L6KQkBD7r3o//vijW3sEODvnmTNnatasWapYsaL8/f1Vv359rVy5Ul26dFGjRo1Uq1YtXblyRXPmzNGLL76oDBkyqGbNmslNw2ErVqzQwoULNWbMGA0ePDjW+iJFimj27NnKkyePBg0apGbNmql+/fqJ3o+njI8oueZ97aiU6PHhinz79eunY8eOad26dfYZqeOyfv16NW/eXAMGDNDHH3+cqDiS6t69e3rttdeUK1cuDRkyRNI/Y1b26dMnVtuAgAC98847D9xmz549NXPmTIf2f+rUKT3//PMxxqV0pZo1a6pPnz5JOi8dVblyZVWuXNll20+Kn3/+WRMnTtThw4eVJk0aPfXUU+rTp4+yZcsm6Z9j3qhRI23evDnOY/9fNpstzhlr7927J+mf8a2jNWjQQG+//baTMnGMK85jT/58+u6772KNXXvt2jVZlqWlS5fq8OHD9uWWZal3795J2s+j/tk0fvx47d27VxMmTFDFihW1YcMGtWnTRjly5FDevHm1YcMGZcyYURs2bFD9+vU1cuRITZ06NTlpOMwV+f7www965plnlDp1ahUuXFi//fabevTooXfeeUeLFy9WkSJFkhNysjkz5+nTp6tw4cLatm2bUqVKpcaNG+uFF15Q3rx5tX79enuPnn379qly5coaNmyYFixYkOwcEsvZ1+off/xRAwYMiNFTXpIKFy5svzOgTJkyGjNmjNavX6+pU6dq0aJFLsntvxYvXqzFixfHue6/xzEx162UmIHZUREREbp3757u37+vyMjIZG9v9+7dse5WGj58uMqXL6/vv//e3sv1zJkzKleunAYPHqxPP/002ftNDFdcuzy1Z7Hkur+bAgMD1bVr1xjjuJYvX1516tTR0KFDU3S8cVfkPHz4cD3++OPatm2bsmTJIumfv8O7deump556St98882jM8Z6SldQ8Y//9ghwRk+A6JnEon95aNy4cYwZi1OaZVn2XwqCg4Mf+HBkdtuBAwcaPz8/M2nSJLNlyxYzcuRIkyVLFlOqVClTrFgxc+HCBWOMMevXrzeBgYGmT58+rk4zBmfnXLx4cVOnTh378549e5ps2bLFmtH+xo0bJleuXKZWrVouySs+TZs2Nfny5bPPmJiQIkWKmEaNGrkhKtdyxfvak3t8uCLfTJkyOTzm4eDBg02mTJmSm4bDPvnkE2Oz2czy5cvty6Kv14GBgTFytdlsZt26dQ/cpmVZpmnTpjF6F8Tlhx9+MFmyZHHrjNz//Ww6e/asCQsLc+lspFFRUebcuXOxesC4y/bt242/v7/x9/c3JUuWNKVLlzb+/v4mR44c5vfffzfHjh0zhQoVMn5+fmb27NkObbNhw4YmX7585saNGzGW9+3b19hsthhjAc+cOdOkSpXKqTk9iCvOY0/9fIoex9bRx8PcS+HfXHGMCxQoYNq3bx9j2bRp04zNZjObNm2Ksbx79+4md+7cTs0pIa7It1atWiYsLCzGePXLly832bJlM+nTp4/Re+lh/zwODQ01Y8aMsT/fsGGDsdls5v3334/V9rXXXjOZM2d2SU4JccW1Ok2aNGbkyJGxlh84cMBYlhXjfT127FiTPXt2p+WTkL/++ivRj4fd3r17zeDBg02+fPmMZVkmKCjING3a1CxatMjs3bs3ST0ws2TJYiZMmGB/fv36dWNZlvnwww9jtR06dKgJDQ1Ndh6J5Yprl6f2LDbGNfnGdadF9Diu0XWRmjVruvS77IPic8XfTnFdu86ePWtKlixp0qRJY7Zt22aMefh7YFLA9BCuKmDOmTMn1gn73HPPufXCFJ8KFSoYy7JM7dq1zTfffOOUbXryl2ljnJ9zcHBwjNtEdu3aZSzLMgsWLIjVdsiQISZNmjTJ3mdihIWFxbpVIz59+/Y1uXLlcur+L168mKgJr5zB2cfY028JcMV5HBgYaKZPn+5Q2+nTp5vAwECn7NcRjRs3NgUL/r/27jsuiuP/H/hrF+ToRSzYQT8SC2oQQYmKRtSIDTvGXrH3FguCDVtiiSVEk9g+tqhREUFj7xXEGj+JvQZ7AxWE9+8Pf9zX8yh3sHt3sO/n43GPh8wNe/N2bmaX2ZmdLzTSMuqv09LSyNXVVas/ysjKlSvJwsKC6tSpQ8+ePcswT/qmIIUKFTLod1qOc1NWwsLCqGDBgiSKIpmZmVGdOnUMfuMtICCAihYtSlevXlWn3bhxg6pXr07Vq1enIkWKkIODA+3evVvnY16+fJns7OyoRIkSNGLECJo5cya1bduWBEGgb7/9ViNv3759ycXFRbJ4dCFHOzbV85PcAwHHjx+nDRs20OHDhyklJSXDPGvXriU3NzcpwtGZHHVsbW1NCxYs0EgbPny4Vh9J9LGvNuTAvBzxmvofiFLGrFKpNDYFvHXrFgmCkOFS44ULFxr0PJxOjr7aVG82HTp0iB4+fGiQz8qKMa6riYjOnj1LY8eOJTc3NxIEgSwsLEgURZo4caJeNztbtGhBfn5+6p+TkpJIFMUMB0Jnz55tlO+1XOdjU32MjRzxZrVJzufjInXq1NHaFFVucsTs6OiodT5O9+rVK6pduzY5ODhQXFwcD2Ayacg5gJnu8wbbqFEj9TOpjOXMmTPUpUsXUqlU5O7uTkuWLMnVbnKmfDGdTsqYbWxsNP5AfPDgAZUsWTLD3Qjnz5+v0x0cKdnY2GTamX5uwYIFkl8obN682SgdtJR1bOozPoikb8fe3t5Up04dSk1NzTJfWloa1a9fn7y9vXP8WfoqVaoUDR48WCMts/568ODB9J///Een4+7bt4+cnJyoYsWKGgMmqampNGTIEBJFkTw8POjmzZu5jkEfcg1guri4UPfu3TXSZsyYQYIgUN26dWn69Ok0cuRIcnNzI1tbW63n58qpUKFCNHbsWK30nTt3kiAI5ObmlqOdLc+dO0d+fn5kZmamnk3Sr18/k9lBVep2bKrnJ7kGAt68eUP16tXTeLZYsWLFaMWKFVp580tfXb58eRo+fLj65xcvXpCTk5NWH0lENGjQICpatGiOPysnpI43L/yBKFXM7u7uNGTIEPXPqampdO3atQyPNXLkSCpXrlyuyp0TcvTVpnqzSRRFk9iZ2ljX1Z86deoUjRw5kkqXLq1e/dKyZUuddlWOjo4mQRBo9uzZ6rQKFSrQ3LlzNfJ9+PCBvLy8qGbNmpKXXxdS912mOrM4ndTx6rLLd/q4iIuLS57uq9N5enpmuZ/E69evqVatWlS0aFH1DZm8igcwTYQhBjDTfT6QWa9ePdq7d2+OP0cKDx8+pNDQUPWgzJgxY3L04GRTv5j+lBQxe3l5aS3Hy0yvXr2ocuXKOSlqjulyAkknx8W+sS+0pKhjU5/x8Smp2vH27dvJzMyMvvzyS5o3bx5FR0fTkSNH6OzZs3Ty5EmKiYmhqVOnUrVq1cjMzExjObfcLCwsaM6cORppmfXXP/zwA1laWup87CtXrpCrqyu5uLhQXFwcPXv2jPz9/UkQBGrRooXWjBBDkGsA08LCQmsgwNnZmZo0aaKRlpiYSBUqVKCWLVvm6vP0YWZmRkuXLtVKT980IDfft/3799P06dNp2rRpRr+BmBmp2rGpnp/kGgiYOHEiiaJIo0ePpujoaPr555/VG9z07dtX44ZMfumrR44cSQ4ODhQdHU3Xr1+nli1bkqWlJZUtW5aOHj2qznf69GmytrbWGgAyFKnizUt/IOY25l69epGLiwslJydnmS8tLY3KlStHzZs3z22R9SZXX22KN5v0uZ6Wk7Gvqz937NgxGjZsGJUoUULnco0ePZoEQaDatWvTihUraPjw4eTl5UVPnz6lO3fuUHR0NH399dckiqJks+NySqq+y1RnFn9Oqnj1aS9v377VebKNHKSKeeDAgWRnZ0fPnz/PNM+LFy+oatWqef7xODyAaSIEQaAGDRrQkCFDaMiQIdSrVy8SRZFatGihTvv0NXToUJ2OmVXj/Xwg09fXV8qQciQ5OZmWLl1KKpWKChQoQH/88Ydev59XLqY/lZuYx40bR9bW1vTkyZMs871584acnJyoS5cuuS2uXgRBoEmTJlFsbGy2rwkTJujUmXp4eFCVKlV0epUpU8YkOujc1HFemPHxudy2YyKi3bt3k4+Pj9ZOiek/C4JAPj4+ei0Nk4KDg4NWfaSkpNDZs2e1dtj9/vvv9V4Wm5CQQD4+PmRnZ6deKjVu3DidniMrB7kGMEuUKEFTp05V//z8+XMSBIHWrFmjlXfatGkGffxFdrtUR0dH633M5ORkCgwM1Nr9s02bNkZ71md2ctuOTfX8JNdAQIUKFahTp04aaWlpaTRlyhQSRZHat2+vHsTML3318+fPqUKFChrf66VLl9KGDRvIzMyMfHx8yNfXlywsLMjR0ZH+/vtvmSLRTW7jzYt/IOY05vj4eBo9enS2dbZ+/XpydHTMcKax3OToq9OZ2s0mOQcw8+J1dUaOHDmic94NGzaQu7u7up1+ep0pCAIVL16cNm7cKGNp9ZPbvuvTmcUjR440mZnFmcltvPXr18/wcRemLLcx37hxgyIiIuivv/7KMt+///5LAQEBVL9+/dwU16h4F3ITcuDAAa2d4KKiojLMKwiCzrvWZsbS0hLDhw/X2J3LmG7fvo0ff/xRvctWp06dUL16db2OERISgujoaPWOuUSEJUuWoGDBgqhXrx68vLxgZmaG2NhYWFtbY8qUKZLHoY/cxjxq1Cj17upZ2b9/P+rVq4e+ffvmqrw5ER4ejvDw8GzzEREEQcg2X+PGjbF06VIkJyejcuXKcHZ2zjRvRrv+Glpu69jNzQ1xcXEZvmdnZ4ddu3ahUaNGCAgIQJcuXSQpc25I0Y6Bj/XcuHFjPHnyBFeuXEFCQgLev38PlUqFIkWKoFKlSihcuLDUxc9WuXLlcOHCBY00c3NzeHl5aeU9ffo0ypYtq9fxixQpgkOHDqFTp07Ytm0bxo8fjxkzZuSqzLm1ceNGXLp0CcDH3bIFQUBERAR27dqllVfXc1OdOnWwdetWhISEAABsbGxgZmam3tn282Maui1/GnO69Nh/+uknxMTEaJQvu5jnzp2LyMhIDBgwAF27doWVlRW2bduG8PBw9O/fH2vWrJEljpySoh3nhfOTlO7cuYMBAwZopAmCgMmTJ8PV1RV9+vRBly5djLJTc0akqGNHR0fExcVh48aNePz4Mfz8/FCzZk0AgK2tLebPn4/79++jdevWCA0NRfny5SWPQ1dSxDt69GhUrVoV//77LxwdHTPM4+DggD///BM9e/bE27dvc1vsXMlNzNWqVUO1atWyzdexY0d07NgxV+XMDan76pSUFLRv3x47duwA8H/Xpq1atcLq1au1dic3JF2ukXPCVK+r7e3ts81jYWEBd3d3DBo0CJ07d842f/p1ZFBQEIKCgnDhwgXEx8cjISEBKSkpcHR0RJUqVeDr6wtzc9MYJpGi76pUqRIOHz6MYcOGYeHChUhLS4O1tTWCg4Mxf/58jbzLli2TrOw5IUW8n4+nmDopYnZzc0O/fv2yzVe0aFFER0fnqJwmw8gDqExG+t6ty2hXY0M4evQotW3blszMzMjFxYXCwsIoISEhx8dLSkqiFStW0Jw5czQ2K4qKiiJ/f3+qUKECBQUF0ZUrV6Qofo5IHbOpWrlypd4vXSxdupREUcx2FtimTZuMdqdYqjrOKzM+lPKdDgkJIRsbm2w39rh8+TJZWFjQuHHjsj1mRjOSz5w5Qw0bNiQ7Oztav3691vuGIteOzSdOnCBRFGnQoEH0/v17IiKqXr26xq63RB9n55UvX54aNGggeWyZkSNmDw8PCggI0EqfNGkSmZmZ0dOnT+UIRW9KaMeCINCyZcskP66rqyuFhIRk+v7GjRvJ3NycgoODjToDUwl1/CmlxUuknJjl6KvTn8U8cOBAOnHiBMXHx1NYWBhZWFgYfBXTpwRBoEKFCpGbm5tOr7Jly+p1fFO8rhYEgXx9falHjx6ZvoKCgqhSpUokiqJOs4Dt7e2pa9euFBUVle3jEYxNrnb85s0bevDgQaabzBmLUvqtTykxZinwAGYe8PbtW9qzZw9t2bJFry/1wYMH6cGDB3p9jiGtXbuWvL29SRRF8vLyolWrVqn/kM2vlBizHP766y+dlrFu3ryZBEEwUKk+krqOTX1JgNzf6du3b1N0dDStW7eOtmzZQkeOHKFXr15Jdnx9PXnyhAoXLkzu7u6ZDiQeO3aMypQpQ05OTjr1wZ8vkf98qfznaaa6fEtfCxcuJFEUqWzZshQaGkp9+vShL774gmJjY+nw4cO0dOlSqlixIqlUKjp+/Lixi5srVlZWGhvapLt48SIJgmD0JYpKOjfJNRDQpEkT8vLyyjLPf//7XxJFUb3s2pCMUccpKSn06NEjWT8jM0r6TqeTMuavv/6a/P396aeffsr3/2+fMtWbTYIgUNWqVal58+Y6v/RhitfV+kzEadq0KVWtWjXbfN9++y3Z29uTKIrk5OREPXv2pF27dtGHDx9yW1zJKK3vkiNeQRDIzMyMgoKCDLoJpK7kquM7d+5QRESERjtOTEykyZMnU0BAALVp04aWLl2a579PAhGRsWeBso9TnTt37ozu3btj5syZ6vRbt26hfv36uHv3LogIKpUKc+fOxeDBg7M9Zvo0cV3cvXsXrVq1QmxsbI5j0JcoijAzM0OzZs3g5+eXbX5BEDBixIhcfeaHDx/w/Plzoyw9BaSPWRRFiKKIdu3aYdKkSfDw8JCyuLlWtmxZhIeH67S8aN26dZg0aRJu3LiRbd7k5GScOHECVapUQcGCBTPN9+TJE1y+fBn16tXTq9y5YYzvtTHJFe+RI0cwfPhwxMfHI/00lb58ShRFNGjQAKGhofjqq69yF0AOnDx5Ei1atMCzZ89Qq1YteHt7w97eHi9evMDJkycRGxsLe3t7bNmyBQ0aNMj2eKtWrdK7DN27d89J0SX17t07HD16FK9evUKdOnVQpEgRvY9x4sQJTJ48GQcOHFAvTUuvZyJCjRo1MH/+fNSuXVvSsudUTmMWRREREREIDg7WSL937x5Kly6NnTt3IiAgQI4i60SOdmyq5ydRFFGlShWULl1a599JX06alV9//RWTJk3C5s2bs/y+Ll68GEOHDoUgCEhNTdW5DLklV1996tQpLFq0CImJiWjVqpW6b5owYQLmzZuHlJQUuLq6YsmSJWjSpEmu49CVXPHevXsX0dHRKFWqFJo2bQoASEpKwuzZs3HmzBlYWVmhYcOG6N27NywsLHIdhz6kjFkURbi6uuLWrVsoXrw4xo4di+Dg4Awf82GKctpXW1tbY8aMGVr/L5cuXULVqlVx6NAh1K1bV44iZymzc4hUTPG6Wp+YFy1ahHHjxiEpKSnbvO/fv0dMTAw2bdqEqKgovHnzBk5OTmjbti06dOiAr7/+GqIoShFCjsjRdzVo0AAjR45UP2It3cmTJ7Fw4UJcv34dDg4OCAgIQP/+/WFtbZ2rGPQh1/VH48aNceHCBSQkJCAwMBCTJk3K0SOt5CBHzBcuXEDdunXx+vVrCIKAwMBA/P777/D19UVcXBxcXFzw+vVrvHnzBjVq1MCBAweM+kiMXDHq8ClTGzBgAFlZWdHjx4810lu0aEGCINCECRNox44d1LFjRzIzM6O4uLhsjykIArVr147evXuXZb5jx46Ri4sLqVSqXMWgL7mWJp48eZI6d+5MrVq10liSPH78eFKpVOoZPzExMXKFlimpYxYEgb755hv1RkytW7c26PLS7AiCQP3796ctW7Zk++rXr58sM1Li4+MlP2ZW5Ppemyo54j127BgVKFCAypcvT/PmzaP9+/fTqVOnqEmTJuTv70/h4eFUtWpVMjMzo4ULFxogSm0PHz6k4cOHU5kyZTTiK1WqFA0aNIju3LljlHJJbf/+/VSsWDH67rvvNNJv3rypfpi/IAhkaWlJixYtyvHnPH/+nA4dOkS///47rV27lnbu3Knxf2jIFQJyxCwIGW9oFhMTQ4Ig0I8//mi0xwSkl0/qdmyq5yd9ZvXI5fTp0wbf9VSOOo6NjSWVSkUqlYrKlStH5ubmNGjQIFq8eDEJgkCtW7emUaNG0RdffEEWFhZ0+vRpA0T6kRzxnj9/nuzt7dX5W7duTSkpKerd5osXL052dnYkCAJ5e3sbfIMuKWNObyfnzp2jwMBAEgSBihYtSnPnzjWZjcfk6qsz6h/u3r1LgpC7jYFywxT6rXSGuq6+evUqvXjxQqe8U6ZMIScnJ70/4927d7Rt2zbq3LkzOTg4kCiKVKRIERo4cCAdPHhQ7+NJQa7z8effn3379pGFhQWpVCry9PSksmXLkiB8nOn7+aaUcpIz3s83LG7WrJnG4+WMRY6YO3ToQM7OzrR79246ePAgubu7U8+ePalMmTLqx+alpaWpHxcxZswYucOUDc/ANBEeHh4oX748tm7dqk57+vQpXFxcUL9+fezZs0ed7uXlhZIlS2L79u1ZHnPVqlUIDg6Gj48PIiMj4eTkpJXn119/xaBBg2BnZ4ctW7bodBdAKrdv39b7d8qUKZPl+3FxceoZWSVLlsTt27fRr18/VKxYEUOGDEGrVq1QtmxZREVF4ebNmzh69Ci8vb1zVP6ckDrm9LuT3bp1U2/ElJCQgICAAISEhKgfpm8soihCEATo2s0IEs9IiYmJQceOHfHy5UvJjpkdOerYFGcxpZOjHX/zzTe4du0aLly4oHF3MDIyEp06dcKjR4+gUqkQGhqKWbNm4dixY0b9rr958wavXr2Cvb09bG1tjVYOOQwcOBArV67EnTt3UKhQIXV6y5YtERUVhfHjx8PX1xdr167Fpk2bcObMGXh6emZ7XFNeISBHzOl9YUbosw3M0n825Ow8OdqxqZ6f5J7JpIv8cG4CPraJc+fO4cSJEyhZsiRu3ryJwMBA3Lp1C506dUJERASAjzMUPT094e7urtNsVinIEW9QUBD27duHdevWQaVSITg4GLVr18b+/fsRExODihUrgogQERGBwYMHY9SoUQbdIFPKmD9vJ+fOncOUKVMQGRkJZ2dnjBgxAkOGDIGdnV2uypwbcvXVEydOROvWrTXSHz16hKZNm2LhwoVaM6wNMavLFPotwDh9ly7Cw8Px+PFjrQ1p9JGcnIxdu3Zh06ZN2LFjB16/fg0XFxfcv39fwpJmT87z8affnxo1auDBgwc4evSoesPJXbt2oW3btujXrx/mzZundzlywhDxvnv3TuM6xN/fHyEhIUaZTQ3IE3Px4sXRtWtXzJ49GwCwZcsWdOjQAXPmzMGoUaM08nbo0AFnz57VadWjSTLm6Cn7P7a2thQeHq6RFhUVRYIg0KxZszTSQ0NDqVChQjodd9++feTk5EQVK1bU2HQiNTWVhgwZQqIokoeHB928eTPXMZiCFi1aUMmSJenu3btE9PH5gVWqVCE7Ozvq16+fOl9iYiK5u7vr/YwYU/P5HbXP7zQ1atTIqM9WEwSBRo0aRQcPHsz2NWLECJ1nI3748IEiIyNpzpw5FBERkeHzTSIiIsjc3Jy++OILqcMyKFOdxSQnW1tbmjZtmlb633//TYIg0JkzZ9RptWrVovbt2xuyeIpSuXJlatWqlUbakydPyNzcnBo2bKiRXr16dWrZsqVOxxUE3VcIFC1a1KArBOSIWa4NzUyZqZ6f5JzJpOu5qUCBAnn+3EREVKxYMa2Ni3777TcqUKAAPXz4UCM9LCyMHB0dDVk8yRUrVozGjh2r/nnz5s0kiiJ9//33Wnnbt29Pbm5uhiyepDJrJ3FxceoZmU5OThQaGkrPnj0zQgnl6avTZztl9lxqYz2TWu4ZmEq6rs7O48ePaeHCheTk5JTnV0mly+h8LIoihYaGauXt168flSlTxnCFk0Fm7eXz65B69erR3r17jVBC6VlYWNDixYvVP//zzz8kCAJt27ZNK++8efMMvvJWSubGHkBlHyUnJ2vN3ImPj4cgCKhRo4ZGurOzs853vho0aIBjx46hadOmqFWrFqKjo+Hq6or27dtj//79aN68OdatW5dvZg2dPXsWffr0QcmSJQEAbm5uGDFiBPr164ewsDB1Pmtra3Tq1AkLFiwwTkFlYmlpieHDh6N///7qO03169dH3bp1ERISAn9/f4OWp0yZMvDy8tLpWTn379/X6ZlkL1++RKNGjRAbG6vxbMRu3bph+fLlMDc3x9ixY/H999+jbt26GrOa86o2bdpg27Zt6jr19vY2+iwmOYmiqH4e4qeSk5MBAImJieq0Fi1aYMmSJQYrm9Lcvn0bnTt31kg7efIkUlNT0bBhQ410fepixYoVCA4ORsOGDXVaIfDnn3/mPAg9yRGzKTyz1NhM7fwkNSWemxITE7Xa7pYtW+Dv7w8XFxeNdGtra7x9+9aQxZPc06dPNa5TqlWrBiLCf/7zH628vr6+iIyMNGTxDMLT0xPbtm1Tz8icOnUq5s+fj4EDB2o8w98Q5OirV6xYIWkZpZLRNZFUlNh3fe7Ro0f4448/sGnTJhw5cgSpqakoXrw4evToYeyiyeLFixcgIlSqVEnrvYoVK2LlypWGL5QBZHQd0rhxY9SsWRPHjx83dvFypXDhwnjw4IH65xIlSmD58uVa40jAxzb/6az1vIYHME1E8eLFce3aNY20kydPQhAEreUODx8+1OvhuhUrVsSpU6fQokUL1KtXD4UKFcKtW7cwduxYzJw5M9NlbXLr1auXXvkFQcCvv/6aZR5Tv5iWI+aMmEoHffPmTZ3zdurUCZ06dco2X2hoKOLi4jBp0iS0bdsWoihi+/btCA8PR8GCBXH37l1s3rwZXbt2xS+//IICBQrkJgS9yVXHGdXpV199ZfRlEHLEW79+faxZswYjR47UuLmyYsUKCIKAcuXKqdNsbW3x5MkT/QrNdCbXzbXu3bujVKlSaNeuHWrXro2YmBj18pi0tDQMHz4cS5YsQaVKlbBjxw64urpKEo8u5IrZlBnq3AQY//wk10CAEs9NFSpUwP79+9WbC5w6dQoxMTFo27atVt49e/bAzc1NrzLkhhzxmvofiIZsx58PZM6ZM8fgA5hy9NVKvNlk6n2XXBISErBlyxZs2rQJR48eRWpqKooVK4b+/fsjKCjIaJsHytWOk5KS8OzZMwAfBzDTf/dzT548gbOzs15lyA1D9lvpMroOMSQ5Yq5ZsyZ27NiBGTNmAACsrKzQu3fvDPPu2bNH42+pPMeo8z+ZWv/+/algwYLqpdz/+9//yNLSkurUqaOV19fXlzw9PfX+jLdv31Lr1q1JED5uCmRsNjY2ZGtrm+3L0tJSvVwjOz4+PhrLwk+ePEmiKGa4xLRRo0ZUoUIFSWPKjtQx67qk5NMp83mdm5sbdezYUSt90aJF6uU8GS0/NhRD1bGpLIOQox1fvnyZ7OzsqESJEjRixAiaOXMmtW3blgRBoG+//VYjb9++fcnFxUWu8BTP1dWVhg4dqpHWvHlzMjMzo6dPn2qkjx8/nhwcHPQ6/pUrV8jV1ZVcXFwoLi6Onj17Rv7+/iQIArVo0YJev36d2xD0JnfMpkiOdqy085PSzk1EROvWrSNBEKhDhw4UEhJCRYoUIXd3d6pbty716dOH/vrrL/rf//5HQ4cONXj8csTbpk0bqlKlik6f/9VXX5Gfn19uw9CLlDHru2T53LlzEkSgHyX21XIw9b5LSg8ePKBFixaRn58fmZubqzenSt+0Jy0tzdhFlO18nNFjETLaCNPT0zPD8Qe5GPP6I50hN4okkifmbdu2UY0aNejChQtZ5jt06BAJgkDTp0+XKhyD4wFME/HgwQMqXLgw2drako+PD9nY2JAoiloDErGxsSQIAo0cOTLbY36+o2lsbCydOXOGGjZsSHZ2drR+/Xqj7nqanWfPntHkyZPJycmJzMzMqEOHDtn+jilfTOtC35hNvYOWg0qlonnz5mmlnz9/ngRBoB9++MEIpdKd1HX8+UCmr6+v1EXOlZy0Y6KPfwz5+fmRmZkZCYJANjY21K9fP0pKSpK5xOxThri5lpCQQD4+PmRnZ0dubm4kCAKNGzfOaH9IGCLmvCYn7Vhp5yelnZvShYSEkJWVFQmCQJUqVaKrV6/SjRs3qGzZshrPCvzmm28oOTlZ5ih0l5N48/ofiPrEPG7cOJP7u+Bz3FdLI6/3XfpI75MKFSpEwcHBtG/fPkpNTTV2sfSSk74rLCwsw9fu3bs18kVHR5tc35WTeA8ePEgPHjzQ+TPi4+NzU0TJ5fR8rIt79+5RfHw8HT16VLJjGhoPYJqQf/75h9q1a0cVKlSgxo0b065duzTeT0tLI39/f3J0dNTpTmdmd1sEQTDqg6iz8+jRIxo7dizZ29uThYUFdevWja5evarz7+fFi+mcxpzXO+icyOyP4nv37pEgCLRz504jlCp7Oa3jvDqLKbftmIho//79NH36dJo2bZpRN6NSMjlurmXElFYIGCrmvCA37Vhp5yelnZs+9fbtW3r06JFGWlJSEq1evZpmzZpFMTExRGQadSxFvNkxtT8QcxKzm5sbbd261TAFzCHuq6WRV/uunOjduzft3r2bPnz4YOyi6M0QfVc6pfTVRB8Hbe3t7SU/bk4oMeac4AHMfCyv7Xp67949Gjp0KFlbW5NKpaLg4GC6ceNGjo6VVy6mpYw5O3m9s0onCAJNmjRJa/ZwTEwMCYJAP/74o0nNLM5tHee1WUxSfKeTk5MpMDBQ6wZLmzZt6M2bNzKVnGVG6ptrRKa/QkCOmPMSQ56biPLH+Ulp5yZ9GbuOlRYvUe5irlGjBomiSA0bNqTLly/LXNKcU3pfLYW81ncpjaH7rpiYmHzRV3/48IEiIyNpzpw5FBERQRcvXtTKExERQebm5vTFF19IUfQcU2LMucEDmMzobty4QX379iWVSkXW1tY0dOhQunfvnuyfa8yLSyljVkpnlS6zmcUZzS425sxiqeo4r8xikvI7PWPGDBIEgQYOHEgnTpyg+Ph4CgsLIwsLC+rSpYvEJWfGkFdXCOR3Up+PlXR+Utq5KZ2p17HS4iWSLuaVK1dSiRIlqECBAjR06FB6/vy59IVlRpdX+i6lUVrfJWW8L168IG9vb63vcI8ePSglJYWIiMaMGUOCIJCfn5/WM3MNRYkxS4F3IWdG89dffyE8PBwbNmyAtbU1hg8fjlGjRqFw4cK5Om5qaiqio6Nx9epV2Nvbo3bt2vDw8NDI8/PPP2Pw4MEG34FL6phfvnyJRo0aITY2FkQE4ONOZd26dcPy5cthbm6OsWPH4vvvv0fdunWxdetWKcMxihUrVhi7CFmSuo7r1aunc96YmBh07NjRoLsiy9GO169fjyZNmmDJkiXqtGrVquHDhw+YOXMmFi5ciIIFC0pRfGYkpt6OlUaOdqy085Opf6eVVsdKixeQPubu3bujffv2mDlzJubNm4d169Zh+vTpCA4OznD3YpY3mXrfpTRK67vkiDc0NBRxcXGYNGkS2rZtC1EUsX37doSHh6NgwYK4e/cuNm/ejK5du+KXX35BgQIFJIwoe0qMWVJGHT5liiaKIjk7O1NYWJhkd3VN/e6D1DEPGzaMzMzMaPLkyXT+/Hm6ePEiTZ8+naytrWnkyJHUvn17EgSBunXrZjLP+8zv5Phem/JdUznitbKyyvCB8hcvXiRBEPh5mIxJTI52zOcn06K0OlZavETyxJzuzp071LJlSxJFkaytrcnOzk7jZezl8ozlF0rru+SI183NjTp27KiVvmjRIvXYgDE38lVizFISiP7/MDxjBiaKIlQqlc53AARByHZm2fDhw7F48WJMnDhR6+5D//79jX73QeqYy5Yti5o1a2L9+vUa6YsXL8bQoUMhCAKmTJmCSZMm5arcTHdS13FO7poacnaiHO1YFEVEREQgODhYI/3evXsoXbo0du7ciYCAgByXmTGmSY52zOcn06K0OlZavIA8MQNAWloali5dirCwMLx48QLNmzeHk5OTVj6eycdY7imt75IjXktLS8ycORMjRozQSL9w4QK+/PJLfP/99xg5cmSOy5xbSoxZSryEnBlN9+7dJT9mZGQk2rdvjylTpqjTPDw84ODgoO6gp06darSLS6ljfvDgAXx8fLTS/fz8AABz587NN51VXiF1HZv6kgA52jEA3L17F3FxcRppjx49AgBcu3ZN673q1avLUg7GlECOdsznJ9OitDpWWryAPDHv3LkTY8aMwdWrV+Hn54f58+fD09NT8s9hjH2ktL5LjniTk5NhY2Ojle7s7AwAqFChguSfqQ8lxiwlHsBkRiPHnVpT7qAB6WNWUmeVV0hdx6Y+KC/XjIvw8HCEh4dn+N7w4cPV/yYiCIKA1NRUWcrBmBLI0Y75/GRalFbHSosXkDbmS5cuYeTIkdi3bx9cXV2xadMmtG3bVrLjM8YyprS+S66/I7KaCHH9+nWjToRQYsxS4gFMZhKuX7+O33//HZcvX8bTp09RuHBhfPnllwgKCkKJEiV0Po4pd9CfkypmpXRWeZEUdWzqg/Kfkuo7zcvQGDMeqdoxwOcnU6W0OlZavOllyU3MX375JWxsbBAeHo4RI0bAwsLCAKVmjH1KaX2XlPFmNRFi2LBh6o3IjD0RQokx5xY/A5MZ1YcPHzB69GgsWbIEqampsLKygpOTE168eIGkpCSYm5tj9OjRmDZtGszMzLI9niiKmDhxIlq3bq2R/ujRIzRt2hQLFy5E7dq1Nd4zdActZcyiKGa6E+Snz0tM/zkvd1Z5idR1nNHzIO/fv49SpUohKioKTZs2lTOcbEndjhljhifH+ZjPT6ZFaXWstHgB6WIODg7G9OnTUaRIEQOWnjEGKK/vkjreVatW6V0GuR6JlRklxiwVnoHJjKpbt27YsGEDevbsiREjRsDDw0P93pUrV/Djjz9i1qxZuH//vs4N09TvPkgZM89UM01Sf69N/a6pHO2YMWZYUrdjPj+ZHqXVsdLiBaSLedmyZYYoLmMsA0rru6SONy8MzCkxZskYeNdzxtQiIyNJEASaMWNGlvlmz55NoijSjh07sj3mypUr9X4ZkhwxM9MidR0LgkCiKGb4EgRB4/30fxsSf6cZy/u4Hed/SqtjpcVLpMyYGctvlNaOlRYvkTJjlhIvIWdG0759e8THx+Pvv//OdFp7usqVK8Pd3R1bt241UOnkocSYlUbqOjb1JQH8nWYs7+N2nP8prY6VFi+gzJgZy2+U1o6VFi+gzJilxEvImdHExsaiZcuW2TZcAAgICMDmzZsNUCp5KTFmpZG6jk19SQB/pxnL+7gd539Kq2OlxQsoM2bG8hultWOlxQsoM2YpicYuAFOuR48ewc3NTae8pUqVUj/zLy9TYsxKo7Q6Vlq8jOVH3I7zP6XVsdLiBZQZM2P5jdLasdLiBZQZs5R4AJMZTVJSEqysrHTKa2lpiffv38tcIvkpMWalUVodKy1exvIjbsf5n9LqWGnxAsqMmbH8RmntWGnxAsqMWUq8hJwZVUa7K2fkzp07BiiNYSgxZqVRWh0rLV7G8iNux/mf0upYafECyoyZsfxGae1YafECyoxZKryJDzMaURR1evYDABARBEFAamqqzKWSlxJjVhql1bHS4mUsP+J2nP8prY6VFi+gzJgZy2+U1o6VFi+gzJilxDMwmdGsWLHC2EUwOCXGrDRKq2OlxctYfsTtOP9TWh0rLV5AmTEzlt8orR0rLV5AmTFLiWdgMsYYY4wxxhhjjDHGTBZv4sMYY4wxxhhjjDHGGDNZPIDJGGOMMcYYY4wxxhgzWTyAyRhjjDHGGGOMMcYYM1k8gMkYY4wxxhhjjDHGGDNZPIDJGGOMMaZgt27dgiAIaNKkibGLwhhjjDHGWIZ4AJMxxhhjzIjOnz+PAQMGoGLFirC1tYVKpUKZMmXQtm1brFmzBu/evVPnDQsLgyAI2LBhQ6bH0yWPoVlaWsLV1dXYxcjSv//+C0EQUL9+fWMXRS8rV66EIAjo0aOHQT83L9QpY4wxxvIPc2MXgDHGGGNMiYgI48aNw9y5cyGKImrXro2vv/4aaWlpuHbtGnbu3Ik//vgDc+bMwcWLF41dXGaiPDw8MGrUKHh7exu7KIwxxhhjsuEBTMYYY4wxIwgJCcHcuXNRvnx5bNq0CdWqVdN4/9mzZ5g3bx5+//13I5WQ5QU1atRAjRo1jF0MxhhjjDFZ8RJyxhhjjDEDu3btGmbNmgVbW1vs3r1ba/ASAAoWLIjp06djz549BivXpUuX0KVLFxQtWhQqlQpeXl74888/M8wbFxeH1q1bo1ChQlCpVHB3d0dISAgSExPVedKfr/n+/Xvcvn0bgiCoX58ueT527BgCAwNRuHBhFChQAMWLF0dgYCDWrFmjdawmTZrg7NmzaNOmDZydnWFlZYU6derg2LFjGZZz3759aNy4MRwdHWFlZYWqVavihx9+wIcPH9R5Dh48iGLFigEADh06pFHOsLCwLP/PwsPDIQgCZs2aleH7//3vfyEIAiZPngwAePr0Kdq3bw8PDw8UKlQI5ubmcHZ2RsOGDREZGan1+z169IAgCDh27BiWLVuG6tWrQ6VSwdraWl12QRDQv39/jd+bNWsWfH19UbJkSahUKtjY2MDDwwOhoaFISkrSyPvpMQ4dOoTWrVujYMGCsLKyQr169RAbG6tVD9nVKWOMMcaYlHgGJmOMMcaYga1atQqpqano3bs33NzcssxbpkwZg5Tp9OnT8PT0RMWKFdGuXTs8fvwYmzdvRosWLfDPP/+gdOnS6rzR0dFo3bo1BEFAo0aNUKxYMZw6dQrTp09HTEwMDh8+DGtrazg4OGDUqFFYsGABbGxs0LdvX/Ux0pc879q1C82bN4eNjQ2aNWsGZ2dn3LhxA3v37sW+ffvQtWtXjXIeOXIEPj4+8PLyQlBQEG7fvo1du3ahYcOGOH78ODw9PdV5f/nlFwQHB8PW1haNGjWCo6MjDh8+jNGjR+PAgQOIjIyEKIooVaoUBgwYgJ9++gklS5ZEUFCQ+hhfffVVlv9vPXr0wOTJk7Fq1Sp89913Wu+nP6OyV69eAIDnz59j586d8PHxgbe3N+zt7ZGQkIA9e/YgMDAQa9euRadOnbSOExQUhGfPnqFp06aoW7cu/v333yzLFR0djZcvX6JOnTooUqQI3r17hxMnTmDq1Kk4deoUdu3apfU7v//+O37++Wf4+vqic+fOuH79OmJiYtCkSRPcuHEDdnZ2OtUpY4wxxpjkiDHGGGOMGdTXX39NAGjXrl16/V5oaCgBoObNm9OoUaMyfPn6+hIAWr9+vU7HvHnzJgEgR0dHWrlypcZ74eHhBIC+++47dVpiYiIVLlyY7O3t6cyZM+r0tLQ0GjVqFAGgMWPGaBxHpVJRmTJlMvz8OnXqEAC6dOmSRvqLFy9o9uzZWuUsUqQIbd++XSPvr7/+SgAoICBAnXb37l1SqVRUunRpunbtmjr9/fv31LFjRwJAS5YsUac/fPiQAFC9evUy+Z/KXPPmzQkAnThxQiP9zp07JIoiNWrUSJ2WkpJC79+/1zrGtWvXSBRF8vT01Ejv3r07ASA/Pz+6ceOG1u8dOHCAAFC/fv000hMTE7Xypqamqr9758+f1zpGiRIlKCoqSuN3goODCQBFRERopGdVp4wxxhhjUuMZmIwxxhhjBpY+e+7TWY36iIqKQlRUlJRFQs2aNdG9e3eNtICAAEyYMAF///23Om3btm14/Pgxxo8fr/HsxfRl1GvWrMEvv/yCWbNmQRSzf1rRy5cvAQCOjo4a6Q4ODhg7dqxWfk9PT7Rs2VIjrUePHpgyZQr27NmDd+/ewdLSEqtXr8b79+8xceJElCtXTp3XwsICixcvxubNm7Fs2TIMHDgw2zJmp3fv3oiKisLKlStRq1Ytdfrq1auRlpaGPn36qNPMzT9eficmJuLUqVO4du0aXr58iVevXsHa2hrXrl3L8DNmz56d7WzdT6UvMT9//jwuXryIJ0+e4NWrV3j//j2Aj48xqFq1qsbvNG/eHM2aNdNICwgIwLJlyzS+A4wxxhhjhsYDmIwxxhhjecz69evRsWPHDN8LCwvDlClTJPkcZ2dnAB83FEoXFxcHAKhdu7ZWfnNzc/j4+CAqKgo3b97UGDjMTIsWLXDx4kV4e3ujS5cuqFOnDqpXr46SJUvqXE5RFFGpUiXcuXMHN27cQKVKlbIsp7OzMypUqICLFy8iJSUFBQoU0PmzMtK8eXO4uLhg48aNWLBgASwtLQF8HMAsVKgQWrVqpc6bnJyMCRMm4KefftJ6FiUAqFSqXJUlXVRUFEaMGJHpgGhycrJOx8noO8AYY4wxZmi8iQ9jjDHGmIEVKVIEAHDnzh0jlyRr6QN7RKROe/HiBQDA3t4+w99JT0/Pl52pU6eqB1znzp2LwMBAlCpVCtWrV8fRo0d1Lmv6DM63b9/qXM60tDS8evVK58/IjLm5Obp3744XL15g27ZtAIDjx4/j77//RteuXWFhYaHOO2zYMPzwww8oWbIkli5ditjYWCQkJCA5OVmy550eP34crVq1wt27dzFmzBjs27cPt27dQmJionozIV1l9B1gjDHGGDM0HsBkjDHGGDOwmjVrAgB27txp5JLoz8HBAQDw8OHDDN9/8OABAO0l4ZkxMzPD5MmTcf/+fVy6dAm//fYbOnbsiAsXLiAgICDTz8nscwsVKqRzOUVRzHSAU1+9e/cG8HHTHuDjRk0ANJaPp6Sk4LfffoOzszNOnz6NAQMGoHr16ihSpEiuZ4F+avny5UhNTcUvv/yCOXPmoEGDBihTpgysra0hCIJkn8MYY4wxZig8gMkYY4wxZmCdO3eGIAj49ddfcf369SzzPn361ECl0o2XlxcA4NixY1rvvXnzBufPn4eTk5PG8xrNzc2zXbIsCAIqV66Mnj17Yv369Rg4cCDevHmD48ePZ1umly9fIj4+HsWKFVPPYsyqnHfu3MHdu3fh4eGhHjhMfzalrkurP1e+fHn4+flhz549uH79OjZu3AhfX19UqlRJnef58+dITk5G5cqV1QOsckhISACQ8fJ5qehSp4wxxhhjUuEBTMYYY4wxA6tatSoGDhyIpKQkfPPNNzh79qxWnrdv32LBggUaz080BYGBgXB2dsby5ctx+vRpdXpaWhrGjx+P58+fo1evXhob+JQoUQIPHz7ElStXtI7Xs2fPDDeIuX//PoD/m1GZLqNBs5CQELx69Qp9+/ZVp3Xp0gUWFhaYM2cO/vnnH3X6+/fvMWLECKSmpmrkd3JygrW1NeLj43M8aNynTx+kpaXh22+/xcuXLzVmXwIfHx3g6OiIS5cu4fnz5+r01NRUzJ49Wz2LNLfc3d0BAEeOHNFIP378ONauXSvJZ2RVp4wxxhhjUuNNfBhjjDHGjGDBggVISkrCihUr4O3tjVq1auHLL7+EKIq4desWjhw5gtevX8PPz8/YRdVgY2ODlStXok2bNqhbty4aN26MYsWK4dSpU7hw4QI8PT21NhEKCgrCtGnT4O/vj2bNmiExMRFVq1bF+PHjsXXrVqxatQo+Pj7q+E+fPo3Y2Fg0adJEK/4DBw7A3d0dtWvXhq2tLY4dO4Zz586hVq1aGD9+vDpf6dKl8eOPP2LAgAHw8vJCo0aN4OjoiMOHD+PatWto0qQJBgwYoM5vZmaGtm3bYs2aNfDx8YG/vz+ePXuGZs2aoWfPnjr937Rr1w5DhgzBmTNnYGdnh6CgIK083333Hb777jtUrVoVAQEBAIB9+/apl7RLYejQoVi+fDn69OmDyMhIFCtWDH/99RcOHDgg2ZL5rOqUMcYYY0xqPAOTMcYYY8wIzM3N8dtvv2H//v349ttvkZCQgJUrV2LFihX43//+h8DAQGzatAl79+41dlG1NG/eHMePH0dAQACOHz+OVatWISkpCRMmTMDhw4dhY2OjkT8kJASTJ0+GlZUV1qxZg6NHj6rz/PHHH+jWrRsePXqEFStWYOPGjRBFET/99BO2b9+u9czGatWqoUGDBjhw4ACWLVuG58+fY8KECdi7d6969+90/fr1w+7du1GzZk3s3bsXa9euhUqlwpw5cxAZGQkzMzON/EuWLMHgwYORkpKC1atXIy4uDnZ2djr/v1hZWaFTp04AgI4dO2r9PwDA2LFjsXDhQlhZWWH16tXYvXs3/P39cfnyZbi4uOj8WVkpW7YsDh8+jNq1a2P37t1Yt24dBEHA5s2bMWzYMEk+I6s6ZYwxxhiTmkC8pSBjjDHGGDNxt27dgpubG7755hvs2rXL2MVhjDHGGGMGxDMwGWOMMcYYY4wxxhhjJosHMBljjDHGGGOMMcYYYyaLBzAZY4wxxhhjjDHGGGMmi5+ByRhjjDHGGGOMMcYYM1k8A5MxxhhjjDHGGGOMMWayeACTMcYYY4wxxhhjjDFmsngAkzHGGGOMMcYYY4wxZrJ4AJMxxhhjjDHGGGOMMWayeACTMcYYY4wxxhhjjDFmsngAkzHGGGOMMcYYY4wxZrJ4AJMxxhhjjDHGGGOMMWayeACTMcYYY4wxxhhjjDFmsv4fYpS0YVJrXk4AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1600x400 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# Prepare\n",
    "stacked_df_carriers = merged_df_percents_melt[merged_df_percents_melt['group'] == 'carriers'].pivot_table(\n",
    "    index='gene_var2', columns='binding level', values='value', aggfunc='sum').reset_index()\n",
    "stacked_df_noncarriers = merged_df_percents_melt[merged_df_percents_melt['group'] == 'noncarriers'].pivot_table(\n",
    "    index='gene_var2', columns='binding level', values='value', aggfunc='sum').reset_index()\n",
    "bar_width = 0.4\n",
    "\n",
    "# Plotting\n",
    "colors1 = ['#1c79cd', '#f72313', '#9c180e']\n",
    "colors2 = ['#76aee1', '#fb9189', '#ce8d88']\n",
    "\n",
    "plt.figure(figsize = (16, 4))\n",
    "\n",
    "for i, var in enumerate(order_all):\n",
    "\n",
    "    plt.bar(2*i, stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['no binding'], color=colors1[0], width=bar_width, label='no binding, carriers' if i == 0 else None)\n",
    "    plt.bar(2*i+0.5, stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['no binding'], color=colors2[0], width=bar_width, label='no binding, non-carriers' if i == 0 else None)\n",
    "    \n",
    "    plt.bar(2*i, stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['weak binding'], color=colors1[1], width=bar_width, label='weak binding, carriers' if i == 0 else None, bottom=stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['no binding'])\n",
    "    plt.bar(2*i+0.5, stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['weak binding'], color=colors2[1], width=bar_width, label='weak binding, non-carriers' if i == 0 else None, bottom=stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['no binding'])\n",
    "    \n",
    "    plt.bar(2*i, stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['strong binding'], color=colors1[2], width=bar_width, label='strong binding, carriers' if i == 0 else None, bottom=stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['no binding']+stacked_df_carriers[stacked_df_carriers['gene_var2']==var]['weak binding'])\n",
    "    plt.bar(2*i+0.5, stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['strong binding'], color=colors2[2], width=bar_width, label='strong binding, non-carriers' if i == 0 else None, bottom=stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['no binding']+stacked_df_noncarriers[stacked_df_noncarriers['gene_var2']==var]['weak binding'])\n",
    "\n",
    "plt.xlim(-1, 2*i+2)\n",
    "order_all2 = [item for item in order_all for _ in range(2)]\n",
    "tick_positions = [tick + 0.25 if i % 2 == 0 else tick for i, tick in enumerate(range(len(order_all) * 2))]\n",
    "tick_labels = [f'{var}' if i % 2 == 0 else '' for i, var in enumerate(order_all2)]\n",
    "plt.xticks(tick_positions, tick_labels, rotation=90, fontsize = 12)\n",
    "plt.xlabel(f'CH hotspot variant', fontsize = 14)\n",
    "plt.ylabel('Percentage (%)', fontsize = 14)\n",
    "plt.yticks(fontsize = 13)\n",
    "plt.title('MHC I', fontsize = 16)\n",
    "plt.legend().remove()\n",
    "plt.tick_params(left = True, bottom = False) \n",
    "plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_{param}_percentages_by_threshold.pdf', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at VAF scores for cases where you have both binding and no binding \n",
    "ch_all_el_rank_log_labels = ch_all_el_rank_log_labels[ch_all_el_rank_log_labels['gene_var2'].isin(order)]\n",
    "\n",
    "# let's add to everyone group based on score\n",
    "conditions = [\n",
    "    (ch_all_el_rank_log_labels['score'] < -1 * np.log10(2)),\n",
    "    ((ch_all_el_rank_log_labels['score'] >= -1 * np.log10(2)) & (ch_all_el_rank_log_labels['score'] < -1 * np.log10(0.5))),\n",
    "    (ch_all_el_rank_log_labels['score'] >= -1 * np.log10(0.5))\n",
    "]\n",
    "values = ['no binding', 'weak binding', 'strong binding']\n",
    "ch_all_el_rank_log_labels['binding'] = np.select(conditions, values)\n",
    "ch_all_el_rank_log_labels['binding']= pd.Categorical(ch_all_el_rank_log_labels['binding'], categories=['no binding', 'weak binding', 'strong binding'], ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# add number of cases to the dataframe to be able to include on the x axis \n",
    "\n",
    "df_with_nr_carriers = df_compare_to_ref2[['gene_var2', 'CH variant3']]\n",
    "ch_all_el_rank_log_labels_with_nr = pd.merge(ch_all_el_rank_log_labels, df_with_nr_carriers, on = 'gene_var2')\n",
    "ch_all_el_rank_log_labels_with_nr = ch_all_el_rank_log_labels_with_nr[~ch_all_el_rank_log_labels_with_nr.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# okay now binding / no binding\n",
    "# look at VAF scores for cases where you have both binding and no binding \n",
    "ch_all_el_rank_log_labels = ch_all_el_rank_log_labels[ch_all_el_rank_log_labels['gene_var2'].isin(order)]\n",
    "\n",
    "# let's add to everyone group based on score\n",
    "conditions = [\n",
    "    (ch_all_el_rank_log_labels_with_nr['score'] < -1 * np.log10(2)),\n",
    "    (ch_all_el_rank_log_labels_with_nr['score'] >= -1 * np.log10(2))\n",
    "]\n",
    "values = ['no binding', 'binding']\n",
    "\n",
    "ch_all_el_rank_log_labels_with_nr['binding_any'] = np.select(conditions, values)\n",
    "ch_all_el_rank_log_labels_with_nr['binding_any']= pd.Categorical(ch_all_el_rank_log_labels_with_nr['binding_any'], categories=['no binding', 'binding'], ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# okay now repeat this but binding / no binding \n",
    "\n",
    "# define function to plot cumulative probability\n",
    "def plot_cumulative_mhci(data, x, y, hue, colors, xlabel, ylabel, var, score, score_title, name):\n",
    "\n",
    "    var_name = var.split('_')[0:2]\n",
    "    var_name = ' '.join(var_name)\n",
    "\n",
    "    nr_cases = data.shape[0]\n",
    "\n",
    "    plt.figure(figsize=(4,4)) \n",
    "    sns.lineplot(x=x, y=y, data=data, hue=hue, palette=colors, legend = True)\n",
    "    sns.scatterplot(x=x, y=y, data=data, hue=hue, palette=colors, legend = False, alpha = 0.3, linewidth=0)\n",
    "\n",
    "    # specify title and axes labels \n",
    "    plt.title(f'{var_name} (n={nr_cases})', fontsize = 12)\n",
    "    plt.xlabel(xlabel, fontsize = 10)\n",
    "    plt.ylabel(ylabel, fontsize = 10)\n",
    "\n",
    "    # convert both axes to log scale \n",
    "    plt.gca().set_xscale('log')\n",
    "    plt.gca().set_yscale('log')\n",
    "\n",
    "    # adjust legend \n",
    "    if var == 'JAK2_V617F':\n",
    "        legend = plt.legend()\n",
    "        legend.prop.set_size(10) \n",
    "        plt.legend(title = 'MHC I presentation', loc = 'lower left', fontsize = 10, title_fontsize = 11)\n",
    "    else:   \n",
    "        plt.legend().remove()\n",
    "    # specify ticks on the x axis \n",
    "    x_ticks = [1, 3, 10, 30]\n",
    "    plt.xticks(x_ticks, x_ticks, fontsize = 9)\n",
    "\n",
    "    y_ticks = [1, 0.3, 0.1, 0.03]\n",
    "    plt.yticks(y_ticks, y_ticks, fontsize = 9)\n",
    "    \n",
    "    plt.savefig(f'/Users/barbarawalkowiak/Desktop/msc_thesis/project_report/folders_for_git/figure 8/UKBB_VAF_{hue}_{name}_{score}_{var}_threshold_mhci.pdf', bbox_inches='tight')\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# okay now do the cumulative plots for this \n",
    "\n",
    "# plots scores for each group on the same plot \n",
    "# note: an alternative is to split by thresholds for binding  \n",
    "# script to do this is included in the complete analysis, but I am not using this way due to issues described there \n",
    "\n",
    "# for extreme groups in each split \n",
    "dfs_list = ['ch_all_el_rank_log_labels_with_nr']\n",
    "\n",
    "for name in dfs_list:\n",
    "\n",
    "    df = globals()[name]\n",
    "\n",
    "    score = df.score_type[0]\n",
    "    score_title = score.replace('_', ' ')\n",
    "    \n",
    "    df['VAF_percent'] = df['VAF'] * 100\n",
    "    colors = colors1\n",
    "\n",
    "    variants = df['gene_var'].unique()\n",
    "\n",
    "    # plot for each variant separately \n",
    "    for var in variants: \n",
    "\n",
    "        data = df[df['gene_var'] == var]        \n",
    "        data = data.sort_values(by='VAF')\n",
    "        N = data.shape[0]\n",
    "        # add number of people in each category \n",
    "        data['group_size'] = data.groupby('binding_any')['binding_any'].transform('count')\n",
    "        data['index_vaf_group'] = data.groupby('binding_any')['VAF'].rank(ascending=False) # index by VAF\n",
    "        data['fraction_index_presenter_group'] = data['index_vaf_group'] / data['group_size']\n",
    "    \n",
    "        plot_cumulative_mhci(data, 'VAF_percent', 'fraction_index_presenter_group', 'binding_any', colors, 'VAF (%)', f'Cumulative probability', var, score, score_title, f'bind_nonbind')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
